---
title: <b style="font-size:45px;">Data Filtering and Quality Control</b>
date: "August 13<sup>th</sup>, 2019"
output:
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 1
    vignette: >
      %\VignetteEngine{knitr::rmarkdown}
      %\VignetteEncoding{UTF-8}
---

<br>

<br>

# **Creating a Seurat object**

```{r "Load packages", include = FALSE}

library(Seurat)
library(tidyverse)

knitr::opts_chunk$set(
  message   = FALSE,
  warning   = FALSE,
  comment   = "#>",
  fig.align = "center"
)

```

```{r "Load data"}

# Data URL
data_url = "https://scrnaseq-workshop.s3-us-west-2.amazonaws.com"

pbmc_mtx <- file.path(data_url, "PBMC_cDNA.csv.gz") %>%
  read_csv() %>%
  column_to_rownames("X1") %>%
  as.sparse()

pbmc_mtx[1:10, 1:10]

# Create Seurat object using gene expression data
sobj <- pbmc_mtx %>%
  CreateSeuratObject(
    min.cells = 5,      # Remove genes that are detected in <5 cells
    project   = "PBMC"
  )

```

<br>

<br>

# **Interacting with the Seurat Object**

## Retrieving raw and normalized counts
```{r "Retrieving raw and normalized counts"}

# Main slot where raw and normalized data are stored
sobj@assays

# Raw counts
sobj@assays$RNA@counts[1:10, 1:10]

# Normalized counts
# This matrix is the same as counts since we haven't normalized the data yet
sobj@assays$RNA@data[1:10, 1:10]

```

## Accessing project meta.data
```{r "Accessing meta.data"}

# The Seurat object stores the current default assay and cell identities
# By default the cell identity is set to the project name
sobj@active.assay
sobj@active.ident[1:10]

# The meta.data table contains stats for each cell
sobj@meta.data %>%
  head()

# Modify orig.ident to include the sample number
# This is useful later when creating separate plots for each sample
sobj@meta.data <- sobj@meta.data %>%
  rownames_to_column("cell_bc") %>%
  mutate(sample_id = str_extract(cell_bc, "[0-9]+$")) %>%
  unite(orig.ident, orig.ident, sample_id, sep = "-") %>%
  column_to_rownames("cell_bc")

sobj@meta.data %>%
  head()

```

## Other data stored in the Seurat object
```{r "Other data stored in the Seurat object", eval = FALSE}

# Slots for downstream analysis
sobj@graphs
sobj@neighbors
sobj@reductions

```

<br>

<br>

# **Using Seurat functions to interact with the data**

## Retrieving cell and features names
```{r "Retrieving cell names"}

# Get cell names
Cells(sobj) %>%
  head(20)

# Get feature names
rownames(sobj) %>%
  head(20)

# Get cell (colums) and feature (rows) counts
ncol(sobj)
nrow(sobj)

```

## Setting cell identities
```{r "Setting cell identities"}

# Retrieve cell identities
Idents(sobj) %>%
  head(20)

# Set cell identities
Idents(sobj) <- sobj@meta.data$orig.ident

# Rename cell identities
sobj <- sobj %>%
  RenameIdents(
    "PBMC-1" = "control",
    "PBMC-2" = "treated"
  )

# Add cell identities to the meta.data table
sobj[["orig.ident"]] <- Idents(sobj)

sobj <- sobj %>%
  AddMetaData(
    metadata = Idents(sobj),
    col.name = "orig.ident"
  )

```

## Subsetting the Seurat object
```{r "Subsetting the Seurat object"}

# Subset based on gene expression
sobj %>%
  subset(CD8A > 1) %>%
  ncol()

# Subset based on cell identity
sobj %>%
  subset(idents = "control") %>%
  ncol()

# Subset based on expression and cell identity
sobj %>%
  subset(CD8A > 1, idents = "control") %>%
  ncol()

# Subset based on meta.data columns
sobj %>%
  subset(nFeature_RNA > 250)

```

<br>

<br>

# **Assessing data quality**

## Calculate the percentage mitochondrial reads for each cell
```{r "Calculate percent mito"}

# Calculate percentage of mitochondrial reads for each cell
sobj <- sobj %>%
  PercentageFeatureSet(
    pattern  = "^MT-", 
    col.name = "percent_mito"
  )

sobj@meta.data %>%
  head()

```

## EXERCISE: Create violin/box plots comparing cell metrics
```{r}

# Create violin/boxplots comparing nCount, nFeature, and percent_mito for each sample

# sobj@meta.data %>%



```

## ANSWER
```{r echo = FALSE, fig.width = 8, fig.height = 6}

# Using Seurat functions
# sobj %>%
#   VlnPlot(
#     features = c("nCount_RNA", "nFeature_RNA", "percent_mito"),
#     ncol     = 3,
#     pt.size  = 0.25
#   )

# Using ggplot2
sobj@meta.data %>%
  gather(key, value, -orig.ident) %>%
  ggplot(aes(orig.ident, value, color = orig.ident)) +
  geom_violin(size = 2) +
  geom_jitter(size = 0.5, color = "black", alpha = 0.4) +
  facet_wrap(~key, scales = "free") +
  theme(legend.position = "none") +
  cowplot::theme_cowplot()

```

## EXERCISE: Create scatter plots comparing cell metrics
```{r}

# Create a scatter plot comparing nCount and percent_mito

# sobj@meta.data %>%



# Create a scatter plot comparing nCount and nFeature

# sobj@meta.data %>%



```

## ANSWER
```{r echo = FALSE, fig.width = 8, fig.height = 8}

# Using Seurat functions
# scatter_1 <- sobj %>%
#   FeatureScatter(
#     feature1 = "nCount_RNA", 
#     feature2 = "percent_mito"
#   )

# scatter_2 <- sobj %>%
#   FeatureScatter(
#     feature1 = "nCount_RNA",
#     feature2 = "nFeature_RNA"
#   )

# Using ggplot2
scatter_1 <- sobj@meta.data %>%
  ggplot(aes(nCount_RNA, percent_mito, color = orig.ident)) +
  geom_point() +
  cowplot::theme_cowplot()

scatter_2 <- sobj@meta.data %>%
  ggplot(aes(nCount_RNA, nFeature_RNA, color = orig.ident)) +
  geom_point() +
  cowplot::theme_cowplot()

CombinePlots(
  plots = list(scatter_1, scatter_2),
  ncol  = 1 
)

```

<br>

<br>

# **Filtering cells**

```{r "Filtering cells"}

# Filter cells based on number of genes and percent mito UMIs
sobj <- sobj %>%
  subset(
    nFeature_RNA > 250 &   # Remove cells with < 250 detected genes
    nFeature_RNA < 2500 &  # Remove cells with > 2500 detected genes (could be doublets)
    percent_mito < 15      # Remove cells with >15% mitochondrial reads
  )

```

## Saving the Seurat object
```{r "Saving Seurat object", eval = FALSE}

sobj %>%
  write_rds("filtered_sobj.rds", compress = "gz")

```



