---
title: <b style="font-size:45px;">Cluster Markers and Cell Type Assignment</b></b>
author: "Rui Fu, Kent Riemondy"
date: "November 11<sup>th</sup>, 2020"
output:
  distill::distill_article:
    toc: false
---

```{r "settings", include = FALSE}
knitr::opts_chunk$set(
  message   = FALSE,
  warning   = FALSE,
  echo      = TRUE,
  comment   = "#>",
  fig.align = "center"
)
```

## finding markers and differential expression analysis
After clustering, differential expression testing (DE analysis, similar to bulk RNA seq) finds gene signatures of each cluster/cell population, to give more insights into biology.

### many proposed methods  
(In general, we recommend default "wilcox" or "t", good balances between speed and  accuracy)

![Soneson and Robinson, 2018](4_markers_files/detests.jpg)

In Seurat, we can use the `FindAllMarkers()` function, which will perform the `wilcox.test()` by default . Specifically, the function will iterate through each cluster, comparing the cells in one cluster to the cells in all of the other clusters. The test is run for every gene that is detectable above certain thresholds. The output is then filtered to identify significant genes with a positive fold-change (i.e. higher in abundance the cluster).

We recommend using the `wilcoxauc()` function from the `presto` package (installed from github), which will do the same thing but in a (tiny) fraction of the time, due to implementation with C++ backend.

### important considerations
1. Keep things simple first and look at the results (try not integrating, not regressing), look at the output, and determine what variables to regress out (batch, nCount_RNA, percent_mito, cell cycle, if needed) during scaling (note this only affects dimension reduction and clustering).

2. Note that the p-values from these tests will be absurdly low (near or at 0). This is in part due to the large number of samples tested (e.g. each cell is considered a sample) but also due to the data being clustered based on the gene expression. Testing for differential expression between clusters will always result in some differntially expressed genes... because that's what makes them different clusters in the first place, hence the logical is somewhat circular and inflate p-values. Nevertheless we can use the p-values to rank the genes, but you shouldn't just blindly trust that a p-value < 0.05 is something meaningful in single cell data.

A couple more words on preprocessing:

```{r "Load preprocessing data"}
library(Seurat)
library(tidyverse)
# load data from saved RDS
sobj <- readRDS("/Users/rf/class/scRNAseq/cellar_v2/cellar/_vignettes/data/filtered_sobj.rds")

# log normalize
sobj_l <- NormalizeData(sobj) # <- skip this step if using scran normalization
sobj_l <- ScaleData(sobj_l,
                    vars.to.regress = c("nCount_RNA",
                                        "percent_mito"))

# alternatively, sctransform is a one step normalization and scaling process
sobj_sc <- suppressWarnings(SCTransform(object = sobj, # sctransform has some unhelpful warnings 
                                        vars.to.regress = c("percent_mito"), # already corrects for transcript number
                                        verbose = FALSE))
```

3. Use normalized data for DE (slot is dependent on normalization method, also don't use integrated assay)

```{r "explore normalization"}
# before normalization, "data" slot is the same as counts
sobj@assays$RNA@data[101:105,101:105] # raw counts

# after log normalization, results are stored in "data" slot
sobj_l@assays$RNA@data[101:105,101:105] # normalized

# after sctransform, results are stored in new assay "SCT"
sobj_sc@assays$SCT@data[101:105,101:105] # normalized
sobj_sc@assays$RNA@data[101:105,101:105] # still same as counts, make sure not to use this!
```

4. Note that marker genes found is very dependent on clustering and the compared populations.

### find all markers for each cluster
`FindAllMarkers` compares cells in each cluster to all other cells in the dataset. Typically, focus is given to genes upregulated in each cluster, ie markers.

```{r "FindAllMarkers"}
# load clustered object
sobj_clusters <- readRDS("/Users/rf/class/scRNAseq/cellar_v2/cellar/_vignettes/data/clustered_sobj.rds")

# also check/set ident to the desired comparison
Idents(sobj_clusters) %>% head()
Idents(sobj_clusters) <- sobj_clusters@meta.data$clusters

markers_df <- FindAllMarkers(sobj_clusters,
                             assay = "RNA", # be careful with sobj_sc and intergrated objects
                             slot = "data",
                             only.pos = TRUE)
markers_df %>% head()
```

pct.1: fraction detected in the cluster in question, pct.2: fraction detected in the rest of the cells
avg_logFC: expression comparison of the cluster in question (> 0 means higher) vs rest of the cells

And the equivalent but faster way using `presto::wilcoxauc`

```{r "using presto instead"}
library(presto)
markers <- wilcoxauc(sobj_clusters, "clusters") # compute markers
markers %>% head()
```

### find DE genes for specific cell groups
For more control in the comparisons, use `FindMarkers`. Positive average log (natural) fold change represents higher expression of the gene in cells of `ident.1`. pct1 and 2 are percent detected in each ident/population.

```{r "specific DE testing between groups"}
# DE analysis for 2 clusters
markers_df2 <- FindMarkers(sobj_clusters,
                           assay = "RNA",
                           slot = "data",
                           ident.1 = "1",
                           ident.2 = "2",
                           test.use = "t")
markers_df2 %>% head()
```

pct.1: fraction detected in the cells of ident.1, pct.2: fraction detected in the cells of ident.2
avg_logFC: expression comparison of ident.1 vs ident.2 (> 0 means higher in ident.1)

`FindMarkers` defaults to the current active ident. To use other value groups, set idents to the intended column, or use the `group.by` argument.

```{r "specific DE testing between groups, more complicated example"}
# compare control vs treated in only cluster 1
markers_df3 <- FindMarkers(sobj_clusters,
                           assay = "RNA",
                           slot = "data",
                           subset.ident = "1", # <- if needed, subset on current ident first, then switch idents
                           group.by = "orig.ident", # <- grouping cells by this metadata column
                           ident.1 = "control",
                           ident.2 = "treated")
```

Genes of interest can then be visualized as violin plots or feature plots.

```{r "plotting"}
# violin plots
Idents(sobj_clusters) <- sobj_clusters@meta.data$clusters # plots are grouped by active ident
VlnPlot(sobj_clusters, "LYZ")
VlnPlot(sobj_clusters, c("LYZ", "S100A4")) # can be a vector of gene names

# project on UMAP
FeaturePlot(sobj_clusters, "LYZ")
FeaturePlot(sobj_clusters, c("LYZ", "S100A4")) # can be a vector of gene names
FeaturePlot(sobj_clusters, "LYZ", split.by = "orig.ident") # <- split into panels based on metadata column
```
### Sidebar: How to regress out cell cycle / ribosomal gene heterogeneity (and do you need to?)

```{r "cell cycle"}
# 1. assess phase and PCA
# Seurat stores a list of cell cycle specific genes for humans
s.genes <- Seurat::cc.genes$s.genes # obviously unsuitable for nonhuman
g2m.genes <- Seurat::cc.genes$g2m.genes

s.genes
g2m.genes

# score and phase call is added to metadata
sobj_l <- CellCycleScoring(sobj_l, 
                           s.features = s.genes,
                           g2m.features = g2m.genes,
                           set.ident = FALSE)
sobj_l@meta.data %>% head()

# check PCA to see if cell cycle has strong effects
sobj_l <- FindVariableFeatures(sobj_l)
sobj_l <- RunPCA(sobj_l, verbose = FALSE, npcs = 10)
DimPlot(sobj_l, group.by = "Phase") # again, in this case, cell cycle does not have strong effects
```

```{r "PCA genes"}
# 2. look for cell cycle-specific genes as main drivers of PCA
topPCAgenes <- apply(
  sobj_l@reductions$pca@feature.loadings, # contribution of each gene to each PC is stored here
  MARGIN = 2, 
  FUN = function(x) names(sort(abs(x), decreasing = TRUE)[1:10]) # finds the most important genes in each PC
  )

topPCAgenes

topPCAgenes %>%
  as.vector() %>%
  intersect(c(s.genes, g2m.genes)) # see how many S and G2M genes intersect with that list

# or just check top variable genes
```

## cluster identities

Without venturing into the realm of philosphical debates on what a "cell type" constitutes, standard pratice is to use certain gene expression features to classify cells. This is often done manually, by visual inspection of key genes. Automated approaches that utilize a broader range of features are currently being developed.

### manual inspection of key genes using expertise in the lab

```{r "manual inspection"}
# from the Seurat pancreas example
# we now have pan_celseq2, fully annotated in metadata column "celltype", and pan_smartseq2, ready to be annotated
data_url = "https://scrnaseq-workshop.s3-us-west-2.amazonaws.com"
pan_celseq2 <- readRDS(url(file.path(data_url, "pan_celseq2.rds")))
pan_celseq2@meta.data %>% head()
pan_smartseq2 <- readRDS(url(file.path(data_url, "pan_smartseq2.rds")))
pan_smartseq2@meta.data %>% head()
a <- DimPlot(pan_smartseq2, label = TRUE) + NoLegend()
a

FeaturePlot(pan_smartseq2, c("IRX2","GC")) # marker genes for alpha
FeaturePlot(pan_smartseq2, c("IAPP","MAFA")) # marker genes for beta
FeaturePlot(pan_smartseq2, c("HHEX","LEPR")) # marker genes for delta

# optional, save the cluster numbers as "cluster.id"
pan_smartseq2 <- StashIdent(object = pan_smartseq2, save.name = "cluster.id") # or use AddMetaData

# Use RenameIdents to remap the idents from the current IDs to the new IDs 
pan_smartseq2 <- RenameIdents(pan_smartseq2,
                              "0" = "alpha",
                              "1" = "alpha",
                              "2" = "alpha",
                              "3" = "alpha",
                              "7" = "alpha",
                              "9" = "alpha",
                              "11" = "alpha", 
                              "5" = "beta",
                              "6" = "beta",
                              "8" = "beta",
                              "10" = "beta",
                              "4" = "delta")

# Plot UMAP with new cluster IDs
b <- DimPlot(object = pan_smartseq2,
        label = TRUE) + NoLegend()

cowplot::plot_grid(a,b)

# remember to save the new idents into meta.data
pan_smartseq2 <- StashIdent(object = pan_smartseq2, save.name = "cluster_name")
pan_smartseq2@meta.data %>% head()
```

### towards a more automated approach of identity assignment

1. using `Seurat`, inference from previous seurat object (requires very similar seurat object), see [Seurat Integration Vignette](https://satijalab.org/seurat/v3.2/integration.html) for more details.

```{r "Seurat integration"}
# reset idents of pan_smartseq2
Idents(pan_smartseq2) <- "cluster.id"

pancreas <- FindTransferAnchors(reference = pan_celseq2, query = pan_smartseq2, 
    dims = 1:30)
predictions <- TransferData(anchorset = pancreas, refdata = pan_celseq2$celltype, 
    dims = 1:30)
pan_smartseq2 <- AddMetaData(pan_smartseq2, metadata = predictions)
Idents(pan_smartseq2) <- "predicted.id"
DimPlot(pan_smartseq2, label = TRUE) + NoLegend()
```

2. using other classification packages such as `clustifyr`.

`clustifyr` works by comparing the average gene expression in each cluster to a reference matrix that contains average gene signatures of reference cell types. The reference can be built from other single cell data, bulk-rna-seq, or other sources. Ranked Spearman correlation is used to compare the reference to the clusters. Only the variable genes are used for the correlation. 

In order to compare our dataset we need to use a publically available reference dataset. In this example we use the other annotated experiment.Please see [clustifyrdatahub](https://github.com/rnabioco/clustifyrdatahub) for prebuilt general references available.

```{r "clustifyr annotation"}
library(clustifyr)
# build a reference from expression matrix data
ref <- average_clusters(pan_celseq2@assays$RNA@data,
                        metadata = pan_celseq2@meta.data,
                        cluster_col = "celltype")
dim(ref)
res <- clustify(input = pan_smartseq2@assays$RNA@data,
                ref_mat = ref,
                metadata = pan_smartseq2@meta.data,
                cluster_col = "cluster.id",
                query_genes = pan_smartseq2@assays$RNA@var.features, # using Seurat computed variable genes
                )
res # a matrix of correlation values between each cluster and reference cell types
res2 <- cor_to_call(res, 
                    metadata = pan_smartseq2@meta.data,
                    cluster_col = "cluster.id",
                    threshold = 0.5) # anything below 0.5 correlation are labeled as "unknown"
res2

# same as last section
new_meta <- call_to_metadata(res2, 
                             metadata = pan_smartseq2@meta.data,
                             cluster_col = "cluster.id")
pan_smartseq2@meta.data <- new_meta
DimPlot(pan_smartseq2, label = T, group.by = "type") + NoLegend()
```
`clustifyr` also takes seurat objects as input directly, finds various needed data, and output another object with identities assigned.

```{r "clustifyr with wrappers"}
ref2 <- seurat_ref(pan_celseq2,
                  cluster_col = "celltype")
identical(ref,ref2) # exactly the same as chunk above

res <- clustify(input = pan_smartseq2,
                ref_mat = ref2,
                cluster_col = "cluster.id",
                seurat_out = TRUE)

res # returns a new seurat object

# look at UMAP
DimPlot(res, label = TRUE, group.by = "type") + NoLegend() # saved in "type" metadata column
```

Also see more tutorials at [`clustifyr`](https://rnabioco.github.io/clustifyr/).

### cell type composition
Insight into different samples can be gained from the proportion of cells that fall into each cell type. Unfortunately, no dedicated tools are available for statistical testing.

```{r "evaluate composition"}
tab1 <- sobj_clusters@meta.data %>% group_by(orig.ident, clusters) %>% tally() # counting up all combinations
tab1
tab2 <- tab1 %>% spread(key = clusters, value = n) # spread out into "wide" form
tab2
tab3 <- tab1 %>% group_by(orig.ident) %>%
  mutate(n = n/sum(n)) %>% # convert counts to proportions first
  spread(key = clusters, value = n) 
tab3
```

and save tables to disk with `write_csv`

```{r "saving files",eval=F}
write_csv(tab3, "data/perc_clusters.csv")
tab4 <- read_csv("data/perc_clusters.csv")
```


## other things to do with marker genes 
1. gene list to pathway activity score, via [`AUCell`](https://github.com/aertslab/AUCell)

2. if TF expression is too low, consider [`SCENIC`](https://github.com/aertslab/SCENIC) for TF activity inference

3. standard GO term enrichment tools [`gProfiler2`](https://biit.cs.ut.ee/gprofiler/page/r), [`enrichR`](https://github.com/wjawaid/enrichR), [`fgsea`](https://github.com/ctlab/fgsea]), etc etc
