---
title: "Trajectory inference"
author: "Kent Riemondy, modified by Caitlin Winkler"
date: "`r Sys.Date()`"
output: 
  distill::distill_article:
    toc: true
    toc_depth: 1
    self_contained: true
    highlight: "tango"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo      = TRUE,
  cache     = TRUE,
  message   = FALSE,
  warning   = FALSE,
  comment   = "#>",
  fig.align = "center"
)
```

Rmarkdown: <a href="https://raw.githubusercontent.com/rnabioco/cellar/master/5_trajectories.Rmd" style="text-decoration: none;"> <i class="fa fa-file-code-o"></i> </a> 

## Trajectory inference, aka pseudotime

Trajectory inference assumes that cells are sampled during various stages of a transition from a cell type or state to another type or state. By identifying trajectories that connect cells based on similarilty in gene expression, one can gain insights into lineage relationships and developmental trajectories. To quantitate the measure of biological progress outside of defined time-points we use a metric called 'pseudotime', which is defined as a distance metric between the 'starting' cell and 'ending' cell along the trajectory. 

Analysis of discrete clusters can hide interesting continous behaviors in cell populations. By ordering cells based on an expression trajectory one can uncover novel patterns of gene expression. 

## There are many methods

```{r}
knitr::include_graphics("img/dynverse.webp")
```

Check out the [`dynverse`](https://dynverse.org/) for help with algorithm selection. They benchmarked > 60 methods, and offer some tools to run multiple algorithms on the same data. 

## Simplest method (PCA)

In some datasets, particularly developmental datasets it is often the case that a principal component may separate out cells based on known developmental time. This can be used a 'pseudotime' metric, which is simply the PC1 scores for each cell. As an example, scRNA-seq analyses of [liver development](http://dx.doi.org/10.1038/nature22796) and [Beta-cell maturation](http://dx.doi.org/10.1016/j.cmet.2017.04.014) used PC1 to examine pseudotime. 

```{r pkgs}
library(Seurat)
library(tidyverse)
library(Matrix)
library(slingshot)
library(tradeSeq)
library(SingleCellExperiment)
library(RColorBrewer)
library(gam)
library(cowplot)
library(scales)
library(gridExtra)
library(destiny)
library(ComplexHeatmap)
theme_set(theme_cowplot())
```

For this first analysis we will use a dataset of mouse early embryogenesis that was curated into a SingleCellExperiment object from here
https://hemberg-lab.github.io/scRNA.seq.datasets/. 

```{r}
# read in data
sce <- readRDS(url("https://scrnaseq-public-datasets.s3.amazonaws.com/scater-objects/deng-reads.rds"))

# convert SingleCellExperiment to Seurat
sobj <- CreateSeuratObject(counts(sce), meta.data = as.data.frame(colData(sce)))
sobj <- NormalizeData(sobj)
sobj <- FindVariableFeatures(sobj)
sobj <- ScaleData(sobj)
sobj <- RunPCA(sobj, verbose = FALSE)
Idents(sobj) <- "cell_type1"

# plot the PCA reduction
DimPlot(sobj, reduction = "pca")
```

Next we will plot out the PC1 scores for each cell type (developmental stage), using the meta data stored in the `cell_type` column of `meta.data`.

```{r}
# get the PC1 values for each cell, and store in a data frame
plt_dat <- FetchData(sobj, c("PC_1", "cell_type1"))

# reorder cell_type based on known developmental time
cell_type <- factor(plt_dat$cell_type1,
                    levels = c("zygote",
                               "2cell",
                               "4cell",
                               "8cell",
                               "16cell",
                               "blast"))

# add the reordered cell_type info as a new column to plt_data
plt_dat$cell_type <- cell_type

# plot the cells ordered by pseudotime (PC1)
ggplot(plt_dat, aes(cell_type, PC_1)) +
  geom_jitter(aes(color = cell_type)) +
  labs(y = "PC1 (aka pseudotime)") +
  ggtitle("Cells ordered by PC1")
```

```{r}
# Visualize the top genes associated with PC1 (pseudotime)
VizDimLoadings(sobj, dims = 1, balanced = T)

# Plot example genes
FeatureScatter(sobj, "PC_1", "Actb")
FeatureScatter(sobj, "PC_1", "Parp12")
```

We can overlay the pseudotime scores onto a projection by storing them as a `meta.data` column. 

```{r}
# Add UMAP dimensional reduction
sobj <- RunUMAP(sobj, dims = 1:15)

# get PC1 values and rank to generate a 'pseudotime'
ptime <- FetchData(sobj, "PC_1")
ptime$ptime <- rank(ptime$PC_1)

# add pseudotime to the metadata
sobj <- AddMetaData(sobj, 
                    ptime$ptime, 
                    col.name = "Pseudotime")

# color single cells on a UMAP plot according to pseudotime
p1 <- FeaturePlot(sobj, "Pseudotime")

# color single cells on a UMAP plot according to cell_type (i.e., timepoint)
p2 <- DimPlot(sobj, 
             group.by  = "cell_type1", 
             reduction = "umap") +
  ggtitle("Cell type")

plot_grid(p1, p2, align = "hv", ncol = 2, rel_widths = c(1, 1))
```

```{r}
sobj <- RunTSNE(sobj, dims = 1:15)

p1 <- DimPlot(sobj, 
             group.by  = "cell_type1", 
             reduction = "tsne") +
  ggtitle("tSNE")

p2 <- DimPlot(sobj, 
             group.by  = "cell_type1", 
             reduction = "umap") +
  ggtitle("UMAP")

plot_grid(p1, p2, align = "hv", ncol = 2, rel_widths = c(1, 1))
```


## Destiny

Non-linear dynamics are not always captured by PCA. 'Diffusion maps' offer an alternative, non-linear transformation approach that infers the temporal order of differentiating cells by modeling it as a diffusion process. This can better resolve complex trajectories and branching than linear methods like PCA. We'll use the R package `destiny` to look at diffusion maps and diffusion pseudotime.

```{r}
# Generating a diffusion map requires a matrix of expression data:
# prepare a counts matrix
logcounts <- GetAssayData(sobj, "data")

# transpose matrix (genes as columns, cells as rows)
input_matrix <- t(logcounts[VariableFeatures(sobj), ])

# generate a diffusion map
set.seed(42)
dm <- DiffusionMap(as.matrix(input_matrix))

# store the diffusion map as a custom dimensional reduction in the Seurat object
sobj[["DM"]] <- CreateDimReducObject(embeddings = dm@eigenvectors, key = "DM_", assay = DefaultAssay(sobj))

# plot the diffusion map
DimPlot(sobj, reduction = "DM", group.by = "cell_type1") + ggtitle("Diffusion Map")
```

Next we calculate diffusion pseudotime using `DPT`.  

```{r}
# calculate the diffusion pseudotime (DPT)
dpt <- DPT(dm)

# color single cells on diffusion map plot according to DPT
p1 <- plot(dpt, 1:2) + ggtitle("Diffusion Pseudotime (DPT)")

# create data.frame for easy plotting of DPT
tmp <- data.frame(DC1 = dm$DC1,
                  DC2 = dm$DC2,
                  timepoint = cell_type,
                  dpt = dpt$DPT1)

# color single cells on diffusion map plot according to timepoint
p2 <- ggplot(tmp, aes(x = DC1, y = DC2, colour = timepoint)) +
    geom_point() +
    theme_classic() +
    ggtitle("Timepoint")

plot_grid(p1, p2, align = "hv", ncol = 2, rel_widths = c(1, 1))
```

```{r}
# plot timepoint vs. DPT
ggplot(tmp, aes(timepoint, dpt, colour = timepoint)) +
  geom_point() + 
  geom_jitter() +
  ggtitle ("Timepoint vs. DPT")
```

PCA and diffusion maps can be good for simple trajectories or pathway activation systems e.g. T-cell activation, drug pertubation, single cell type responses. 

## Slingshot 

For more complex trajectories, simply using a single PC or a diffusion componenet may not be sufficient. Systems with bi or tri-furcating trajectories won't be well fit within a single dimension. 

For this next analysis we will use a dataset taken from a single cell RNA-seq study of [hepatocyte development](https://doi.org/10.1002/hep.29353). 

```{r}
sobj <- readRDS("obj/slingshot.rds")
```

```{r, echo = FALSE}
# color single cells by timepoint vs. cluster id in PCA vs. UMAP space

# Timepoint - PCA
p1 <- DimPlot(sobj, group.by = "timepoint", reduction = "pca") + 
  NoLegend() + 
  ggtitle("Timepoint - PCA")
p1 <- LabelClusters(plot = p1, id = "timepoint")

# Clusters - PCA
p2 <- DimPlot(sobj, group.by = "seurat_clusters", reduction = "pca") + 
  NoLegend() +
  ggtitle("Clusters - PCA")
p2 <- LabelClusters(plot = p2, id = "seurat_clusters")

# Timepoint - UMAP
p3 <- DimPlot(sobj, group.by = "timepoint", reduction = "umap") +
  NoLegend() +
  ggtitle("Timepoint - UMAP")
p3 <- LabelClusters(plot = p3, id = "timepoint")

# Clusters - UMAP
p4 <- DimPlot(sobj, group.by = "seurat_clusters", reduction = "umap") +
  NoLegend() +
  ggtitle("Clusters - UMAP")
p4 <- LabelClusters(plot = p4, id = "seurat_clusters")

# This is a busy figure!
plot_grid(p1, p2, p3, p4, align = "hv", nrow = 2, ncol = 2, rel_heights = c(1, 1), rel_widths = c(1, 1))
```

[Slingshot](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-4772-0) is a [Bioconductor package](https://bioconductor.org/packages/release/bioc/html/slingshot.html) that draws curved trajectories through a low dimensional embedding to infer developmental dynamics. It provides functionality for computing pseudotimes through multiple trajectories.

```{r, echo = FALSE}
knitr::include_graphics("img/slingshot.png")
```

```{r}
#convert to SingleCellExperiment
sce <- as.SingleCellExperiment(sobj)

#subset to only a few PCA dimensions
reducedDim(sce) <- reducedDim(sce)[, 1:10]

sce <- suppressWarnings(slingshot(
  sce,
  reducedDim = 'PCA',
  clusterLabels = 'seurat_clusters',
  start.clus = "4"
))

# extract info about pseudotimes from sce
sds <- SlingshotDataSet(sce)
sds
```

`slingshot` provides some minimal plotting utilities in base R plotting: 

```{r}
# Plot clusters with identified lineages overlayed
par(xpd=TRUE)
par(mar=c(4.5,5.5,2,7))
plot(reducedDims(sce)$PCA, col = brewer.pal(9,'Set1')[sce$seurat_clusters], pch=16)
lines(SlingshotDataSet(sce), lwd=2, col='black')
legend(x=25, y=20, legend=unique(colData(sce)$seurat_clusters), fill=brewer.pal(11,'Set1')[as.integer(unique(colData(sce)$seurat_clusters))])

# Plot pseudotime for a lineage (we'll plot lineage #3 in this example)
colors <- colorRampPalette(brewer.pal(11,'Spectral')[-6])(100)
plotcol <- colors[cut(sce$slingPseudotime_3, breaks=100)]
plot(reducedDims(sce)$PCA, col = plotcol, pch=16)
lines(SlingshotDataSet(sce), lwd=2, col='black')
```

### Plotting `slingshot` pseudotime onto UMAP

The lineage trees are generated in the same dimensionality as the input data. To visualize the pseudotime estimates in UMAP space we can add these to the meta.data slot of the seurat object. 

```{r}
# add slingshot pseudotimes to seurat object `meta.data`
sobj$pseudotime1 <- sce$slingPseudotime_1
sobj$pseudotime2 <- sce$slingPseudotime_2
sobj$pseudotime3 <- sce$slingPseudotime_3

FeaturePlot(sobj, c("pseudotime1", "pseudotime2", "pseudotime3"))
```

## Identifying differentially expressed genes along a trajectory

We can use a `GAM` (or other models) to compare pseudotimes to gene expression values to identify genes that vary significantly with pseudotime. Using a GAM can uncover non-linear expression patterns (i.e. not just up or down). See this [blog](https://multithreaded.stitchfix.com/blog/2015/07/30/gam/) for a nice description. 

```{r}
# select the ptime values 
ptime <- sce$slingPseudotime_2

# get cells in that lineage
lineage_cells <- colnames(sce)[!is.na(ptime)]

# remove values for cells not in the lineage
ptime <- ptime[!is.na(ptime)]

# just test variable genes to save some time
genes_to_test <- VariableFeatures(sobj)[1:1000]

# get log normalized data to test
cnts <- logcounts(sce)[genes_to_test, lineage_cells]

# fit a GAM with a loess term for pseudotime
gam.pval <- apply(cnts, 1, function(z){
    d <- data.frame(z = z, ptime = ptime)
    tmp <- suppressWarnings(gam(z ~ lo(ptime), data=d))
    p <- summary(tmp)[4][[1]][1, 5]
    p
})

# adjust pvalues 
res <- tibble(
  id = names(gam.pval),
  pvals = gam.pval,
  qval = p.adjust(gam.pval, method = "fdr")) %>% 
  arrange(qval)

head(res)
```

Finally make a heatmap to plot out these genes that vary over pseudotime.

```{r, fig.height = 6}

# get log normalized counts 
to_plot <- as.matrix(logcounts(sce)[res$id[1:100], lineage_cells])

# arrange cells by pseudotime
ptime_order <- colnames(to_plot)[order(ptime)]

# add useful annotations
annotations <- colData(sce)[lineage_cells, 
                            c("slingPseudotime_2", 
                            "seurat_clusters", 
                            "timepoint")] %>% 
  as.data.frame()

ha <- HeatmapAnnotation(df = annotations)

Heatmap(to_plot,
        column_order = ptime_order,
        show_column_names = FALSE,
        show_row_names = FALSE,
        top_annotation = ha)
```

## tradeSeq

`tradeSeq` is a package that allows analysis of gene expression along trajectories (such as those found with `Slingshot`). For every gene in the dataset, `tradeSeq` fits a GAM (as explained in the previous section). It then allows statistical inference on the GAM by assessing contrasts of the parameters of the fitted GAM, aiding in interpreting complex datasets. For more details, see the `tradeSeq` [paper](https://www.nature.com/articles/s41467-020-14766-3). The following is from the `tradeSeq` [tutorial](https://bioconductor.org/packages/release/bioc/vignettes/tradeSeq/inst/doc/tradeSeq.html).

Let's use `tradeSeq` to look at our `Slingshot` data set from the previous section:

```{r}
# tradeSeq needs a counts matrix and lineage curves as minimal input
crv <- sds
counts <- as.matrix(assays(sce)$counts)
clusters <- sce$seurat_clusters
```

The following two code blocks we will **NOT** run, as they both take several minutes. These two code blocks are fitting the `tradeSeq` negative binomial generalized additive model (NB-GAM). For an in-depth description of what's going on, see [this vignette](https://statomics.github.io/tradeSeq/articles/fitGAM.html). Basically, we are using the NB-GAM framework in `tradeSeq` to smooth each gene's expression in each lineage. Smoothers can be decomposed into a set of basis functions, which are joined together at knot points (simply called knots).

We run the `evaluateK` function in order to determine the correct number of knots. This returns the AIC (Akaike Informaction Criterion) for each gene fitted with each number of knots defined in the function (typically the interesting range is 3 to 10 knots) in a series of diagnostic plots.

```{r, eval=FALSE}
set.seed(5)
icMat <- evaluateK(counts = counts, sds = crv, k = 3:10, nGenes = 200, verbose = T)
```

```{r, echo = FALSE}
knitr::include_graphics("img/evaluateK.png")
```

The middles panels show that the drop in AIC levels off if the number of knots is increased beyond 6, so we'll choose that number of knots to fit the `tradeSeq` models using the `fitGAM` function. `nknots = 6` is also the default for the `fitGAM` function:

```{r, eval=FALSE}
set.seed(7)
sce <- fitGAM(counts = counts, sds = sds, nknots = 6, genes = 1:1000, verbose = TRUE, sce=TRUE)
```

```{r}
# Load in the SingleCellExperiment created by running `fitGAM` above
sce <- readRDS("obj/tradeSeq.rds")
```

```{r}
# plot our Slingshot lineage trajectories, this time illustrating the new tradeSeq knots
plotGeneCount(curve = sds, counts = counts,
              clusters = clusters,
              models = sce)
```

With `tradeSeq`, we can examine differential gene expression in a variety of ways. We can check if a gene is associated with a particular lineage using `associationTest`, discover progenitor marker genes using `startVsEndTest`, compare specific pseudotime values with a lineage, discover differentiated cell type markers using `diffEndTest`, discover genes with different expression patterns using `patternTest`, and identify early drivers of differentiation using `earlyDETest`.

### Discovering differentiated cell type markers

```{r}
# discover marker genes for the differentiated cell types
endRes <- diffEndTest(sce, pairwise = TRUE)
head(endRes)
```

We can plot the most significant gene using the `plotSmoothers` function.

```{r}
o <- order(endRes$waldStat, decreasing = TRUE)
sigGene <- names(sce)[o[1]]
plotSmoothers(sce, counts, sigGene)
```

Alternatively, we can color the cells in PC space with that geneâ€™s expression.

```{r}
plotGeneCount(crv, counts, gene = sigGene)
```

### Discovering early drivers of differentiation

We might want to find a list of genes that are differentially expressed between lineages at a particular region, e.g. around the separation of two or multiple lineages.

```{r}
earlyDERes <- earlyDETest(sce, pairwise = TRUE, knots = c(2, 4))
oEarly <- order(earlyDERes$waldStat, decreasing = TRUE)
head(rownames(earlyDERes)[oEarly])
```

```{r}
plotSmoothers(sce, counts, gene = rownames(earlyDERes)[oEarly][2])
```

```{r}
plotGeneCount(crv, counts, gene = rownames(earlyDERes)[oEarly][2])
```

## PAGA

[PAGA](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x) (partition-based graph abstraction) is a python module available in the python counterpart to seurat known as `scanpy`. `PAGA` provides both a very nice visualization technique, and a method for trajectory analysis. Whereas PC and DPT analyses are good for simple trajectories, and `Slingshot` is good for bi- and multifuracting trajectories, PAGA is good for more complex trajectories.

```{r, echo = FALSE}
knitr::include_graphics("img/paga.webp")
```

```{r, echo = FALSE}
knitr::include_graphics("img/paga_graph.png")
```

## How to deal with multiple timepoints

[URD](https://github.com/farrellja/URD)
[STITCH](https://github.com/wagnerde/STITCH)
[Waddington-OT](https://broadinstitute.github.io/wot/)

```{r, echo = FALSE}
knitr::include_graphics("img/urd.jpg")
```

## RNA velocity

Based on a cool concept that while the exonic transcriptome represents the current state of a cell, the intronic transcriptome represents what the cell is going to be in the near future. By introducing a transcriptional dynamic model, RNA velocity analysis predicts the directional 'flow' of cell state transition, which greatly expands the application of scRNA-seq to capture dynamics of molecular changes. There is a python and R package implementation for performing analysis, both of which requires processing a bam file to quantitate unspliced and spliced mRNAs per cell (which we can't do within the constraints of the workshop). Recently a scalable (python) toolkit for RNA velocity analysis in single cells called `scVelo` was published, which can do additional analysis such as estimating the directional velocity pseudotime.

http://velocyto.org/
https://github.com/velocyto-team/velocyto.R
https://scvelo.readthedocs.io/

![](https://user-images.githubusercontent.com/31883718/80227452-eb822480-864d-11ea-9399-56886c5e2785.gif)

```{r, echo = FALSE}
knitr::include_graphics("img/rna_velo.jpg")
```

## Recommendations:

1) Limit analysis to relevant cell populations (i.e. don't try to infer trajectories where you know they don't exist)

2) Start with simple methods and trajectories. 

3) Complement pseudotime trajectories with RNA velocity analyses to provide directionality and orthogonal validation. 

## Credits:

Much of this material was either copied or motivated by the `slingshot`, `destiny`, and `tradeSeq` vignettes. Some wording was also borrowed/copied from [this](https://github.com/quadbiolab/scRNAseq_analysis_vignette/blob/master/Tutorial.pdf) scRNA-seq analyis vignette (which I recommend).