---
title: "single cell RNA-seq workshop"
site: distill::distill_website
preview: img/rbi-twitter-card.png
---

```{r setup, include = FALSE, eval = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### **November 1st - 5th, 2021 **

5-day workshop covering methods used to analyze single cell RNA-seq data using R/RStudio. The workshop will be held remotely with lectures conducted over Zoom, with assistance available via a slack channel.

### **Prerequisites**

A basic understanding of the R programming language and a single cell RNA-seq dataset to analyze. Individuals without R experience will need to complete online courses suggested by instructors. The single cell RNA-seq dataset can be from a public resource, or private data from a lab on campus. The instructors will provide specific recommendations for preferred data formats, dataset size and complexity.

Attendees will be expected to devote time outside of the workshop lectures to review pre-class materials, and apply the concepts and techniques discussed in class to their chosen single cell dataset. 

### **Register**

Please fill out this [form](https://forms.gle/WBHJ2EQmR9Fttjdz6) to indicate your interest in the workshop.  Selected attendees will be notified by October 1. 

### **Schedule**

Pre-class videos (< 30 min) with lecture and background material will be provided prior to each day's workshop. 

* **Session 0** | November 1<sup>th</sup>, 9:00am - 11:00am
  + Course on-boarding session for those with questions on prerequisite material 
  
* **Session 1** | November 2<sup>th</sup>, 9:00am - 11:00am
  + Introduction to single cell RNA-seq
  + Processing scRNA-seq data: from FASTQ to a count matrix
  + Interacting with single cell data in R
  + Quality control
      
* **Session 2** | November 3<sup>th</sup>, 9:00am - 11:00am
  + Dimensionality reduction techniques for scRNA-seq (PCA, UMAP, tSNE, etc)
  + Clustering scRNA-seq data
  + Differential expression and marker gene discovery
  + Cell-type annotation

* **Session 3** | November 4<sup>th</sup>, 9:00am - 11:00am
  + Using pseudotime approaches to find trajectories or gene expression gradients   
  + Working with multi-sample datasets and mitigating batch effects

* **Session 4** | November 5<sup>th</sup>, 9:00am - 11:00am
  + Incorporating additional single cell modalities, e.g. CITE-seq
  + Additional time to discuss personal datasets
  
### **Assistance**

An instructor will be available after class for an hour to assist with additional questions. Additional office hours may be scheduled by contacting workshop instuctors, or scheduled through the [RBI office hours](https://medschool.cuanschutz.edu/rbi/training-and-education/rbi-office-hours). 

A slack workspace will be set up for the workshop to allow attendees to seek assistance from other attendees and instructors. 

### **Instructors**

Please contact the RBI fellows (rbi.fellows@cuanschutz.edu) if you have questions. 

- Ryan Sheridan  
- Caitlin Winkler  
- Tyler Matheny  
- Kristen Wells-Wrasman
- Rui Fu  
- Kent Riemondy  


