---
title: <b style="font-size:45px;">Working with Multi-modal Data</b>
author: "Ryan Sheridan"
date: "August 15<sup>th</sup>, 2019"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 1
    self_contained: true
    highlight: "tango"
---

<br>

<br>

<br>

#**Creating a Seurat object with multiple assays**

## Loading the counts matrices
```{r "Load packages", include = F}

library(Seurat)
library(tidyverse)
library(cowplot)

knitr::opts_chunk$set(
  message = F, 
  warning = F,
  comment = ""
)

```

```{r "Load data"}

# Data URL
data_url = "https://scrnaseq-workshop.s3-us-west-2.amazonaws.com"

# Import RNA matrix
rna_mtx <- file.path(data_url, "CITEseq_cDNA.csv.gz") %>%
  read_csv() %>%
  column_to_rownames("X1") %>%
  as.sparse()

rna_mtx[1:10, 1:10]

# Import ADT matrix
adt_mtx <- file.path(data_url, "CITEseq_ADT.csv.gz") %>%
  read_csv() %>%
  column_to_rownames("X1") %>%
  as.sparse()

rownames(adt_mtx) <- str_c("adt-", rownames(adt_mtx))

adt_mtx[, 1:10]

# Import HTO matrix
hto_mtx <- file.path(data_url, "CITEseq_HTO.csv.gz") %>%
  read_csv() %>%
  column_to_rownames("X1") %>%
  as.sparse()

hto_mtx[, 1:10]

```

## Creating a Seurat object
```{r "Create Seurat object"}

# Get list of common cell barcodes
rna_bcs <- colnames(rna_mtx)
adt_bcs <- colnames(adt_mtx)
hto_bcs <- colnames(hto_mtx)

merged_bcs <- rna_bcs %>%
  intersect(adt_bcs) %>%
  intersect(hto_bcs)

# Create Seurat object
sobj <- rna_mtx %>%
  .[, rna_bcs %in% merged_bcs] %>%
  CreateSeuratObject(min.cells = 5)

# Add antibody and cell hashing data to Seurat object
sobj[["ADT"]] <- adt_mtx %>%
  .[, adt_bcs %in% merged_bcs] %>%
  CreateAssayObject()

sobj[["HTO"]] <- hto_mtx %>%
  .[, hto_bcs %in% merged_bcs] %>%
  CreateAssayObject()

sobj

```

```{r "Normalize HTO counts"}

# Normalize HTO counts
sobj <- sobj %>%
  NormalizeData(
    assay                = "HTO",
    normalization.method = "CLR"      # Centered log-ratio normalization
  )

```

<br>

<br>

#**Demultiplexing hashed samples**

## Sample demultiplexing
```{r "Demultiplex samples", fig.width = 8, fig.height = 6}

# Demultiplex samples
# By default HTODemux will look for the "HTO" assay
sobj <- sobj %>%
  HTODemux()

sobj@meta.data %>%
  head()

# Create ridge plots showing HTO signal
sobj %>%
  RidgePlot(
    assay    = "HTO",
    features = rownames(hto_mtx),
    ncol     = 2
  )

# Rename cell identities with sample names
sobj <- sobj %>%
  RenameIdents(
    "HTO28" = "PBMC-1",
    "HTO29" = "PBMC-2",
    "HTO30" = "PBMC-3",
    "HTO44" = "PBMC-4"
  )

# Add sample names to meta.data table
sobj <- sobj %>%
  AddMetaData(
    metadata = Idents(sobj),
    col.name = "sample"
  )

```

## EXERCISE: Compare the number of cells with each hash.ID
```{r}

# Use the meta.data table to compare the number of cells with each hash.ID

# sobj@meta.data %>%



```

## ANSWER
```{r echo = F, fig.width = 7, fig.height = 4}

# Create bar graphs comparing cell count for each sample
HTO_bars_1 <- sobj@meta.data %>%
  rownames_to_column("cell_id") %>%
  ggplot(aes(hash.ID, fill = hash.ID)) +
  geom_bar() +
  labs(y = "Cell Count") +
  theme(
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.text.x     = element_text(hjust = 1, angle = 45)
  )

# Create stacked bar graph showing fraction of cells
HTO_bars_2 <- sobj@meta.data %>%
  rownames_to_column("cell_id") %>%
  group_by(hash.ID) %>%
  summarize(hash_frac = n() / nrow(.)) %>%
  ggplot(aes("CITE-seq", hash_frac, fill = hash.ID)) +
  geom_bar(stat = "identity", color = "white", size = 0.5) +
  labs(y = "Fraction of Cells") +
  theme(
    legend.title = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x  = element_blank(),
    axis.ticks.x = element_blank()
  )

# Combine plots
plot_grid(
  HTO_bars_1, HTO_bars_2,
  nrow        = 1,
  rel_widths  = c(0.5, 0.5),
  rel_heights = c(1, 0.8)
)

```

<br>

<br>

#**Filtering data and assessing quality**

## Assessing data quality
```{r "Create RNA violin plots", fig.width = 10}

# Add mitochondrial percentage to meta.data table
sobj <- sobj %>%
  PercentageFeatureSet(
    assay    = "RNA",
    pattern  = "^MT-", 
    col.name = "percent_mito"
  )

# Create violin plots for gene expression data
sobj %>%
  VlnPlot(
    features = c("nCount_RNA", "nFeature_RNA", "percent_mito"), 
    ncol     = 3,
    pt.size  = 0.25
  )

```

```{r "Create ADT and HTO violin plots", fig.height = 8}

sobj %>%
  VlnPlot(
    features = c(
      "nCount_ADT", "nFeature_ADT",
      "nCount_HTO", "nFeature_HTO"
    ),
    ncol    = 2,
    pt.size = 0.25,
    log     = T
  )

```

## EXERCISE: How many cells are left after filtering?
```{r}

# Filter cells based on number of genes detected, percent mito UMIs, and HTO_classification.global

# filt_sobj <- sobj %>%
#   subset(
#     
#     
#     
#     
#   )

```

## ANSWER
```{r echo = F}

# Filter cells based on HTO class, number of genes, and percent mito UMIs
filt_sobj <- sobj %>%
  subset(
    nFeature_RNA > 250 &   # Remove cells with < 250 detected genes
    nFeature_RNA < 2500 &  # Remove cells with > 2500 detected genes (could be doublets)
    percent_mito < 15 &    # Remove cells with >0.15 mito/total reads
    HTO_classification.global == "Singlet"
  )

filt_sobj

```

<br>

<br>

#**Normalizing gene expression and antibody data**

```{r "Normalize data", results = F}

# Normalize expression data
filt_sobj <- filt_sobj %>%
  SCTransform(
    vars.to.regress     = "percent_mito",  
    variable.features.n = 2000,
    verbose             = F
  )

# Normalize and scale antibody data
filt_sobj <- filt_sobj %>%
  NormalizeData(
    assay                = "ADT",
    normalization.method = "CLR"   # Centered log-ratio normalization
  ) %>%
  ScaleData(assay = "ADT")

```

<br>

<br>

#**Clustering cells based on gene expression**

## Perform pricipal component analysis
```{r "Perform PCA", fig.width = 10, fig.height = 4}

# Perform PCA
filt_sobj <- filt_sobj %>%
  RunPCA(verbose = F)       # By default only variable features are used for PCA

# Plot standard deviations of principal components
elbow_plot <- filt_sobj %>%
  ElbowPlot(ndims = 50)

# Create scatter plot comparing PC-1 and PC-2
PCA_plot <- filt_sobj %>%
  DimPlot(reduction = "pca")

plot_grid(
  elbow_plot, PCA_plot,
  nrow       = 1,
  rel_widths = c(0.4, 0.6)
)

```

## Cluster cells using gene expression data
```{r "Cluster cells and run UMAP"}

# Cluster cells
filt_sobj <- filt_sobj %>%
  FindNeighbors(dims = 1:40) %>%
  FindClusters(
    resolution = 0.6,
    verbose    = F
  )

# Run UMAP
# Use same PCs as clustering, by default reduction = "pca"
filt_sobj <- filt_sobj %>%
  RunUMAP(dims = 1:40)

# Find marker genes for each cluster
RNA_markers <- filt_sobj %>%
  FindAllMarkers(only.pos = T) %>%
  filter(p_val_adj < 0.05)

# Add RNA clusters to meta.data
filt_sobj <- filt_sobj %>%
  AddMetaData(
    metadata = Idents(filt_sobj),
    col.name = "RNA_clusters"
  )

# Create UMAP
filt_sobj %>%
  DimPlot(reduction = "umap")

```

<br>

<br>

#**Clustering cells based on antibody signal**

## Cluster cells using antibody signal
```{r "Cluster data using antibody data", results = F}

# Since there are only a few antibodies, instead of doing PCA just use a standard
# euclidean distance matrix
adt_data <- filt_sobj@assays$ADT@data
adt_dist <- dist(t(adt_data))

# Find clusters using distance matrix
filt_sobj[["adt_snn"]] <- FindNeighbors(adt_dist)$snn

filt_sobj <- filt_sobj %>%
  FindClusters(
    assay      = "ADT",
    resolution = 0.2,
    graph.name = "adt_snn"
  )

# Add ADT clusters to meta.data
filt_sobj <- filt_sobj %>%
  AddMetaData(
    metadata = Idents(filt_sobj),
    col.name = "ADT_clusters"
  )

```

## Create UMAP using antibody signal
```{r "Run UMAP for antibody data"}

# Run UMAP
filt_sobj <- filt_sobj %>%
  RunUMAP(
    reduction.name  = "adt_umap",
    reduction.key   = "ADTUMAP_",
    graph           = "adt_snn"
  )

# Create UMAP
filt_sobj %>%
  DimPlot(reduction = "adt_umap")

```

## EXERCISE: Which cluster has CD19 as its top marker?
```{r}

# Identify differentially expressed proteins for each cluster

# filt_sobj %>%
#   FindAllMarkers(
# 
#   )

```

## ANSWER
```{r echo = F}

# Identify differentially expressed proteins for each cluster
ADT_markers <- filt_sobj %>%
  FindAllMarkers(
    assay    = "ADT",
    only.pos = T
  ) %>%
  filter(p_val_adj < 0.05)

ADT_markers

```

<br>

<br>

#**Visualizing antibody signal**

## Overlay antibody signal on UMAPs
```{r "Overlay antibody signal on UMAPs", fig.width = 10, fig.height = 8}

# Set active.assay to ADT
filt_sobj@active.assay <- "ADT"

# Overlay antibody signal on gene expression UMAP
filt_sobj %>%
  FeaturePlot(
    reduction  = "umap",
    features   = c("adt-CD3", "adt-CD4", "adt-CD8", "adt-CD19"),
    cols       = c("#3288bd", "#bd0026")
  )

# Overlay antibody signal on antibody UMAP
filt_sobj %>%
  FeaturePlot(
    reduction  = "adt_umap",
    features   = c("adt-CD3", "adt-CD4", "adt-CD8", "adt-CD19"),
    cols       = c("#3288bd", "#bd0026")
  )

```

## Create ridge plots
```{r "Create ridge plots", fig.width = 10, fig.height = 6}

# Set ADT as the active assay
filt_sobj@active.assay <- "ADT"

# Create ridge plot
filt_sobj %>%
  RidgePlot(features = c(
    "adt-CD14", "adt-CD45",
    "adt-CD19", "adt-CD3",  
    "adt-CD4",  "adt-CD8"
  ))

```

## Create heatmaps
```{r "Create heatmaps"}

filt_sobj %>%
  DoHeatmap(
    features = rownames(filt_sobj), 
    angle    = 0
  ) + 
  NoLegend()

```

<br>

<br>

#**Classifying cells based on antibody signal**

## Identify CD19+ cells
```{r "Filter using antibody signal"}

# Set ADT as the active assay
filt_sobj@active.assay <- "ADT"

# Plot CD3 and CD19 signal
filt_sobj %>%
  FeatureScatter("adt-CD3", "adt-CD19")

# Identify CD19+ cells using antibody signal
CD19_cells <- filt_sobj %>%
  subset(`adt-CD19` > 2.5 & `adt-CD3` < 1) %>%
  Cells()

# Set cell identities
labeled_sobj <- filt_sobj %>%
  SetIdent(value = "Other") %>%
  SetIdent(
    value = "CD19+", 
    cells = CD19_cells
  )

labeled_sobj@active.ident %>%
  head()

# Label UMAP with new cell identities
labeled_sobj %>%
  DimPlot(reduction = "adt_umap")

```

## Filter cells using CellSelector()
```{r "Filter using CellSelector()", eval = F}

# Plot CD3 and CD19 signal
CD19_plot <- filt_sobj %>%
  FeatureScatter("adt-CD3", "adt-CD19")

CD19_plot

# Identify CD19+ cells using antibody signal
labeled_sobj <- filt_sobj %>%
  SetIdent(value = "Other")

labeled_sobj <- CellSelector(
  plot   = CD19_plot,
  object = labeled_sobj,
  ident  = "B cells"
)

labeled_sobj@active.ident %>%
  head()

# Label UMAP with new cell identities
labeled_sobj %>%
  DimPlot(reduction = "adt_umap")

```

## EXERCISE: Identify CD4+ and CD8+ cells
```{r}

# Plot CD4 and CD8 signal

# filt_sobj %>%
#   FeatureScatter("adt_CD4", "adt_CD8")

# Identify CD4+ and CD8+ cells using antibody signal

# Set cell identities

# Label UMAP with new cell identities



```

## ANSWER
```{r echo = F}

# Compare CD3 and CD19 signal
filt_sobj %>%
  FeatureScatter("adt-CD4", "adt-CD8")

# Identify CD4+ and CD8+ cells using antibody signal
CD4_cells <- labeled_sobj %>%
  subset(`adt-CD4` > 2.5 & `adt-CD8` < 1) %>%
  Cells()

CD8_cells <- labeled_sobj %>%
  subset(`adt-CD8` > 1.5 & `adt-CD4` < 1) %>%
  Cells()

# Set cell identities
labeled_sobj <- labeled_sobj %>%
  SetIdent(
    value = "CD4+", 
    cells = CD4_cells
  ) %>%
  SetIdent(
    value = "CD8+", 
    cells = CD8_cells
  )

# Add cell identities to meta.data table
labeled_sobj <- labeled_sobj %>%
  AddMetaData(
    metadata = Idents(labeled_sobj),
    col.name = "cell_label"
  )

# Label UMAP with new cell identities
labeled_sobj %>%
  DimPlot(reduction = "adt_umap")

```

<br>

<br>

#**Viewing results with the Cell Browser**

## Merge gene expression and antibody matrices
```{r "Merge RNA and ADT matrices"}

# Combine RNA and ADT matrices
merged_sobj <- labeled_sobj
RNA_data    <- merged_sobj@assays$SCT@data
ADT_data    <- merged_sobj@assays$ADT@data
merged_data <- rbind(RNA_data, ADT_data)

# Add merged matrix to Seurat object
merged_sobj@assays$SCT@data <- merged_data

# Set active assay
merged_sobj@active.assay <- "SCT"

```

## Create Cell Browser files
```{r "Create gene expression Cell Browser files", eval = F}

# Create Cell Browser directories
dir.create(
  path = "cellbrowser/SCT",
  recursive = T
)

# Export marker genes
RNA_markers %>%
  rename(score = p_val_adj) %>%
  select(cluster, gene, score) %>%
  write_tsv("cellbrowser/SCT/RNA_markers.tsv")

# Create Cell Browser files for gene expression data
merged_sobj %>%
  ExportToCellbrowser(
    dir          = "cellbrowser/SCT",
    reductions   = "umap",
    dataset.name = "SCT",
    markers.file = "cellbrowser/SCT/RNA_markers.tsv",
    
    sample       = "Sample",
    RNA_clusters = "Cluster",
    ADT_clusters = "ADT cluster",
    cell_label   = "Cell label",
    
    nCount_RNA   = "RNA UMI count",
    nFeature_RNA = "Gene count",
    percent_mito = "Percent mito UMIs",
    nCount_ADT   = "ADT UMI count",
    nFeature_ADT = "Antibody count",
    nCount_HTO   = "HTO UMI count",
    nFeature_HTO = "HTO count"
  )

```

```{r "Create ADT Cell Browser files", eval = F}

# Create Cell Browser directories
dir.create(
  path = "cellbrowser/ADT",
  recursive = T
)

# Export marker genes
ADT_markers %>%
  rename(score = p_val_adj) %>%
  select(cluster, gene, score) %>%
  write_tsv("cellbrowser/ADT/ADT_markers.tsv")

# Create Cell Browser files for antibody data
merged_sobj %>%
  ExportToCellbrowser(
    dir          = "cellbrowser/ADT",
    reductions   = "adt_umap",
    dataset.name = "ADT",
    markers.file = "cellbrowser/ADT/ADT_markers.tsv",
    
    sample       = "Sample",
    ADT_clusters = "Cluster",
    RNA_clusters = "RNA cluster",
    cell_label   = "Cell label",
    
    nCount_RNA   = "RNA UMI count",
    nFeature_RNA = "Gene count",
    percent_mito = "Percent mito UMIs",
    nCount_ADT   = "ADT UMI count",
    nFeature_ADT = "Antibody count",
    nCount_HTO   = "HTO UMI count",
    nFeature_HTO = "HTO count"
  )

```

## Build Cell Browser session
```{bash "Build Cell Browser session", eval = F}

mkdir -p cellbrowser/html

cbBuild \
  -i cellbrowser/SCT/cellbrowser.conf \
  -i cellbrowser/ADT/cellbrowser.conf \
  -o cellbrowser/html

```


