<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

  <!--radix_placeholder_meta_tags-->
  <title>Single-cell RNA-seq Workshop: &lt;b style="font-size:45px;"&gt;Working with Multi-modal Data&lt;/b&gt;</title>
  
  
  <link rel="icon" type="image/png" href="img/logo.png"/>
  
  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2020-11-13"/>
  <meta property="article:created" itemprop="dateCreated" content="2020-11-13"/>
  <meta name="article:author" content="Ryan Sheridan, Kent Riemondy"/>
  
  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="Single-cell RNA-seq Workshop: &lt;b style=&quot;font-size:45px;&quot;&gt;Working with Multi-modal Data&lt;/b&gt;"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="Single-cell RNA-seq Workshop"/>
  
  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="Single-cell RNA-seq Workshop: &lt;b style=&quot;font-size:45px;&quot;&gt;Working with Multi-modal Data&lt;/b&gt;"/>
  <meta property="twitter:site" content="@rnabioco"/>
  
  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","author","date","output"]}},"value":[{"type":"character","attributes":{},"value":["<b style=\"font-size:45px;\">Working with Multi-modal Data<\/b>"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name"]}},"value":[{"type":"character","attributes":{},"value":["Ryan Sheridan, Kent Riemondy"]}]}]},{"type":"character","attributes":{},"value":["November 13<sup>th<\/sup>, 2020"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["toc"]}},"value":[{"type":"logical","attributes":{},"value":[false]}]}]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  
  <script type="application/javascript">
  
    window.headroom_prevent_pin = false;
  
    window.document.addEventListener("DOMContentLoaded", function (event) {
  
      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');
  
      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });
  
      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });
  
      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>
  
  <style type="text/css">
  
  /* Theme (user-documented overrideables for nav appearance) */
  
  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #455a64;
    font-size: 15px;
    font-weight: 300;
  }
  
  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }
  
  .distill-site-nav a:hover {
    color: white;
  }
  
  @media print {
    .distill-site-nav {
      display: none;
    }
  }
  
  .distill-site-header {
  
  }
  
  .distill-site-footer {
  
  }
  
  
  /* Site Header */
  
  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }
  
  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }
  
  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }
  
  
  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }
  
  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }
  
  .distill-site-header .title {
    font-size: 18px;
  }
  
  .distill-site-header .logo {
    padding: 0;
  }
  
  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }
  
  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }
  
  
  
  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }
  
  
  .distill-site-header .nav-toggle {
    display: none;
  }
  
  .nav-dropdown {
    display: inline-block;
    position: relative;
  }
  
  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }
  
  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }
  
  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .nav-dropdown-active {
    display: block;
  }
  
  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }
  
  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }
  
  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }
  
  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }
  
  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }
  
  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }
  
  
  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative;}
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }
  
  /* Site Footer */
  
  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }
  
  /* Headroom */
  
  d-title {
    padding-top: 6rem;
  }
  
  @media print {
    d-title {
      padding-top: 4rem;
    }
  }
  
  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }
  
  .headroom--transition {
    transition: all .4s ease-in-out;
  }
  
  .headroom--unpinned {
    top: -100px;
  }
  
  .headroom--pinned {
    top: 0;
  }
  
  </style>
  
  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->
  
  <style type="text/css">
  
  body {
    background-color: white;
  }
  
  .pandoc-table {
    width: 100%;
  }
  
  .pandoc-table>caption {
    margin-bottom: 10px;
  }
  
  .pandoc-table th:not([align]) {
    text-align: left;
  }
  
  .pagedtable-footer {
    font-size: 15px;
  }
  
  .html-widget {
    margin-bottom: 2.0em;
  }
  
  .l-screen-inset {
    padding-right: 16px;
  }
  
  .l-screen .caption {
    margin-left: 10px;
  }
  
  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .shaded .shaded-content {
    background: white;
  }
  
  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }
  
  .hidden {
    display: none !important;
  }
  
  d-article {
    padding-bottom: 30px;
  }
  
  d-appendix {
    padding-top: 30px;
  }
  
  d-article>p>img {
    width: 100%;
  }
  
  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }
  
  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }
  
  /* CSS for table of contents */
  
  .d-toc {
    color: rgba(0,0,0,0.8);
    font-size: 0.8em;
    line-height: 1em;
  }
  
  .d-toc-header {
    font-size: 0.6rem;
    font-weight: 400;
    color: rgba(0, 0, 0, 0.5);
    text-transform: uppercase;
    margin-top: 0;
    margin-bottom: 1.3em;
  }
  
  .d-toc a {
    border-bottom: none;
  }
  
  .d-toc ul {
    padding-left: 0;
  }
  
  .d-toc li>ul {
    padding-top: 0.8em;
    padding-left: 16px;
    margin-bottom: 0.6em;
  }
  
  .d-toc ul,
  .d-toc li {
    list-style-type: none;
  }
  
  .d-toc li {
    margin-bottom: 0.9em;
  }
  
  .d-toc-separator {
    margin-top: 20px;
    margin-bottom: 2em;
  }
  
  .d-article-with-toc {
    border-top: none;
    padding-top: 0;
  }
  
  
  
  /* Tweak code blocks (note that this CSS is repeated above in an injection
     into the d-code shadow dom) */
  
  d-code {
    overflow-x: auto !important;
  }
  
  pre.d-code code.d-code {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }
  
  pre.text-output {
  
    font-size: 12px;
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;
  
    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;
  
    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }
  
  @media(min-width: 768px) {
  
  d-code {
    overflow-x: visible !important;
  }
  
  pre.d-code code.d-code  {
      padding-left: 18px;
      font-size: 14px;
  }
  pre.text-output {
    font-size: 14px;
  }
  }
  
  /* Figure */
  
  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }
  
  .figure img {
    width: 100%;
  }
  
  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }
  
  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }
  
  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }
  
  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }
  
  
  
  /* Tweak 1000px media break to show more text */
  
  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }
  
    .grid {
      grid-column-gap: 16px;
    }
  
    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }
  
  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }
  
    .grid {
      grid-column-gap: 32px;
    }
  }
  
  
  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */
  
  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }
  
  
  /* Social footer */
  
  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }
  
  .disqus-comments {
    margin-right: 30px;
  }
  
  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }
  
  #disqus_thread {
    margin-top: 30px;
  }
  
  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }
  
  .article-sharing a:hover {
    border-bottom: none;
  }
  
  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }
  
  .subscribe p {
    margin-bottom: 0.5em;
  }
  
  
  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }
  
  
  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }
  
  .custom p {
    margin-bottom: 0.5em;
  }
  
  
  /* Improve display for browsers without grid (IE/Edge <= 15) */
  
  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }
  
  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }
  
  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }
  
  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }
  
  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }
  
  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }
  
  
  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }
  
  .downlevel .footnotes ol {
    padding-left: 13px;
  }
  
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }
  
  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }
  
  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }
  
  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;
  
    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;
  
    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }
  
  </style>
  
  <script type="application/javascript">
  
  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }
  
  // show body when load is complete
  function on_load_complete() {
  
    // set body to visible
    document.body.style.visibility = 'visible';
  
    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }
  
    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }
  
  function init_distill() {
  
    init_common();
  
    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);
  
    // create d-title
    $('.d-title').changeElementType('d-title');
  
    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);
  
    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();
  
    // move posts container into article
    $('.posts-container').appendTo($('d-article'));
  
    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');
  
    // create d-bibliography
    var bibliography = $('<d-bibliography></d-bibliography>');
    $('#distill-bibliography').wrap(bibliography);
  
    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;
  
    // replace citations with <d-cite>
    $('.citation').each(function(i, val) {
      appendix = true;
      var cites = $(this).attr('data-cites').split(" ");
      var dt_cite = $('<d-cite></d-cite>');
      dt_cite.attr('key', cites.join());
      $(this).replaceWith(dt_cite);
    });
    // remove refs
    $('#refs').remove();
  
    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();
  
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-toc a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });
  
    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');
  
    // replace code blocks with d-code
    $('pre>code').each(function(i, val) {
      var code = $(this);
      var pre = code.parent();
      var clz = "";
      var language = pre.attr('class');
      if (language) {
        // map unknown languages to "clike" (without this they just dissapear)
        if ($.inArray(language, ["bash", "clike", "css", "go", "html",
                                 "javascript", "js", "julia", "lua", "markdown",
                                 "markup", "mathml", "python", "svg", "xml"]) == -1)
          language = "clike";
        language = ' language="' + language + '"';
        var dt_code = $('<d-code block' + language + clz + '></d-code>');
        dt_code.text(code.text());
        pre.replaceWith(dt_code);
      } else {
        code.addClass('text-output').unwrap().changeElementType('pre');
      }
    });
  
    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {
  
      // capture layout
      var layout = $(this).attr('data-layout');
  
      // apply layout to markdown level block elements
      var elements = $(this).children().not('d-code, pre.text-output, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });
  
  
      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });
  
    // load distill framework
    load_distill_framework();
  
    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {
  
      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;
  
      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');
  
      // table of contents
      if (have_authors) // adjust border if we are in authors
        $('.d-toc').parent().addClass('d-article-with-toc');
  
      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');
  
      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }
  
      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');
  
      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");
  
      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }
  
       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }
  
      // inject pre code styles (can't do this with a global stylesheet b/c a shadow root is used)
      $('d-code').each(function(i, val) {
        var style = document.createElement('style');
        style.innerHTML = 'pre code { padding-left: 10px; font-size: 12px; border-left: 2px solid rgba(0,0,0,0.1); } ' +
                          '@media(min-width: 768px) { pre code { padding-left: 18px; font-size: 14px; } }';
        if (this.shadowRoot)
          this.shadowRoot.appendChild(style);
      });
  
      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();
  
      // clear polling timer
      clearInterval(tid);
  
      // show body now that everything is ready
      on_load_complete();
    }
  
    var tid = setInterval(distill_post_process, 50);
    distill_post_process();
  
  }
  
  function init_downlevel() {
  
    init_common();
  
     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));
  
    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
  
    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();
  
    // remove toc
    $('.d-toc-header').remove();
    $('.d-toc').remove();
    $('.d-toc-separator').remove();
  
    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });
  
  
    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);
  
    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);
  
    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();
  
    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();
  
    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));
  
    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });
  
    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));
  
    $('body').addClass('downlevel');
  
    on_load_complete();
  }
  
  
  function init_common() {
  
    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};
  
        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });
  
        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);
  
    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});
  
    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });
  
      }
    });
  
    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });
  
    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }
  
    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');
  
    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");
  
    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();
  
    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }
  
  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });
  
  </script>
  
  <!--/radix_placeholder_distill-->
  <script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <style type="text/css">
  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #000000;
      font-size: 15px;
    font-weight: 300;
  }
  
  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }
  
  .distill-site-nav a:hover {
    color: white;
    text-decoration:underline;
  }
  
  .distill-site-header {
  }
  
  .distill-site-footer {
  }
  
  @media print {
    .distill-site-nav {
      display: none;
    }
  }
  
  .l-body {
    display: none;
  }
  
  @media screen and (min-width: 768px) {
    .l-body {display: inline;}
  }
  
  </style>
  <!--/radix_placeholder_site_in_header-->

  <link rel="stylesheet" href="styles.css" type="text/css"/>

</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"<b style=\"font-size:45px;\">Working with Multi-modal Data<\/b>","authors":[{"author":"Ryan Sheridan, Kent Riemondy","authorURL":"#","affiliation":"&nbsp;","affiliationURL":"#"}],"publishedDate":"2020-11-13T00:00:00.000-07:00","citationText":"Riemondy, 2020"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a href="index.html" class="title">Single-cell RNA-seq Workshop</a>
<a href="prerequisite.html">Prerequisite-Material</a>
<a href="tba.html">Introduction</a>
<a href="2_filtering_QC.html">Data Filtering and Quality Control</a>
<a href="tba.html">Clustering and Visualization</a>
<a href="4_markers.html">Cluster Markers and Cell Type Assignment</a>
<a href="tba.html">Trajectories</a>
<a href="tba.html">Alignment</a>
<a href="7_multimodal.html">Multi-modal</a>
<a href="tba.html">VDJ</a>
<a href="previous.html">Previous-workshops</a>
<a href="r_install.html">Installing R</a>
</div>
<div class="nav-right">
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Working with Multi-modal Data</h1>

</div>

<div class="d-byline">
  
  Ryan Sheridan, Kent Riemondy
  
<br/>November 13<sup>th</sup>, 2020
</div>

<div class="d-article">
<p>Recorded lecture: <a href="https://www.dropbox.com/s/j24aordt33ew6tg/zoom_0.mp4?dl=0" > <i class="fa fa-video-camera"></i> </a></p>
<p>Rmarkdown: <a href="https://raw.githubusercontent.com/rnabioco/cellar/master/7_multimodal.Rmd" style="text-decoration: none;"> <i class="fa fa-file-code-o"></i> </a></p>
<h2 id="experimental-design">Experimental design</h2>
<h3 id="cite-seq-enables-detection-of-cell-surface-proteins-and-gene-expression">CITE-seq enables detection of cell surface proteins AND gene expression</h3>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-1"></span>
<img src="img/cite-seq.png" alt="taken from https://cite-seq.com/" width="700" />
<p class="caption">
Figure 1: taken from <a href="https://cite-seq.com/" class="uri">https://cite-seq.com/</a>
</p>
</div>
</div>
<h3 id="cite-seq-reagents">CITE-seq reagents</h3>
<p>Biolegend is the main company selling CITE-seq and cell hashing antibodies (<a href="https://www.biolegend.com/en-us/totalseq">TotalSeq</a>). Biolegend reagents are divided into three product lines:</p>
<ul>
<li>TotalSeq-A: 3’ gene expression, v2 and v3 chemistry</li>
<li>TotalSeq-B: 3’ gene expression, v3 chemistry</li>
<li>TotalSeq-C: 5’ gene expression and V(D)J</li>
</ul>
<p><img src="img/Totalseq_compatibility_figure_V1.png" /></p>
<h3 id="cell-hashing-reagents">Cell hashing reagents</h3>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1603-1">Cell hashing</a> allows for sample multiplexing and “super-loaded” runs with &gt;10,000 captured cells. Super-loading results in higher doublet rates (~10% for 10,000 captured cells), but these doublets can be removed by identifying cell barcodes that are associated with multiple hashtag oligos.</p>
<p><img src="img/cell_hashing_diagram.png" /></p>
<p>Biolegend cell hashing reagents for human cells include a mix of two antibodies that recognize CD298 and β2 microglobulin. Mouse hashing antibodies recognize CD45 and H-2 MHC class I.</p>
<p>TotalSeq-A reagents use a different PCR handle for CITE-seq and cell hashing antibodies. This means two separate libraries have to be prepared. To ensure that the correct libraries are created, it is import to tell the sequencing core which types of antibodies were included in the experiment.</p>
<p>TotalSeq-C reagents use the same PCR handle for CITE-seq and cell hashing antibodies, which means that only a single library will be prepared. However, to ensure that the correct libraries are created the core should be notified of all reagents used for the experiment.</p>
<p><a href="https://www.nature.com/articles/s41592-019-0433-8">MULTI-seq</a> uses lipid- and cholesterol-modified oligonucleotides.</p>
<h2 id="creating-a-seurat-object-with-multiple-assays">Creating a Seurat object with multiple assays</h2>
<h3 id="loading-counts-matrices">Loading counts matrices</h3>
<p>The <code>Read10X</code> function can be used with the output directory generated by Cell Ranger. However, public datasets can be formatted in many different ways, so it is very useful to become familar with converting between formats. Our UMI count data is stored as comma-separated files, which we can load as data.frames and then convert to sparse matrices.</p>
<p>Our data today will be a collection of 4 PBMC samples that have each been “hashed” with different cell-hashtags, stained with 8 different CITE-seq antibodies, then combined and captured on the 10x platform in 1 sample.</p>
<p>We have three files that we will work with: CITEseq_cDNA.csv.gz: UMI count gene expression data CITEseq_ADT.csv.gz: Antibody count data CITEseq_HTO.csv.gz: Hashtag count data</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Data URL
data_url &lt;- &quot;https://scrnaseq-workshop.s3-us-west-2.amazonaws.com&quot;

# Function to import counts
import_counts &lt;- function(file_name, file_url = data_url) {
  mtx &lt;- file.path(file_url, file_name) %&gt;%
    read_csv() %&gt;%
    column_to_rownames(&quot;X1&quot;) %&gt;%
    as.sparse()

  mtx
}

# Import gene expression matrix
rna_mtx &lt;- import_counts(&quot;CITEseq_cDNA.csv.gz&quot;)

# Import CITE-seq matrix
adt_mtx &lt;- import_counts(&quot;CITEseq_ADT.csv.gz&quot;)

rownames(adt_mtx) &lt;- str_c(&quot;adt-&quot;, rownames(adt_mtx))
adt_mtx[, 1:10]</code></pre>
<pre><code>
#&gt; 8 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                               
#&gt; adt-CD14   11   2  14   1   .   1   1  . .   1
#&gt; adt-CD19    9   . 165 295   .   3   6  4 2   1
#&gt; adt-CD3     .  19  79   2   3   6   9  3 3   4
#&gt; adt-CD4    69 360 293   7 159 321   1  7 7   4
#&gt; adt-CD45    .   .   3   4   .   1   .  . 1   .
#&gt; adt-CD45RA 25   3 246  16   8   5  18  6 4   9
#&gt; adt-CD45RO  6   4 108   6   .   2   1  2 1   3
#&gt; adt-CD8     7   .  36   6   .   . 473 21 7 197</code></pre>
<pre class="r"><code>
# Import HTO matrix
hto_mtx &lt;- import_counts(&quot;CITEseq_HTO.csv.gz&quot;)

hto_mtx[, 1:10]</code></pre>
<pre><code>
#&gt; 4 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                             
#&gt; HTO28 351   .   .  .   2 161   .   .   .   .
#&gt; HTO29   6   2   3  1   1   2   2 107   2   .
#&gt; HTO30   . 131 177 60   .   .   .   . 239 155
#&gt; HTO44   1   .   .  . 122   . 172   .   1   .</code></pre>
</div>
<h3 id="creating-a-seurat-object">Creating a Seurat object</h3>
<p>When adding multiple assays to a Seurat object, we first must identify cell barcodes that are present in all of the datasets. If one of the assays has a different number of cell barcodes Seurat will throw an error.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Get list of common cell barcodes
rna_bcs &lt;- colnames(rna_mtx)
adt_bcs &lt;- colnames(adt_mtx)
hto_bcs &lt;- colnames(hto_mtx)

merged_bcs &lt;-  intersect(rna_bcs, adt_bcs) %&gt;%
  intersect(hto_bcs)

# Create Seurat object
sobj &lt;-  CreateSeuratObject(rna_mtx[, merged_bcs], 
                            min.cells = 5)

# Add CITE-seq and cell hashing data to Seurat object
sobj[[&quot;ADT&quot;]] &lt;-  CreateAssayObject(adt_mtx[, merged_bcs])

sobj[[&quot;HTO&quot;]] &lt;- CreateAssayObject(hto_mtx[, merged_bcs])
  
sobj</code></pre>
<pre><code>
#&gt; An object of class Seurat 
#&gt; 15032 features across 4292 samples within 3 assays 
#&gt; Active assay: RNA (15020 features, 0 variable features)
#&gt;  2 other assays present: ADT, HTO</code></pre>
</div>
<h2 id="demultiplexing-hashed-samples">Demultiplexing hashed samples</h2>
<p>We will first use the cell hashing data to assign cells to their sample of origin. This is referred to as demuliplexing as we are using the cell hashing antibody barcodes to assign each cell to a sample.</p>
<h3 id="normalizing-hto-counts">Normalizing HTO counts</h3>
<p>To account for differences in antibody binding efficiencies, CITE-seq and cell hashing data can be normalized by performing a centered log-ratio transformation for each individual antibody.</p>
<p>Note that the best approach for normalizing CITE-seq data has not been settled. For a more detailed discussion, and alternative approaches, see the Normalization chapter from the <a href="https://bioconductor.org/books/release/OSCA/integrating-with-protein-abundance.html#normalization-1">Orchestrating Single Cell Analysis eBook</a> from Bioconductor.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Normalize HTO counts
sobj &lt;- sobj %&gt;%
  NormalizeData(
    assay = &quot;HTO&quot;,
    normalization.method = &quot;CLR&quot;
  )</code></pre>
</div>
<h3 id="sample-demultiplexing-and-identification-of-doublets">Sample demultiplexing and identification of doublets</h3>
<p>To demultiplex hashed samples, the <code>HTODemux</code> function uses the normalized HTO counts for k-medoids clustering. This results in a cluster for each HTO. A background signal is then calculated for each HTO using cells that are not present in the HTO-specific cluster. Outlier cells from this background signal are then classified as being “positive” for the HTO if they are above a cutoff quantile value (~99% of background).</p>
<p>Cells that are positive for multiple HTOs are classified as doublets and cells that are not positive for any HTO are classified as “negative” cells.</p>
<p>The <code>HTODemux</code> function automatically adds several columns to the meta.data table.</p>
<ul>
<li>HTO_classification: shows positive HTOs that were identified for the cell</li>
<li>HTO_classification.global: singlet classification (singlet, doublet, negative)</li>
<li>hash.ID: final HTO assignment including doublet and negative classifications</li>
</ul>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Demultiplex samples
# By default HTODemux will look for the &quot;HTO&quot; assay
sobj &lt;-  HTODemux(sobj, positive.quantile = 0.97)

head(sobj@meta.data)</code></pre>
<pre><code>
#&gt;                     orig.ident nCount_RNA nFeature_RNA nCount_ADT
#&gt; AAACCTGAGACAAAGG SeuratProject       3054         1217        388
#&gt; AAACCTGAGCTACCGC SeuratProject       2411          989        337
#&gt; AAACCTGCACCAGGTC SeuratProject       4182         1394        509
#&gt; AAACCTGCACTCTGTC SeuratProject       3681         1456         43
#&gt; AAACCTGGTATGAAAC SeuratProject       2503         1255        530
#&gt; AAACCTGGTGAGTATA SeuratProject       1241          692        352
#&gt;                  nFeature_ADT nCount_HTO nFeature_HTO HTO_maxID
#&gt; AAACCTGAGACAAAGG            5        358            3     HTO28
#&gt; AAACCTGAGCTACCGC            8        180            2     HTO30
#&gt; AAACCTGCACCAGGTC            7        163            2     HTO28
#&gt; AAACCTGCACTCTGTC            6        174            2     HTO44
#&gt; AAACCTGGTATGAAAC            6        638            2     HTO28
#&gt; AAACCTGGTGAGTATA            6         71            2     HTO44
#&gt;                  HTO_secondID HTO_margin HTO_classification
#&gt; AAACCTGAGACAAAGG        HTO29   3.211027              HTO28
#&gt; AAACCTGAGCTACCGC        HTO29   3.448779              HTO30
#&gt; AAACCTGCACCAGGTC        HTO29   2.802755              HTO28
#&gt; AAACCTGCACTCTGTC        HTO29   3.148890              HTO44
#&gt; AAACCTGGTATGAAAC        HTO29   3.649778              HTO28
#&gt; AAACCTGGTGAGTATA        HTO29   2.285112              HTO44
#&gt;                  HTO_classification.global hash.ID
#&gt; AAACCTGAGACAAAGG                   Singlet   HTO28
#&gt; AAACCTGAGCTACCGC                   Singlet   HTO30
#&gt; AAACCTGCACCAGGTC                   Singlet   HTO28
#&gt; AAACCTGCACTCTGTC                   Singlet   HTO44
#&gt; AAACCTGGTATGAAAC                   Singlet   HTO28
#&gt; AAACCTGGTGAGTATA                   Singlet   HTO44</code></pre>
<pre class="r"><code>
# Summarize cell classifications
table(sobj$HTO_classification.global)</code></pre>
<pre><code>
#&gt; 
#&gt;  Doublet Negative  Singlet 
#&gt;      386        1     3905</code></pre>
<pre class="r"><code>
# Create ridge plots showing HTO signal
RidgePlot(sobj,
          assay    = &quot;HTO&quot;,
          features = rownames(hto_mtx),
          ncol     = 2
)</code></pre>
<p><img src="7_multimodal_files/figure-html5/Demultiplex%20samples-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<h3 id="compare-the-number-of-cells-with-each-hash.id-and-calculate-the-doublet-rate">Compare the number of cells with each hash.ID and calculate the doublet rate</h3>
<div class="layout-chunk" data-layout="l-body">
<pre><code>
#&gt; # A tibble: 6 x 2
#&gt;   hash.ID     fract
#&gt;   &lt;fct&gt;       &lt;dbl&gt;
#&gt; 1 Doublet  0.0899  
#&gt; 2 HTO28    0.251   
#&gt; 3 HTO30    0.199   
#&gt; 4 HTO44    0.230   
#&gt; 5 HTO29    0.229   
#&gt; 6 Negative 0.000233</code></pre>
<p><img src="7_multimodal_files/figure-html5/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<h2 id="filtering-data-and-assessing-quality">Filtering data and assessing quality</h2>
<p>Now that we have assigned each cell to the appropriate sample we can continue with processing the <code>RNA</code> data in the same manner as before.</p>
<h3 id="assessing-data-quality">Assessing data quality</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Add mitochondrial percentage to meta.data table
sobj &lt;- sobj %&gt;%
  PercentageFeatureSet(
    assay    = &quot;RNA&quot;,
    pattern  = &quot;^MT-&quot;, 
    col.name = &quot;percent_mito&quot;
  )

# Create violin plots for gene expression data
sobj %&gt;%
  VlnPlot(
    features = c(&quot;nCount_RNA&quot;, &quot;nFeature_RNA&quot;, &quot;percent_mito&quot;), 
    ncol     = 3,
    pt.size  = 0.25
  )</code></pre>
<p><img src="7_multimodal_files/figure-html5/Create%20RNA%20violin%20plots-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Aim to sequence CITE-seq libraries at 2k-5k reads/cell, cell hashing 1k-2k reads/cell

sobj %&gt;%
  VlnPlot(
    features = c(&quot;nCount_ADT&quot;, &quot;nCount_HTO&quot;),
    ncol    = 2,
    pt.size = 0.25,
    log     = T
  )</code></pre>
<p><img src="7_multimodal_files/figure-html5/Create%20ADT%20and%20HTO%20violin%20plots-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Filter cells based on HTO class, number of genes, and percent mito UMIs
filt_so &lt;-  subset(sobj,
    nFeature_RNA &gt; 250 &amp;   # Remove cells with &lt; 250 detected genes
    nFeature_RNA &lt; 2500 &amp;  # Remove cells with &gt; 2500 detected genes (could be doublets)
    percent_mito &lt; 15 &amp;    # Remove cells with &gt; 0.15 mito/total reads
    HTO_classification.global == &quot;Singlet&quot;
  )

filt_so</code></pre>
<pre><code>
#&gt; An object of class Seurat 
#&gt; 15032 features across 3686 samples within 3 assays 
#&gt; Active assay: RNA (15020 features, 0 variable features)
#&gt;  2 other assays present: ADT, HTO</code></pre>
<pre class="r"><code>
# Rename cell identities with sample names
filt_so &lt;- filt_so %&gt;%
  RenameIdents(
    &quot;HTO28&quot; = &quot;PBMC-1&quot;,
    &quot;HTO29&quot; = &quot;PBMC-2&quot;,
    &quot;HTO30&quot; = &quot;PBMC-3&quot;,
    &quot;HTO44&quot; = &quot;PBMC-4&quot;
  )

# Add sample names to meta.data table
filt_so &lt;- filt_so %&gt;%
  AddMetaData(
    metadata = Idents(filt_so),
    col.name = &quot;sample&quot;
  )

head(filt_so@meta.data)</code></pre>
<pre><code>
#&gt;                     orig.ident nCount_RNA nFeature_RNA nCount_ADT
#&gt; AAACCTGAGACAAAGG SeuratProject       3054         1217        388
#&gt; AAACCTGAGCTACCGC SeuratProject       2411          989        337
#&gt; AAACCTGCACCAGGTC SeuratProject       4182         1394        509
#&gt; AAACCTGCACTCTGTC SeuratProject       3681         1456         43
#&gt; AAACCTGGTATGAAAC SeuratProject       2503         1255        530
#&gt; AAACCTGGTGAGTATA SeuratProject       1241          692        352
#&gt;                  nFeature_ADT nCount_HTO nFeature_HTO HTO_maxID
#&gt; AAACCTGAGACAAAGG            5        358            3     HTO28
#&gt; AAACCTGAGCTACCGC            8        180            2     HTO30
#&gt; AAACCTGCACCAGGTC            7        163            2     HTO28
#&gt; AAACCTGCACTCTGTC            6        174            2     HTO44
#&gt; AAACCTGGTATGAAAC            6        638            2     HTO28
#&gt; AAACCTGGTGAGTATA            6         71            2     HTO44
#&gt;                  HTO_secondID HTO_margin HTO_classification
#&gt; AAACCTGAGACAAAGG        HTO29   3.211027              HTO28
#&gt; AAACCTGAGCTACCGC        HTO29   3.448779              HTO30
#&gt; AAACCTGCACCAGGTC        HTO29   2.802755              HTO28
#&gt; AAACCTGCACTCTGTC        HTO29   3.148890              HTO44
#&gt; AAACCTGGTATGAAAC        HTO29   3.649778              HTO28
#&gt; AAACCTGGTGAGTATA        HTO29   2.285112              HTO44
#&gt;                  HTO_classification.global hash.ID percent_mito
#&gt; AAACCTGAGACAAAGG                   Singlet   HTO28     7.105435
#&gt; AAACCTGAGCTACCGC                   Singlet   HTO30     9.622563
#&gt; AAACCTGCACCAGGTC                   Singlet   HTO28     7.747489
#&gt; AAACCTGCACTCTGTC                   Singlet   HTO44     6.166802
#&gt; AAACCTGGTATGAAAC                   Singlet   HTO28     7.311227
#&gt; AAACCTGGTGAGTATA                   Singlet   HTO44    12.006446
#&gt;                  sample
#&gt; AAACCTGAGACAAAGG PBMC-1
#&gt; AAACCTGAGCTACCGC PBMC-3
#&gt; AAACCTGCACCAGGTC PBMC-1
#&gt; AAACCTGCACTCTGTC PBMC-4
#&gt; AAACCTGGTATGAAAC PBMC-1
#&gt; AAACCTGGTGAGTATA PBMC-4</code></pre>
</div>
<h2 id="normalizing-gene-expression-and-antibody-data">Normalizing gene expression and antibody data</h2>
<p>Like the cell hashing data, CITE-seq counts can be normalized by performing a centered log-ratio transformation.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Normalize expression data
filt_so &lt;-  NormalizeData(filt_so,
                          normalization.method = &quot;LogNormalize&quot;,
                          verbose = FALSE) %&gt;%
  FindVariableFeatures(verbose = FALSE) %&gt;%
  ScaleData(verbose = FALSE)

# Normalize CITE-seq data
filt_so &lt;-  NormalizeData(filt_so,
                          assay = &quot;ADT&quot;,
                          normalization.method = &quot;CLR&quot;) %&gt;%  
  ScaleData(assay = &quot;ADT&quot;, verbose = FALSE)</code></pre>
</div>
<h2 id="clustering-cells-based-on-gene-expression">Clustering cells based on gene expression</h2>
<h3 id="perform-pricipal-component-analysis">Perform pricipal component analysis</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Perform PCA
filt_so &lt;-  RunPCA(filt_so, verbose = FALSE)       # By default only variable features are used for PCA, uses only the RNA data

# Plot standard deviations of principal components
elbow_plot &lt;- ElbowPlot(filt_so, ndims = 50)

# Create scatter plot comparing PC-1 and PC-2
PCA_plot &lt;- DimPlot(filt_so, reduction = &quot;pca&quot;)

plot_grid(elbow_plot, PCA_plot,
          nrow = 1,
          rel_widths = c(0.45, 0.55))</code></pre>
<p><img src="7_multimodal_files/figure-html5/Perform%20PCA-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<h3 id="cluster-cells-using-gene-expression-data">Cluster cells using gene expression data</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Cluster cells
filt_so &lt;- FindNeighbors(filt_so, dims = 1:20) %&gt;%
  FindClusters(resolution = 0.6, verbose = FALSE)

# Store clusters as RNA clusters in meta.data for clarity 
filt_so@meta.data$RNA_clusters &lt;- filt_so@meta.data$RNA_snn_res.0.6 

# Run UMAP
# Use same PCs as clustering, by default reduction = &quot;pca&quot;
filt_so &lt;- RunUMAP(filt_so, dims = 1:20)

# Create UMAP
DimPlot(filt_so, reduction = &quot;umap&quot;, group.by = &quot;RNA_clusters&quot;)</code></pre>
<p><img src="7_multimodal_files/figure-html5/Cluster%20cells%20and%20run%20UMAP-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<h2 id="clustering-cells-based-on-antibody-signal">Clustering cells based on antibody signal</h2>
<h3 id="cluster-cells-using-antibody-signal">Cluster cells using antibody signal</h3>
<p>Since there are only a few antibodies and not many dimensions, instead of performing PCA we can just use the scaled matrix for clustering directly. In contrast to the gene expression data, we have already done feature selection and dimensionality reduction by choosing (for very practical reasons) to only stain cells with a few antibodies that discriminate cell populations. However, if you stained cells with many (&gt;100s) antibodies, then you may want to select variable antibodies, and use PCA for clustering.</p>
<p>We will directly pass the <code>scaled.data</code> matrix from the ADT assay to the <code>FindNeighbors</code> function, which creates a Shared Nearest Neighbor (SNN) graph that is used for clustering.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
filt_so &lt;- FindNeighbors(filt_so,
                         assay = &quot;ADT&quot;, 
                         features = rownames(filt_so@assays$ADT),
                         graph.name = &quot;ADT_snn&quot;)

filt_so &lt;- FindClusters(filt_so,
                        resolution = 0.2,
                        graph.name = &quot;ADT_snn&quot;)

filt_so@meta.data$ADT_clusters &lt;- filt_so@meta.data$ADT_snn_res.0.2</code></pre>
</div>
<h3 id="run-umap-using-the-antibody-signal">Run UMAP using the antibody signal</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Run UMAP
filt_so &lt;- RunUMAP(filt_so,
                   assay = &quot;ADT&quot;,
                   features = rownames(filt_so@assays$ADT),
                   reduction.name = &quot;adt_umap&quot;,
                   reduction.key = &quot;ADTUMAP_&quot;)

# Plot UMAPs
DimPlot(filt_so,
    reduction = &quot;adt_umap&quot;, 
    group.by = c(&quot;sample&quot;, &quot;ADT_clusters&quot;),
    ncol = 2)</code></pre>
<p><img src="7_multimodal_files/figure-html5/Run%20UMAP%20for%20antibody%20data-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<h3 id="identify-marker-proteins">Identify marker proteins</h3>
<div class="layout-chunk" data-layout="l-body">
<pre><code>
#&gt;                     p_val avg_log2FC pct.1 pct.2     p_val_adj
#&gt; adt-CD8      0.000000e+00  1.9813885 0.996 0.940  0.000000e+00
#&gt; adt-CD4     1.634829e-213  2.4744195 0.972 0.912 1.307864e-212
#&gt; adt-CD14    6.625301e-139  2.2351278 0.871 0.346 5.300241e-138
#&gt; adt-CD45     1.715050e-94  1.1299631 0.915 0.613  1.372040e-93
#&gt; adt-CD45RO   9.890876e-90  1.6445220 0.972 0.865  7.912700e-89
#&gt; adt-CD45RA   4.323820e-07  0.4249255 0.987 0.982  3.459056e-06
#&gt; adt-CD19     2.318370e-79  4.9637748 1.000 0.916  1.854696e-78
#&gt; adt-CD45RA1  1.496662e-18  0.5880825 0.992 0.982  1.197330e-17
#&gt; adt-CD45RO1  7.607724e-17  0.7118029 0.984 0.871  6.086180e-16
#&gt;             cluster       gene
#&gt; adt-CD8           0    adt-CD8
#&gt; adt-CD4           1    adt-CD4
#&gt; adt-CD14          3   adt-CD14
#&gt; adt-CD45          3   adt-CD45
#&gt; adt-CD45RO        3 adt-CD45RO
#&gt; adt-CD45RA        3 adt-CD45RA
#&gt; adt-CD19          5   adt-CD19
#&gt; adt-CD45RA1       5 adt-CD45RA
#&gt; adt-CD45RO1       5 adt-CD45RO</code></pre>
</div>
<h2 id="visualizing-antibody-signal">Visualizing antibody signal</h2>
<h3 id="overlay-antibody-signal-on-umaps">Overlay antibody signal on UMAPs</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Set active.assay to ADT
filt_so@active.assay &lt;- &quot;ADT&quot;

# Overlay antibody signal on gene expression UMAP
filt_so %&gt;%
  FeaturePlot(
    reduction = &quot;umap&quot;,
    features  = c(&quot;adt-CD4&quot;, &quot;CD4&quot;, &quot;adt-CD8&quot;, &quot;CD8A&quot;)
  )</code></pre>
<p><img src="7_multimodal_files/figure-html5/Overlay%20antibody%20signal%20on%20UMAPs-1.png" width="960" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
# Overlay antibody signal on antibody UMAP
filt_so %&gt;%
  FeaturePlot(
    reduction = &quot;adt_umap&quot;,
    features  = c(&quot;adt-CD4&quot;, &quot;CD4&quot;, &quot;adt-CD8&quot;, &quot;CD8A&quot;)
  )</code></pre>
<p><img src="7_multimodal_files/figure-html5/Overlay%20antibody%20signal%20on%20UMAPs-2.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<h3 id="ridge-plots">Ridge plots</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Create ridge plot
filt_so %&gt;%
  RidgePlot(features = c(
    &quot;adt-CD14&quot;, &quot;adt-CD45&quot;,
    &quot;adt-CD19&quot;, &quot;adt-CD3&quot;,  
    &quot;adt-CD4&quot;,  &quot;adt-CD8&quot;
  ))</code></pre>
<p><img src="7_multimodal_files/figure-html5/Create%20ridge%20plots-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<h3 id="violin-plots">Violin plots</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
filt_so %&gt;%
  VlnPlot(
    features = c(&quot;adt-CD4&quot;, &quot;rna_CD4&quot;, &quot;adt-CD19&quot;, &quot;rna_CD19&quot;),
    ncol = 2
  )</code></pre>
<p><img src="7_multimodal_files/figure-html5/Create%20violinplots-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<h2 id="classifying-cells-based-on-antibody-signal">Classifying cells based on antibody signal</h2>
<p>We can use 2-d scatterplots to derive “gates” to assign cell types in the same manner as done for flow cytometry analysis. Generally the CITE-seq signal has lower dynamic range and signal-to-noise compared to flow cytometry, but the overall patterns seen with flow cytometry are also seen with CITE-seq.</p>
<h3 id="identify-cd19-cells">Identify CD19+ cells</h3>
<p>CD19 is a marker of B-cells, whereas CD3 is a marker of T-cells. We can compare their expression to classify B-cells.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Plot CD3 and CD19 signal
FeatureScatter(filt_so, &quot;adt-CD3&quot;, &quot;adt-CD19&quot;)</code></pre>
<p><img src="7_multimodal_files/figure-html5/Filter%20using%20antibody%20signal-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Identify CD19+ cells using antibody signal
CD19_cells &lt;- filt_so %&gt;%
  subset(`adt-CD19` &gt; 2.5 &amp; `adt-CD3` &lt; 1) %&gt;%
  Cells()

head(CD19_cells)</code></pre>
<pre><code>
#&gt; [1] &quot;AAACCTGAGCTACCGC&quot; &quot;AAAGCAATCTGTCCGT&quot; &quot;AACACGTGTCGCTTCT&quot;
#&gt; [4] &quot;AACCATGGTCTGGTCG&quot; &quot;AACGTTGTCAATCACG&quot; &quot;AACTTTCCAGCCTTGG&quot;</code></pre>
<pre class="r"><code>
# Set cell identities
labeled_so &lt;- filt_so %&gt;%
  SetIdent(value = &quot;Other&quot;) %&gt;%
  SetIdent(value = &quot;CD19+&quot;, cells = CD19_cells)

head(labeled_so@active.ident)</code></pre>
<pre><code>
#&gt; AAACCTGAGACAAAGG AAACCTGAGCTACCGC AAACCTGCACCAGGTC AAACCTGCACTCTGTC 
#&gt;            Other            CD19+            Other            Other 
#&gt; AAACCTGGTATGAAAC AAACCTGGTGAGTATA 
#&gt;            Other            Other 
#&gt; Levels: CD19+ Other</code></pre>
<pre class="r"><code>
# Label UMAP with new cell identities
DimPlot(labeled_so, reduction = &quot;adt_umap&quot;)</code></pre>
<p><img src="7_multimodal_files/figure-html5/unnamed-chunk-4-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<h3 id="filter-cells-using-cellselector">Filter cells using CellSelector()</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Plot CD3 and CD19 signal
CD19_plot &lt;- filt_so %&gt;%
  FeatureScatter(&quot;adt-CD3&quot;, &quot;adt-CD19&quot;)

CD19_plot

# Identify CD19+ cells using antibody signal
labeled_so &lt;- filt_so %&gt;%
  SetIdent(value = &quot;Other&quot;)

labeled_so &lt;- CellSelector(
  plot   = CD19_plot,
  object = labeled_so,
  ident  = &quot;B cells&quot;
)

labeled_so@active.ident %&gt;%
  head()

# Label UMAP with new cell identities
labeled_so %&gt;%
  DimPlot(reduction = &quot;adt_umap&quot;)</code></pre>
</div>
<h3 id="identify-cd4-and-cd8-cells">Identify CD4+ and CD8+ cells</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Compare CD3 and CD19 signal
FeatureScatter(filt_so, &quot;adt-CD4&quot;, &quot;adt-CD8&quot;)</code></pre>
<p><img src="7_multimodal_files/figure-html5/unnamed-chunk-5-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Identify CD4+ and CD8+ cells using antibody signal
CD4_cells &lt;- labeled_so %&gt;%
  subset(`adt-CD4` &gt; 2.5 &amp; `adt-CD8` &lt; 1) %&gt;%
  Cells()

CD8_cells &lt;- labeled_so %&gt;%
  subset(`adt-CD8` &gt; 1.5 &amp; `adt-CD4` &lt; 1) %&gt;%
  Cells()

# Set cell identities
labeled_so &lt;- labeled_so %&gt;%
  SetIdent(value = &quot;CD4+&quot;, cells = CD4_cells) %&gt;%
  SetIdent(value = &quot;CD8+&quot;, cells = CD8_cells)

# Add cell identities to meta.data table
labeled_so@meta.data$cell_label &lt;- Idents(labeled_so)

# Label UMAP with new cell identities
DimPlot(labeled_so, reduction = &quot;adt_umap&quot;)</code></pre>
<p><img src="7_multimodal_files/figure-html5/unnamed-chunk-6-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<h2 id="examine-cite-seq-derived-cell-labels-on-gene-expression-umap">Examine CITE-seq derived cell labels on gene expression umap</h2>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
DimPlot(labeled_so, 
        reduction = &quot;umap&quot;,
        group.by = &quot;cell_label&quot;)</code></pre>
<p><img src="7_multimodal_files/figure-html5/unnamed-chunk-7-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
filt_so %&gt;%
  FeaturePlot(
    reduction = &quot;umap&quot;,
    features  = c(&quot;adt-CD4&quot;, &quot;CD4&quot;,
                  &quot;adt-CD8&quot;, &quot;CD8A&quot;, 
                  &quot;adt-CD19&quot;, &quot;CD19&quot;)
  )</code></pre>
<p><img src="7_multimodal_files/figure-html5/unnamed-chunk-8-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<h2 id="how-to-combine-antibody-and-rna-data-for-clustering-and-umap">How to combine antibody and RNA data for clustering and UMAP?</h2>
<p>The authors of Seurat have implemented a new algorithm for combined data from different types of assays (e.g. antibody data and RNA data), described in a <a href="https://www.biorxiv.org/content/10.1101/2020.10.12.335331v1">preprint</a> and presented in a <a href="https://satijalab.org/seurat/v4.0/weighted_nearest_neighbor_analysis.html">vignette</a>.</p>
<p>The basic idea is that we want to cluster and visualize the data using <strong>both</strong> the RNA and antibody data. How to do this?</p>
<p>We could literally just merge the two datasets together then run them through Seurat. This will however end up downweighting the contribution of the antibody data because there are so few antibodies(~8) compared to RNA features (~10,000)</p>
<p>Seurat has implemented a “Weighted Nearest Neighbor” approach that will combine the nearest neighbor graphs from the RNA data with the antibody data. The algorithm will calculate relative weights for the RNA or the Protein data for each cell and use these new weights to constuct a shared graph. The related weights are calculated based on the relative information content of neighboring cells in each modalitly. So if the data from 1 modality provides more information, then it is weighted higher.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# need to use PCA on the antibody data
labeled_so &lt;- RunPCA(labeled_so, 
                     assay = &quot;ADT&quot;, 
                     features = rownames(labeled_so@assays$ADT),
                     reduction.name = &#39;apca&#39;)

labeled_so &lt;- FindMultiModalNeighbors(labeled_so, 
                                      reduction.list = list(&quot;pca&quot;,
                                                            &quot;apca&quot;), 
                                      dims.list = list(1:20,
                                                       1:6),
                                      modality.weight.name = &quot;RNA.weight&quot;)

labeled_so &lt;- RunUMAP(labeled_so, 
                      nn.name = &quot;weighted.nn&quot;, 
                      reduction.name = &quot;wnn.umap&quot;, 
                      reduction.key = &quot;wnnUMAP_&quot;)</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
rna_umap &lt;- DimPlot(labeled_so, reduction = &quot;umap&quot;)
adt_umap &lt;- DimPlot(labeled_so, reduction = &quot;adt_umap&quot;)
both_umap &lt;- DimPlot(labeled_so, reduction = &quot;wnn.umap&quot;)

plot_grid(rna_umap,
          adt_umap,
          both_umap,
          labels = c(&quot;RNA&quot;, &quot;ADT&quot;, &quot;Both&quot;),
          ncol = 3)</code></pre>
<p><img src="7_multimodal_files/figure-html5/unnamed-chunk-10-1.png" width="1440" style="display: block; margin: auto;" /></p>
</div>
<h2 id="viewing-results-with-the-ucsc-cell-browser">Viewing results with the UCSC Cell Browser</h2>
<p>The UCSC <a href="http://www.genome.ucsc.edu/singlecell.html">Cell Browser</a> allows you to easily explore and share single-cell data. With the Cell Browser you can:</p>
<ul>
<li>View t-SNE or UMAP projections</li>
<li>Color cells by metadata and gene expression</li>
<li>View cluster marker genes</li>
<li>Rename clusters and add custom annotations to selected sets of cells</li>
</ul>
<p>UCSC hosts a <a href="http://cells.ucsc.edu/">large collection of cellbrowsers</a> for various single cell datasets.</p>
<h3 id="merge-gene-expression-and-antibody-matrices">Merge gene expression and antibody matrices</h3>
<p>We won’t be generating these browsers in the class, however below is some example code of how to build browsers.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Combine RNA and ADT matrices
merged_so   &lt;- labeled_so
RNA_data    &lt;- merged_so@assays$RNA@data
ADT_data    &lt;- merged_so@assays$ADT@data
merged_data &lt;- rbind(RNA_data, ADT_data)

# Add merged matrix to Seurat object
merged_so@assays$RNA@data &lt;- merged_data

# Set active assay
merged_so@active.assay &lt;- &quot;RNA&quot;</code></pre>
</div>
<h3 id="create-cell-browser-files">Create Cell Browser files</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Create Cell Browser directories for gene expression data
dir.create(
  path = &quot;cellbrowser/RNA&quot;,
  recursive = T
)

# Create Cell Browser files for gene expression data
merged_so %&gt;%
  ExportToCellbrowser(
    dir          = &quot;cellbrowser/RNA&quot;,
    reductions   = &quot;umap&quot;,
    dataset.name = &quot;RNA&quot;,

    sample       = &quot;Sample&quot;,
    RNA_clusters = &quot;Cluster&quot;,
    ADT_clusters = &quot;ADT cluster&quot;,
    cell_label   = &quot;Cell label&quot;,
    
    nCount_RNA   = &quot;RNA UMI count&quot;,
    nFeature_RNA = &quot;Gene count&quot;,
    percent_mito = &quot;Percent mito UMIs&quot;,
    nCount_ADT   = &quot;ADT UMI count&quot;,
    nFeature_ADT = &quot;Antibody count&quot;,
    nCount_HTO   = &quot;HTO UMI count&quot;,
    nFeature_HTO = &quot;HTO count&quot;
  )</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# Create Cell Browser directories for antibody data
dir.create(
  path = &quot;cellbrowser/ADT&quot;,
  recursive = T
)

# Export marker genes
ADT_markers %&gt;%
  rename(score = p_val_adj) %&gt;%
  select(cluster, gene, score) %&gt;%
  write_tsv(&quot;cellbrowser/ADT/ADT_markers.tsv&quot;)

# Create Cell Browser files for antibody data
merged_so %&gt;%
  ExportToCellbrowser(
    dir          = &quot;cellbrowser/ADT&quot;,
    reductions   = &quot;adt_umap&quot;,
    dataset.name = &quot;ADT&quot;,
    markers.file = &quot;cellbrowser/ADT/ADT_markers.tsv&quot;,
    
    sample       = &quot;Sample&quot;,
    ADT_clusters = &quot;Cluster&quot;,
    RNA_clusters = &quot;RNA cluster&quot;,
    cell_label   = &quot;Cell label&quot;,
    
    nCount_RNA   = &quot;RNA UMI count&quot;,
    nFeature_RNA = &quot;Gene count&quot;,
    percent_mito = &quot;Percent mito UMIs&quot;,
    nCount_ADT   = &quot;ADT UMI count&quot;,
    nFeature_ADT = &quot;Antibody count&quot;,
    nCount_HTO   = &quot;HTO UMI count&quot;,
    nFeature_HTO = &quot;HTO count&quot;
  )</code></pre>
</div>
<h3 id="build-cell-browser-session">Build Cell Browser session</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="bash"><code>
mkdir -p cellbrowser/html

cbBuild \
  -i cellbrowser/RNA/cellbrowser.conf \
  -i cellbrowser/ADT/cellbrowser.conf \
  -o cellbrowser/html \
  -p 8888
</code></pre>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="updates-and-corrections">Corrections</h3>
<p>If you see mistakes or want to suggest changes, please <a href="https://github.com/rnabioco/cellar/issues/new">create an issue</a> on the source repository.</p>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
