---
title: <b style="font-size:45px;">Normalization, Dimensionality Reduction and clustering</b>
author: "Kent Riemondy"
date: "August 14<sup>th</sup>, 2019"
output: 
  html_document:
    toc: true
    toc_depth: 1
bibliography: bib/refs.bib
---

```{r pkgs, include=FALSE}

library(Seurat)
library(tidyverse)
library(cowplot)

knitr::opts_chunk$set(
  message   = FALSE,
  warning   = FALSE,
  comment   = "#>",
  fig.align = "center"
)

```

# Processing single cell RNA-seq data

- Normalization and Dimensionality reduction methods
- Discuss visualization strategies for exploring single cell seq data
- Discuss clustering methods commonly used for single cell data


## Data matrices generated during processing:

1) UMIs (`counts`)
2) log normalized counts (`data`)
3) Scaled counts (`scaled.data`) 
4) Dim reductions

etc.


## Reload previously generated dataset

```{r read_data}
so <- readRDS("data/filtered_sobj.rds")

so
```

## Normalization

Normalization of the UMI count data is necessary due to the large variability in total counts observed across cells. This variability is generally thought to be due to inefficiencies in library preparation and variable sequenicng depths between cells. Without normalization the largest source of variabilty in the data would be due to technical differences rather than the biological differences of interest.


### Global size factor

A commonly used approach for normalization of UMI data is a simple scaling normalization similar to the commonly used counts per million normalization methods. UMI counts per cell are scaled to the total # of UMIs per cell and multipled by a constant for ease to prevent small numbers (10e4). The data are then log transformed (wtih a pseudocount) to stabilize the variance. e.g.

$\log(10000 \frac{x_{i}}{\sum_{x_i}} + 1)$

This approach makes the assumption that cells all have the same amount of RNA and that sequencing depth is the reason for the variability. These assumptions aren't really valid, especially when comparing diverse cell types, however in practice this approach can still provide reasonable results and is the default normalization method used in Seurat. 


```{r}
so <- NormalizeData(so, 
                    normalization.method = "LogNormalize")

# data slot now contains log normalized values
GetAssayData(so, "data")[1:10, 1:10]

```


### Scran

Normalization methods used in bulk RNA-seq data (e.g. DESeq2 size-factors ) are not appropriate due to the large numbers of zeros in the data. However, by pooling similar cells together, one can obtain enough data to estimate a size factor. The Bioconductor package [`scran`](https://www.bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html#1_introduction) implements such a strategy  [@Lun2016-fe]. 

The basic approach is as follows:

1) clusters cells to group cells with similar gene expression
2) pool cells within these clusters and calculate size factors
3) repeat step 2 with varying sets of cells and pool sizes
4) derive cell-specific size factors using linear algebra


First we need to convert out seurat object to a Bioconductor single cell data structure, the [SingleCellExperiment](ref) class. This class if similar to other bioC data strucutes (e.g. `SummarizedExperiment`). Seurat provides a conversion funciton to convert to an SingleCellExperiment object (and other formats, such as `loom` and `CellDataSet`).


```{r}
library(scran)

# convert to SingleCellExperiment
sce <- as.SingleCellExperiment(so)

# get raw UMI counts
counts(sce)[40:50, 40:50]

# get log Normalized counts form Seurat
logcounts(sce)[40:50, 40:50]

#genes
rownames(sce)[1:5]

#cells
colnames(sce)[1:5]

# get cell metadata
colData(sce)[1:5, ]

# get gene metadata
rowData(sce)[1:5, ]
```

First we'll cluster the cells. We'll discuss clustering approach more in depth later in the tutorial. 

```{r}
# takes a few minutes
clusters <- quickCluster(sce,
                         use.ranks = FALSE, # suggested by the authors
                         min.size = 100) # require at least 100 cells per cluster

table(clusters)

# Calculate size factors per cell
sce <- computeSumFactors(sce, 
                         clusters = clusters, 
                         min.mean = 0.1) # ignore low abundance genes 
```


```{r}
summary(sizeFactors(sce))

# apply size factors to generate log normalized data
sce <- normalize(sce)

logcounts(sce)[40:50, 40:50]
```



Lastly we'll replace the `Seurat` normalized values with those from `scran` and store the sizeFactors in the meta.data

```{r}
so[["RNA"]] <- SetAssayData(so[["RNA"]],
                              slot = "data", 
                              new.data = logcounts(sce))

so$sizeFactors <- sizeFactors(sce)

```


Now we that have normalized values we can visualize expression across groups. 

```{r}
VlnPlot(so, "CD3E", slot = "counts")
VlnPlot(so, c("CD4", "CD8A", "CD19"), slot = "data")
```


## Identifying highly variable genes (i.e. feature selection)

We will next select important features to use for dimensionality reduction, clustering and tSNE/uMAP projection. We can in theory use all ~20K genes in the dataset for these steps, however this is often computationally expensive and unneccesary. Most genes in a dataset are going to be expressed at relatively similar values across different cells, or are going to be so lowly expressed that they contirbute more noise than signal to the analysis. For determining the relationships between cells in gene-expression space, we want to focus on gene expression differences that are driven primarily by biological variation rather than technical variation. 

Seurat provides a function to help identify these genes, `FindVariableGenes`. Ranking genes by their variance alone will bias towards selecting highly expressed genes. To help mitigate this Seurat uses a `vst` method to identify genes. Briefly, a curve is fit to model the mean and variance for each gene in log space. Variances are then standarized against this model, and ranked. The top 2000 genes with the highest standarized variances are then called as highly variable, by default. 


```{r, pkg_data, message = F, fig.height=5, fig.width=6}
so <- FindVariableFeatures(so,
                             selection.method = "vst")

VariableFeaturePlot(so) 

so@assays$RNA@var.features[1:10]
```

An alternative approach uses the Bioconductor package [`M3Drop`](https://www.bioconductor.org/packages/release/bioc/html/M3Drop.html) [ref] . This approach does not model the variance in the data, but rather uses the information about dropouts to identify genes that have high variance. Specifically there is a clear relationship between the dropout rate and the abundance of a gene. 

Exercise: Plot the relationship...


```{r}
# dropout rate per gene

dropout <- data.frame(
  dropout_rate = 1 - (rowSums(GetAssayData(so, "counts") > 0) / ncol(so)),
  log_abundance = log(rowMeans(GetAssayData(so, "counts")))
)

ggplot(dropout, aes(log_abundance, dropout_rate)) +
  geom_point() +
  theme_cowplot()

```

```{r}
library(M3Drop)

# input either sce or counts slot from Seurat object
counts <- NBumiConvertData(sce)

# fits to model
fit <- NBumiFitModel(counts)

# Identify genes with high dropout rate for a given abundance
drop_hvg <- NBumiFeatureSelectionCombinedDrop(fit, 
                                              ntop = 2000,
                                              suppress.plot=FALSE)

head(drop_hvg)
```

How similar are the two appraoches?

```{r}
shared_hvgs <- intersect(rownames(drop_hvg), VariableFeatures(so))
```


Do these approaches really extract meaningful genes?

```{r, fig.width=8, fig.height=8}
library(ComplexHeatmap)

# get log normalized values for top 50 variable genes
to_plot <- GetAssayData(so, slot = "data")[shared_hvgs[1:50], ]

Heatmap(as.matrix(to_plot),
        show_column_names = FALSE)

VariableFeatures(so) <- shared_hvgs
```


## Scaling

Next, we will scale the normalized expression values. The scaled values will only be used for dimensionality reduction and clustering, and not differential expression. The purpose of scaling is to scale the expression values for each gene onto a similar numeric scale. This avoid issues of have a gene that is highly expressed being given more weight in clustering simply because it has larger numbers. Scaling converts the normalized data into z-scores by default and the values are stored in the `scale.data` slot.

Single cell data sets often contain various types of 'uninteresting' variation. These can include technical noise, batch effects and confounding biological variation (such as cell cycle stage). We can use modeling to regress these signals out of the analysis using ` ScaleData`. By default a linear model is fit between the the supplied variables (i.e `nUMI`, `proportion.mito`, etc.) and gene expression value. The residuals from this model are then scaled instead of the normalized gene expression.

In general we recommend performing exploratory data analysis first without regressing any nuisance variables. Later, if you discover a nuisance variable (i.e. batch or cell cycle), then you can regress it out. 

```{r scale}

# scale all of the data, useful if you want to make heatmaps later
so <- ScaleData(object = so, 
                 features = rownames(so))

# for large datasets, just scale the variable genes:
#so <- ScaleData(object = so, 
#                 features = VariableFeatures(so))
```



```{r, fig.width=8, fig.height=8}

# get scaled values for top 50 variable genes
to_plot <- GetAssayData(so, slot = "scale.data")[VariableFeatures(so)[1:50], ]

Heatmap(as.matrix(to_plot),
        show_column_names = FALSE)
```


## Performing dimensionality reduction and clustering: PCA

Heatmaps are useful for visualizing small numbers of genes (10s or 100s), but quickly become impractical for larger single cell datasets due to high dimensionality (1000s of cells X 20K genes). 

Dimensionality reduction techinques are used in single cell data analysis generally for two purposes:

1) Provide visualizations in 2 dimensions that convey information about relationships between cells

2) Reduce to computational burden of working with 20K dimenions.

By identifying the highly variable genes we've already reduced the dimensions from 20K to ~1000. Next we will use Principal Componenets Analysis to identify a small subset of dimensions (10-100) that will be used for clustering and visualization. 



```{r, pca, message = F}
so <- RunPCA(so,
               features = VariableFeatures(so),
               verbose = FALSE, set.seed = NULL)


DimPlot(so, reduction = "pca")
```

The `RunPCA` function will calculate the top 50 principal components using a modified form of PCA (see `?irlba`) that doesn't need to compute all of the PCs. By default it will also print the genes most highly associated with each principal component, we will visualize these later. Often these genes will be recognizable as markers of different cell populations. 

**Note** 
Some functions in Seurat (`RunPCA`, `RunTSNE`, `RunUMAP`, `FindClusters`), have a non-deterministic component that will give different results (usually slightly) each run due to a randomization step in the algorithm. However, you can set an integer seed to make the output reproducible. This is useful so that you get the same clustering results if you rerun your analysis code. In many functions seurat has done this for you with the `seed.use` parameter, set to `42`. If you set this to NULL you will get a non-deterministic result. 


### PCA Plots {.tabset .tabsetfade}
  Shown in the tabs are 2 dimensional plots showing the prinicpal component scores of cells. 
  
```{r pc_plots, results ='asis'}
pcs <- list(
  c(1, 2),
  c(1, 3),
  c(1, 4),
  c(2, 3),
  c(2, 4),
  c(3, 4)
)


for(i in seq_along(pcs)){
  cat('\n#### ', 'PC', pcs[[i]][1], ' vs PC', pcs[[i]][2], '\n', sep = "")
  DimPlot(so, 
          dims = pcs[[i]], 
          reduction = "pca",
          cols = RColorBrewer::brewer.pal(10, "Paired"))
  cat('\n')
}
```


To see which genes are most strongly associated with the PCs you can use the `VizDimLoadings` function.


```{r, pca_loadings, message = F, fig.height=7, fig.width=6}
VizDimLoadings(object = so, dims = 1:3, balanced = TRUE, ncol = 3)
```


Finally, we can generate an 'elbow plot' of the standard deviation explained by each of the principle components. 

```{r, elbow_plot, message = F, fig.height=5, fig.width=6}
ElbowPlot(so, ndims = 40)
```

Note that most of the variation in the data is captured in a small number of PCs (< 20)

## Dimensionality reduction

### PCA

Early single cell studies used PCA to visualize cell clusters, and by plotting the data in scatterplots reduced the dimensionality of the data to 2D. This is a fine appraoch for simple datasets, but often there is much more information present in higher prinicapl components.

We can use PC1 and PC2 to begin to assess the structure of the dataset. 

```{R}
FeaturePlot(so, 
            c("nFeature_RNA", "nCount_RNA"), 
            reduction = "pca")

# Color by CD8, marker of CD8 t-cells
FeaturePlot(so, 
            "CD8A", 
            reduction = "pca")
```

Use wrapper functions for plotting to save some typing.

```{r}
plot_pca <- function(sobj, features, ...){
  FeaturePlot(sobj, 
            features,
            reduction = "pca", 
            ...)
}


# Color by top genes associated with Pc1 and Pc2
plot_pca(so, c("CST3", "NKG7")) 


# or use ggplot directly if need more custom viz
var_df <- FetchData(so, c("PC_1", "PC_2", "CD19", "CD8A")) %>% 
  gather(gene, expr, -PC_1, -PC_2)

ggplot(var_df, aes(PC_1, PC_2)) +
  geom_point(aes(color = expr)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(9, "Reds")) + 
  facet_grid(~gene) +
  theme_cowplot()
```

Note how other PCs separate different cell populations (PPBP is a marker of megakaryocytes).

```{r}
plot_pca(so, c("PPBP", "NKG7"), dims = c(2,3))

```




### t-SNE 

The first two dimensions of PCA are not sufficient to capture the variation in these data due to the complexity. Howver PCA is useful in identifying a new set of dimensions (10-50) that capture most of the interesting variation in the data. This is useful because now with this reduced dataset we can use newer visualization technique to examine relationships between cells in 2 dimenions. 

 t-distributed stochastic neighbor embedding (tSNE) is a popular techinique for visualizing single cell datsets including mass cytometry data. A useful interactive tool to help you understand tSNE and interpret tSNE plots can be found [here](https://distill.pub/2016/misread-tsne/). 
 

Seurat provides a wrapper (`RunTSNE`) around the `Rtsne` package that generates the tSNE projection. By default RunTSNE uses the PCA matrix as input for Rtsne. The tSNE data will be stored in the `so@reductions$tsne` slot. 

```{r}
# specify to use the first 15 prinicpal components
so <- RunTSNE(so, dims = 1:15)

# again make wrapper to plot tsnes easily
plot_tsne <- function(sobj, features = "", ...){
  FeaturePlot(sobj, 
            features,
            reduction = "tsne", 
            ...)
}

# use TSNEplot for factors, use FeaturePlot for everything else
TSNEPlot(so)

plot_tsne(so, "CD8A")
```

EXERCISE: What effect does including all genes and all PCs (at least 50 PCs) have on the tSNE projection? Do clusters seem separated similarly?

```{r echo = FALSE}
so <- RunPCA(so, features = rownames(so))
so <- RunTSNE(so, dims = 1:50)

TSNEPlot(so)
```


EXERCISE: The perplexity paramter is key to producing a good visualization. Generate tSNE plots at a range of perplexities (min = 2, max = 100). What do you notice about the projections?

```{r echo = FALSE}
# run on multiple perplexities
ps <- seq(2, 52, by = 10)

plts <- lapply(ps, 
    function(x) {
      so <- RunTSNE(so, 
               dims = 1:10, 
               perplexity = x)
      
      p <- TSNEPlot(so)
      p
    })


```

### UMAP

[`UMAP`](https://umap-learn.readthedocs.io/en/latest/parameters.html) is a newer algorithm for projecting data in 2D and have become very popular for single cell visualization. 

Similar to RunTSNE we specify the number of PCA dimesions to pass to RunUMAP.

```{r}
# specify to use the first 15 prinicpal components
so <- RunUMAP(so, dims = 1:15)

UMAPPlot(so)


# again make wrapper to plot umaps easily
plot_umap <- function(sobj, features = "", ...){
  FeaturePlot(sobj, 
            features,
            reduction = "umap", 
            ...)
}

plot_umap(so, c("CST3", "PPBP"))
```



EXERCISE: How does the projection change with varying parameters (min_dist, n_neighbors, etc)?


```{r}

plts <- lapply(c(5, 50, 100), function(n){
  so <- RunUMAP(object = so, 
                  n.neighbors = n, dims = 1:20)
  p <- UMAPPlot(object = so)
  p
})


plts <- lapply(c(0.1, 0.25, 0.5), 
            function(n){
              
  so <- RunUMAP(object = so, 
                  n.neighbors = 30L, 
                  min.dist= n,
                  dims = 1:20)
  
  p <- UMAPPlot(object = so)
  p
})

```

### Force directed graphs

Include Graph as image...

```{r, eval = FALSE}

library(igraph)

net <- graph.adjacency(
          adjmatrix = as.matrix(so@graphs$RNA_snn),
          mode = "undirected",
          weighted = TRUE,
          diag = FALSE
        )

lo <- layout_with_fr(net)

plot.igraph(net, 
            layout = lo,  
            edge.width = E(net)$weight,
            vertex.label = NA,
            vertex.color = so$seurat_clusters,
            vertex.size = 0, 
            curved = T)
```


## Diffusion Map

```{r}
library(destiny)
df_map <- DiffusionMap(as.matrix(GetAssayData(so, "data")[VariableFeatures(so), ]))

palette(cube_helix(6))
plot(df_map, 1:2)
```


## Clustering single cell data

The PCA/UMAP/tSNE plots above indicate that there are likely different cell types present in the data as we can see clear clusters. Next we will formally cluster the dataset to assign cells to clusters. 

[Duo et al, 2018 A systematic performance evaluation of clustering methods for single-cell RNA-seq data](https://f1000research.com/articles/7-1141)


### Graph based clustering

Description of graph based clustering. 

KNN grpah -> SNN graph using jaccard index of shared neighbors -> community detection algo (Louvain). originally used in Levain et al 2015.

First we compute a set of nearest neighbor graphs. Then we call the clustering algorithm.

```{r}
so <- FindNeighbors(so, 
              reduction = "pca",
              dims = 1:15,
              k.param = 15)

so <- FindClusters(so, resolution = 0.5)
```

## Exercise: Resolution and k.param are key parameters, what effect does varying them have on the clustering?

```{r echo = FALSE}

ns <- c(5, 30, 100)
plts <- list()
for (n in ns){
  so <- FindNeighbors(so, 
              reduction = "pca",
              dims = 1:15,
              k.param = n)
  so <- FindClusters(so, resolution = 0.5)
  p <- UMAPPlot(so)
  plts[[n]] <- p
}

plts 


so <- FindNeighbors(so,
                    reduction = "pca",
                    dims = 1:15,
                    k.param = 30)
  
res <- seq(0.1, 0.5, 0.1)
plts <- list()
for (n in res){
  so <- FindClusters(so, resolution = n)
  p <- UMAPPlot(so)
  plts[[paste0("res", n)]] <- p
}

plts 

```


## How many clusters?

Clustering algorithms produce clusters, even if there isn't anything meaningfully different between cells. Determining hte optimal number of clusters can be tricky and also dependent on the biological question. 

Some guidelines:
1) Cluster the data into a low number of clusters to identify cell types, then recluster to generate additional clusters to define sub-populations of cell types. 

2) To determine if the data is overclustered examine differentially expressed genes between clusters. If the clusters have few or no differentially expressed genes then the data is overclustered. Similar clusters can also be merged post-hoc if necessary as sometimes it is difficult to use 1 clustering setting for many diverse cell populations.

## Assessing relationships between clusters

Hierarchical clustering can be used to visualize the relationships between clusters. The average expression of each cluster is computed then the distances between the clusters are used for hierarchical clustering.

```{r}
so <- BuildClusterTree(so)

PlotClusterTree(so)
```


The [`clustree`](https://lazappi.github.io/clustree/index.html) package provides a nice plotting utility to visualize the relationship of cells at different resolution settings. 


```{r, fig.width=98 fig.height=8}
library(clustree)

clustree(so)

```


## Rename to clusters

Finally let's define the "cluster" as those generated with a resolution of 0.5 with 20 neighbors
```{r}
so$clusters <- so$RNA_snn_res.0.5

```
## Save object and metadata

It's always a good idea to store the cell metadata, particularly the clustering and projections, as these may change with reruns of the same data, for example if a package is updated or if the `seed.use` argument changes. 

```{r save}
dir.create("data", showWarnings = FALSE)

saveRDS(so, "data/clustered_sobj.rds")



to_keep <- c(colnames(so@meta.data), "UMAP_1", "UMAP_2")

df_out <- FetchData(so, to_keep) %>% 
  rownames_to_column("cell")

out_name <- paste0("data/", Sys.Date(), "_cell_metadata.tsv.gz")

write_tsv(df_out, out_name)
```


