<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>


  <!--radix_placeholder_meta_tags-->
<title>single cell RNA-seq workshop: Class 2: Dimensionality reduction, clustering, differential expression, cell type annotation</title>

<meta property="description" itemprop="description" content="In this class we will discuss scRNA-seq data processing for visualization and defining transcriptome features."/>


<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2021-11-03"/>
<meta property="article:created" itemprop="dateCreated" content="2021-11-03"/>
<meta name="article:author" content="Rui Fu"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="single cell RNA-seq workshop: Class 2: Dimensionality reduction, clustering, differential expression, cell type annotation"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="In this class we will discuss scRNA-seq data processing for visualization and defining transcriptome features."/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="single cell RNA-seq workshop"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary"/>
<meta property="twitter:title" content="single cell RNA-seq workshop: Class 2: Dimensionality reduction, clustering, differential expression, cell type annotation"/>
<meta property="twitter:description" content="In this class we will discuss scRNA-seq data processing for visualization and defining transcriptome features."/>

<!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","output"]}},"value":[{"type":"character","attributes":{},"value":["Class 2: Dimensionality reduction, clustering, differential expression, cell type annotation"]},{"type":"character","attributes":{},"value":["In this class we will discuss scRNA-seq data processing for visualization and defining transcriptome features."]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name"]}},"value":[{"type":"character","attributes":{},"value":["Rui Fu"]}]}]},{"type":"character","attributes":{},"value":["11-03-2021"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained"]}},"value":[{"type":"logical","attributes":{},"value":[false]}]}]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["class-2_files/anchor-4.2.2/anchor.min.js","class-2_files/bowser-1.9.3/bowser.min.js","class-2_files/distill-2.2.21/template.v2.js","class-2_files/figure-html5/clustree-1.png","class-2_files/figure-html5/elbow_plot-1.png","class-2_files/figure-html5/manual inspection-1.png","class-2_files/figure-html5/manual inspection-2.png","class-2_files/figure-html5/pc_plots-1.png","class-2_files/figure-html5/pc_plots-2.png","class-2_files/figure-html5/pc_plots-3.png","class-2_files/figure-html5/pc_plots-4.png","class-2_files/figure-html5/pc_plots-5.png","class-2_files/figure-html5/pc_plots-6.png","class-2_files/figure-html5/pca_loadings-1.png","class-2_files/figure-html5/pca-1.png","class-2_files/figure-html5/pkg_data-1.png","class-2_files/figure-html5/plotting-1.png","class-2_files/figure-html5/plotting-2.png","class-2_files/figure-html5/plotting-3.png","class-2_files/figure-html5/plotting-4.png","class-2_files/figure-html5/plotting-5.png","class-2_files/figure-html5/plotting-6.png","class-2_files/figure-html5/rename-1.png","class-2_files/figure-html5/umap-1.png","class-2_files/figure-html5/unnamed-chunk-3-1.png","class-2_files/figure-html5/unnamed-chunk-4-1.png","class-2_files/figure-html5/unnamed-chunk-4-2.png","class-2_files/figure-html5/unnamed-chunk-5-1.png","class-2_files/figure-html5/unnamed-chunk-5-2.png","class-2_files/figure-html5/unnamed-chunk-6-1.png","class-2_files/figure-html5/unnamed-chunk-7-1.png","class-2_files/figure-html5/unnamed-chunk-8-1.png","class-2_files/header-attrs-2.11/header-attrs.js","class-2_files/jquery-3.6.0/jquery-3.6.0.js","class-2_files/jquery-3.6.0/jquery-3.6.0.min.js","class-2_files/jquery-3.6.0/jquery-3.6.0.min.map","class-2_files/popper-2.6.0/popper.min.js","class-2_files/tippy-6.2.7/tippy-bundle.umd.min.js","class-2_files/tippy-6.2.7/tippy-light-border.css","class-2_files/tippy-6.2.7/tippy.css","class-2_files/tippy-6.2.7/tippy.umd.min.js","class-2_files/webcomponents-2.0.0/webcomponents.js"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure img {
  width: 100%;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme
  $('code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      var refHtml = $('#ref-' + $(this).attr('data-cites')).html();
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.11/header-attrs.js"></script>
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Class 2: Dimensionality reduction, clustering, differential expression, cell type annotation","description":"In this class we will discuss scRNA-seq data processing for visualization and defining transcriptome features.","authors":[{"author":"Rui Fu","authorURL":"#","affiliation":"&nbsp;","affiliationURL":"#","orcidID":""}],"publishedDate":"2021-11-03T00:00:00.000+00:00","citationText":"Fu, 2021"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo">
<img src="../../img/logo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">single cell RNA-seq workshop</a>
</div>
<div class="nav-right">
<a href="../../index.html">Workshop overview</a>
<a href="../../classes.html">Classes</a>
<a href="../../previous.html">Previous-workshops</a>
<a href="https://github.com/rnabioco/cellar" aria-label="Link to source">
<i class="fab fa-github" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Class 2: Dimensionality reduction, clustering, differential expression, cell type annotation</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>In this class we will discuss scRNA-seq data processing for visualization and defining transcriptome features.</p></p>
</div>

<div class="d-byline">
  
  Rui Fu
  
<br/>11-03-2021
</div>

<div class="d-article">
<h2 id="topics-for-this-session">Topics for this session:</h2>
<ul>
<li>Normalization and dimensionality reduction methods</li>
<li>Visualization strategies well-suited to single cell seq data</li>
<li>Clustering methods commonly used for single cell data</li>
<li>Differential gene expression discovery</li>
<li>Cell identity assignment</li>
</ul>
<h2 id="common-analysis-paradigm-for-scrna-seq">Common analysis paradigm for scRNA-seq</h2>
<ol type="1">
<li>Normalize counts and log transform. Data used for differential gene expression</li>
<li>Identify subset of informative genes to use for clustering (feature selection)</li>
<li>Run PCA to reduce the dimensionality of the dataset</li>
<li>Identify clusters using the PCA reduced data</li>
<li>Project the data into 2-D using UMAP/tSNE or just PCA to visualize</li>
<li>Find marker genes enriched in clusters</li>
<li>Define cell types</li>
<li>Repeat Steps 4-7 until satisfied</li>
</ol>
<p>Seurat accomplishes these steps using a succinct set of functions provided by <code>Seurat</code>:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='op'>?</span><span class='fu'>Seurat</span><span class='fu'>::</span><span class='va'><a href='https://satijalab.org/seurat/reference/NormalizeData.html'>NormalizeData</a></span>
<span class='op'>?</span><span class='fu'>Seurat</span><span class='fu'>::</span><span class='va'><a href='https://satijalab.org/seurat/reference/ScaleData.html'>ScaleData</a></span>
<span class='op'>?</span><span class='fu'>Seurat</span><span class='fu'>::</span><span class='va'><a href='https://satijalab.org/seurat/reference/RunPCA.html'>RunPCA</a></span>
<span class='op'>?</span><span class='fu'>Seurat</span><span class='fu'>::</span><span class='va'><a href='https://satijalab.org/seurat/reference/RunUMAP.html'>RunUMAP</a></span>
<span class='op'>?</span><span class='fu'>Seurat</span><span class='fu'>::</span><span class='va'><a href='https://satijalab.org/seurat/reference/FindNeighbors.html'>FindNeighbors</a></span>
<span class='op'>?</span><span class='fu'>Seurat</span><span class='fu'>::</span><span class='va'><a href='https://satijalab.org/seurat/reference/FindClusters.html'>FindClusters</a></span>
<span class='op'>?</span><span class='fu'>Seurat</span><span class='fu'>::</span><span class='va'><a href='https://satijalab.org/seurat/reference/FindMarkers.html'>FindMarkers</a></span>
</code></pre>
</div>
</div>
<h2 id="reloading-saved-objects">Reloading saved objects</h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://satijalab.org/seurat'>Seurat</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://wilkelab.org/cowplot/'>cowplot</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://Matrix.R-forge.R-project.org/'>Matrix</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/jokergoo/ComplexHeatmap'>ComplexHeatmap</a></span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>For this class, we load a basic QC filtered Seurat object, saved in RDS format, uploaded on Amazon S3.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/connections.html'>url</a></span><span class='op'>(</span><span class='st'>"https://scrnaseq-workshop.s3.us-west-2.amazonaws.com/filtered_sobj.rds"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>so</span>
</code></pre>
</div>
<pre><code>An object of class Seurat 
12572 features across 2686 samples within 1 assay 
Active assay: RNA (12572 features, 0 variable features)</code></pre>
</div>
<h2 id="normalization">Normalization</h2>
<p>Normalization of the UMI count data is necessary due to the large variability in total counts observed across cells. This variability is generally thought to be due to inefficiencies in library preparation and variable sequencing depths between cells. Without normalization the largest source of variability in the data would be due to technical differences rather than the biological differences of interest.</p>
<h3 id="global-size-factor-approach">Global size factor approach</h3>
<p>A commonly used approach for normalization of UMI data is a simple scaling normalization similar to the commonly used counts per million normalization methods. UMI counts per cell are scaled to the total # of UMIs per cell and multiplied by a constant for ease to prevent small numbers (10,000). The data are then log transformed (with a pseudocount) to stabilize the variance. e.g.</p>
<p><span class="math inline">\(\log((10000 * \frac{x_{i}}{\sum_{x_i}}) + 1)\)</span></p>
<p>This approach makes the assumption that cells all have the same amount of RNA and that sequencing depth is the reason for the variability. These assumptions aren’t really valid, especially when comparing diverse cell types, however in practice this approach can still provide reasonable results and is the default normalization method used in <code>Seurat</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/NormalizeData.html'>NormalizeData</a></span><span class='op'>(</span><span class='va'>so</span>, normalization.method <span class='op'>=</span> <span class='st'>"LogNormalize"</span><span class='op'>)</span>
<span class='co'># data slot now contains log normalized values</span>
<span class='fu'><a href='https://satijalab.org/seurat/reference/AssayData.html'>GetAssayData</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='st'>"data"</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span><span class='op'>]</span>
</code></pre>
</div>
<pre><code>10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
                                                             
AL627309.1    . . . . . .        . .        .        .       
RP11-206L10.2 . . . . . .        . .        .        .       
LINC00115     . . . . . .        . .        .        .       
NOC2L         . . . . . .        . .        1.254193 .       
KLHL17        . . . . . .        . .        .        .       
PLEKHN1       . . . . . .        . .        .        .       
HES4          . . . . . 1.766183 . .        .        .       
ISG15         . . . . . 2.369960 . 1.679169 .        2.836986
AGRN          . . . . . .        . .        .        .       
C1orf159      . . . . . .        . .        .        .       </code></pre>
</div>
<p>Normalization methods used in bulk RNA-seq data (e.g. <code>DESeq2</code> size-factors ) are not appropriate due to the large numbers of zeros in the data. However, by pooling similar cells together, one can obtain enough data to estimate a normalization size factor. The Bioconductor package <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html#1_introduction"><code>scran</code></a> implements such a strategy.</p>
<h2 id="identifying-highly-variable-genes-i.e.-feature-selection">Identifying highly variable genes (i.e. feature selection)</h2>
<p>We will next select important features to use for dimensionality reduction, clustering and tSNE/uMAP projection. We can in theory use all ~20K genes in the dataset for these steps, however this is often computationally expensive and unnecessary. Most genes in a dataset are going to be expressed at relatively similar values across different cells, or are going to be so lowly expressed that they contribute more noise than signal to the analysis. For determining the relationships between cells in gene-expression space, we want to focus on gene expression differences that are driven primarily by biological variation rather than technical variation.</p>
<p>Seurat provides a function to help identify these genes, <code>FindVariableGenes</code>. Ranking genes by their variance alone will bias towards selecting highly expressed genes. To help mitigate this Seurat uses a <code>vst</code> method to identify genes. Briefly, a curve is fit to model the mean and variance for each gene in log space. Variances are then standardized against this model, and ranked. The top 2000 genes with the highest standardized variances are then called as highly variable, by default.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FindVariableFeatures.html'>FindVariableFeatures</a></span><span class='op'>(</span><span class='va'>so</span>, selection.method <span class='op'>=</span> <span class='st'>"vst"</span><span class='op'>)</span>
<span class='fu'><a href='https://satijalab.org/seurat/reference/VariableFeaturePlot.html'>VariableFeaturePlot</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span> 
</code></pre>
</div>
<img src="class-2_files/figure-html5/pkg_data-1.png" width="576" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>so</span><span class='op'>@</span><span class='va'>assays</span><span class='op'>$</span><span class='va'>RNA</span><span class='op'>@</span><span class='va'>var.features</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span><span class='op'>]</span>
</code></pre>
</div>
<pre><code> [1] &quot;PPBP&quot;   &quot;S100A9&quot; &quot;LYZ&quot;    &quot;IGLL5&quot;  &quot;GNLY&quot;   &quot;FTL&quot;    &quot;PF4&quot;   
 [8] &quot;FTH1&quot;   &quot;FCER1A&quot; &quot;GNG11&quot; </code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/VariableFeatures.html'>VariableFeatures</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span><span class='op'>]</span>
</code></pre>
</div>
<pre><code> [1] &quot;PPBP&quot;   &quot;S100A9&quot; &quot;LYZ&quot;    &quot;IGLL5&quot;  &quot;GNLY&quot;   &quot;FTL&quot;    &quot;PF4&quot;   
 [8] &quot;FTH1&quot;   &quot;FCER1A&quot; &quot;GNG11&quot; </code></pre>
</div>
<p>An alternative approach uses the Bioconductor package <a href="https://www.bioconductor.org/packages/release/bioc/html/M3Drop.html"><code>M3Drop</code></a>. This approach does not model the variance in the data, but rather uses the information about dropouts to identify genes that have high variance. Specifically, there is a clear relationship between the dropout rate and the abundance of a gene.</p>
<h2 id="scaling">Scaling</h2>
<p>Next, we will scale the normalized expression values. The scaled values will only be used for dimensionality reduction and clustering, and not differential expression. The purpose of scaling is to set the expression values for each gene onto a similar numeric scale. This avoid issues of have a gene that is highly expressed being given more weight in clustering simply because it has larger numbers. Scaling converts the normalized data into z-scores by default and the values are stored in the <code>scale.data</code> slot.</p>
<p>Single cell data sets often contain various types of ‘uninteresting’ variation. These can include technical noise, batch effects and confounding biological variation (such as cell cycle stage). We can use modeling to regress these signals out of the analysis using <code>ScaleData</code>. Variables present in the <code>meta.data</code> can be supplied as variables to regress. A linear model is fit between the the supplied variables (i.e <code>nUMI</code>, <code>proportion.mito</code>, etc.) and gene expression value. The residuals from this model are then scaled instead of the normalized gene expression.</p>
<p>In general, we recommend performing exploratory data analysis first without regressing any nuisance variables. Later, if you discover a nuisance variable (i.e. batch or cell cycle), then you can regress it out.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># scale all of the data, useful if you want to make heatmaps later</span>
<span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/ScaleData.html'>ScaleData</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>so</span>, features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># for large datasets, just scale the variable genes:</span>
<span class='co'># so &lt;- ScaleData(object = so, </span>
<span class='co'>#                 features = VariableFeatures(so))</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># get scaled values for top 50 variable genes</span>
<span class='va'>to_plot</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/AssayData.html'>GetAssayData</a></span><span class='op'>(</span><span class='va'>so</span>, slot <span class='op'>=</span> <span class='st'>"scale.data"</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://satijalab.org/seurat/reference/VariableFeatures.html'>VariableFeatures</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>50</span><span class='op'>]</span>, <span class='op'>]</span>
<span class='fu'><a href='https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html'>Heatmap</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>to_plot</span><span class='op'>)</span>,
        show_column_names <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/unnamed-chunk-3-1.png" width="768" /></p>
</div>
<h2 id="performing-dimensionality-reduction-and-clustering">Performing dimensionality reduction and clustering</h2>
<h3 id="pca">PCA</h3>
<p>Heatmaps are useful for visualizing small numbers of genes (10s or 100s), but quickly become impractical for larger single cell datasets due to high dimensionality (1000s of cells X 20000 genes).</p>
<p>Dimensionality reduction techniques are used in single cell data analysis generally for two purposes:</p>
<ol type="1">
<li><p>Provide visualizations in 2 dimensions that convey information about relationships between cells.</p></li>
<li><p>Reduce the computational burden of working with 20K dimensions to a smaller set of dimensions that capture signal.</p></li>
</ol>
<p>By identifying the highly variable genes we’ve already reduced the dimensions from 20K to ~1500. Next we will use <a href="http://setosa.io/ev/principal-component-analysis/">Principal Component Analysis</a> to identify a smaller subset of dimensions (10-100) that will be used for clustering and visualization.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/RunPCA.html'>RunPCA</a></span><span class='op'>(</span><span class='va'>so</span>,
             features <span class='op'>=</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/VariableFeatures.html'>VariableFeatures</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span>, 
             verbose <span class='op'>=</span> <span class='cn'>FALSE</span>, 
             ndims.print <span class='op'>=</span> <span class='fl'>0</span><span class='op'>)</span>
<span class='fu'><a href='https://satijalab.org/seurat/reference/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>so</span>, reduction <span class='op'>=</span> <span class='st'>"pca"</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/pca-1.png" width="624" /></p>
</div>
<p>The <code>RunPCA</code> function will calculate the top 50 principal components using a modified form of PCA (see <code>?irlba</code>) that doesn’t need to compute all of the PCs. By default it will also print the genes most highly associated with each principal component, we will visualize these later. Often these genes will be recognizable as markers of different cell populations and they are also often highly variable genes.</p>
<p><strong>Note</strong> Some functions in Seurat (<code>RunPCA</code>, <code>RunTSNE</code>, <code>RunUMAP</code>, <code>FindClusters</code>), have a non-deterministic component that will give different results (usually slightly) each run due to a randomization step in the algorithm. However, you can set an integer seed to make the output reproducible. This is useful so that you get the same clustering results if you rerun your analysis code. In many functions seurat has done this for you with the <code>seed.use</code> parameter, set to <code>42</code>. If you set this to NULL you will get a non-deterministic result.</p>
<p>Shown in the tabs are 2 dimensional plots showing the principal component scores of cells.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>pcs</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>2</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>3</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>4</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>3</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>4</span><span class='op'>)</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>3</span>, <span class='fl'>4</span><span class='op'>)</span>
<span class='op'>)</span>
<span class='kw'>for</span><span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>pcs</span><span class='op'>)</span><span class='op'>)</span><span class='op'>{</span>
  <span class='fu'><a href='https://rdrr.io/r/base/cat.html'>cat</a></span><span class='op'>(</span><span class='st'>'\n#### '</span>, <span class='st'>'PC'</span>, <span class='va'>pcs</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>, <span class='st'>' vs PC'</span>, <span class='va'>pcs</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span>, <span class='st'>'\n'</span>, sep <span class='op'>=</span> <span class='st'>""</span><span class='op'>)</span>
  <span class='va'>p</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>so</span>, 
          dims <span class='op'>=</span> <span class='va'>pcs</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span>, 
          reduction <span class='op'>=</span> <span class='st'>"pca"</span>,
          cols <span class='op'>=</span> <span class='fu'>RColorBrewer</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html'>brewer.pal</a></span><span class='op'>(</span><span class='fl'>10</span>, <span class='st'>"Paired"</span><span class='op'>)</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>p</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/cat.html'>cat</a></span><span class='op'>(</span><span class='st'>'\n'</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
<h4 id="pc1-vs-pc2">PC1 vs PC2</h4>
<p><img src="class-2_files/figure-html5/pc_plots-1.png" width="624" /></p>
<h4 id="pc1-vs-pc3">PC1 vs PC3</h4>
<p><img src="class-2_files/figure-html5/pc_plots-2.png" width="624" /></p>
<h4 id="pc1-vs-pc4">PC1 vs PC4</h4>
<p><img src="class-2_files/figure-html5/pc_plots-3.png" width="624" /></p>
<h4 id="pc2-vs-pc3">PC2 vs PC3</h4>
<p><img src="class-2_files/figure-html5/pc_plots-4.png" width="624" /></p>
<h4 id="pc2-vs-pc4">PC2 vs PC4</h4>
<p><img src="class-2_files/figure-html5/pc_plots-5.png" width="624" /></p>
<h4 id="pc3-vs-pc4">PC3 vs PC4</h4>
<p><img src="class-2_files/figure-html5/pc_plots-6.png" width="624" /></p>
</div>
<p>To see which genes are most strongly associated with the PCs you can use the <code>VizDimLoadings</code> function.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/VizDimLoadings.html'>VizDimLoadings</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>so</span>, 
               dims <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span>, 
               balanced <span class='op'>=</span> <span class='cn'>TRUE</span>, 
               ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/pca_loadings-1.png" width="864" /></p>
</div>
<p>Finally, we can generate an ‘elbow plot’ of the standard deviation explained by each of the principle components.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/ElbowPlot.html'>ElbowPlot</a></span><span class='op'>(</span><span class='va'>so</span>, ndims <span class='op'>=</span> <span class='fl'>40</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/elbow_plot-1.png" width="576" /></p>
</div>
<p>Note that most of the variation in the data is captured in a small number of PCs (&lt; 20)</p>
<p>Early single cell studies simply used PCA to visualize single cell data, and by plotting the data in scatterplots reduced the dimensionality of the data to 2D. This is a fine approach for simple datasets, but often there is much more information present in higher principal components.</p>
<p>We can use PC1 and PC2 to begin to assess the structure of the dataset.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>so</span>, 
            <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"nFeature_RNA"</span>, <span class='st'>"nCount_RNA"</span><span class='op'>)</span>, 
            reduction <span class='op'>=</span> <span class='st'>"pca"</span><span class='op'>)</span>
</code></pre>
</div>
<img src="class-2_files/figure-html5/unnamed-chunk-4-1.png" width="624" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Color by CD8, marker of CD8 t-cells</span>
<span class='fu'><a href='https://satijalab.org/seurat/reference/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>so</span>, 
            <span class='st'>"CD8A"</span>, 
            reduction <span class='op'>=</span> <span class='st'>"pca"</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/unnamed-chunk-4-2.png" width="624" /></p>
</div>
<p>Use wrapper functions for plotting to save some typing.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>plot_pca</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>sobj</span>, <span class='va'>features</span>, <span class='va'>...</span><span class='op'>)</span><span class='op'>{</span>
  <span class='fu'><a href='https://satijalab.org/seurat/reference/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>sobj</span>, 
            <span class='va'>features</span>,
            reduction <span class='op'>=</span> <span class='st'>"pca"</span>, 
            <span class='va'>...</span><span class='op'>)</span>
<span class='op'>}</span>
<span class='co'># Color by top genes associated with Pc1 and Pc2</span>
<span class='fu'>plot_pca</span><span class='op'>(</span><span class='va'>so</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CST3"</span>, <span class='st'>"NKG7"</span><span class='op'>)</span><span class='op'>)</span> 
</code></pre>
</div>
<img src="class-2_files/figure-html5/unnamed-chunk-5-1.png" width="624" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># or use ggplot directly if need more custom viz</span>
<span class='va'>var_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FetchData.html'>FetchData</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"PC_1"</span>, <span class='st'>"PC_2"</span>, <span class='st'>"CD19"</span>, <span class='st'>"CD8A"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>gather</span><span class='op'>(</span><span class='va'>gene</span>, <span class='va'>expr</span>, <span class='op'>-</span><span class='va'>PC_1</span>, <span class='op'>-</span><span class='va'>PC_2</span><span class='op'>)</span>

<span class='fu'>ggplot</span><span class='op'>(</span><span class='va'>var_df</span>, <span class='fu'>aes</span><span class='op'>(</span><span class='va'>PC_1</span>, <span class='va'>PC_2</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_point</span><span class='op'>(</span><span class='fu'>aes</span><span class='op'>(</span>color <span class='op'>=</span> <span class='va'>expr</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>scale_color_gradientn</span><span class='op'>(</span>colours <span class='op'>=</span> <span class='fu'>RColorBrewer</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html'>brewer.pal</a></span><span class='op'>(</span><span class='fl'>9</span>, <span class='st'>"Reds"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'>facet_grid</span><span class='op'>(</span><span class='op'>~</span><span class='va'>gene</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/unnamed-chunk-5-2.png" width="624" /></p>
</div>
<p>Note how other PCs separate different cell populations (PPBP is a marker of megakaryocytes).</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>plot_pca</span><span class='op'>(</span><span class='va'>so</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"PPBP"</span>, <span class='st'>"NKG7"</span><span class='op'>)</span>, dims <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>,<span class='fl'>3</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/unnamed-chunk-6-1.png" width="624" /></p>
</div>
<h3 id="umap">UMAP</h3>
<p>The first two dimensions of PCA are not sufficient to capture the variation in these data due to the complexity. However, PCA is useful in identifying a new set of dimensions (10-100) that capture most of the interesting variation in the data. This is useful because now with this reduced dataset we can use newer visualization technique to examine relationships between cells in 2 dimensions.</p>
<p><a href="https://umap-learn.readthedocs.io/en/latest/parameters.html"><code>UMAP</code></a> is a newer algorithm for projecting data in 2D (or higher) and has become very popular for single cell visualization. The UMAP algorithm derives from topological methods for data analysis, and is nicely described by the authors in the UMAP <a href="https://umap-learn.readthedocs.io/en/latest/how_umap_works.html">documentation</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/RunUMAP.html'>RunUMAP</a></span><span class='op'>(</span><span class='va'>so</span>, dims <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>15</span><span class='op'>)</span>
<span class='fu'><a href='https://satijalab.org/seurat/reference/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>so</span>, reduction <span class='op'>=</span> <span class='st'>"umap"</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/umap-1.png" width="624" /></p>
</div>
<h2 id="clustering-single-cell-data">Clustering single cell data</h2>
<p>The PCA/UMAP/tSNE plots above indicate that there are likely different cell types present in the data as we can see clusters. Next we will formally cluster the dataset to assign cells to clusters. See this recent benchmarking paper for discussion of best clustering methods. <a href="https://f1000research.com/articles/7-1141">Duo et al, 2018 A systematic performance evaluation of clustering methods for single-cell RNA-seq data</a></p>
<p>Graph based clustering methods are commonly used for single cell data, because they can scale to millions of cells, produce reasonable assignments, and have tunable parameters. The approach that is implemented in Seurat is performed as follows:</p>
<ol type="1">
<li>construct a K-nearest (KNN) neighbor matrix.</li>
<li>calculate shared nearest neighbors (SNN) using the jaccard index.</li>
<li>use the <a href="https://en.wikipedia.org/wiki/Louvain_modularity">Louvain</a> community detection algorithm to assign clusters.</li>
</ol>
<p>In Seurat the clustering is done using two functions:<code>FindNeighbors</code> which computes the KNN and SNN graphs, and <code>FindClusters</code> which finds clusters.</p>
<p><code>FindNeighbors</code> calculates the KNN graph using the PCA matrix as input (or other dimensionality reductions).</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FindNeighbors.html'>FindNeighbors</a></span><span class='op'>(</span><span class='va'>so</span>, 
              reduction <span class='op'>=</span> <span class='st'>"pca"</span>,
              dims <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>15</span>,
              k.param <span class='op'>=</span> <span class='fl'>15</span><span class='op'>)</span>
<span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FindClusters.html'>FindClusters</a></span><span class='op'>(</span><span class='va'>so</span>, resolution <span class='op'>=</span> <span class='fl'>0.5</span>, verbose <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>so</span><span class='op'>$</span><span class='va'>RNA_snn_res.0.5</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
               0                3                4                4 
AAAGTTTGATCACG-1 AAATCATGACCACA-1 
               3                5 
Levels: 0 1 2 3 4 5 6 7 8</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>so</span><span class='op'>$</span><span class='va'>seurat_clusters</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
               0                3                4                4 
AAAGTTTGATCACG-1 AAATCATGACCACA-1 
               3                5 
Levels: 0 1 2 3 4 5 6 7 8</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>so</span>, reduction <span class='op'>=</span> <span class='st'>"umap"</span>, label <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/unnamed-chunk-7-1.png" width="624" /></p>
</div>
<h2 id="how-many-clusters">How many clusters?</h2>
<p>Clustering algorithms produce clusters, even if there isn’t anything meaningfully different between cells. Determining the optimal number of clusters can be tricky and also dependent on the biological question.</p>
<p>Some guidelines:</p>
<ol type="1">
<li><p>Cluster the data into a small number of clusters to identify cell types, then recluster to generate additional clusters to define sub-populations of cell types.</p></li>
<li><p>To determine if the data is overclustered, examine differentially expressed genes between clusters. If the clusters have few or no differentially expressed genes then the data is overclustered. Similar clusters can be merged post-hoc if necessary as sometimes it is difficult to use one clustering approach for many diverse cell populations.</p></li>
</ol>
<h2 id="assessing-relationships-between-clusters">Assessing relationships between clusters</h2>
<p>Hierarchical clustering can be used to visualize the relationships between clusters. The average expression of each cluster is computed then the distances between the clusters are used for hierarchical clustering.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/Idents.html'>Idents</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='st'>"RNA_snn_res.0.5"</span>
<span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/BuildClusterTree.html'>BuildClusterTree</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span>
<span class='fu'><a href='https://satijalab.org/seurat/reference/PlotClusterTree.html'>PlotClusterTree</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/unnamed-chunk-8-1.png" width="624" /></p>
</div>
<p>The <a href="https://lazappi.github.io/clustree/index.html"><code>clustree</code></a> package provides a nice plotting utility to visualize the relationship of cells at different resolution settings.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/lazappi/clustree'>clustree</a></span><span class='op'>)</span>
<span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FindClusters.html'>FindClusters</a></span><span class='op'>(</span><span class='va'>so</span>, resolution <span class='op'>=</span> <span class='fl'>0.2</span>, verbose <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FindClusters.html'>FindClusters</a></span><span class='op'>(</span><span class='va'>so</span>, resolution <span class='op'>=</span> <span class='fl'>0.8</span>, verbose <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/clustree/man/clustree.html'>clustree</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/clustree-1.png" width="624" /></p>
</div>
<p>Finally let’s define a meta data column as <code>clusters</code> as those generated with a resolution of 0.5 with 30 neighbors.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>so</span><span class='op'>$</span><span class='va'>clusters</span> <span class='op'>&lt;-</span> <span class='va'>so</span><span class='op'>$</span><span class='va'>RNA_snn_res.0.5</span>
</code></pre>
</div>
</div>
<h2 id="save-object-and-metadata">Save object and metadata</h2>
<p>It’s always a good idea to store the cell metadata, particularly the clustering and projections, as these may change with reruns of the same data, for example if a package is updated or if the <code>seed.use</code> argument changes. Further down the road, when sequencing data is submitted to NCBI GEO, depositing metadata will help future bioinformaticians reproduce/reanalyze findings.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='st'>"data"</span>, showWarnings <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='st'>"data/clustered_sobj.rds"</span><span class='op'>)</span>
<span class='va'>to_keep</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>@</span><span class='va'>meta.data</span><span class='op'>)</span>, <span class='st'>"UMAP_1"</span>, <span class='st'>"UMAP_2"</span><span class='op'>)</span>
<span class='va'>df_out</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FetchData.html'>FetchData</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='va'>to_keep</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>rownames_to_column</span><span class='op'>(</span><span class='st'>"cell"</span><span class='op'>)</span>
<span class='va'>out_name</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"data/"</span>, <span class='fu'><a href='https://rdrr.io/r/base/Sys.time.html'>Sys.Date</a></span><span class='op'>(</span><span class='op'>)</span>, <span class='st'>"_cell_metadata.tsv.gz"</span><span class='op'>)</span>
<span class='fu'>write_tsv</span><span class='op'>(</span><span class='va'>df_out</span>, <span class='va'>out_name</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h2 id="finding-markers-and-differential-expression-analysis">Finding markers and differential expression analysis</h2>
<p>After clustering, differential expression testing (DE analysis, similar to bulk RNA-seq) finds gene signatures of each cluster/cell population, to give more insights into biology.</p>
<h3 id="many-proposed-methods">Many proposed methods</h3>
<p>(In general, we recommend default “wilcox” or “t”, good balances between speed and accuracy)</p>
<figure>
<img src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fnmeth.4612/MediaObjects/41592_2018_Article_BFnmeth4612_Fig5_HTML.jpg" alt="Soneson and Robinson, 2018" /><figcaption aria-hidden="true">Soneson and Robinson, 2018</figcaption>
</figure>
<p>In Seurat, we can use the <code>FindAllMarkers()</code> function, which will perform the <code>wilcox.test()</code> by default . Specifically, the function will iterate through each cluster, comparing the cells in one cluster to the cells in all of the other clusters. The test is run for every gene that is detectable above certain thresholds. The output is then filtered to identify significant genes with a positive fold-change (i.e. higher in abundance the cluster).</p>
<p>We recommend using the <code>wilcoxauc()</code> function from the <code>presto</code> package (installed from GitHub), which will do the same thing but in a (tiny) fraction of the time, due to implementation with C++ backend.</p>
<h3 id="important-considerations">Important considerations</h3>
<ol type="1">
<li><p>Keep things simple first and look at the results (try not integrating, not regressing), look at the output, and determine what variables need to be regressed out (batch, nCount_RNA, percent_mito, cell cycle, etc) during scaling (note this only affects dimension reduction and clustering).</p></li>
<li><p>Note that the p-values from these tests will be absurdly low (near or at 0). This is in part due to the large number of samples tested (e.g. each cell is considered a sample) but also due to the data being clustered based on the gene expression. Testing for differential expression between clusters will always result in some differentially expressed genes… because that’s what makes them different clusters in the first place, hence the logical is somewhat circular and result in inflated p-values. Nevertheless, we can use the p-values to rank the genes, but you shouldn’t just blindly trust that a p-value &lt; 0.05 is something meaningful in single cell data.</p></li>
<li><p>Use normalized data for DE (slot is dependent on normalization method, also don’t use integrated assay)</p></li>
<li><p>Note that marker genes found is very dependent on clustering and the compared populations.</p></li>
</ol>
<h3 id="find-all-markers-for-each-cluster">Find all markers for each cluster</h3>
<p><code>FindAllMarkers</code> compares cells in each cluster to all other cells in the dataset. Typically, focus is given to genes upregulated in each cluster, i.e. markers.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Load clustered object if needed</span>
<span class='co'># so &lt;- readRDS(url("https://scrnaseq-workshop.s3.us-west-2.amazonaws.com/clustered_sobj.rds"))</span>
<span class='co'># Also check/set ident to the desired comparison</span>
<span class='fu'><a href='https://satijalab.org/seurat/reference/Idents.html'>Idents</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
               0                3                4                4 
AAAGTTTGATCACG-1 AAATCATGACCACA-1 
               3                5 
Levels: 0 1 2 3 4 5 6 7 8 9</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/Idents.html'>Idents</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='st'>"clusters"</span>
<span class='va'>markers_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FindAllMarkers.html'>FindAllMarkers</a></span><span class='op'>(</span><span class='va'>so</span>,
                             assay <span class='op'>=</span> <span class='st'>"RNA"</span>, <span class='co'># be careful with sobj_sc and intergrated objects</span>
                             slot <span class='op'>=</span> <span class='st'>"data"</span>,
                             only.pos <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>markers_df</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>              p_val avg_log2FC pct.1 pct.2     p_val_adj cluster
RPS12 1.687445e-141  0.7223366 1.000 0.991 2.121455e-137       0
RPL32 6.448401e-140  0.6191260 0.999 0.995 8.106930e-136       0
RPS27 1.489397e-137  0.7075273 0.999 0.992 1.872469e-133       0
RPS6  2.128866e-137  0.6546818 0.997 0.995 2.676410e-133       0
RPS25 2.281824e-132  0.7760917 0.999 0.974 2.868710e-128       0
RPS14 1.962082e-127  0.5966430 0.999 0.995 2.466730e-123       0
       gene
RPS12 RPS12
RPL32 RPL32
RPS27 RPS27
RPS6   RPS6
RPS25 RPS25
RPS14 RPS14</code></pre>
</div>
<p>pct.1: fraction detected in the cluster in question, pct.2: fraction detected in the rest of the cells avg_logFC: expression comparison of the cluster in question (natural log, so exp(x) to reverse) vs rest of the cells</p>
<p>And save tables to disk with <code>write_csv</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>write_csv</span><span class='op'>(</span><span class='va'>markers_df</span>, <span class='st'>"data/allmarkers.csv"</span><span class='op'>)</span>
<span class='va'>markers_df</span> <span class='op'>&lt;-</span> <span class='fu'>read_csv</span><span class='op'>(</span><span class='st'>"data/allmarkers.csv"</span><span class='op'>)</span>
<span class='va'>markers_df</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Consider the equivalent but faster way, using <code>presto::wilcoxauc</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>presto</span><span class='op'>)</span>
<span class='va'>markers_presto</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/presto/man/wilcoxauc.html'>wilcoxauc</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='st'>"clusters"</span><span class='op'>)</span> <span class='co'># give metadata column name</span>
<span class='va'>markers_presto</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span> <span class='co'># also contains negative markers, how would you keep only the positive?</span>
</code></pre>
</div>
<pre><code>        feature group     avgExpr         logFC statistic       auc
1    AL627309.1     0 0.002306096 -0.0041467779    706481 0.4986579
2 RP11-206L10.2     0 0.002530023  0.0001246319    707925 0.4996771
3     LINC00115     0 0.005074268 -0.0077729488    704592 0.4973245
4         NOC2L     0 0.148937014 -0.0133363124    697980 0.4926576
5        KLHL17     0 0.002593014 -0.0026994222    706844 0.4989141
6       PLEKHN1     0 0.002349838 -0.0029621530    707202 0.4991668
       pval      padj    pct_in   pct_out
1 0.2862773 0.4893801 0.1386963 0.4071247
2 0.7310980 0.8510458 0.1386963 0.2035623
3 0.1321250 0.3025638 0.2773925 0.8142494
4 0.2516458 0.4484960 8.3217753 9.9745547
5 0.3602607 0.5645237 0.1386963 0.3562341
6 0.4531110 0.6544015 0.1386963 0.3053435</code></pre>
</div>
<h3 id="find-de-genes-for-specific-cell-groups">Find DE genes for specific cell groups</h3>
<p>For more control in the comparisons, use <code>FindMarkers</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># DE analysis for cluster 1 vs 2</span>
<span class='va'>markers_df2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FindMarkers.html'>FindMarkers</a></span><span class='op'>(</span><span class='va'>so</span>,
                           assay <span class='op'>=</span> <span class='st'>"RNA"</span>,
                           slot <span class='op'>=</span> <span class='st'>"data"</span>,
                           ident.1 <span class='op'>=</span> <span class='st'>"1"</span>,
                           ident.2 <span class='op'>=</span> <span class='st'>"2"</span>,
                           test.use <span class='op'>=</span> <span class='st'>"t"</span><span class='op'>)</span>
<span class='va'>markers_df2</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>               p_val avg_log2FC pct.1 pct.2     p_val_adj
S100A9  0.000000e+00   6.277399 0.994 0.186  0.000000e+00
LYZ     0.000000e+00   5.498958 0.998 0.540  0.000000e+00
CST3    0.000000e+00   4.881313 0.992 0.220  0.000000e+00
TYROBP  0.000000e+00   5.076521 0.994 0.136  0.000000e+00
FTL     0.000000e+00   3.711881 1.000 0.992  0.000000e+00
FTH1   2.978166e-317   2.835164 1.000 0.994 3.744151e-313</code></pre>
</div>
<p>pct.1: fraction detected in the cells of ident.1, pct.2: fraction detected in the cells of ident.2 avg_logFC: expression comparison (natural log fold change) of ident.1 vs ident.2 (&gt; 0 means higher in ident.1)</p>
<p><code>FindMarkers</code> defaults to the current active ident. To use other value groups, set idents to the intended column, or use the <code>group.by</code> argument.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Compare control vs treated in only cluster 1</span>
<span class='va'>markers_df3</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FindMarkers.html'>FindMarkers</a></span><span class='op'>(</span><span class='va'>so</span>,
                           assay <span class='op'>=</span> <span class='st'>"RNA"</span>,
                           slot <span class='op'>=</span> <span class='st'>"data"</span>,
                           subset.ident <span class='op'>=</span> <span class='st'>"1"</span>, <span class='co'># &lt;- if needed, subset on current ident first, then switch idents</span>
                           group.by <span class='op'>=</span> <span class='st'>"orig.ident"</span>, <span class='co'># &lt;- grouping cells by this metadata column</span>
                           ident.1 <span class='op'>=</span> <span class='st'>"control"</span>,
                           ident.2 <span class='op'>=</span> <span class='st'>"treated"</span><span class='op'>)</span>
<span class='co'># More complicated example</span>
<span class='co'># Compare cluster 1 control vs cluster 2 treated</span>
<span class='co'># Make new metadata column for this need</span>
<span class='va'>so</span><span class='op'>@</span><span class='va'>meta.data</span><span class='op'>$</span><span class='va'>newid</span> <span class='op'>&lt;-</span> <span class='va'>so</span><span class='op'>@</span><span class='va'>meta.data</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>mutate</span><span class='op'>(</span>newid <span class='op'>=</span> <span class='fu'>str_c</span><span class='op'>(</span><span class='va'>orig.ident</span>, <span class='va'>clusters</span>, sep <span class='op'>=</span> <span class='st'>"_"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>pull</span><span class='op'>(</span><span class='va'>newid</span><span class='op'>)</span>
<span class='va'>markers_df4</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FindMarkers.html'>FindMarkers</a></span><span class='op'>(</span><span class='va'>so</span>,
                           assay <span class='op'>=</span> <span class='st'>"RNA"</span>,
                           slot <span class='op'>=</span> <span class='st'>"data"</span>,
                           group.by <span class='op'>=</span> <span class='st'>"newid"</span>, 
                           ident.1 <span class='op'>=</span> <span class='st'>"control_1"</span>,
                           ident.2 <span class='op'>=</span> <span class='st'>"treated_2"</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Genes of interest can then be visualized as violin plots or feature plots.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Violin plots</span>
<span class='fu'><a href='https://satijalab.org/seurat/reference/VlnPlot.html'>VlnPlot</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='st'>"LYZ"</span><span class='op'>)</span>
</code></pre>
</div>
<img src="class-2_files/figure-html5/plotting-1.png" width="624" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/VlnPlot.html'>VlnPlot</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"LYZ"</span>, <span class='st'>"S100A4"</span><span class='op'>)</span><span class='op'>)</span> <span class='co'># can be a vector of gene names</span>
</code></pre>
</div>
<img src="class-2_files/figure-html5/plotting-2.png" width="624" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/VlnPlot.html'>VlnPlot</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='st'>"LYZ"</span>, split.by <span class='op'>=</span> <span class='st'>"orig.ident"</span>, split.plot <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
<img src="class-2_files/figure-html5/plotting-3.png" width="624" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Project on UMAP</span>
<span class='fu'><a href='https://satijalab.org/seurat/reference/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='st'>"LYZ"</span><span class='op'>)</span>
</code></pre>
</div>
<img src="class-2_files/figure-html5/plotting-4.png" width="624" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"LYZ"</span>, <span class='st'>"S100A4"</span><span class='op'>)</span><span class='op'>)</span> <span class='co'># can be a vector of gene names</span>
</code></pre>
</div>
<img src="class-2_files/figure-html5/plotting-5.png" width="624" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>so</span>, <span class='st'>"LYZ"</span>, split.by <span class='op'>=</span> <span class='st'>"orig.ident"</span><span class='op'>)</span> <span class='co'># &lt;- split into panels based on metadata column</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/plotting-6.png" width="624" /></p>
</div>
<h2 id="cluster-identities">Cluster identities</h2>
<p>Without venturing into the realm of philosophical debates on what a “cell type” constitutes, standard practice is to use certain gene expression features to classify cells. This is often done manually, by visual inspection of key genes. Automated/less-biased approaches that utilize a broader range of features are currently being developed.</p>
<h3 id="manual-inspection-of-key-genes-using-expertise-in-the-lab">Manual inspection of key genes using expertise in the lab</h3>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>so</span>, 
            features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"MS4A1"</span>, <span class='st'>"GNLY"</span>, <span class='st'>"CD3E"</span>, <span class='st'>"CD14"</span>, <span class='st'>"FCER1A"</span>, <span class='st'>"FCGR3A"</span>, <span class='st'>"LYZ"</span>, <span class='st'>"PPBP"</span>, <span class='st'>"CD8A"</span><span class='op'>)</span>,
            ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
</code></pre>
</div>
<img src="class-2_files/figure-html5/manual%20inspection-1.png" width="624" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://satijalab.org/seurat/reference/VlnPlot.html'>VlnPlot</a></span><span class='op'>(</span><span class='va'>so</span>, 
        features <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"MS4A1"</span>, <span class='st'>"GNLY"</span>, <span class='st'>"CD3E"</span>, <span class='st'>"CD14"</span>, <span class='st'>"FCER1A"</span>, <span class='st'>"FCGR3A"</span>, <span class='st'>"LYZ"</span>, <span class='st'>"PPBP"</span>, <span class='st'>"CD8A"</span><span class='op'>)</span>,
        ncol <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="class-2_files/figure-html5/manual%20inspection-2.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>a</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>so</span>, label <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/SeuratTheme.html'>NoLegend</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='co'># Use RenameIdents to remap the idents from the current IDs to the new IDs </span>
<span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/Idents.html'>RenameIdents</a></span><span class='op'>(</span><span class='va'>so</span>,
                   <span class='st'>"0"</span> <span class='op'>=</span> <span class='st'>"CD4+ T"</span>,
                   <span class='st'>"1"</span> <span class='op'>=</span> <span class='st'>"CD14+ Mono"</span>,
                   <span class='st'>"2"</span> <span class='op'>=</span> <span class='st'>"CD4+ T"</span>,
                   <span class='st'>"3"</span> <span class='op'>=</span> <span class='st'>"B"</span>,
                   <span class='st'>"4"</span> <span class='op'>=</span> <span class='st'>"CD8+ T"</span>,
                   <span class='st'>"5"</span> <span class='op'>=</span> <span class='st'>"FCGR3A+ Mono"</span>,
                   <span class='st'>"6"</span> <span class='op'>=</span> <span class='st'>"NK"</span>,
                   <span class='st'>"7"</span> <span class='op'>=</span> <span class='st'>"DC"</span>,
                   <span class='st'>"8"</span> <span class='op'>=</span> <span class='st'>"Platelet"</span><span class='op'>)</span>
<span class='co'># Plot UMAP with new cluster IDs</span>
<span class='va'>b</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/DimPlot.html'>DimPlot</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>so</span>,
             label <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'><a href='https://satijalab.org/seurat/reference/SeuratTheme.html'>NoLegend</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='fu'>cowplot</span><span class='fu'>::</span><span class='fu'><a href='https://wilkelab.org/cowplot/reference/plot_grid.html'>plot_grid</a></span><span class='op'>(</span><span class='va'>a</span>,<span class='va'>b</span><span class='op'>)</span>
</code></pre>
</div>
<img src="class-2_files/figure-html5/rename-1.png" width="624" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Remember to save the new idents into meta.data</span>
<span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/Idents.html'>StashIdent</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>so</span>, save.name <span class='op'>=</span> <span class='st'>"cluster_name"</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h3 id="towards-a-or-20-more-automated-approach-of-identity-assignment">Towards a (or 20) more automated approach of identity assignment</h3>
<ol type="1">
<li><p>Using <code>Seurat</code>, inference from previous Seurat object (requires very similar Seurat object), see <a href="https://satijalab.org/seurat/articles/integration_mapping.html">Seurat Integration Vignette</a> for more details.</p></li>
<li><p>Using other classification packages such as <code>clustifyr</code> (disclosure: developed by the RBI).</p></li>
</ol>
<p><code>clustifyr</code> works by comparing the average gene expression in each cluster to a reference matrix that contains average gene signatures of reference cell types. The reference can be built from other single cell data, bulk rna-seq, microarray data, or other sources. Ranked Spearman correlation is used to compare the reference to the clusters. Only the highly variable genes are used for the correlation.</p>
<h3 id="cell-type-composition">cell type composition</h3>
<p>Insight into different samples can be gained from the proportion of cells that fall into each cell type.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>tab1</span> <span class='op'>&lt;-</span> <span class='va'>so</span><span class='op'>@</span><span class='va'>meta.data</span> <span class='op'>%&gt;%</span>
  <span class='fu'>group_by</span><span class='op'>(</span><span class='va'>orig.ident</span>, <span class='va'>cluster_name</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>tally</span><span class='op'>(</span><span class='op'>)</span> <span class='co'># counting up all combinations</span>
<span class='va'>tab1</span>
</code></pre>
</div>
<pre><code># A tibble: 16 x 3
# Groups:   orig.ident [2]
   orig.ident cluster_name     n
   &lt;fct&gt;      &lt;fct&gt;        &lt;int&gt;
 1 control    CD4+ T         586
 2 control    CD14+ Mono     241
 3 control    B              176
 4 control    CD8+ T         155
 5 control    FCGR3A+ Mono    83
 6 control    NK              74
 7 control    DC              12
 8 control    Platelet         7
 9 treated    CD4+ T         613
10 treated    CD14+ Mono     247
11 treated    B              174
12 treated    CD8+ T         125
13 treated    FCGR3A+ Mono    84
14 treated    NK              88
15 treated    DC              15
16 treated    Platelet         6</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>tab2</span> <span class='op'>&lt;-</span> <span class='va'>tab1</span> <span class='op'>%&gt;%</span> <span class='fu'>pivot_wider</span><span class='op'>(</span>names_from <span class='op'>=</span> <span class='va'>cluster_name</span>, values_from <span class='op'>=</span> <span class='va'>n</span><span class='op'>)</span> <span class='co'># spread out "long" into "wide" form</span>
<span class='va'>tab2</span>
</code></pre>
</div>
<pre><code># A tibble: 2 x 9
# Groups:   orig.ident [2]
  orig.ident `CD4+ T` `CD14+ Mono`     B `CD8+ T` `FCGR3A+ Mono`    NK
  &lt;fct&gt;         &lt;int&gt;        &lt;int&gt; &lt;int&gt;    &lt;int&gt;          &lt;int&gt; &lt;int&gt;
1 control         586          241   176      155             83    74
2 treated         613          247   174      125             84    88
# … with 2 more variables: DC &lt;int&gt;, Platelet &lt;int&gt;</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>tab3</span> <span class='op'>&lt;-</span> <span class='va'>tab1</span> <span class='op'>%&gt;%</span> <span class='fu'>group_by</span><span class='op'>(</span><span class='va'>orig.ident</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>mutate</span><span class='op'>(</span>n <span class='op'>=</span> <span class='va'>n</span><span class='op'>/</span><span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> <span class='co'># convert counts to proportions first</span>
  <span class='fu'>pivot_wider</span><span class='op'>(</span>names_from <span class='op'>=</span> <span class='va'>cluster_name</span>, values_from <span class='op'>=</span> <span class='va'>n</span><span class='op'>)</span>
<span class='va'>tab3</span>
</code></pre>
</div>
<pre><code># A tibble: 2 x 9
# Groups:   orig.ident [2]
  orig.ident `CD4+ T` `CD14+ Mono`     B `CD8+ T` `FCGR3A+ Mono`
  &lt;fct&gt;         &lt;dbl&gt;        &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;          &lt;dbl&gt;
1 control       0.439        0.181 0.132   0.116          0.0622
2 treated       0.453        0.183 0.129   0.0925         0.0621
# … with 3 more variables: NK &lt;dbl&gt;, DC &lt;dbl&gt;, Platelet &lt;dbl&gt;</code></pre>
</div>
<h2 id="other-things-to-do-with-marker-genes">Other things to do with marker genes</h2>
<ol type="1">
<li><p>Gene list to pathway activity score, via <code>Seurat::AddModuleScore</code> or <a href="https://github.com/aertslab/AUCell"><code>AUCell</code></a></p></li>
<li><p>If TF expression is too low for detection, consider <a href="https://github.com/aertslab/SCENIC"><code>SCENIC</code></a> for TF activity inference</p></li>
<li><p>Standard GO term enrichment tools <a href="https://biit.cs.ut.ee/gprofiler/page/r"><code>gProfiler2</code></a>, <a href="https://github.com/wjawaid/enrichR"><code>enrichR</code></a>, <a href="https://github.com/ctlab/fgsea%5D"><code>fgsea</code></a>, etc</p></li>
</ol>
<div class="sourceCode" id="cb14"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom"></div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
