---
title: <b style="font-size:45px;">Working with Multi-modal Data</b>
author: "Ryan Sheridan"
date: "August 15<sup>th</sup>, 2019"
output:
  distill::distill_article:
    toc: false  
---

```{r "Load packages", include = FALSE}

library(Seurat)
library(tidyverse)
library(cowplot)

knitr::opts_chunk$set(
  message   = FALSE,
  warning   = FALSE,
  echo      = TRUE,
  comment   = "#>",
  fig.align = "center"
)

```

## Experimental design

### CITE-seq reagents

Biolegend is the main company selling CITE-seq and cell hashing antibodies ([TotalSeq](https://www.biolegend.com/en-us/totalseq)). Biolegend reagents are divided into three product lines:

* TotalSeq-A: 3' gene expression, v2 and v3 chemistry
* TotalSeq-B: 3' gene expression, v3 chemistry
* TotalSeq-C: 5' gene expression and V(D)J

![](7_multimodal_files2/Totalseq_compatibility_figure_V1.png)

### Cell hashing reagents

[Cell hashing](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1603-1) allows for sample multiplexing and "super-loaded" runs with >10,000 captured cells. Super-loading results in higher doublet rates (~10% for 10,000 captured cells), but these doublets can be removed by identifying cell barcodes that are associated with multiple hashtag oligos.

![](7_multimodal_files2/cell_hashing_diagram.png)

Biolegend cell hashing reagents for human cells include a mix of two antibodies that recognize CD298 and Î²2 microglobulin. Mouse hashing antibodies recognize CD45 and H-2 MHC class I.

TotalSeq-A reagents use a different PCR handle for CITE-seq and cell hashing antibodies. This means two separate libraries have to be prepared. To ensure that the correct libraries are created, it is import to tell the sequencing core which types of antibodies were included in the experiment.

TotalSeq-C reagents use the same PCR handle for CITE-seq and cell hashing antibodies, which means that only a single library will be prepared. However, to ensure that the correct libraries are created the core should be notified of all reagents used for the experiment.

[MULTI-seq](https://www.nature.com/articles/s41592-019-0433-8) uses lipid- and cholesterol-modified oligonucleotides.

## Creating a Seurat object with multiple assays

### Loading counts matrices

The `Read10X` function can be used with the output directory generated by Cell Ranger. However, our count data is stored as comma-separated files, which we can load as data.frames and then convert to sparse matrices.

```{r "Load data"}

# Data URL
data_url <- "https://scrnaseq-workshop.s3-us-west-2.amazonaws.com"

# Function to import counts
import_counts <- function(file_name, file_url = data_url) {
  mtx <- file.path(file_url, file_name) %>%
    read_csv() %>%
    column_to_rownames("X1") %>%
    as.sparse()

  mtx
}

# Import gene expression matrix
rna_mtx <- import_counts("CITEseq_cDNA.csv.gz")

# Import CITE-seq matrix
adt_mtx <- import_counts("CITEseq_ADT.csv.gz")

rownames(adt_mtx) <- str_c("adt-", rownames(adt_mtx))
adt_mtx[, 1:10]

# Import HTO matrix
hto_mtx <- import_counts("CITEseq_HTO.csv.gz")

hto_mtx[, 1:10]

```

### Creating a Seurat object

When adding multiple assays to a Seurat object, we first must identify cell barcodes that are present in all of the datasets. If one of the assays has a different number of cell barcodes Seurat will throw an error.

```{r "Create Seurat object"}

# Get list of common cell barcodes
rna_bcs <- colnames(rna_mtx)
adt_bcs <- colnames(adt_mtx)
hto_bcs <- colnames(hto_mtx)

merged_bcs <- rna_bcs %>%
  intersect(adt_bcs) %>%
  intersect(hto_bcs)

# Create Seurat object
sobj <- rna_mtx[, merged_bcs] %>%
  CreateSeuratObject(min.cells = 5)

# Add CITE-seq and cell hashing data to Seurat object
sobj[["ADT"]] <- adt_mtx[, merged_bcs] %>%
  CreateAssayObject()

sobj[["HTO"]] <- hto_mtx[, merged_bcs] %>%
  CreateAssayObject()

sobj

```

## Demultiplexing hashed samples

### Normalizing HTO counts

To account for differences in antibody binding efficiencies, CITE-seq and cell hashing data can be normalized by performing a centered log-ratio transformation for each individual antibody.

```{r "Normalize HTO counts"}

# Normalize HTO counts
sobj <- sobj %>%
  NormalizeData(
    assay = "HTO",
    normalization.method = "CLR"
  )

```

### Sample demultiplexing and identification of doublets

To demultiplex hashed samples, the `HTODemux` function uses the normalized HTO counts for k-medoids clustering. This results in a cluster for each HTO A background signal is then calculated for each HTO using cells that are not present in the HTO-specific cluster. Outlier cells from this background signal are then classified as being "positive" for the HTO.

Cells that are positive for multiple HTOs are classified as doublets and cells that are not positive for any HTO are classified as "negative" cells.

The `HTODemux` function automatically adds several columns to the meta.data table. 

* HTO_classification: shows positive HTOs that were identified for the cell
* HTO_classification.global: singlet classification (singlet, doublet, negative)
* hash.ID: final HTO assignment including doublet and negative classifications

```{r "Demultiplex samples", fig.width = 8, fig.height = 6}

# Demultiplex samples
# By default HTODemux will look for the "HTO" assay
sobj <- sobj %>%
  HTODemux(positive.quantile = 0.97)

sobj@meta.data %>%
  head()

# Summarize cell classifications
table(sobj$HTO_classification.global)

# Create ridge plots showing HTO signal
sobj %>%
  RidgePlot(
    assay    = "HTO",
    features = rownames(hto_mtx),
    ncol     = 2
  )

```

### EXERCISE: Compare the number of cells with each hash.ID and calculate the doublet rate

```{r}

# Use the meta.data table to compare the number of cells with each hash.ID

# sobj@meta.data %>%



```

### ANSWER

```{r echo = FALSE, fig.width = 7, fig.height = 4}

# Calculate doublet rate
sobj@meta.data %>%
  group_by(hash.ID) %>% 
  summarize(fract = n() / nrow(.))

# Create bar graphs comparing cell count for each sample
HTO_bars_1 <- sobj@meta.data %>%
  rownames_to_column("cell_id") %>%
  
  ggplot(aes(hash.ID, fill = hash.ID)) +
  geom_bar() +
  labs(y = "Cell Count") +
  cowplot::theme_cowplot() +
  theme(
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.text.x     = element_text(hjust = 1, angle = 45)
  )

# Create stacked bar graph showing fraction of cells
HTO_bars_2 <- sobj@meta.data %>%
  rownames_to_column("cell_id") %>%
  group_by(hash.ID) %>%
  summarize(hash_frac = n() / nrow(.)) %>%
  
  ggplot(aes("CITE-seq", hash_frac, fill = hash.ID)) +
  geom_bar(stat = "identity", color = "white", size = 0.5) +
  labs(y = "Fraction of Cells") +
  cowplot::theme_cowplot() +
  theme(
    legend.title = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x  = element_blank(),
    axis.ticks.x = element_blank()
  )

# Combine plots
plot_grid(
  HTO_bars_1, HTO_bars_2,
  nrow        = 1,
  rel_widths  = c(0.5, 0.5),
  rel_heights = c(1, 0.8)
)

```

## Filtering data and assessing quality

### Assessing data quality

```{r "Create RNA violin plots", fig.width = 8, fig.height = 4}

# Add mitochondrial percentage to meta.data table
sobj <- sobj %>%
  PercentageFeatureSet(
    assay    = "RNA",
    pattern  = "^MT-", 
    col.name = "percent_mito"
  )

# Create violin plots for gene expression data
sobj %>%
  VlnPlot(
    features = c("nCount_RNA", "nFeature_RNA", "percent_mito"), 
    ncol     = 3,
    pt.size  = 0.25
  )

```

```{r "Create ADT and HTO violin plots", fig.width = 6, fig.height = 4}

# Aim to sequence CITE-seq libraries at 2k-5k reads/cell, cell hashing 1k-2k reads/cell

sobj %>%
  VlnPlot(
    features = c("nCount_ADT", "nCount_HTO"),
    ncol    = 2,
    pt.size = 0.25,
    log     = T
  )

```

```{r "Filter cells"}

# Filter cells based on HTO class, number of genes, and percent mito UMIs
filt_so <- sobj %>%
  subset(
    nFeature_RNA > 250 &   # Remove cells with < 250 detected genes
    nFeature_RNA < 2500 &  # Remove cells with > 2500 detected genes (could be doublets)
    percent_mito < 15 &    # Remove cells with > 0.15 mito/total reads
    HTO_classification.global == "Singlet"
  )

filt_so

# Rename cell identities with sample names
filt_so <- filt_so %>%
  RenameIdents(
    "HTO28" = "PBMC-1",
    "HTO29" = "PBMC-2",
    "HTO30" = "PBMC-3",
    "HTO44" = "PBMC-4"
  )

# Add sample names to meta.data table
filt_so <- filt_so %>%
  AddMetaData(
    metadata = Idents(filt_so),
    col.name = "sample"
  )

filt_so@meta.data %>%
  head()

```

## Normalizing gene expression and antibody data

Like the cell hashing data, CITE-seq counts can be normalized by performing a centered log-ratio transformation.

```{r "Normalize data", results = FALSE}

# Normalize expression data
filt_so <- filt_so %>%
  NormalizeData(
    normalization.method = "LogNormalize",
    verbose = FALSE
  ) %>%
  FindVariableFeatures(verbose = FALSE) %>%
  ScaleData(verbose = FALSE)

# Normalize CITE-seq data
filt_so <- filt_so %>%
  NormalizeData(
    assay = "ADT",
    normalization.method = "CLR"   # Centered log-ratio normalization
  ) %>%
  ScaleData(
    assay   = "ADT",
    verbose = FALSE
  )

```

## Clustering cells based on gene expression

### Perform pricipal component analysis

```{r "Perform PCA", fig.width = 10, fig.height = 4}

# Perform PCA
filt_so <- filt_so %>%
  RunPCA(verbose = FALSE)       # By default only variable features are used for PCA

# Plot standard deviations of principal components
elbow_plot <- filt_so %>%
  ElbowPlot(ndims = 50)

# Create scatter plot comparing PC-1 and PC-2
PCA_plot <- filt_so %>%
  DimPlot(reduction = "pca")

plot_grid(
  elbow_plot, PCA_plot,
  nrow       = 1,
  rel_widths = c(0.45, 0.55)
)

```

### Cluster cells using gene expression data

```{r "Cluster cells and run UMAP"}

# Cluster cells
filt_so <- filt_so %>%
  FindNeighbors(dims = 1:20) %>%
  FindClusters(
    resolution = 0.6,
    verbose    = FALSE
  )

# Run UMAP
# Use same PCs as clustering, by default reduction = "pca"
filt_so <- filt_so %>%
  RunUMAP(dims = 1:20)

# Add RNA clusters to meta.data
filt_so <- filt_so %>%
  AddMetaData(
    metadata = Idents(filt_so),
    col.name = "RNA_clusters"
  )

# Create UMAP
filt_so %>%
  DimPlot(reduction = "umap")

```

## Clustering cells based on antibody signal

### Cluster cells using antibody signal

Since there are only a few antibodies and not many dimensions, instead of performing PCA we can just create a distance matrix to use for clustering. To do this we can use the `dist` function which computes distances between the rows. To calculate distances between cells, we need to transpose the matrix.

We can then use the resulting distance matrix as input to the `FindNeighbors` function, which creates a Shared Nearest Neighbor (SNN) graph that is used for clustering.

```{r "Cluster cells using antibody data", results = FALSE}

# Create a standard euclidean distance matrix for clustering
adt_data <- filt_so %>%
  GetAssayData(
    assay = "ADT", 
    slot = "data"
  )

adt_data[, 1:5]

adt_dist <- adt_data %>%
  t() %>%
  dist()
  
# Find clusters using distance matrix
adt_graphs <- FindNeighbors(adt_dist)

filt_so[["ADT_snn"]] <- adt_graphs$snn

filt_so <- filt_so %>%
  FindClusters(
    resolution = 0.2,
    graph.name = "ADT_snn"
  )

# Add ADT clusters to meta.data
filt_so <- filt_so %>%
  AddMetaData(
    metadata = Idents(filt_so),
    col.name = "ADT_clusters"
  )

```

### Run UMAP using the antibody signal

```{r "Run UMAP for antibody data"}

# Run UMAP
filt_so <- filt_so %>%
  RunUMAP(
    reduction.name = "adt_umap",
    reduction.key  = "ADTUMAP_",
    graph          = "ADT_snn"
  )

# Create UMAP
filt_so %>%
  DimPlot(
    reduction = "adt_umap", 
    group.by  = "sample",
    # split.by = "sample",
    ncol = 2
  )

```

### Identify marker proteins

```{r echo = FALSE}

# Identify differentially expressed proteins for each cluster
ADT_markers <- filt_so %>%
  FindAllMarkers(
    assay    = "ADT",
    only.pos = T
  )

ADT_markers

```

## Visualizing antibody signal

### Overlay antibody signal on UMAPs

```{r "Overlay antibody signal on UMAPs", fig.width = 10, fig.height = 8}

# Set active.assay to ADT
filt_so@active.assay <- "ADT"

# Overlay antibody signal on gene expression UMAP
filt_so %>%
  FeaturePlot(
    reduction = "umap",
    features  = c("adt-CD4", "CD4", "adt-CD8", "CD8A")
  )

# Overlay antibody signal on antibody UMAP
filt_so %>%
  FeaturePlot(
    reduction = "adt_umap",
    features  = c("adt-CD4", "CD4", "adt-CD8", "CD8A")
  )

```

### Ridge plots

```{r "Create ridge plots", fig.width = 10, fig.height = 6}

# Create ridge plot
filt_so %>%
  RidgePlot(features = c(
    "adt-CD14", "adt-CD45",
    "adt-CD19", "adt-CD3",  
    "adt-CD4",  "adt-CD8"
  ))

```

### Heatmaps

```{r "Create heatmaps"}

filt_so %>%
  DoHeatmap(
    features = rownames(filt_so), 
    angle    = 0
  ) + 
  NoLegend()

```

### Violin plots

```{r "Create violinplots", fig.width = 8, fig.height = 8}

filt_so %>%
  VlnPlot(
    features = c("adt-CD4", "rna_CD4", "adt-CD19", "rna_CD19"),
    ncol     = 2
  )

```

## Classifying cells based on antibody signal

### Identify CD19+ cells

```{r "Filter using antibody signal"}

# Plot CD3 and CD19 signal
filt_so %>%
  FeatureScatter("adt-CD3", "adt-CD19")

# Identify CD19+ cells using antibody signal
CD19_cells <- filt_so %>%
  subset(`adt-CD19` > 2.5 & `adt-CD3` < 1) %>%
  Cells()

CD19_cells %>%
  head()

# Set cell identities
labeled_so <- filt_so %>%
  SetIdent(value = "Other") %>%
  SetIdent(
    value = "CD19+",
    cells = CD19_cells
  )

labeled_so@active.ident %>%
  head()

# Label UMAP with new cell identities
labeled_so %>%
  DimPlot(reduction = "adt_umap")

```

### Filter cells using CellSelector()

```{r "Filter using CellSelector()", eval = FALSE}

# Plot CD3 and CD19 signal
CD19_plot <- filt_so %>%
  FeatureScatter("adt-CD3", "adt-CD19")

CD19_plot

# Identify CD19+ cells using antibody signal
labeled_so <- filt_so %>%
  SetIdent(value = "Other")

labeled_so <- CellSelector(
  plot   = CD19_plot,
  object = labeled_so,
  ident  = "B cells"
)

labeled_so@active.ident %>%
  head()

# Label UMAP with new cell identities
labeled_so %>%
  DimPlot(reduction = "adt_umap")

```

### EXERCISE: Identify CD4+ and CD8+ cells

```{r}

# Plot CD4 and CD8 signal

# filt_so %>%
#   FeatureScatter("adt_CD4", "adt_CD8")

# Identify CD4+ and CD8+ cells using antibody signal

# Set cell identities

# Label UMAP with new cell identities



```

### ANSWER

```{r echo = FALSE}

# Compare CD3 and CD19 signal
filt_so %>%
  FeatureScatter("adt-CD4", "adt-CD8")

# Identify CD4+ and CD8+ cells using antibody signal
CD4_cells <- labeled_so %>%
  subset(`adt-CD4` > 2.5 & `adt-CD8` < 1) %>%
  Cells()

CD8_cells <- labeled_so %>%
  subset(`adt-CD8` > 1.5 & `adt-CD4` < 1) %>%
  Cells()

# Set cell identities
labeled_so <- labeled_so %>%
  SetIdent(
    value = "CD4+", 
    cells = CD4_cells
  ) %>%
  SetIdent(
    value = "CD8+", 
    cells = CD8_cells
  )

# Add cell identities to meta.data table
labeled_so <- labeled_so %>%
  AddMetaData(
    metadata = Idents(labeled_so),
    col.name = "cell_label"
  )

# Label UMAP with new cell identities
labeled_so %>%
  DimPlot(reduction = "adt_umap")

```

## Viewing results with the UCSC Cell Browser

The UCSC [Cell Browser](http://www.genome.ucsc.edu/singlecell.html) allows you to easily explore and share single-cell data. With the Cell Browser you can:

* View t-SNE or UMAP projections
* Color cells by metadata and gene expression
* View cluster marker genes
* Rename clusters and add custom annotations to selected sets of cells

### Merge gene expression and antibody matrices

```{r "Merge RNA and ADT matrices", eval = FALSE}

# Combine RNA and ADT matrices
merged_so   <- labeled_so
RNA_data    <- merged_so@assays$RNA@data
ADT_data    <- merged_so@assays$ADT@data
merged_data <- rbind(RNA_data, ADT_data)

# Add merged matrix to Seurat object
merged_so@assays$RNA@data <- merged_data

# Set active assay
merged_so@active.assay <- "RNA"

```

### Create Cell Browser files

```{r "Create gene expression Cell Browser files", eval = FALSE}

# Create Cell Browser directories for gene expression data
dir.create(
  path = "cellbrowser/RNA",
  recursive = T
)

# Create Cell Browser files for gene expression data
merged_so %>%
  ExportToCellbrowser(
    dir          = "cellbrowser/RNA",
    reductions   = "umap",
    dataset.name = "RNA",

    sample       = "Sample",
    RNA_clusters = "Cluster",
    ADT_clusters = "ADT cluster",
    cell_label   = "Cell label",
    
    nCount_RNA   = "RNA UMI count",
    nFeature_RNA = "Gene count",
    percent_mito = "Percent mito UMIs",
    nCount_ADT   = "ADT UMI count",
    nFeature_ADT = "Antibody count",
    nCount_HTO   = "HTO UMI count",
    nFeature_HTO = "HTO count"
  )

```

```{r "Create ADT Cell Browser files", eval = FALSE}

# Create Cell Browser directories for antibody data
dir.create(
  path = "cellbrowser/ADT",
  recursive = T
)

# Export marker genes
ADT_markers %>%
  rename(score = p_val_adj) %>%
  select(cluster, gene, score) %>%
  write_tsv("cellbrowser/ADT/ADT_markers.tsv")

# Create Cell Browser files for antibody data
merged_so %>%
  ExportToCellbrowser(
    dir          = "cellbrowser/ADT",
    reductions   = "adt_umap",
    dataset.name = "ADT",
    markers.file = "cellbrowser/ADT/ADT_markers.tsv",
    
    sample       = "Sample",
    ADT_clusters = "Cluster",
    RNA_clusters = "RNA cluster",
    cell_label   = "Cell label",
    
    nCount_RNA   = "RNA UMI count",
    nFeature_RNA = "Gene count",
    percent_mito = "Percent mito UMIs",
    nCount_ADT   = "ADT UMI count",
    nFeature_ADT = "Antibody count",
    nCount_HTO   = "HTO UMI count",
    nFeature_HTO = "HTO count"
  )

```

### Build Cell Browser session

```{bash "Build Cell Browser session", eval = FALSE}

mkdir -p cellbrowser/html

cbBuild \
  -i cellbrowser/RNA/cellbrowser.conf \
  -i cellbrowser/ADT/cellbrowser.conf \
  -o cellbrowser/html \
  -p 8888

```


