<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

  <!--radix_placeholder_meta_tags-->
  <title>Single-cell RNA-seq Workshop: &lt;b style="font-size:45px;"&gt;Introduction to scRNA-seq Analysis&lt;/b&gt;</title>
  
  
  <link rel="icon" type="image/png" href="img/logo.png"/>
  
  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2019-08-13"/>
  <meta property="article:created" itemprop="dateCreated" content="2019-08-13"/>
  <meta name="article:author" content="Rui Fu"/>
  
  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="Single-cell RNA-seq Workshop: &lt;b style=&quot;font-size:45px;&quot;&gt;Introduction to scRNA-seq Analysis&lt;/b&gt;"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="Single-cell RNA-seq Workshop"/>
  
  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="Single-cell RNA-seq Workshop: &lt;b style=&quot;font-size:45px;&quot;&gt;Introduction to scRNA-seq Analysis&lt;/b&gt;"/>
  <meta property="twitter:site" content="@rnabioco"/>
  
  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","author","date","output"]}},"value":[{"type":"character","attributes":{},"value":["<b style=\"font-size:45px;\">Introduction to scRNA-seq Analysis<\/b>"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name"]}},"value":[{"type":"character","attributes":{},"value":["Rui Fu"]}]}]},{"type":"character","attributes":{},"value":["August 13<sup>th<\/sup>, 2019"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["html_document"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["toc","toc_depth","self_contained"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"integer","attributes":{},"value":[3]},{"type":"logical","attributes":{},"value":[true]}]}]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  
  <script type="application/javascript">
  
    window.headroom_prevent_pin = false;
  
    window.document.addEventListener("DOMContentLoaded", function (event) {
  
      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');
  
      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });
  
      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });
  
      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>
  
  <style type="text/css">
  
  /* Theme (user-documented overrideables for nav appearance) */
  
  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #455a64;
    font-size: 15px;
    font-weight: 300;
  }
  
  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }
  
  .distill-site-nav a:hover {
    color: white;
  }
  
  @media print {
    .distill-site-nav {
      display: none;
    }
  }
  
  .distill-site-header {
  
  }
  
  .distill-site-footer {
  
  }
  
  
  /* Site Header */
  
  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }
  
  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }
  
  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }
  
  
  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }
  
  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }
  
  .distill-site-header .title {
    font-size: 18px;
  }
  
  .distill-site-header .logo {
    padding: 0;
  }
  
  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }
  
  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }
  
  
  
  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }
  
  
  .distill-site-header .nav-toggle {
    display: none;
  }
  
  .nav-dropdown {
    display: inline-block;
    position: relative;
  }
  
  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }
  
  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }
  
  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .nav-dropdown-active {
    display: block;
  }
  
  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }
  
  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }
  
  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }
  
  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }
  
  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }
  
  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }
  
  
  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative;}
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }
  
  /* Site Footer */
  
  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }
  
  /* Headroom */
  
  d-title {
    padding-top: 6rem;
  }
  
  @media print {
    d-title {
      padding-top: 4rem;
    }
  }
  
  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }
  
  .headroom--transition {
    transition: all .4s ease-in-out;
  }
  
  .headroom--unpinned {
    top: -100px;
  }
  
  .headroom--pinned {
    top: 0;
  }
  
  </style>
  
  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->
  
  <style type="text/css">
  
  body {
    background-color: white;
  }
  
  .pandoc-table {
    width: 100%;
  }
  
  .pandoc-table>caption {
    margin-bottom: 10px;
  }
  
  .pandoc-table th:not([align]) {
    text-align: left;
  }
  
  .pagedtable-footer {
    font-size: 15px;
  }
  
  .html-widget {
    margin-bottom: 2.0em;
  }
  
  .l-screen-inset {
    padding-right: 16px;
  }
  
  .l-screen .caption {
    margin-left: 10px;
  }
  
  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .shaded .shaded-content {
    background: white;
  }
  
  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }
  
  .hidden {
    display: none !important;
  }
  
  d-article {
    padding-bottom: 30px;
  }
  
  d-appendix {
    padding-top: 30px;
  }
  
  d-article>p>img {
    width: 100%;
  }
  
  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }
  
  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }
  
  /* CSS for table of contents */
  
  .d-toc {
    color: rgba(0,0,0,0.8);
    font-size: 0.8em;
    line-height: 1em;
  }
  
  .d-toc-header {
    font-size: 0.6rem;
    font-weight: 400;
    color: rgba(0, 0, 0, 0.5);
    text-transform: uppercase;
    margin-top: 0;
    margin-bottom: 1.3em;
  }
  
  .d-toc a {
    border-bottom: none;
  }
  
  .d-toc ul {
    padding-left: 0;
  }
  
  .d-toc li>ul {
    padding-top: 0.8em;
    padding-left: 16px;
    margin-bottom: 0.6em;
  }
  
  .d-toc ul,
  .d-toc li {
    list-style-type: none;
  }
  
  .d-toc li {
    margin-bottom: 0.9em;
  }
  
  .d-toc-separator {
    margin-top: 20px;
    margin-bottom: 2em;
  }
  
  .d-article-with-toc {
    border-top: none;
    padding-top: 0;
  }
  
  
  
  /* Tweak code blocks (note that this CSS is repeated above in an injection
     into the d-code shadow dom) */
  
  d-code {
    overflow-x: auto !important;
  }
  
  pre.d-code code.d-code {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }
  
  pre.text-output {
  
    font-size: 12px;
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;
  
    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;
  
    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }
  
  @media(min-width: 768px) {
  
  d-code {
    overflow-x: visible !important;
  }
  
  pre.d-code code.d-code  {
      padding-left: 18px;
      font-size: 14px;
  }
  pre.text-output {
    font-size: 14px;
  }
  }
  
  /* Figure */
  
  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }
  
  .figure img {
    width: 100%;
  }
  
  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }
  
  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }
  
  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }
  
  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }
  
  
  
  /* Tweak 1000px media break to show more text */
  
  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }
  
    .grid {
      grid-column-gap: 16px;
    }
  
    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }
  
  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }
  
    .grid {
      grid-column-gap: 32px;
    }
  }
  
  
  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */
  
  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }
  
  
  /* Social footer */
  
  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }
  
  .disqus-comments {
    margin-right: 30px;
  }
  
  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }
  
  #disqus_thread {
    margin-top: 30px;
  }
  
  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }
  
  .article-sharing a:hover {
    border-bottom: none;
  }
  
  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }
  
  .subscribe p {
    margin-bottom: 0.5em;
  }
  
  
  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }
  
  
  /* Improve display for browsers without grid (IE/Edge <= 15) */
  
  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }
  
  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }
  
  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }
  
  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }
  
  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }
  
  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }
  
  
  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }
  
  .downlevel .footnotes ol {
    padding-left: 13px;
  }
  
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }
  
  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }
  
  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }
  
  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;
  
    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;
  
    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }
  
  </style>
  
  <script type="application/javascript">
  
  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }
  
  // show body when load is complete
  function on_load_complete() {
  
    // set body to visible
    document.body.style.visibility = 'visible';
  
    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }
  
    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }
  
  function init_distill() {
  
    init_common();
  
    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);
  
    // create d-title
    $('.d-title').changeElementType('d-title');
  
    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);
  
    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();
  
    // move posts container into article
    $('.posts-container').appendTo($('d-article'));
  
    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');
  
    // create d-bibliography
    var bibliography = $('<d-bibliography></d-bibliography>');
    $('#distill-bibliography').wrap(bibliography);
  
    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;
  
    // replace citations with <d-cite>
    $('.citation').each(function(i, val) {
      appendix = true;
      var cites = $(this).attr('data-cites').split(" ");
      var dt_cite = $('<d-cite></d-cite>');
      dt_cite.attr('key', cites.join());
      $(this).replaceWith(dt_cite);
    });
    // remove refs
    $('#refs').remove();
  
    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();
  
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-toc a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });
  
    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');
  
    // replace code blocks with d-code
    $('pre>code').each(function(i, val) {
      var code = $(this);
      var pre = code.parent();
      var clz = "";
      var language = pre.attr('class');
      if (language) {
        // map unknown languages to "clike" (without this they just dissapear)
        if ($.inArray(language, ["bash", "clike", "css", "go", "html",
                                 "javascript", "js", "julia", "lua", "markdown",
                                 "markup", "mathml", "python", "svg", "xml"]) == -1)
          language = "clike";
        language = ' language="' + language + '"';
        var dt_code = $('<d-code block' + language + clz + '></d-code>');
        dt_code.text(code.text());
        pre.replaceWith(dt_code);
      } else {
        code.addClass('text-output').unwrap().changeElementType('pre');
      }
    });
  
    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {
  
      // capture layout
      var layout = $(this).attr('data-layout');
  
      // apply layout to markdown level block elements
      var elements = $(this).children().not('d-code, pre.text-output, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });
  
  
      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });
  
    // load distill framework
    load_distill_framework();
  
    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {
  
      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;
  
      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');
  
      // table of contents
      if (have_authors) // adjust border if we are in authors
        $('.d-toc').parent().addClass('d-article-with-toc');
  
      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');
  
      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }
  
      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');
  
      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");
  
      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }
  
       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }
  
      // inject pre code styles (can't do this with a global stylesheet b/c a shadow root is used)
      $('d-code').each(function(i, val) {
        var style = document.createElement('style');
        style.innerHTML = 'pre code { padding-left: 10px; font-size: 12px; border-left: 2px solid rgba(0,0,0,0.1); } ' +
                          '@media(min-width: 768px) { pre code { padding-left: 18px; font-size: 14px; } }';
        if (this.shadowRoot)
          this.shadowRoot.appendChild(style);
      });
  
      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();
  
      // clear polling timer
      clearInterval(tid);
  
      // show body now that everything is ready
      on_load_complete();
    }
  
    var tid = setInterval(distill_post_process, 50);
    distill_post_process();
  
  }
  
  function init_downlevel() {
  
    init_common();
  
     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));
  
    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
  
    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();
  
    // remove toc
    $('.d-toc-header').remove();
    $('.d-toc').remove();
    $('.d-toc-separator').remove();
  
    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });
  
  
    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);
  
    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);
  
    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();
  
    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();
  
    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));
  
    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });
  
    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));
  
    $('body').addClass('downlevel');
  
    on_load_complete();
  }
  
  
  function init_common() {
  
    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};
  
        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });
  
        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);
  
    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});
  
    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });
  
      }
    });
  
    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });
  
    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }
  
    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');
  
    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");
  
    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();
  
    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }
  
  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });
  
  </script>
  
  <!--/radix_placeholder_distill-->
  <script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <style type="text/css">
  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #000000;
      font-size: 15px;
    font-weight: 300;
  }
  
  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }
  
  .distill-site-nav a:hover {
    color: white;
    text-decoration:underline;
  }
  
  .distill-site-header {
  }
  
  .distill-site-footer {
  }
  
  @media print {
    .distill-site-nav {
      display: none;
    }
  }
  
  .l-body {
    display: none;
  }
  
  @media screen and (min-width: 768px) {
    .l-body {display: inline;}
  }
  
  </style>
  <!--/radix_placeholder_site_in_header-->

  <link rel="stylesheet" href="styles.css" type="text/css"/>

</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"<b style=\"font-size:45px;\">Introduction to scRNA-seq Analysis<\/b>","authors":[{"author":"Rui Fu","authorURL":"#","affiliation":"&nbsp;","affiliationURL":"#"}],"publishedDate":"2019-08-13T00:00:00.000-06:00","citationText":"Fu, 2019"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a href="index.html" class="title">Single-cell RNA-seq Workshop</a>
<a href="1_intro.html">Introduction</a>
<a href="2_filtering_QC.html">Quality Control</a>
<a href="3_norm_viz_clustering.html">Clustering</a>
<a href="4_markers.html">Cell Type Annotation</a>
<a href="5_trajectories.html">Pseudotime</a>
<a href="6_alignment.html">Dataset Alignment</a>
<a href="7_multimodal.html">Multi-modal Data</a>
</div>
<div class="nav-right">
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Introduction to scRNA-seq Analysis</h1>

</div>

<div class="d-byline">
  
  Rui Fu
  
<br/>August 13<sup>th</sup>, 2019
</div>

<div class="d-article">
<div class="layout-chunk" data-layout="l-body">

</div>
<h2 id="intro-to-rbi">intro to RBI</h2>
<p><strong><a href="mailto:rbi.fellows@ucdenver.edu" class="email">rbi.fellows@ucdenver.edu</a></strong>,</p>
<p>RC1 South room9101</p>
<p><a href="https://rnabioco.github.io/rbi-project-summary/office-hours.html">office hours on Thursday afternoons</a></p>
<h2 id="outline-of-the-workshop">outline of the workshop</h2>
<ul>
<li>Prereq: check all required packages are installed</li>
<li>Day 1 (pm):
<ul>
<li>Say Hi</li>
<li>Intro to scRNAseq</li>
<li>Data preprocessing</li>
<li>Import into R</li>
<li>Quality control</li>
</ul></li>
<li>Day 2 (am):
<ul>
<li>Projections</li>
<li>Clustering</li>
<li>Finding markers</li>
<li>Annotating cell types</li>
</ul></li>
<li>Day 2 (pm):
<ul>
<li>Pseudotime analysis</li>
<li>Dataset alignment</li>
</ul></li>
<li>Day 3 (am):
<ul>
<li>Multimodal data (CITE-seq, hashing)</li>
<li>Miscellaneous</li>
</ul></li>
<li>Day 3 (pm) : working through your own data</li>
<li>Survey after the workshop for feedback</li>
</ul>
<h2 id="limitations-of-scrnaseq">limitations of scRNAseq</h2>
<figure>
<img src="1_intro_files2/cd4vs8.png" alt="from seurat pbmc example" width="600" /><figcaption>from seurat pbmc example</figcaption>
</figure>
<ol type="1">
<li><p>dropout effect, only a small portion of the mRNAs from each cell is captured (see: GFP and other supposed markers)</p></li>
<li><p>low number of detectable genes, might not detect low expressing genes at all (see: TFs)</p></li>
<li><p>little info outside of gene counts (see: no tail-seq, isoform info)</p></li>
<li><p>discrepancies between RNA and protein, especially surface proteins (see: CD4+ vs CD8+)</p></li>
<li><p>constantly evolving chemistry and bioinformatics tools (see: 10x v2/v3/v3.1, <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview#cell_calling">bad cell calling in cellrangerV2</a>)</p></li>
</ol>
<h2 id="experimental-design-informatics-corrections">experimental design &gt;&gt; informatics corrections</h2>
<figure>
<img src="1_intro_files2/10xV3_model.png" alt="cell isolation process for 10x" width="400" /><figcaption>cell isolation process for 10x</figcaption>
</figure>
<ol type="1">
<li><p>too few or too many cells - 1000 cell lower limit for capture, and waste of money</p></li>
<li><p>doublets - 10000 cells ~ 10% doublet rate<br />
        (try: <a href="https://github.com/AllonKleinLab/scrublet">scrublet</a>)</p></li>
<li><p>cell death - worse RNA capture, potentially different expression profile<br />
        (try: MT-RNA cutoff, regress out MT-RNA score)</p></li>
<li><p>lysed/free RNA - background for all cells, interferes with clustering and markers<br />
        (try: <a href="https://github.com/constantAmateur/SoupX">SoupX</a>)</p></li>
</ol>
<figure>
<img src="1_intro_files2/contaminant.png" alt="example data with lysed red blood cells" width="600" /><figcaption>example data with lysed red blood cells</figcaption>
</figure>
<ol start="5" type="1">
<li><p>different sex of mice in samples - makes sample comparisons harder<br />
        (try: ignore known sex-dependent genes)</p></li>
<li><p>batch effects - still preferable to have all, or at least the direct comparison samples, captured at the same time<br />
        (try: various alignment methods)</p></li>
<li><p>n = ? - no consensus in the field, but aggregating 2 or more biological repeats, or hash multiple samples at the same time, should be valid</p></li>
</ol>
<h2 id="single-cell-rna-sequencing-overview">single cell RNA sequencing overview</h2>
<p>cell suspension -&gt; cDNA library -&gt; fastq file -&gt; gene expression matrix -&gt;</p>
<p>    filtering -&gt; preprocessing -&gt; dimension reduction -&gt; clustering -&gt; identity and markers -&gt; pseudotime</p>
<h3 id="sequencing-methods">1. sequencing methods</h3>
<p>main platforms: 10x vs Smart-seq2 vs sci-RNA-seq3</p>
<div class="layout-chunk" data-layout="l-body">
<table>
<colgroup>
<col style="width: 12%" />
<col style="width: 18%" />
<col style="width: 13%" />
<col style="width: 16%" />
<col style="width: 14%" />
<col style="width: 24%" />
</colgroup>
<thead>
<tr class="header">
<th>method</th>
<th>cell isolation</th>
<th>coverage</th>
<th>saturation read #</th>
<th>gene detection</th>
<th>throughput and cost</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>10x Chromium</td>
<td>droplet-based</td>
<td>polyA+ 3’ only</td>
<td>10^5 per cell</td>
<td>~2000 per cell</td>
<td>~5k cells, low cost</td>
</tr>
<tr class="even">
<td>Smart-seq2</td>
<td>FACS-sorted</td>
<td>full length</td>
<td>10^6 per cell</td>
<td>~4000 per cell</td>
<td>manual pipetting, high cost</td>
</tr>
<tr class="odd">
<td>sci-RNA-seq3</td>
<td>combinatorial index</td>
<td>polyA+ 3’ only</td>
<td>10^5 per cell</td>
<td>~1000 per cell</td>
<td>~1m cells, lowest cost per</td>
</tr>
</tbody>
</table>
</div>
<p>variations within 10x system: standard 3’, 5’ + V(D)J, antibody hashing, CITE-seq, variant-calling (very different sample preps, may require up to 4 library preps per sample)</p>
<h3 id="alignment-pipelines">2. alignment pipelines</h3>
<p><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/installation">cellranger</a> performs alignment, filtering, UMI counting, clustering, and gene expression analysis.</p>
<p>Alternatives such as <a href="https://salmon.readthedocs.io/en/latest/alevin.html">Alevin</a>, <a href="https://github.com/alexdobin/STAR/releases">STARsolo</a>, <a href="https://www.kallistobus.tools/">Kallisto</a></p>
<div class="layout-chunk" data-layout="l-body">
<pre class="bash"><code>
# in terminal
cellranger count --id=123 \ 
                 --transcriptome=/refdata-cellranger-GRCh38-3.0.0 \ 
                 # build new transcriptome if you have GFP/RFP/transgene
                 --fastqs=/home/runs/HAT7ADXX/outs/fastq_path \ 
                 # a list of fastqs, a folder, or pass table via csv file
                 --sample=mysample 
                 # will need additional arguments for feature barcoding</code></pre>
</div>
<p>output folder structure</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
KO_1_CDNA
|-- KO_1_CDNA.mri.tgz
|-- SC_RNA_COUNTER_CS
|   |-- CLOUPE_PREPROCESS
|   |-- EXPAND_SAMPLE_DEF
|   |-- SC_RNA_COUNTER
|   `-- fork0
|-- _cmdline
|-- _filelist
|-- _finalstate
|-- _invocation
|-- _jobmode
|-- _log
|-- _mrosource
|-- _perf
|-- _sitecheck
|-- _tags
|-- _timestamp
|-- _uuid
|-- _vdrkill
|-- _versions
`-- outs
    |-- analysis
    |-- cloupe.cloupe # cloupe file for browser view
    |-- filtered_feature_bc_matrix # use this fold for seurat
    |-- filtered_feature_bc_matrix.h5
    |-- metrics_summary.csv
    |-- molecule_info.h5
    |-- possorted_genome_bam.bam
    |-- possorted_genome_bam.bam.bai
    |-- raw_feature_bc_matrix # or this for seurat
    |-- raw_feature_bc_matrix.h5
    `-- web_summary.html # qc summary</code></pre>
</div>
<p>~ 5-8 hours per sample on biochem department cluster “Bodhi”</p>
<p>theoretically can be ran locally on linux, but will require at least 32GB of RAM</p>
<p>other campus options include Rosalind, AWS</p>
<p>cellranger output <a href="http://cf.10xgenomics.com/samples/cell-exp/3.1.0/5k_pbmc_protein_v3/5k_pbmc_protein_v3_web_summary.html">html</a> and <a href="http://cf.10xgenomics.com/samples/cell-exp/3.1.0/5k_pbmc_protein_v3/5k_pbmc_protein_v3_cloupe.cloupe">loupe</a> files</p>
<h3 id="key-things-to-check-for-in-the-initial-output-html-report">3. key things to check for in the initial output html report</h3>
<ol type="1">
<li>number of cells close to expectations? reasonable elbow plot?</li>
<li>reads per cell (this is merely calculated as reads/cells, so somewhat inaccurate) - ideally ~50-100k</li>
<li>genes per cell, ideally ~2000</li>
<li>sequencing saturation?</li>
<li>any structure in the dimension reduction tSNE?</li>
</ol>
<h3 id="downstream-processing-in-r-or-python">4. downstream processing in R or python</h3>
<ol type="1">
<li>more accurate cell calling</li>
<li>more stringent filtering</li>
<li>regress out unwanted sources of variance</li>
<li>appropriate choice of normalization method</li>
<li>UMAP instead of tSNE projections, and fine-tuning</li>
<li>fine-tuning of dataset alignment methods and parameters (by default cellranger does align during aggregation)</li>
<li>find marker genes and assign cluster identities (may merge some)</li>
<li>more flexible visualizations</li>
<li>GO term analysis and other things</li>
<li>export to a hosted browser, similar to cloupe</li>
</ol>
<h2 id="reading-and-making-umi-barcode-elbow-plots">reading and making UMI-barcode elbow plots</h2>
<figure>
<img src="1_intro_files2/10xV3_seq.png" alt="model of final 10x library" width="600" /><figcaption>model of final 10x library</figcaption>
</figure>
<p>The total UMI (unique molecular identifier - represent each transcript) of a cell barcode is used to rank the barcodes determine the UMI threshold for signal vs noise. A plot is generated for cellranger html output, but will also be useful in other situations like hashing and CITE-seq.</p>
<figure>
<img src="1_intro_files2/elbowcomp.png" alt="example of good and bad data" width="400" /><figcaption>example of good and bad data</figcaption>
</figure>
<p>for a standard 10x scRNAseq run:</p>
<p>        expected range of x axis (barcode) : ~ 10^5 (if using ggplot to visualize, filter)</p>
<p>        expected inflection point of x axis (cell number): ~ # of loaded cells / 2</p>
<p>        expected range of y axis (UMI_counts) : ~ 10^4</p>
<p>        expected inflection point of y axis (cutoff UMI count): ~ 1000</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
library(tidyverse)
# use &quot;raw&quot; instead of &quot;filtered&quot; cellranger output folder
data_url = &quot;https://scrnaseq-workshop.s3-us-west-2.amazonaws.com&quot;
m1 &lt;- readRDS(url(file.path(data_url, &quot;raw_matrix.rds&quot;)))
# all genes x all barcodes

counts &lt;- Matrix::colSums(m1) # calculate total UMI read number for each cell barcode
countdf &lt;- as.data.frame(counts) %&gt;% 
  as_tibble(rownames = &quot;barcode&quot;) %&gt;% 
  filter(counts &gt;= 2) %&gt;% # throw out cell barcodes with 1 or less UMI, this is mainly for time purposes
  arrange(desc(counts)) %&gt;% # arrange by descending order
  mutate(rank = 1:n()) # rank

head(countdf) # barcodes now ranked by UMI counts
#&gt; # A tibble: 6 x 3
#&gt;   barcode          counts  rank
#&gt;   &lt;chr&gt;             &lt;dbl&gt; &lt;int&gt;
#&gt; 1 GCAGCCACATACCGTA  40745     1
#&gt; 2 ATGGAGGGTGGTAACG  20916     2
#&gt; 3 GTAACCATCGCTTGAA  19958     3
#&gt; 4 TTCATGTGTCGTGTTA  19765     4
#&gt; 5 GAATAGACATCCTGTC  18494     5
#&gt; 6 GTCTCACGTTGGCCGT  17828     6

ggplot(countdf, aes(x = rank, y = counts)) +
  geom_point() +
  labs(x = &quot;barcodes&quot;, y = &quot;UMI_counts&quot;) +
  theme_classic() +
  scale_x_log10() + 
  scale_y_log10()</code></pre>
<p><img src="1_intro_files/figure-html5/unnamed-chunk-3-1.png" width="288" style="display: block; margin: auto;" /></p>
</div>
<h2 id="briefly-tidyversedplyr-verbs">briefly, tidyverse/dplyr verbs</h2>
<p>The tidyverse is a collection of R packages designed for data science. All packages share design philosophy, grammar, and data structures.</p>
<p>Seurat also uses some of the grammar, such as <code>group.by</code>.</p>
<p>A large amount of code introduced in this workshop in based on matrix and dataframe manipulation. Therefore some basic understanding of tidyverse/dplyr will be helpful.</p>
<p>Also, “%&gt;%” is used as “pipe”, similar to unix <code>|</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# we will look at metadata from pbmc_small from Seurat
library(Seurat)
library(tidyverse)
# note that tidyverse is philosophically against rownames

meta &lt;- pbmc_small@meta.data %&gt;% as_tibble(rownames = &quot;cell_id&quot;)
meta %&gt;% head() %&gt;% print() # &lt;- same as print(head(meta))
#&gt; # A tibble: 6 x 8
#&gt;   cell_id orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8
#&gt;   &lt;chr&gt;   &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;          
#&gt; 1 ATGCCA… SeuratPro…         70           47 0              
#&gt; 2 CATGGC… SeuratPro…         85           52 0              
#&gt; 3 GAACCT… SeuratPro…         87           50 1              
#&gt; 4 TGACTG… SeuratPro…        127           56 0              
#&gt; 5 AGTCAG… SeuratPro…        173           53 0              
#&gt; 6 TCTGAT… SeuratPro…         70           48 0              
#&gt; # … with 3 more variables: letter.idents &lt;fct&gt;, groups &lt;chr&gt;,
#&gt; #   RNA_snn_res.1 &lt;fct&gt;</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# &quot;select&quot; certain columns of data
meta2 &lt;- meta %&gt;% select(cell_id, 
                         nCount_RNA, 
                         RNA_snn_res.0.8, 
                         letter.idents)
meta2 %&gt;% head() %&gt;% print()
#&gt; # A tibble: 6 x 4
#&gt;   cell_id        nCount_RNA RNA_snn_res.0.8 letter.idents
#&gt;   &lt;chr&gt;               &lt;dbl&gt; &lt;fct&gt;           &lt;fct&gt;        
#&gt; 1 ATGCCAGAACGACT         70 0               A            
#&gt; 2 CATGGCCTGTGCAT         85 0               A            
#&gt; 3 GAACCTGATGAACC         87 1               B            
#&gt; 4 TGACTGGATTCTCA        127 0               A            
#&gt; 5 AGTCAGACTGCACA        173 0               A            
#&gt; 6 TCTGATACACGTGT         70 0               A</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# &quot;filter&quot; data to select specific rows
meta2 %&gt;% filter(nCount_RNA &gt;= 71, 
                 letter.idents == &quot;A&quot;) %&gt;%
  head() %&gt;%
  print()
#&gt; # A tibble: 6 x 4
#&gt;   cell_id        nCount_RNA RNA_snn_res.0.8 letter.idents
#&gt;   &lt;chr&gt;               &lt;dbl&gt; &lt;fct&gt;           &lt;fct&gt;        
#&gt; 1 CATGGCCTGTGCAT         85 0               A            
#&gt; 2 TGACTGGATTCTCA        127 0               A            
#&gt; 3 AGTCAGACTGCACA        173 0               A            
#&gt; 4 GCAGCTCTGTTTCT         72 0               A            
#&gt; 5 AATGTTGACAGTCA        100 0               A            
#&gt; 6 AGAGATGATCTCGC        191 0               A</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# &quot;arrange&quot; the rows of your data into an order
meta2 %&gt;% arrange(nCount_RNA) %&gt;%
  head() %&gt;%
  print()
#&gt; # A tibble: 6 x 4
#&gt;   cell_id        nCount_RNA RNA_snn_res.0.8 letter.idents
#&gt;   &lt;chr&gt;               &lt;dbl&gt; &lt;fct&gt;           &lt;fct&gt;        
#&gt; 1 CTTCATGACCGAAT         41 0               A            
#&gt; 2 CATGAGACACGGGA         51 0               A            
#&gt; 3 GATATAACACGCAT         52 0               A            
#&gt; 4 AGGTCATGAGTGTC         62 0               A            
#&gt; 5 TGGTATCTAAACAG         64 0               A            
#&gt; 6 GTAAGCACTCATTC         67 0               A</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# &quot;mutate&quot; your data frame to contain new columns
meta3 &lt;- meta2 %&gt;% mutate(seurat_clusters = str_c(&quot;cluster_&quot;, letter.idents), 
                          nCount_RNA_norm = nCount_RNA/max(nCount_RNA))
meta3 %&gt;% head() %&gt;% print()
#&gt; # A tibble: 6 x 6
#&gt;   cell_id nCount_RNA RNA_snn_res.0.8 letter.idents seurat_clusters
#&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt;          
#&gt; 1 ATGCCA…         70 0               A             cluster_A      
#&gt; 2 CATGGC…         85 0               A             cluster_A      
#&gt; 3 GAACCT…         87 1               B             cluster_B      
#&gt; 4 TGACTG…        127 0               A             cluster_A      
#&gt; 5 AGTCAG…        173 0               A             cluster_A      
#&gt; 6 TCTGAT…         70 0               A             cluster_A      
#&gt; # … with 1 more variable: nCount_RNA_norm &lt;dbl&gt;</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# &quot;summarise&quot; chunks of you data (by group) in some way.
meta3 %&gt;% group_by(seurat_clusters) %&gt;%
  summarise(n = n())
#&gt; # A tibble: 2 x 2
#&gt;   seurat_clusters     n
#&gt;   &lt;chr&gt;           &lt;int&gt;
#&gt; 1 cluster_A          53
#&gt; 2 cluster_B          27</code></pre>
</div>
<h2 id="reading-material">reading material</h2>
<ol type="1">
<li><a href="https://www.nature.com/articles/s12276-018-0071-8">Single-cell RNA sequencing technologies and bioinformatics pipelines</a></li>
<li><a href="https://www.embopress.org/doi/10.15252/msb.20188746">Current best practices in single‐cell RNA‐seq analysis: a tutorial</a></li>
<li><a href="https://scrnaseq-course.cog.sanger.ac.uk/website/index.html">“Analysis of single cell RNA-seq data” course</a></li>
<li><a href="https://dplyr.tidyverse.org/articles/dplyr.html">Introduction to dplyr</a></li>
</ol>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="updates-and-corrections">Corrections</h3>
<p>If you see mistakes or want to suggest changes, please <a href="https://github.com/rnabioco/cellar/issues/new">create an issue</a> on the source repository.</p>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
