<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Data Filtering and Quality Control</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="2_filtering_QC.html">Quality Control</a>
</li>
<li>
  <a href="7_multimodal.html">Multi-modal data</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore"><b style="font-size:45px;">Data Filtering and Quality Control</b></h1>
<h4 class="date">August 13<sup>th</sup>, 2019</h4>

</div>

<div id="TOC">
<ul>
<li><a href="#creating-a-seurat-object"><strong>Creating a Seurat object</strong></a></li>
<li><a href="#interacting-with-the-seurat-object"><strong>Interacting with the Seurat Object</strong></a></li>
<li><a href="#using-seurat-functions-to-interact-with-the-data"><strong>Using Seurat functions to interact with the data</strong></a></li>
<li><a href="#assessing-data-quality"><strong>Assessing data quality</strong></a></li>
<li><a href="#filtering-cells"><strong>Filtering cells</strong></a></li>
</ul>
</div>

<p><br></p>
<p><br></p>
<div id="creating-a-seurat-object" class="section level1">
<h1><strong>Creating a Seurat object</strong></h1>
<pre class="r"><code># Data URL
data_url = &quot;https://scrnaseq-workshop.s3-us-west-2.amazonaws.com&quot;

pbmc_mtx &lt;- file.path(data_url, &quot;PBMC_cDNA.csv.gz&quot;) %&gt;%
  read_csv() %&gt;%
  column_to_rownames(&quot;X1&quot;) %&gt;%
  as.sparse()

pbmc_mtx[1:10, 1:10]</code></pre>
<pre><code>#&gt; 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                  
#&gt; AL627309.1    . . . . . . . . . .
#&gt; RP11-206L10.2 . . . . . . . . . .
#&gt; LINC00115     . . . . . . . . . .
#&gt; NOC2L         . . . . . . . . 1 .
#&gt; KLHL17        . . . . . . . . . .
#&gt; PLEKHN1       . . . . . . . . . .
#&gt; HES4          . . . . . 2 . . . .
#&gt; ISG15         . . . . . 4 . 1 . 2
#&gt; AGRN          . . . . . . . . . .
#&gt; C1orf159      . . . . . . . . . .</code></pre>
<pre class="r"><code># Create Seurat object using gene expression data
sobj &lt;- pbmc_mtx %&gt;%
  CreateSeuratObject(
    min.cells = 5,      # Remove genes that are detected in &lt;5 cells
    project   = &quot;PBMC&quot;
  )</code></pre>
<p><br></p>
<p><br></p>
</div>
<div id="interacting-with-the-seurat-object" class="section level1">
<h1><strong>Interacting with the Seurat Object</strong></h1>
<div id="retrieving-raw-and-normalized-counts" class="section level2">
<h2>Retrieving raw and normalized counts</h2>
<pre class="r"><code># Main slot where raw and normalized data are stored
sobj@assays</code></pre>
<pre><code>#&gt; $RNA
#&gt; Assay data with 12572 features for 2700 cells
#&gt; First 10 features:
#&gt;  AL627309.1, RP11-206L10.2, LINC00115, NOC2L, KLHL17, PLEKHN1,
#&gt; HES4, ISG15, AGRN, C1orf159</code></pre>
<pre class="r"><code># Raw counts
sobj@assays$RNA@counts[1:10, 1:10]</code></pre>
<pre><code>#&gt; 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                  
#&gt; AL627309.1    . . . . . . . . . .
#&gt; RP11-206L10.2 . . . . . . . . . .
#&gt; LINC00115     . . . . . . . . . .
#&gt; NOC2L         . . . . . . . . 1 .
#&gt; KLHL17        . . . . . . . . . .
#&gt; PLEKHN1       . . . . . . . . . .
#&gt; HES4          . . . . . 2 . . . .
#&gt; ISG15         . . . . . 4 . 1 . 2
#&gt; AGRN          . . . . . . . . . .
#&gt; C1orf159      . . . . . . . . . .</code></pre>
<pre class="r"><code># Normalized counts
# This matrix is the same as counts since we haven&#39;t normalized the data yet
sobj@assays$RNA@data[1:10, 1:10]</code></pre>
<pre><code>#&gt; 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                  
#&gt; AL627309.1    . . . . . . . . . .
#&gt; RP11-206L10.2 . . . . . . . . . .
#&gt; LINC00115     . . . . . . . . . .
#&gt; NOC2L         . . . . . . . . 1 .
#&gt; KLHL17        . . . . . . . . . .
#&gt; PLEKHN1       . . . . . . . . . .
#&gt; HES4          . . . . . 2 . . . .
#&gt; ISG15         . . . . . 4 . 1 . 2
#&gt; AGRN          . . . . . . . . . .
#&gt; C1orf159      . . . . . . . . . .</code></pre>
</div>
<div id="accessing-project-meta.data" class="section level2">
<h2>Accessing project meta.data</h2>
<pre class="r"><code># The Seurat object stores the current default assay and cell identities
# By default the cell identity is set to the project name
sobj@active.assay</code></pre>
<pre><code>#&gt; [1] &quot;RNA&quot;</code></pre>
<pre class="r"><code>sobj@active.ident[1:10]</code></pre>
<pre><code>#&gt; AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AAAGTTTGATCACG-1 AAATCATGACCACA-1 AACATTGATGGGAG-1 AACCGATGCTCCCA-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AACCGCCTCTACGA-1 AACCTTACGCGAGA-1 
#&gt;             PBMC             PBMC 
#&gt; Levels: PBMC</code></pre>
<pre class="r"><code># The meta.data table contains stats for each cell
sobj@meta.data %&gt;%
  head()</code></pre>
<pre><code>#&gt;                  orig.ident nCount_RNA nFeature_RNA
#&gt; AAACATACAACCAC-1       PBMC       2419          779
#&gt; AAACATTGAGCTAC-1       PBMC       4901         1350
#&gt; AAACGCTGACCAGT-1       PBMC       2174          781
#&gt; AAACGCTGGTTCTT-1       PBMC       2259          789
#&gt; AAAGTTTGATCACG-1       PBMC       1265          441
#&gt; AAATCATGACCACA-1       PBMC       4125         1365</code></pre>
<pre class="r"><code># Modify orig.ident to include the sample number
# This is useful later when creating separate plots for each sample
sobj@meta.data &lt;- sobj@meta.data %&gt;%
  rownames_to_column(&quot;cell_bc&quot;) %&gt;%
  mutate(sample_id = str_extract(cell_bc, &quot;[0-9]+$&quot;)) %&gt;%
  unite(orig.ident, orig.ident, sample_id, sep = &quot;-&quot;) %&gt;%
  column_to_rownames(&quot;cell_bc&quot;)

sobj@meta.data %&gt;%
  head()</code></pre>
<pre><code>#&gt;                  orig.ident nCount_RNA nFeature_RNA
#&gt; AAACATACAACCAC-1     PBMC-1       2419          779
#&gt; AAACATTGAGCTAC-1     PBMC-1       4901         1350
#&gt; AAACGCTGACCAGT-1     PBMC-1       2174          781
#&gt; AAACGCTGGTTCTT-1     PBMC-1       2259          789
#&gt; AAAGTTTGATCACG-1     PBMC-1       1265          441
#&gt; AAATCATGACCACA-1     PBMC-1       4125         1365</code></pre>
</div>
<div id="other-data-stored-in-the-seurat-object" class="section level2">
<h2>Other data stored in the Seurat object</h2>
<pre class="r"><code># Slots for downstream analysis
sobj@graphs
sobj@neighbors
sobj@reductions</code></pre>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="using-seurat-functions-to-interact-with-the-data" class="section level1">
<h1><strong>Using Seurat functions to interact with the data</strong></h1>
<div id="retrieving-cell-and-features-names" class="section level2">
<h2>Retrieving cell and features names</h2>
<pre class="r"><code># Get cell names
Cells(sobj) %&gt;%
  head(20)</code></pre>
<pre><code>#&gt;  [1] &quot;AAACATACAACCAC-1&quot; &quot;AAACATTGAGCTAC-1&quot; &quot;AAACGCTGACCAGT-1&quot;
#&gt;  [4] &quot;AAACGCTGGTTCTT-1&quot; &quot;AAAGTTTGATCACG-1&quot; &quot;AAATCATGACCACA-1&quot;
#&gt;  [7] &quot;AACATTGATGGGAG-1&quot; &quot;AACCGATGCTCCCA-1&quot; &quot;AACCGCCTCTACGA-1&quot;
#&gt; [10] &quot;AACCTTACGCGAGA-1&quot; &quot;AACTCACTCAAGCT-1&quot; &quot;AACTCGGAAAGTGA-1&quot;
#&gt; [13] &quot;AAGATGGAGATAAG-1&quot; &quot;AAGCCATGTCTCGC-1&quot; &quot;AAGCGTACGTCTTT-1&quot;
#&gt; [16] &quot;AAGGTCTGACAGTC-1&quot; &quot;AAGTCTCTCGGAGA-1&quot; &quot;AATAGGGAGAATGA-1&quot;
#&gt; [19] &quot;AATGGAGAATCGTG-1&quot; &quot;AATGGCTGTAAAGG-1&quot;</code></pre>
<pre class="r"><code># Get feature names
rownames(sobj) %&gt;%
  head(20)</code></pre>
<pre><code>#&gt;  [1] &quot;AL627309.1&quot;    &quot;RP11-206L10.2&quot; &quot;LINC00115&quot;     &quot;NOC2L&quot;        
#&gt;  [5] &quot;KLHL17&quot;        &quot;PLEKHN1&quot;       &quot;HES4&quot;          &quot;ISG15&quot;        
#&gt;  [9] &quot;AGRN&quot;          &quot;C1orf159&quot;      &quot;TNFRSF18&quot;      &quot;TNFRSF4&quot;      
#&gt; [13] &quot;SDF4&quot;          &quot;B3GALT6&quot;       &quot;UBE2J2&quot;        &quot;ACAP3&quot;        
#&gt; [17] &quot;PUSL1&quot;         &quot;CPSF3L&quot;        &quot;GLTPD1&quot;        &quot;DVL1&quot;</code></pre>
<pre class="r"><code># Get cell (colums) and feature (rows) counts
ncol(sobj)</code></pre>
<pre><code>#&gt; [1] 2700</code></pre>
<pre class="r"><code>nrow(sobj)</code></pre>
<pre><code>#&gt; [1] 12572</code></pre>
</div>
<div id="setting-cell-identities" class="section level2">
<h2>Setting cell identities</h2>
<pre class="r"><code># Retrieve cell identities
Idents(sobj) %&gt;%
  head(20)</code></pre>
<pre><code>#&gt; AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AAAGTTTGATCACG-1 AAATCATGACCACA-1 AACATTGATGGGAG-1 AACCGATGCTCCCA-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AACCGCCTCTACGA-1 AACCTTACGCGAGA-1 AACTCACTCAAGCT-1 AACTCGGAAAGTGA-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AAGATGGAGATAAG-1 AAGCCATGTCTCGC-1 AAGCGTACGTCTTT-1 AAGGTCTGACAGTC-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AAGTCTCTCGGAGA-1 AATAGGGAGAATGA-1 AATGGAGAATCGTG-1 AATGGCTGTAAAGG-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; Levels: PBMC</code></pre>
<pre class="r"><code># Set cell identities
Idents(sobj) &lt;- sobj@meta.data$orig.ident

# Rename cell identities
sobj &lt;- sobj %&gt;%
  RenameIdents(
    &quot;PBMC-1&quot; = &quot;control&quot;,
    &quot;PBMC-2&quot; = &quot;treated&quot;
  )

# Add cell identities to the meta.data table
sobj[[&quot;orig.ident&quot;]] &lt;- Idents(sobj)

sobj &lt;- sobj %&gt;%
  AddMetaData(
    metadata = Idents(sobj),
    col.name = &quot;orig.ident&quot;
  )</code></pre>
</div>
<div id="subsetting-the-seurat-object" class="section level2">
<h2>Subsetting the Seurat object</h2>
<pre class="r"><code># Subset based on gene expression
sobj %&gt;%
  subset(CD8A &gt; 1) %&gt;%
  ncol()</code></pre>
<pre><code>#&gt; [1] 107</code></pre>
<pre class="r"><code># Subset based on cell identity
sobj %&gt;%
  subset(idents = &quot;control&quot;) %&gt;%
  ncol()</code></pre>
<pre><code>#&gt; [1] 1342</code></pre>
<pre class="r"><code># Subset based on expression and cell identity
sobj %&gt;%
  subset(CD8A &gt; 1, idents = &quot;control&quot;) %&gt;%
  ncol()</code></pre>
<pre><code>#&gt; [1] 56</code></pre>
<pre class="r"><code># Subset based on meta.data columns
sobj %&gt;%
  subset(nFeature_RNA &gt; 250)</code></pre>
<pre><code>#&gt; An object of class Seurat 
#&gt; 12572 features across 2697 samples within 1 assay 
#&gt; Active assay: RNA (12572 features)</code></pre>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="assessing-data-quality" class="section level1">
<h1><strong>Assessing data quality</strong></h1>
<div id="calculate-the-percentage-mitochondrial-reads-for-each-cell" class="section level2">
<h2>Calculate the percentage mitochondrial reads for each cell</h2>
<pre class="r"><code># Calculate percentage of mitochondrial reads for each cell
sobj &lt;- sobj %&gt;%
  PercentageFeatureSet(
    pattern  = &quot;^MT-&quot;, 
    col.name = &quot;percent_mito&quot;
  )

sobj@meta.data %&gt;%
  head()</code></pre>
<pre><code>#&gt;                  orig.ident nCount_RNA nFeature_RNA percent_mito
#&gt; AAACATACAACCAC-1    control       2419          779     3.017776
#&gt; AAACATTGAGCTAC-1    control       4901         1350     3.795144
#&gt; AAACGCTGACCAGT-1    control       2174          781     3.817847
#&gt; AAACGCTGGTTCTT-1    control       2259          789     3.098716
#&gt; AAAGTTTGATCACG-1    control       1265          441     3.478261
#&gt; AAATCATGACCACA-1    control       4125         1365     4.581818</code></pre>
</div>
<div id="exercise-create-violinbox-plots-comparing-cell-metrics" class="section level2">
<h2>EXERCISE: Create violin/box plots comparing cell metrics</h2>
<pre class="r"><code># Create violin/boxplots comparing nCount, nFeature, and percent_mito for each sample

# sobj@meta.data %&gt;%</code></pre>
</div>
<div id="answer" class="section level2">
<h2>ANSWER</h2>
<p><img src="2_filtering_QC_files/figure-html/unnamed-chunk-2-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="exercise-create-scatter-plots-comparing-cell-metrics" class="section level2">
<h2>EXERCISE: Create scatter plots comparing cell metrics</h2>
<pre class="r"><code># Create a scatter plot comparing nCount and percent_mito

# sobj@meta.data %&gt;%



# Create a scatter plot comparing nCount and nFeature

# sobj@meta.data %&gt;%</code></pre>
</div>
<div id="answer-1" class="section level2">
<h2>ANSWER</h2>
<p><img src="2_filtering_QC_files/figure-html/unnamed-chunk-4-1.png" width="768" style="display: block; margin: auto;" /></p>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="filtering-cells" class="section level1">
<h1><strong>Filtering cells</strong></h1>
<pre class="r"><code># Filter cells based on number of genes and percent mito UMIs
sobj &lt;- sobj %&gt;%
  subset(
    nFeature_RNA &gt; 250 &amp;   # Remove cells with &lt; 250 detected genes
    nFeature_RNA &lt; 2500 &amp;  # Remove cells with &gt; 2500 detected genes (could be doublets)
    percent_mito &lt; 15      # Remove cells with &gt;15% mitochondrial reads
  )</code></pre>
<div id="saving-the-seurat-object" class="section level2">
<h2>Saving the Seurat object</h2>
<pre class="r"><code>sobj %&gt;%
  write_rds(&quot;filtered_sobj.rds&quot;, compress = &quot;gz&quot;)</code></pre>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
