<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Data Filtering and Quality Control</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link rel="stylesheet" href="/home/rmsheridan/R/x86_64-pc-linux-gnu-library/3.6/rmarkdown/rmarkdown/templates/html_vignette/resources/vignette.css" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore"><b style="font-size:45px;">Data Filtering and Quality Control</b></h1>
<h4 class="date">August 13<sup>th</sup>, 2019</h4>


<div id="TOC">
<ul>
<li><a href="#creating-a-seurat-object"><strong>Creating a Seurat object</strong></a></li>
<li><a href="#interacting-with-the-seurat-object"><strong>Interacting with the Seurat Object</strong></a></li>
<li><a href="#using-seurat-functions-to-interact-with-the-data"><strong>Using Seurat functions to interact with the data</strong></a></li>
<li><a href="#assessing-data-quality"><strong>Assessing data quality</strong></a></li>
<li><a href="#filtering-cells"><strong>Filtering cells</strong></a></li>
</ul>
</div>

<p><br></p>
<p><br></p>
<div id="creating-a-seurat-object" class="section level1">
<h1><strong>Creating a Seurat object</strong></h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># Data URL</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">data_url =<span class="st"> &quot;https://scrnaseq-workshop.s3-us-west-2.amazonaws.com&quot;</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">pbmc_mtx &lt;-<span class="st"> </span><span class="kw">file.path</span>(data_url, <span class="st">&quot;PBMC_cDNA.csv.gz&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="st">  </span><span class="kw">read_csv</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="st">  </span><span class="kw">column_to_rownames</span>(<span class="st">&quot;X1&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="st">  </span><span class="kw">as.sparse</span>()</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"></a>
<a class="sourceLine" id="cb1-9" data-line-number="9">pbmc_mtx[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a></code></pre></div>
<pre><code>#&gt; 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                  
#&gt; AL627309.1    . . . . . . . . . .
#&gt; RP11-206L10.2 . . . . . . . . . .
#&gt; LINC00115     . . . . . . . . . .
#&gt; NOC2L         . . . . . . . . 1 .
#&gt; KLHL17        . . . . . . . . . .
#&gt; PLEKHN1       . . . . . . . . . .
#&gt; HES4          . . . . . 2 . . . .
#&gt; ISG15         . . . . . 4 . 1 . 2
#&gt; AGRN          . . . . . . . . . .
#&gt; C1orf159      . . . . . . . . . .</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># Create Seurat object using gene expression data</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">sobj &lt;-<span class="st"> </span>pbmc_mtx <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="st">  </span><span class="kw">CreateSeuratObject</span>(</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">    <span class="dt">min.cells =</span> <span class="dv">5</span>,      <span class="co"># Remove genes that are detected in &lt;5 cells</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5">    <span class="dt">project   =</span> <span class="st">&quot;PBMC&quot;</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6">  )</a></code></pre></div>
<p><br></p>
<p><br></p>
</div>
<div id="interacting-with-the-seurat-object" class="section level1">
<h1><strong>Interacting with the Seurat Object</strong></h1>
<div id="retrieving-raw-and-normalized-counts" class="section level2">
<h2>Retrieving raw and normalized counts</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Main slot where raw and normalized data are stored</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">sobj<span class="op">@</span>assays</a></code></pre></div>
<pre><code>#&gt; $RNA
#&gt; Assay data with 12572 features for 2700 cells
#&gt; First 10 features:
#&gt;  AL627309.1, RP11-206L10.2, LINC00115, NOC2L, KLHL17, PLEKHN1,
#&gt; HES4, ISG15, AGRN, C1orf159</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># Raw counts</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2">sobj<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>counts[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a></code></pre></div>
<pre><code>#&gt; 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                  
#&gt; AL627309.1    . . . . . . . . . .
#&gt; RP11-206L10.2 . . . . . . . . . .
#&gt; LINC00115     . . . . . . . . . .
#&gt; NOC2L         . . . . . . . . 1 .
#&gt; KLHL17        . . . . . . . . . .
#&gt; PLEKHN1       . . . . . . . . . .
#&gt; HES4          . . . . . 2 . . . .
#&gt; ISG15         . . . . . 4 . 1 . 2
#&gt; AGRN          . . . . . . . . . .
#&gt; C1orf159      . . . . . . . . . .</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co"># Normalized counts</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="co"># This matrix is the same as counts since we haven&#39;t normalized the data yet</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3">sobj<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>data[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a></code></pre></div>
<pre><code>#&gt; 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                  
#&gt; AL627309.1    . . . . . . . . . .
#&gt; RP11-206L10.2 . . . . . . . . . .
#&gt; LINC00115     . . . . . . . . . .
#&gt; NOC2L         . . . . . . . . 1 .
#&gt; KLHL17        . . . . . . . . . .
#&gt; PLEKHN1       . . . . . . . . . .
#&gt; HES4          . . . . . 2 . . . .
#&gt; ISG15         . . . . . 4 . 1 . 2
#&gt; AGRN          . . . . . . . . . .
#&gt; C1orf159      . . . . . . . . . .</code></pre>
</div>
<div id="accessing-project-meta.data" class="section level2">
<h2>Accessing project meta.data</h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># The Seurat object stores the current default assay and cell identities</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="co"># By default the cell identity is set to the project name</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3">sobj<span class="op">@</span>active.assay</a></code></pre></div>
<pre><code>#&gt; [1] &quot;RNA&quot;</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">sobj<span class="op">@</span>active.ident[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a></code></pre></div>
<pre><code>#&gt; AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AAAGTTTGATCACG-1 AAATCATGACCACA-1 AACATTGATGGGAG-1 AACCGATGCTCCCA-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AACCGCCTCTACGA-1 AACCTTACGCGAGA-1 
#&gt;             PBMC             PBMC 
#&gt; Levels: PBMC</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co"># The meta.data table contains stats for each cell</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2">sobj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-3" data-line-number="3"><span class="st">  </span><span class="kw">head</span>()</a></code></pre></div>
<pre><code>#&gt;                  orig.ident nCount_RNA nFeature_RNA
#&gt; AAACATACAACCAC-1       PBMC       2419          779
#&gt; AAACATTGAGCTAC-1       PBMC       4901         1350
#&gt; AAACGCTGACCAGT-1       PBMC       2174          781
#&gt; AAACGCTGGTTCTT-1       PBMC       2259          789
#&gt; AAAGTTTGATCACG-1       PBMC       1265          441
#&gt; AAATCATGACCACA-1       PBMC       4125         1365</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="co"># Modify orig.ident to include the sample number</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="co"># This is useful later when creating separate plots for each sample</span></a>
<a class="sourceLine" id="cb16-3" data-line-number="3">sobj<span class="op">@</span>meta.data &lt;-<span class="st"> </span>sobj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4"><span class="st">  </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;cell_bc&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb16-5" data-line-number="5"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">sample_id =</span> <span class="kw">str_extract</span>(cell_bc, <span class="st">&quot;[0-9]+$&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb16-6" data-line-number="6"><span class="st">  </span><span class="kw">unite</span>(orig.ident, orig.ident, sample_id, <span class="dt">sep =</span> <span class="st">&quot;-&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb16-7" data-line-number="7"><span class="st">  </span><span class="kw">column_to_rownames</span>(<span class="st">&quot;cell_bc&quot;</span>)</a>
<a class="sourceLine" id="cb16-8" data-line-number="8"></a>
<a class="sourceLine" id="cb16-9" data-line-number="9">sobj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb16-10" data-line-number="10"><span class="st">  </span><span class="kw">head</span>()</a></code></pre></div>
<pre><code>#&gt;                  orig.ident nCount_RNA nFeature_RNA
#&gt; AAACATACAACCAC-1     PBMC-1       2419          779
#&gt; AAACATTGAGCTAC-1     PBMC-1       4901         1350
#&gt; AAACGCTGACCAGT-1     PBMC-1       2174          781
#&gt; AAACGCTGGTTCTT-1     PBMC-1       2259          789
#&gt; AAAGTTTGATCACG-1     PBMC-1       1265          441
#&gt; AAATCATGACCACA-1     PBMC-1       4125         1365</code></pre>
</div>
<div id="other-data-stored-in-the-seurat-object" class="section level2">
<h2>Other data stored in the Seurat object</h2>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="co"># Slots for downstream analysis</span></a>
<a class="sourceLine" id="cb18-2" data-line-number="2">sobj<span class="op">@</span>graphs</a>
<a class="sourceLine" id="cb18-3" data-line-number="3">sobj<span class="op">@</span>neighbors</a>
<a class="sourceLine" id="cb18-4" data-line-number="4">sobj<span class="op">@</span>reductions</a></code></pre></div>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="using-seurat-functions-to-interact-with-the-data" class="section level1">
<h1><strong>Using Seurat functions to interact with the data</strong></h1>
<div id="retrieving-cell-and-features-names" class="section level2">
<h2>Retrieving cell and features names</h2>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="co"># Get cell names</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw">Cells</span>(sobj) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb19-3" data-line-number="3"><span class="st">  </span><span class="kw">head</span>(<span class="dv">20</span>)</a></code></pre></div>
<pre><code>#&gt;  [1] &quot;AAACATACAACCAC-1&quot; &quot;AAACATTGAGCTAC-1&quot; &quot;AAACGCTGACCAGT-1&quot;
#&gt;  [4] &quot;AAACGCTGGTTCTT-1&quot; &quot;AAAGTTTGATCACG-1&quot; &quot;AAATCATGACCACA-1&quot;
#&gt;  [7] &quot;AACATTGATGGGAG-1&quot; &quot;AACCGATGCTCCCA-1&quot; &quot;AACCGCCTCTACGA-1&quot;
#&gt; [10] &quot;AACCTTACGCGAGA-1&quot; &quot;AACTCACTCAAGCT-1&quot; &quot;AACTCGGAAAGTGA-1&quot;
#&gt; [13] &quot;AAGATGGAGATAAG-1&quot; &quot;AAGCCATGTCTCGC-1&quot; &quot;AAGCGTACGTCTTT-1&quot;
#&gt; [16] &quot;AAGGTCTGACAGTC-1&quot; &quot;AAGTCTCTCGGAGA-1&quot; &quot;AATAGGGAGAATGA-1&quot;
#&gt; [19] &quot;AATGGAGAATCGTG-1&quot; &quot;AATGGCTGTAAAGG-1&quot;</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="co"># Get feature names</span></a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="kw">rownames</span>(sobj) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb21-3" data-line-number="3"><span class="st">  </span><span class="kw">head</span>(<span class="dv">20</span>)</a></code></pre></div>
<pre><code>#&gt;  [1] &quot;AL627309.1&quot;    &quot;RP11-206L10.2&quot; &quot;LINC00115&quot;     &quot;NOC2L&quot;        
#&gt;  [5] &quot;KLHL17&quot;        &quot;PLEKHN1&quot;       &quot;HES4&quot;          &quot;ISG15&quot;        
#&gt;  [9] &quot;AGRN&quot;          &quot;C1orf159&quot;      &quot;TNFRSF18&quot;      &quot;TNFRSF4&quot;      
#&gt; [13] &quot;SDF4&quot;          &quot;B3GALT6&quot;       &quot;UBE2J2&quot;        &quot;ACAP3&quot;        
#&gt; [17] &quot;PUSL1&quot;         &quot;CPSF3L&quot;        &quot;GLTPD1&quot;        &quot;DVL1&quot;</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="co"># Get cell (colums) and feature (rows) counts</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="kw">ncol</span>(sobj)</a></code></pre></div>
<pre><code>#&gt; [1] 2700</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw">nrow</span>(sobj)</a></code></pre></div>
<pre><code>#&gt; [1] 12572</code></pre>
</div>
<div id="setting-cell-identities" class="section level2">
<h2>Setting cell identities</h2>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="co"># Retrieve cell identities</span></a>
<a class="sourceLine" id="cb27-2" data-line-number="2"><span class="kw">Idents</span>(sobj) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-3" data-line-number="3"><span class="st">  </span><span class="kw">head</span>(<span class="dv">20</span>)</a></code></pre></div>
<pre><code>#&gt; AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AAAGTTTGATCACG-1 AAATCATGACCACA-1 AACATTGATGGGAG-1 AACCGATGCTCCCA-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AACCGCCTCTACGA-1 AACCTTACGCGAGA-1 AACTCACTCAAGCT-1 AACTCGGAAAGTGA-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AAGATGGAGATAAG-1 AAGCCATGTCTCGC-1 AAGCGTACGTCTTT-1 AAGGTCTGACAGTC-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; AAGTCTCTCGGAGA-1 AATAGGGAGAATGA-1 AATGGAGAATCGTG-1 AATGGCTGTAAAGG-1 
#&gt;             PBMC             PBMC             PBMC             PBMC 
#&gt; Levels: PBMC</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="co"># Set cell identities</span></a>
<a class="sourceLine" id="cb29-2" data-line-number="2"><span class="kw">Idents</span>(sobj) &lt;-<span class="st"> </span>sobj<span class="op">@</span>meta.data<span class="op">$</span>orig.ident</a>
<a class="sourceLine" id="cb29-3" data-line-number="3"></a>
<a class="sourceLine" id="cb29-4" data-line-number="4"><span class="co"># Rename cell identities</span></a>
<a class="sourceLine" id="cb29-5" data-line-number="5">sobj &lt;-<span class="st"> </span>sobj <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb29-6" data-line-number="6"><span class="st">  </span><span class="kw">RenameIdents</span>(</a>
<a class="sourceLine" id="cb29-7" data-line-number="7">    <span class="st">&quot;PBMC-1&quot;</span> =<span class="st"> &quot;control&quot;</span>,</a>
<a class="sourceLine" id="cb29-8" data-line-number="8">    <span class="st">&quot;PBMC-2&quot;</span> =<span class="st"> &quot;treated&quot;</span></a>
<a class="sourceLine" id="cb29-9" data-line-number="9">  )</a>
<a class="sourceLine" id="cb29-10" data-line-number="10"></a>
<a class="sourceLine" id="cb29-11" data-line-number="11"><span class="co"># Add cell identities to the meta.data table</span></a>
<a class="sourceLine" id="cb29-12" data-line-number="12">sobj[[<span class="st">&quot;orig.ident&quot;</span>]] &lt;-<span class="st"> </span><span class="kw">Idents</span>(sobj)</a>
<a class="sourceLine" id="cb29-13" data-line-number="13"></a>
<a class="sourceLine" id="cb29-14" data-line-number="14">sobj &lt;-<span class="st"> </span>sobj <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb29-15" data-line-number="15"><span class="st">  </span><span class="kw">AddMetaData</span>(</a>
<a class="sourceLine" id="cb29-16" data-line-number="16">    <span class="dt">metadata =</span> <span class="kw">Idents</span>(sobj),</a>
<a class="sourceLine" id="cb29-17" data-line-number="17">    <span class="dt">col.name =</span> <span class="st">&quot;orig.ident&quot;</span></a>
<a class="sourceLine" id="cb29-18" data-line-number="18">  )</a></code></pre></div>
</div>
<div id="subsetting-the-seurat-object" class="section level2">
<h2>Subsetting the Seurat object</h2>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="co"># Subset based on gene expression</span></a>
<a class="sourceLine" id="cb30-2" data-line-number="2">sobj <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb30-3" data-line-number="3"><span class="st">  </span><span class="kw">subset</span>(CD8A <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb30-4" data-line-number="4"><span class="st">  </span><span class="kw">ncol</span>()</a></code></pre></div>
<pre><code>#&gt; [1] 107</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="co"># Subset based on cell identity</span></a>
<a class="sourceLine" id="cb32-2" data-line-number="2">sobj <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb32-3" data-line-number="3"><span class="st">  </span><span class="kw">subset</span>(<span class="dt">idents =</span> <span class="st">&quot;control&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb32-4" data-line-number="4"><span class="st">  </span><span class="kw">ncol</span>()</a></code></pre></div>
<pre><code>#&gt; [1] 1342</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="co"># Subset based on expression and cell identity</span></a>
<a class="sourceLine" id="cb34-2" data-line-number="2">sobj <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb34-3" data-line-number="3"><span class="st">  </span><span class="kw">subset</span>(CD8A <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">idents =</span> <span class="st">&quot;control&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb34-4" data-line-number="4"><span class="st">  </span><span class="kw">ncol</span>()</a></code></pre></div>
<pre><code>#&gt; [1] 56</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="co"># Subset based on meta.data columns</span></a>
<a class="sourceLine" id="cb36-2" data-line-number="2">sobj <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb36-3" data-line-number="3"><span class="st">  </span><span class="kw">subset</span>(nFeature_RNA <span class="op">&gt;</span><span class="st"> </span><span class="dv">250</span>)</a></code></pre></div>
<pre><code>#&gt; An object of class Seurat 
#&gt; 12572 features across 2697 samples within 1 assay 
#&gt; Active assay: RNA (12572 features)</code></pre>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="assessing-data-quality" class="section level1">
<h1><strong>Assessing data quality</strong></h1>
<div id="calculate-the-percentage-mitochondrial-reads-for-each-cell" class="section level2">
<h2>Calculate the percentage mitochondrial reads for each cell</h2>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="co"># Calculate percentage of mitochondrial reads for each cell</span></a>
<a class="sourceLine" id="cb38-2" data-line-number="2">sobj &lt;-<span class="st"> </span>sobj <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-3" data-line-number="3"><span class="st">  </span><span class="kw">PercentageFeatureSet</span>(</a>
<a class="sourceLine" id="cb38-4" data-line-number="4">    <span class="dt">pattern  =</span> <span class="st">&quot;^MT-&quot;</span>, </a>
<a class="sourceLine" id="cb38-5" data-line-number="5">    <span class="dt">col.name =</span> <span class="st">&quot;percent_mito&quot;</span></a>
<a class="sourceLine" id="cb38-6" data-line-number="6">  )</a>
<a class="sourceLine" id="cb38-7" data-line-number="7"></a>
<a class="sourceLine" id="cb38-8" data-line-number="8">sobj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb38-9" data-line-number="9"><span class="st">  </span><span class="kw">head</span>()</a></code></pre></div>
<pre><code>#&gt;                  orig.ident nCount_RNA nFeature_RNA percent_mito
#&gt; AAACATACAACCAC-1    control       2419          779     3.017776
#&gt; AAACATTGAGCTAC-1    control       4901         1350     3.795144
#&gt; AAACGCTGACCAGT-1    control       2174          781     3.817847
#&gt; AAACGCTGGTTCTT-1    control       2259          789     3.098716
#&gt; AAAGTTTGATCACG-1    control       1265          441     3.478261
#&gt; AAATCATGACCACA-1    control       4125         1365     4.581818</code></pre>
</div>
<div id="exercise-create-violinbox-plots-comparing-cell-metrics" class="section level2">
<h2>EXERCISE: Create violin/box plots comparing cell metrics</h2>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="co"># Create violin/boxplots comparing nCount, nFeature, and percent_mito for each sample</span></a>
<a class="sourceLine" id="cb40-2" data-line-number="2"></a>
<a class="sourceLine" id="cb40-3" data-line-number="3"><span class="co"># sobj@meta.data %&gt;%</span></a></code></pre></div>
</div>
<div id="answer" class="section level2">
<h2>ANSWER</h2>
<p><img src="2_filtering_QC_files/figure-html/unnamed-chunk-2-1.png" style="display: block; margin: auto;" /></p>
</div>
<div id="exercise-create-scatter-plots-comparing-cell-metrics" class="section level2">
<h2>EXERCISE: Create scatter plots comparing cell metrics</h2>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1"><span class="co"># Create a scatter plot comparing nCount and percent_mito</span></a>
<a class="sourceLine" id="cb41-2" data-line-number="2"></a>
<a class="sourceLine" id="cb41-3" data-line-number="3"><span class="co"># sobj@meta.data %&gt;%</span></a>
<a class="sourceLine" id="cb41-4" data-line-number="4"></a>
<a class="sourceLine" id="cb41-5" data-line-number="5"></a>
<a class="sourceLine" id="cb41-6" data-line-number="6"></a>
<a class="sourceLine" id="cb41-7" data-line-number="7"><span class="co"># Create a scatter plot comparing nCount and nFeature</span></a>
<a class="sourceLine" id="cb41-8" data-line-number="8"></a>
<a class="sourceLine" id="cb41-9" data-line-number="9"><span class="co"># sobj@meta.data %&gt;%</span></a></code></pre></div>
</div>
<div id="answer-1" class="section level2">
<h2>ANSWER</h2>
<p><img src="2_filtering_QC_files/figure-html/unnamed-chunk-4-1.png" style="display: block; margin: auto;" /></p>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="filtering-cells" class="section level1">
<h1><strong>Filtering cells</strong></h1>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1"><span class="co"># Filter cells based on number of genes and percent mito UMIs</span></a>
<a class="sourceLine" id="cb42-2" data-line-number="2">sobj &lt;-<span class="st"> </span>sobj <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb42-3" data-line-number="3"><span class="st">  </span><span class="kw">subset</span>(</a>
<a class="sourceLine" id="cb42-4" data-line-number="4">    nFeature_RNA <span class="op">&gt;</span><span class="st"> </span><span class="dv">250</span> <span class="op">&amp;</span><span class="st">   </span><span class="co"># Remove cells with &lt; 250 detected genes</span></a>
<a class="sourceLine" id="cb42-5" data-line-number="5"><span class="st">    </span>nFeature_RNA <span class="op">&lt;</span><span class="st"> </span><span class="dv">2500</span> <span class="op">&amp;</span><span class="st">  </span><span class="co"># Remove cells with &gt; 2500 detected genes (could be doublets)</span></a>
<a class="sourceLine" id="cb42-6" data-line-number="6"><span class="st">    </span>percent_mito <span class="op">&lt;</span><span class="st"> </span><span class="dv">15</span>      <span class="co"># Remove cells with &gt;15% mitochondrial reads</span></a>
<a class="sourceLine" id="cb42-7" data-line-number="7">  )</a></code></pre></div>
<div id="saving-the-seurat-object" class="section level2">
<h2>Saving the Seurat object</h2>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1">sobj <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb43-2" data-line-number="2"><span class="st">  </span><span class="kw">write_rds</span>(<span class="st">&quot;filtered_sobj.rds&quot;</span>, <span class="dt">compress =</span> <span class="st">&quot;gz&quot;</span>)</a></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
