<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

  <!--radix_placeholder_meta_tags-->
  <title>Single-cell RNA-seq Workshop: &lt;b style="font-size:45px;"&gt;Cluster Markers and Cell Type Assignment&lt;/b&gt;&lt;/b&gt;</title>
  
  
  <link rel="icon" type="image/png" href="img/logo.png"/>
  
  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2019-08-14"/>
  <meta property="article:created" itemprop="dateCreated" content="2019-08-14"/>
  <meta name="article:author" content="Rui Fu"/>
  
  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="Single-cell RNA-seq Workshop: &lt;b style=&quot;font-size:45px;&quot;&gt;Cluster Markers and Cell Type Assignment&lt;/b&gt;&lt;/b&gt;"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="Single-cell RNA-seq Workshop"/>
  
  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="Single-cell RNA-seq Workshop: &lt;b style=&quot;font-size:45px;&quot;&gt;Cluster Markers and Cell Type Assignment&lt;/b&gt;&lt;/b&gt;"/>
  <meta property="twitter:site" content="@rnabioco"/>
  
  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","author","date","output"]}},"value":[{"type":"character","attributes":{},"value":["<b style=\"font-size:45px;\">Cluster Markers and Cell Type Assignment<\/b><\/b>"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name"]}},"value":[{"type":"character","attributes":{},"value":["Rui Fu"]}]}]},{"type":"character","attributes":{},"value":["August 14<sup>th<\/sup>, 2019"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["html_document"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["toc","toc_depth","self_contained"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"integer","attributes":{},"value":[1]},{"type":"logical","attributes":{},"value":[true]}]}]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  
  <script type="application/javascript">
  
    window.headroom_prevent_pin = false;
  
    window.document.addEventListener("DOMContentLoaded", function (event) {
  
      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');
  
      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });
  
      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });
  
      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>
  
  <style type="text/css">
  
  /* Theme (user-documented overrideables for nav appearance) */
  
  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #455a64;
    font-size: 15px;
    font-weight: 300;
  }
  
  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }
  
  .distill-site-nav a:hover {
    color: white;
  }
  
  @media print {
    .distill-site-nav {
      display: none;
    }
  }
  
  .distill-site-header {
  
  }
  
  .distill-site-footer {
  
  }
  
  
  /* Site Header */
  
  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }
  
  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }
  
  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }
  
  
  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }
  
  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }
  
  .distill-site-header .title {
    font-size: 18px;
  }
  
  .distill-site-header .logo {
    padding: 0;
  }
  
  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }
  
  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }
  
  
  
  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }
  
  
  .distill-site-header .nav-toggle {
    display: none;
  }
  
  .nav-dropdown {
    display: inline-block;
    position: relative;
  }
  
  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }
  
  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }
  
  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .nav-dropdown-active {
    display: block;
  }
  
  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }
  
  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }
  
  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }
  
  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }
  
  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }
  
  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }
  
  
  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative;}
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }
  
  /* Site Footer */
  
  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }
  
  /* Headroom */
  
  d-title {
    padding-top: 6rem;
  }
  
  @media print {
    d-title {
      padding-top: 4rem;
    }
  }
  
  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }
  
  .headroom--transition {
    transition: all .4s ease-in-out;
  }
  
  .headroom--unpinned {
    top: -100px;
  }
  
  .headroom--pinned {
    top: 0;
  }
  
  </style>
  
  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->
  
  <style type="text/css">
  
  body {
    background-color: white;
  }
  
  .pandoc-table {
    width: 100%;
  }
  
  .pandoc-table>caption {
    margin-bottom: 10px;
  }
  
  .pandoc-table th:not([align]) {
    text-align: left;
  }
  
  .pagedtable-footer {
    font-size: 15px;
  }
  
  .html-widget {
    margin-bottom: 2.0em;
  }
  
  .l-screen-inset {
    padding-right: 16px;
  }
  
  .l-screen .caption {
    margin-left: 10px;
  }
  
  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .shaded .shaded-content {
    background: white;
  }
  
  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }
  
  .hidden {
    display: none !important;
  }
  
  d-article {
    padding-bottom: 30px;
  }
  
  d-appendix {
    padding-top: 30px;
  }
  
  d-article>p>img {
    width: 100%;
  }
  
  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }
  
  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }
  
  /* CSS for table of contents */
  
  .d-toc {
    color: rgba(0,0,0,0.8);
    font-size: 0.8em;
    line-height: 1em;
  }
  
  .d-toc-header {
    font-size: 0.6rem;
    font-weight: 400;
    color: rgba(0, 0, 0, 0.5);
    text-transform: uppercase;
    margin-top: 0;
    margin-bottom: 1.3em;
  }
  
  .d-toc a {
    border-bottom: none;
  }
  
  .d-toc ul {
    padding-left: 0;
  }
  
  .d-toc li>ul {
    padding-top: 0.8em;
    padding-left: 16px;
    margin-bottom: 0.6em;
  }
  
  .d-toc ul,
  .d-toc li {
    list-style-type: none;
  }
  
  .d-toc li {
    margin-bottom: 0.9em;
  }
  
  .d-toc-separator {
    margin-top: 20px;
    margin-bottom: 2em;
  }
  
  .d-article-with-toc {
    border-top: none;
    padding-top: 0;
  }
  
  
  
  /* Tweak code blocks (note that this CSS is repeated above in an injection
     into the d-code shadow dom) */
  
  d-code {
    overflow-x: auto !important;
  }
  
  pre.d-code code.d-code {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }
  
  pre.text-output {
  
    font-size: 12px;
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;
  
    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;
  
    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }
  
  @media(min-width: 768px) {
  
  d-code {
    overflow-x: visible !important;
  }
  
  pre.d-code code.d-code  {
      padding-left: 18px;
      font-size: 14px;
  }
  pre.text-output {
    font-size: 14px;
  }
  }
  
  /* Figure */
  
  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }
  
  .figure img {
    width: 100%;
  }
  
  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }
  
  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }
  
  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }
  
  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }
  
  
  
  /* Tweak 1000px media break to show more text */
  
  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }
  
    .grid {
      grid-column-gap: 16px;
    }
  
    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }
  
  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }
  
    .grid {
      grid-column-gap: 32px;
    }
  }
  
  
  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */
  
  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }
  
  
  /* Social footer */
  
  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }
  
  .disqus-comments {
    margin-right: 30px;
  }
  
  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }
  
  #disqus_thread {
    margin-top: 30px;
  }
  
  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }
  
  .article-sharing a:hover {
    border-bottom: none;
  }
  
  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }
  
  .subscribe p {
    margin-bottom: 0.5em;
  }
  
  
  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }
  
  
  /* Improve display for browsers without grid (IE/Edge <= 15) */
  
  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }
  
  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }
  
  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }
  
  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }
  
  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }
  
  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }
  
  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }
  
  
  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }
  
  .downlevel .footnotes ol {
    padding-left: 13px;
  }
  
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }
  
  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }
  
  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }
  
  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;
  
    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;
  
    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }
  
  </style>
  
  <script type="application/javascript">
  
  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }
  
  // show body when load is complete
  function on_load_complete() {
  
    // set body to visible
    document.body.style.visibility = 'visible';
  
    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }
  
    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }
  
  function init_distill() {
  
    init_common();
  
    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);
  
    // create d-title
    $('.d-title').changeElementType('d-title');
  
    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);
  
    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();
  
    // move posts container into article
    $('.posts-container').appendTo($('d-article'));
  
    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');
  
    // create d-bibliography
    var bibliography = $('<d-bibliography></d-bibliography>');
    $('#distill-bibliography').wrap(bibliography);
  
    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;
  
    // replace citations with <d-cite>
    $('.citation').each(function(i, val) {
      appendix = true;
      var cites = $(this).attr('data-cites').split(" ");
      var dt_cite = $('<d-cite></d-cite>');
      dt_cite.attr('key', cites.join());
      $(this).replaceWith(dt_cite);
    });
    // remove refs
    $('#refs').remove();
  
    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();
  
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-toc a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });
  
    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');
  
    // replace code blocks with d-code
    $('pre>code').each(function(i, val) {
      var code = $(this);
      var pre = code.parent();
      var clz = "";
      var language = pre.attr('class');
      if (language) {
        // map unknown languages to "clike" (without this they just dissapear)
        if ($.inArray(language, ["bash", "clike", "css", "go", "html",
                                 "javascript", "js", "julia", "lua", "markdown",
                                 "markup", "mathml", "python", "svg", "xml"]) == -1)
          language = "clike";
        language = ' language="' + language + '"';
        var dt_code = $('<d-code block' + language + clz + '></d-code>');
        dt_code.text(code.text());
        pre.replaceWith(dt_code);
      } else {
        code.addClass('text-output').unwrap().changeElementType('pre');
      }
    });
  
    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {
  
      // capture layout
      var layout = $(this).attr('data-layout');
  
      // apply layout to markdown level block elements
      var elements = $(this).children().not('d-code, pre.text-output, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });
  
  
      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });
  
    // load distill framework
    load_distill_framework();
  
    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {
  
      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;
  
      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');
  
      // table of contents
      if (have_authors) // adjust border if we are in authors
        $('.d-toc').parent().addClass('d-article-with-toc');
  
      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');
  
      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }
  
      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');
  
      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");
  
      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }
  
       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }
  
      // inject pre code styles (can't do this with a global stylesheet b/c a shadow root is used)
      $('d-code').each(function(i, val) {
        var style = document.createElement('style');
        style.innerHTML = 'pre code { padding-left: 10px; font-size: 12px; border-left: 2px solid rgba(0,0,0,0.1); } ' +
                          '@media(min-width: 768px) { pre code { padding-left: 18px; font-size: 14px; } }';
        if (this.shadowRoot)
          this.shadowRoot.appendChild(style);
      });
  
      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();
  
      // clear polling timer
      clearInterval(tid);
  
      // show body now that everything is ready
      on_load_complete();
    }
  
    var tid = setInterval(distill_post_process, 50);
    distill_post_process();
  
  }
  
  function init_downlevel() {
  
    init_common();
  
     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));
  
    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
  
    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();
  
    // remove toc
    $('.d-toc-header').remove();
    $('.d-toc').remove();
    $('.d-toc-separator').remove();
  
    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });
  
  
    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);
  
    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);
  
    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();
  
    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();
  
    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));
  
    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });
  
    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));
  
    $('body').addClass('downlevel');
  
    on_load_complete();
  }
  
  
  function init_common() {
  
    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};
  
        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });
  
        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);
  
    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});
  
    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });
  
      }
    });
  
    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });
  
    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }
  
    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');
  
    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");
  
    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();
  
    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }
  
  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });
  
  </script>
  
  <!--/radix_placeholder_distill-->
  <script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <style type="text/css">
  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #000000;
      font-size: 15px;
    font-weight: 300;
  }
  
  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }
  
  .distill-site-nav a:hover {
    color: white;
    text-decoration:underline;
  }
  
  .distill-site-header {
  }
  
  .distill-site-footer {
  }
  
  @media print {
    .distill-site-nav {
      display: none;
    }
  }
  
  .l-body {
    display: none;
  }
  
  @media screen and (min-width: 768px) {
    .l-body {display: inline;}
  }
  
  </style>
  <!--/radix_placeholder_site_in_header-->

  <link rel="stylesheet" href="styles.css" type="text/css"/>

</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"<b style=\"font-size:45px;\">Cluster Markers and Cell Type Assignment<\/b><\/b>","authors":[{"author":"Rui Fu","authorURL":"#","affiliation":"&nbsp;","affiliationURL":"#"}],"publishedDate":"2019-08-14T00:00:00.000-06:00","citationText":"Fu, 2019"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a href="index.html" class="title">Single-cell RNA-seq Workshop</a>
<a href="1_intro.html">Introduction</a>
<a href="2_filtering_QC.html">Quality Control</a>
<a href="3_norm_viz_clustering.html">Clustering</a>
<a href="4_markers.html">Cell Type Annotation</a>
<a href="5_trajectories.html">Pseudotime</a>
<a href="6_alignment.html">Dataset Alignment</a>
<a href="7_multimodal.html">Multi-modal Data</a>
</div>
<div class="nav-right">
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Cluster Markers and Cell Type Assignment</h1>

</div>

<div class="d-byline">
  
  Rui Fu
  
<br/>August 14<sup>th</sup>, 2019
</div>

<div class="d-article">
<div class="layout-chunk" data-layout="l-body">

</div>
<p>The RMD file to follow along: <a href="https://scrnaseq-workshop.s3-us-west-2.amazonaws.com/4_markers.Rmd" class="uri">https://scrnaseq-workshop.s3-us-west-2.amazonaws.com/4_markers.Rmd</a></p>
<h2 id="finding-markers-and-differential-expression-analysis">finding markers and differential expression analysis</h2>
<p>After clustering, differential expression testing (DE analysis, similar to bulk RNA seq) finds gene signatures of each cluster/cell population, to give more insights into biology.</p>
<h3 id="many-proposed-methods">many proposed methods</h3>
<p>(for Seurat, we recommend default “wilcox” or “t”, good balances between speed and accuracy)</p>
<figure>
<img src="4_markers_files2/detests.jpg" alt="Soneson and Robinson, 2018" width="600" /><figcaption>Soneson and Robinson, 2018</figcaption>
</figure>
<h3 id="important-considerations">important considerations</h3>
<ol type="1">
<li><p>keep things simple first and look at the results (try not aligning, not regressing)</p></li>
<li><p>determine what variables to regress out (batch, nCount_RNA, percent_mito, cell cycle, if needed) during scaling (note this only affects dimension reduction and clustering)</p></li>
</ol>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
library(Seurat)
library(tidyverse)
# load data from saved RDS
sobj &lt;- readRDS(&quot;_vignettes/data/filtered_sobj.rds&quot;)

# log normalize
sobj_l &lt;- NormalizeData(sobj) # &lt;- skip this step if using scran normalization
sobj_l &lt;- ScaleData(sobj_l,
                    vars.to.regress = c(&quot;nCount_RNA&quot;,
                                        &quot;percent_mito&quot;))

# alternatively, sctransform is a one step normalization and scaling process
sobj_sc &lt;- suppressWarnings(SCTransform(object = sobj, # sctransform has some unhelpful warnings 
                                        vars.to.regress = c(&quot;percent_mito&quot;), # already corrects for transcript number
                                        verbose = FALSE))</code></pre>
</div>
<ol start="3" type="1">
<li>use normalized data for DE (slot is dependent on normalization method, also don’t use integrated assay)</li>
</ol>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# before normalization, &quot;data&quot; slot is the same as counts
sobj@assays$RNA@data[101:105,101:105] # raw counts
#&gt; 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;        ATCTTGACCTCCCA-1 ATCTTTCTTGTCCC-1 ATGAAGGACCTGTC-1
#&gt; PGD                   2                .                .
#&gt; APITD1                .                .                .
#&gt; DFFA                  .                .                .
#&gt; PEX14                 .                .                .
#&gt; CASZ1                 .                .                .
#&gt;        ATGACGTGATCGGT-1 ATGAGAGAAGTAGA-1
#&gt; PGD                   .                .
#&gt; APITD1                .                .
#&gt; DFFA                  .                .
#&gt; PEX14                 .                .
#&gt; CASZ1                 .                .

# after log normalization, results are stored in &quot;data&quot; slot
sobj_l@assays$RNA@data[101:105,101:105] # normalized
#&gt; 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;        ATCTTGACCTCCCA-1 ATCTTTCTTGTCCC-1 ATGAAGGACCTGTC-1
#&gt; PGD            2.727586                .                .
#&gt; APITD1         .                       .                .
#&gt; DFFA           .                       .                .
#&gt; PEX14          .                       .                .
#&gt; CASZ1          .                       .                .
#&gt;        ATGACGTGATCGGT-1 ATGAGAGAAGTAGA-1
#&gt; PGD                   .                .
#&gt; APITD1                .                .
#&gt; DFFA                  .                .
#&gt; PEX14                 .                .
#&gt; CASZ1                 .                .

# after sctransform, results are stored in new assay &quot;SCT&quot;
sobj_sc@assays$SCT@data[101:105,101:105] # normalized
#&gt; 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;        ATCTTGACCTCCCA-1 ATCTTTCTTGTCCC-1 ATGAAGGACCTGTC-1
#&gt; PGD            1.386294                .                .
#&gt; APITD1         .                       .                .
#&gt; DFFA           .                       .                .
#&gt; PEX14          .                       .                .
#&gt; CASZ1          .                       .                .
#&gt;        ATGACGTGATCGGT-1 ATGAGAGAAGTAGA-1
#&gt; PGD                   .                .
#&gt; APITD1                .                .
#&gt; DFFA                  .                .
#&gt; PEX14                 .                .
#&gt; CASZ1                 .                .
sobj_sc@assays$RNA@data[101:105,101:105] # still same as counts, make sure not to use this!
#&gt; 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;        ATCTTGACCTCCCA-1 ATCTTTCTTGTCCC-1 ATGAAGGACCTGTC-1
#&gt; PGD                   2                .                .
#&gt; APITD1                .                .                .
#&gt; DFFA                  .                .                .
#&gt; PEX14                 .                .                .
#&gt; CASZ1                 .                .                .
#&gt;        ATGACGTGATCGGT-1 ATGAGAGAAGTAGA-1
#&gt; PGD                   .                .
#&gt; APITD1                .                .
#&gt; DFFA                  .                .
#&gt; PEX14                 .                .
#&gt; CASZ1                 .                .</code></pre>
</div>
<ol start="4" type="1">
<li><p>Note that marker genes found is very dependent on clustering and the compared populations</p></li>
<li><p>P‐values are inflated, due to the cyclic nature of identifying the same variable genes as markers, which were used for dimension reduction and clustering. However, the ranking of genes based on P‐values is unaffected.</p></li>
</ol>
<h3 id="sidebar-how-to-regress-out-cell-cycle-heterogeneity-and-do-you-need-to">Sidebar: How to regress out cell cycle heterogeneity (and do you need to?)</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# many options to assess cell cycle effects on gene expression
# 1. ridge plot on key genes (from Seurat2 tutorial)
RidgePlot(sobj_l,
          features = c(&quot;MKI67&quot;, # smoothing eaves plots empty if no significant amount of cells express it
                       &quot;MCM6&quot;,
                       &quot;TOP2A&quot;,
                       &quot;PCNA&quot;,
                       &quot;ZFP36&quot;), # how actual robust gene expression should look
          ncol = 2)</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
# see https://satijalab.org/seurat/v3.0/cell_cycle_vignette.html for a case where some cells are actively cycling</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# 2. assess phase and PCA
# Seurat stores a list of cell cycle specific genes for humans
s.genes &lt;- Seurat::cc.genes$s.genes # obviously unsuitable for nonhuman
g2m.genes &lt;- Seurat::cc.genes$g2m.genes

s.genes
#&gt;  [1] &quot;MCM5&quot;     &quot;PCNA&quot;     &quot;TYMS&quot;     &quot;FEN1&quot;     &quot;MCM2&quot;     &quot;MCM4&quot;    
#&gt;  [7] &quot;RRM1&quot;     &quot;UNG&quot;      &quot;GINS2&quot;    &quot;MCM6&quot;     &quot;CDCA7&quot;    &quot;DTL&quot;     
#&gt; [13] &quot;PRIM1&quot;    &quot;UHRF1&quot;    &quot;MLF1IP&quot;   &quot;HELLS&quot;    &quot;RFC2&quot;     &quot;RPA2&quot;    
#&gt; [19] &quot;NASP&quot;     &quot;RAD51AP1&quot; &quot;GMNN&quot;     &quot;WDR76&quot;    &quot;SLBP&quot;     &quot;CCNE2&quot;   
#&gt; [25] &quot;UBR7&quot;     &quot;POLD3&quot;    &quot;MSH2&quot;     &quot;ATAD2&quot;    &quot;RAD51&quot;    &quot;RRM2&quot;    
#&gt; [31] &quot;CDC45&quot;    &quot;CDC6&quot;     &quot;EXO1&quot;     &quot;TIPIN&quot;    &quot;DSCC1&quot;    &quot;BLM&quot;     
#&gt; [37] &quot;CASP8AP2&quot; &quot;USP1&quot;     &quot;CLSPN&quot;    &quot;POLA1&quot;    &quot;CHAF1B&quot;   &quot;BRIP1&quot;   
#&gt; [43] &quot;E2F8&quot;
g2m.genes
#&gt;  [1] &quot;HMGB2&quot;   &quot;CDK1&quot;    &quot;NUSAP1&quot;  &quot;UBE2C&quot;   &quot;BIRC5&quot;   &quot;TPX2&quot;   
#&gt;  [7] &quot;TOP2A&quot;   &quot;NDC80&quot;   &quot;CKS2&quot;    &quot;NUF2&quot;    &quot;CKS1B&quot;   &quot;MKI67&quot;  
#&gt; [13] &quot;TMPO&quot;    &quot;CENPF&quot;   &quot;TACC3&quot;   &quot;FAM64A&quot;  &quot;SMC4&quot;    &quot;CCNB2&quot;  
#&gt; [19] &quot;CKAP2L&quot;  &quot;CKAP2&quot;   &quot;AURKB&quot;   &quot;BUB1&quot;    &quot;KIF11&quot;   &quot;ANP32E&quot; 
#&gt; [25] &quot;TUBB4B&quot;  &quot;GTSE1&quot;   &quot;KIF20B&quot;  &quot;HJURP&quot;   &quot;CDCA3&quot;   &quot;HN1&quot;    
#&gt; [31] &quot;CDC20&quot;   &quot;TTK&quot;     &quot;CDC25C&quot;  &quot;KIF2C&quot;   &quot;RANGAP1&quot; &quot;NCAPD2&quot; 
#&gt; [37] &quot;DLGAP5&quot;  &quot;CDCA2&quot;   &quot;CDCA8&quot;   &quot;ECT2&quot;    &quot;KIF23&quot;   &quot;HMMR&quot;   
#&gt; [43] &quot;AURKA&quot;   &quot;PSRC1&quot;   &quot;ANLN&quot;    &quot;LBR&quot;     &quot;CKAP5&quot;   &quot;CENPE&quot;  
#&gt; [49] &quot;CTCF&quot;    &quot;NEK2&quot;    &quot;G2E3&quot;    &quot;GAS2L3&quot;  &quot;CBX5&quot;    &quot;CENPA&quot;

# score and phase call is added to metadata
sobj_l &lt;- CellCycleScoring(sobj_l, 
                           s.features = s.genes,
                           g2m.features = g2m.genes,
                           set.ident = FALSE)
sobj_l@meta.data %&gt;% head()
#&gt;                  orig.ident nCount_RNA nFeature_RNA percent_mito
#&gt; AAACATACAACCAC-1    control       2419          779     3.017776
#&gt; AAACATTGAGCTAC-1    control       4901         1350     3.795144
#&gt; AAACGCTGACCAGT-1    control       2174          781     3.817847
#&gt; AAACGCTGGTTCTT-1    control       2259          789     3.098716
#&gt; AAAGTTTGATCACG-1    control       1265          441     3.478261
#&gt; AAATCATGACCACA-1    control       4125         1365     4.581818
#&gt;                      S.Score   G2M.Score Phase
#&gt; AAACATACAACCAC-1  0.10140023 -0.05909799     S
#&gt; AAACATTGAGCTAC-1 -0.03669937 -0.05460303    G1
#&gt; AAACGCTGACCAGT-1 -0.04334459 -0.04401501    G1
#&gt; AAACGCTGGTTCTT-1 -0.02891103  0.01944167   G2M
#&gt; AAAGTTTGATCACG-1  0.09548679 -0.07351012     S
#&gt; AAATCATGACCACA-1 -0.06976057 -0.01901402    G1

# check PCA to see if cell cycle has strong effects
sobj_l &lt;- FindVariableFeatures(sobj_l)
sobj_l &lt;- RunPCA(sobj_l, verbose = FALSE, npcs = 10)
DimPlot(sobj_l, group.by = &quot;Phase&quot;) # again, in this case, cell cycle does not have strong effects</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# 3. look for cell cycle-specific genes as main drivers of PCA
topPCAgenes &lt;- apply(
  sobj_l@reductions$pca@feature.loadings, # contribution of each gene to each PC is stored here
  MARGIN = 2, 
  FUN = function(x) names(sort(abs(x), decreasing = TRUE)[1:10]) # finds the most important genes in each PC
  )

topPCAgenes
#&gt;       PC_1     PC_2     PC_3       PC_4        PC_5           
#&gt;  [1,] &quot;CST3&quot;   &quot;NKG7&quot;   &quot;HLA-DQA1&quot; &quot;PF4&quot;       &quot;CKB&quot;          
#&gt;  [2,] &quot;S100A9&quot; &quot;CST7&quot;   &quot;CD79A&quot;    &quot;SDPR&quot;      &quot;FCGR3A&quot;       
#&gt;  [3,] &quot;TYROBP&quot; &quot;GZMA&quot;   &quot;CD79B&quot;    &quot;PPBP&quot;      &quot;MS4A7&quot;        
#&gt;  [4,] &quot;FTL&quot;    &quot;PRF1&quot;   &quot;HLA-DPB1&quot; &quot;HIST1H2AC&quot; &quot;RP11-290F20.3&quot;
#&gt;  [5,] &quot;LST1&quot;   &quot;GZMB&quot;   &quot;CD74&quot;     &quot;GNG11&quot;     &quot;RHOC&quot;         
#&gt;  [6,] &quot;FCN1&quot;   &quot;FGFBP2&quot; &quot;HLA-DPA1&quot; &quot;SPARC&quot;     &quot;SIGLEC10&quot;     
#&gt;  [7,] &quot;AIF1&quot;   &quot;GNLY&quot;   &quot;MS4A1&quot;    &quot;GP9&quot;       &quot;LILRA3&quot;       
#&gt;  [8,] &quot;LYZ&quot;    &quot;CTSW&quot;   &quot;HLA-DQB1&quot; &quot;NRGN&quot;      &quot;LGALS2&quot;       
#&gt;  [9,] &quot;FTH1&quot;   &quot;CD79A&quot;  &quot;HLA-DRB1&quot; &quot;RGS18&quot;     &quot;HMOX1&quot;        
#&gt; [10,] &quot;S100A8&quot; &quot;CCL4&quot;   &quot;PPBP&quot;     &quot;PTCRA&quot;     &quot;MS4A6A&quot;       
#&gt;       PC_6      PC_7      PC_8         PC_9       PC_10  
#&gt;  [1,] &quot;GZMK&quot;    &quot;CCL5&quot;    &quot;FCER1A&quot;     &quot;IFIT1&quot;    &quot;IFIT1&quot;
#&gt;  [2,] &quot;IL32&quot;    &quot;GZMK&quot;    &quot;CLEC10A&quot;    &quot;ISG15&quot;    &quot;ISG15&quot;
#&gt;  [3,] &quot;GZMB&quot;    &quot;FCER1A&quot;  &quot;ENHO&quot;       &quot;MX1&quot;      &quot;ANXA1&quot;
#&gt;  [4,] &quot;SPON2&quot;   &quot;KLRG1&quot;   &quot;SERPINF1&quot;   &quot;GZMK&quot;     &quot;PTTG1&quot;
#&gt;  [5,] &quot;VIM&quot;     &quot;CLEC10A&quot; &quot;CD1C&quot;       &quot;IFI6&quot;     &quot;GBP1&quot; 
#&gt;  [6,] &quot;S100A4&quot;  &quot;SELL&quot;    &quot;GZMK&quot;       &quot;OASL&quot;     &quot;LTB&quot;  
#&gt;  [7,] &quot;AKR1C3&quot;  &quot;LYAR&quot;    &quot;RBP7&quot;       &quot;TNFSF10&quot;  &quot;MX1&quot;  
#&gt;  [8,] &quot;B2M&quot;     &quot;NKG7&quot;    &quot;CLIC2&quot;      &quot;ACTG1&quot;    &quot;PCNA&quot; 
#&gt;  [9,] &quot;S100A10&quot; &quot;AKR1C3&quot;  &quot;RP6-91H8.3&quot; &quot;APOBEC3B&quot; &quot;HOPX&quot; 
#&gt; [10,] &quot;S100B&quot;   &quot;IGFBP7&quot;  &quot;S100A12&quot;    &quot;LYAR&quot;     &quot;CD27&quot;

topPCAgenes %&gt;%
  as.vector() %&gt;%
  intersect(c(s.genes, g2m.genes)) # see how many S and G2M genes intersect with that list
#&gt; [1] &quot;PCNA&quot;

# or just check top variable genes</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# if needed, regress out cell cycle effects
sobj_l &lt;- ScaleData(sobj_l,
                    vars.to.regress = c(&quot;nCount_RNA&quot;, &quot;percent_mito&quot;, &quot;S.Score&quot;, &quot;G2M.Score&quot;))

# alternatively, leave the difference of cycling vs noncycling in place, only regress out phases in actively cycling cells
sobj_l$CC.Difference &lt;- sobj_l$S.Score - sobj_l$G2M.Score
sobj_l &lt;- ScaleData(sobj_l,
                    vars.to.regress = c(&quot;nCount_RNA&quot;,
                                        &quot;percent_mito&quot;,
                                        &quot;CC.Difference&quot;))</code></pre>
</div>
<h3 id="find-all-markers-for-each-cluster">find all markers for each cluster</h3>
<p><code>FindAllMarkers</code> compares cells in each cluster to all other cells in the dataset. Typically, focus is given to genes upregulated in each cluster, ie markers.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# load clustered object
sobj_clusters &lt;- readRDS(&quot;_vignettes/data/clustered_sobj.rds&quot;)

# also check/set ident to the desired comparison
Idents(sobj_clusters) %&gt;% head()
#&gt; AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
#&gt;                0                2                3                3 
#&gt; AAAGTTTGATCACG-1 AAATCATGACCACA-1 
#&gt;                2                5 
#&gt; Levels: 0 1 2 3 4 5 6 7
Idents(sobj_clusters) &lt;- sobj_clusters@meta.data$clusters

markers_df &lt;- FindAllMarkers(sobj_clusters,
                             assay = &quot;RNA&quot;, # &lt;- be careful with sobj_sc and intergrated objects
                             slot = &quot;data&quot;,
                             only.pos = TRUE)
markers_df %&gt;% head()
#&gt;               p_val avg_logFC pct.1 pct.2     p_val_adj cluster  gene
#&gt; LDHB  7.282957e-247 1.1811710 0.926 0.483 9.156134e-243       0  LDHB
#&gt; RPS12 3.142126e-222 0.8524704 1.000 0.988 3.950281e-218       0 RPS12
#&gt; RPS25 1.258715e-211 0.8660816 1.000 0.966 1.582456e-207       0 RPS25
#&gt; CD3D  6.379927e-201 1.0471003 0.873 0.261 8.020845e-197       0  CD3D
#&gt; RPS27 1.100287e-196 0.7788876 0.999 0.990 1.383281e-192       0 RPS27
#&gt; RPS3  2.473657e-192 0.7175880 1.000 0.991 3.109882e-188       0  RPS3</code></pre>
</div>
<h3 id="find-de-genes-for-specific-cell-groups">find DE genes for specific cell groups</h3>
<p>For more control in the comparisons, use <code>FindMarkers</code>. Positive average log (natural) fold change represents higher expression of the gene in cells of <code>ident.1</code>. pct1 and 2 are percent detected in each ident/population.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# DE analysis for 2 clusters
markers_df2 &lt;- FindMarkers(sobj_clusters,
                           assay = &quot;RNA&quot;,
                           slot = &quot;data&quot;,
                           ident.1 = &quot;1&quot;,
                           ident.2 = &quot;2&quot;,
                           test.use = &quot;t&quot;)
markers_df2 %&gt;% head()
#&gt;                p_val avg_logFC pct.1 pct.2     p_val_adj
#&gt; LYZ     0.000000e+00  4.586764 1.000 0.426  0.000000e+00
#&gt; TYROBP  0.000000e+00  3.544515 0.994 0.111  0.000000e+00
#&gt; CST3    0.000000e+00  3.459219 0.992 0.185  0.000000e+00
#&gt; LGALS1 7.140047e-315  3.066783 0.980 0.131 8.976467e-311
#&gt; S100A4 1.217470e-300  3.278204 1.000 0.358 1.530603e-296
#&gt; S100A9 7.302898e-282  4.927262 0.996 0.136 9.181203e-278</code></pre>
</div>
<p><code>FindMarkers</code> defaults to the current active ident. To use other value groups, set idents to the intended column, or use the <code>group.by</code> argument.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# compare control vs treated in only cluster 1
markers_df3 &lt;- FindMarkers(sobj_clusters,
                           assay = &quot;RNA&quot;,
                           slot = &quot;data&quot;,
                           subset.ident = &quot;1&quot;, # &lt;- if needed, subset on current ident first, then switch idents
                           group.by = &quot;orig.ident&quot;, # &lt;- grouping cells by this metadata column
                           ident.1 = &quot;control&quot;,
                           ident.2 = &quot;treated&quot;)</code></pre>
</div>
<p>For greater control, assign new idents or new columns in metadata, and customize DE analysis pairs.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# set new idents for more customized comparisons
Idents(sobj_clusters,
       WhichCells(object = sobj_clusters,
                  idents = &quot;1&quot;,
                  expression = ZFP36 &gt;= 1, # would usually do this for transgenes etc
                  slot = &#39;data&#39;)) &lt;- &#39;ZFP36.pos&#39;
Idents(sobj_clusters,
       WhichCells(object = sobj_clusters,
                  idents = &quot;1&quot;,
                  expression = ZFP36 &lt; 1,
                  slot = &#39;data&#39;)) &lt;- &#39;ZFP36.neg&#39;
Idents(sobj_clusters) %&gt;% head() # note that with this method, some cell idents might not be changed
#&gt; AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
#&gt;                0                2                3                3 
#&gt; AAAGTTTGATCACG-1 AAATCATGACCACA-1 
#&gt;                2                5 
#&gt; Levels: ZFP36.neg ZFP36.pos 0 2 3 4 5 6 7

markers_df4 &lt;- FindMarkers(sobj_clusters,
                           ident.1 = &quot;ZFP36.pos&quot;,
                           ident.2 = &quot;ZFP36.neg&quot;)

# or something like:
# Idents(sobj_clusters,
#        WhichCells(object = sobj_clusters,
#                   idents = &quot;1&quot;,
#                   expression = sizeFactors &gt;= 1, # would usually do this for transgenes etc
#                   slot = &#39;data&#39;)) &lt;- &#39;cluster1_large&#39;</code></pre>
</div>
<p>Genes of interest can then be visualized as violin plots or feature plots.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
Idents(sobj_clusters) &lt;- sobj_clusters@meta.data$clusters # plots are grouped by active ident
VlnPlot(sobj_clusters, &quot;LYZ&quot;)</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
VlnPlot(sobj_clusters, c(&quot;ZFP36&quot;, &quot;ZFP36L2&quot;)) # can be a vector of gene names</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-11-2.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
FeaturePlot(sobj_clusters, &quot;LYZ&quot;)</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-11-3.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
FeaturePlot(sobj_clusters, c(&quot;ZFP36&quot;, &quot;ZFP36L2&quot;)) # can be a vector of gene names</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-11-4.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
FeaturePlot(sobj_clusters, &quot;ZFP36&quot;, split.by = &quot;orig.ident&quot;) # &lt;- split into panels based on metadata column</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-11-5.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<h2 id="cluster-identities">cluster identities</h2>
<p>Without venturing into the realm of philosphical debates on what a “cell type” constitutes, standard pratice is to use certain gene expression features to classify cells. This is often done manually, by visual inspection of key genes. Automated approaches that utilize a broader range of features are currently being developed.</p>
<h3 id="manual-check-and-cluster-identity-assignment">manual check and cluster identity assignment</h3>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# from the Seurat pancreas example
# we now have pan_celseq2, fully annotated in metadata column &quot;celltype&quot;, and pan_smartseq2, ready to be annotated
data_url = &quot;https://scrnaseq-workshop.s3-us-west-2.amazonaws.com&quot;
pan_celseq2 &lt;- readRDS(url(file.path(data_url, &quot;pan_celseq2.rds&quot;)))
pan_celseq2@meta.data %&gt;% head()
#&gt;             orig.ident nCount_RNA nFeature_RNA    tech celltype
#&gt; D28-1_1  SeuratProject  23438.618         5448 celseq2    alpha
#&gt; D28-1_15 SeuratProject  27252.978         5918 celseq2    alpha
#&gt; D28-1_17 SeuratProject  16177.715         4522 celseq2    alpha
#&gt; D28-1_29 SeuratProject  39356.525         7416 celseq2    alpha
#&gt; D28-1_30 SeuratProject  28958.986         6121 celseq2    alpha
#&gt; D28-1_39 SeuratProject   7389.849         2746 celseq2    alpha
#&gt;          RNA_snn_res.0.8 seurat_clusters
#&gt; D28-1_1                8               8
#&gt; D28-1_15               8               8
#&gt; D28-1_17               0               0
#&gt; D28-1_29               8               8
#&gt; D28-1_30               8               8
#&gt; D28-1_39               0               0
pan_smartseq2 &lt;- readRDS(url(file.path(data_url, &quot;pan_smartseq2.rds&quot;)))
pan_smartseq2@meta.data %&gt;% head()
#&gt;           orig.ident nCount_RNA nFeature_RNA      tech celltype
#&gt; AZ_B9  SeuratProject     654549         4433 smartseq2    alpha
#&gt; AZ_A6  SeuratProject     753413         4414 smartseq2    alpha
#&gt; AZ_C1  SeuratProject    2044839         5069 smartseq2    alpha
#&gt; AZ_A11 SeuratProject     705927         3900 smartseq2    alpha
#&gt; AZ_C2  SeuratProject    1338503         5367 smartseq2    alpha
#&gt; AZ_B11 SeuratProject    3378011         7137 smartseq2    alpha
#&gt;        RNA_snn_res.0.8 seurat_clusters
#&gt; AZ_B9               11              11
#&gt; AZ_A6                7               7
#&gt; AZ_C1               11              11
#&gt; AZ_A11              11              11
#&gt; AZ_C2               11              11
#&gt; AZ_B11              11              11
a &lt;- DimPlot(pan_smartseq2, label = TRUE) + NoLegend()
a</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-12-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
FeaturePlot(pan_smartseq2, c(&quot;IRX2&quot;,&quot;GC&quot;)) # marker genes for alpha</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-12-2.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
FeaturePlot(pan_smartseq2, c(&quot;IAPP&quot;,&quot;MAFA&quot;)) # marker genes for beta</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-12-3.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
FeaturePlot(pan_smartseq2, c(&quot;HHEX&quot;,&quot;LEPR&quot;)) # marker genes for delta</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-12-4.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
# optional, save the cluster numbers as &quot;cluster.id&quot;
pan_smartseq2 &lt;- StashIdent(object = pan_smartseq2, save.name = &quot;cluster.id&quot;) # or use AddMetaData

# Use RenameIdents to remap the idents from the current IDs to the new IDs 
pan_smartseq2 &lt;- RenameIdents(pan_smartseq2,
                              &quot;0&quot; = &quot;alpha&quot;,
                              &quot;1&quot; = &quot;alpha&quot;,
                              &quot;2&quot; = &quot;alpha&quot;,
                              &quot;3&quot; = &quot;alpha&quot;,
                              &quot;7&quot; = &quot;alpha&quot;,
                              &quot;9&quot; = &quot;alpha&quot;,
                              &quot;11&quot; = &quot;alpha&quot;, 
                              &quot;5&quot; = &quot;beta&quot;,
                              &quot;6&quot; = &quot;beta&quot;,
                              &quot;8&quot; = &quot;beta&quot;,
                              &quot;10&quot; = &quot;beta&quot;,
                              &quot;4&quot; = &quot;delta&quot;)

# Plot UMAP with new cluster IDs
b &lt;- DimPlot(object = pan_smartseq2,
        label = TRUE) + NoLegend()
b</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-12-5.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
cowplot::plot_grid(a,b)</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-12-6.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>
pan_smartseq2 &lt;- StashIdent(object = pan_smartseq2, save.name = &quot;cluster_name&quot;)
pan_smartseq2@meta.data %&gt;% head()
#&gt;           orig.ident nCount_RNA nFeature_RNA      tech celltype
#&gt; AZ_B9  SeuratProject     654549         4433 smartseq2    alpha
#&gt; AZ_A6  SeuratProject     753413         4414 smartseq2    alpha
#&gt; AZ_C1  SeuratProject    2044839         5069 smartseq2    alpha
#&gt; AZ_A11 SeuratProject     705927         3900 smartseq2    alpha
#&gt; AZ_C2  SeuratProject    1338503         5367 smartseq2    alpha
#&gt; AZ_B11 SeuratProject    3378011         7137 smartseq2    alpha
#&gt;        RNA_snn_res.0.8 seurat_clusters cluster.id cluster_name
#&gt; AZ_B9               11              11         11        alpha
#&gt; AZ_A6                7               7          7        alpha
#&gt; AZ_C1               11              11         11        alpha
#&gt; AZ_A11              11              11         11        alpha
#&gt; AZ_C2               11              11         11        alpha
#&gt; AZ_B11              11              11         11        alpha</code></pre>
</div>
<h3 id="towards-a-more-automated-approach-of-identity-assignment">towards a more automated approach of identity assignment</h3>
<ol type="1">
<li>using <code>Seurat</code>, inference from previous seurat object (requires very similar seurat object)</li>
</ol>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# reset idents of pan_smartseq2
Idents(pan_smartseq2) &lt;- &quot;cluster.id&quot;

pancreas &lt;- FindTransferAnchors(reference = pan_celseq2, query = pan_smartseq2, 
    dims = 1:30)
predictions &lt;- TransferData(anchorset = pancreas, refdata = pan_celseq2$celltype, 
    dims = 1:30)
pan_smartseq2 &lt;- AddMetaData(pan_smartseq2, metadata = predictions)
Idents(pan_smartseq2) &lt;- &quot;predicted.id&quot;
DimPlot(pan_smartseq2, label = TRUE) + NoLegend()</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<ol start="2" type="1">
<li>using <code>clustifyr</code>, and marker gene list (requires markers or gene list), using gene list enrichment methods</li>
</ol>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# reset idents of pan_smartseq2
Idents(pan_smartseq2) &lt;- &quot;cluster.id&quot;

# manually enter gene list
library(clustifyr)
beta &lt;- c(&quot;IAPP&quot;,&quot;MAFA&quot;)
alpha &lt;- c(&quot;IRX2&quot;,&quot;GC&quot;)
delta &lt;- c(&quot;HHEX&quot;,&quot;LEPR&quot;)
genelist &lt;- data.frame(alpha, beta, delta)
genelist
#&gt;   alpha beta delta
#&gt; 1  IRX2 IAPP  HHEX
#&gt; 2    GC MAFA  LEPR

res &lt;- clustify_lists(input = pan_smartseq2@assays$RNA@data,
                      marker = genelist,
                      cluster_info = pan_smartseq2@meta.data$seurat_clusters,
                      method = &quot;jaccard&quot;) # calculate jaccard index

res # a matrix of scores
#&gt;        alpha      beta   delta
#&gt; 0  2.6714700 0.0000000 0.00000
#&gt; 1  0.8017024 0.0000000 0.00000
#&gt; 10 0.0000000 0.8017024 0.00000
#&gt; 11 0.8017024 0.0000000 0.00000
#&gt; 2  2.6714700 0.0000000 0.00000
#&gt; 3  0.8017024 0.0000000 0.00000
#&gt; 4  0.0000000 0.0000000 2.67147
#&gt; 5  0.0000000 0.8017024 0.00000
#&gt; 6  0.0000000 0.8017024 0.00000
#&gt; 7  2.6714700 0.0000000 0.00000
#&gt; 8  0.0000000 0.8017024 0.00000
#&gt; 9  0.8017024 0.0000000 0.00000

res2 &lt;- cor_to_call(res, threshold = 0.5) # take the highest, and if lower than cutoff, assign &quot;unknown&quot;)
res2 # dataframe of identities
#&gt; # A tibble: 12 x 3
#&gt; # Groups:   cluster [12]
#&gt;    cluster type      r
#&gt;    &lt;chr&gt;   &lt;chr&gt; &lt;dbl&gt;
#&gt;  1 0       alpha 2.67 
#&gt;  2 1       alpha 0.802
#&gt;  3 11      alpha 0.802
#&gt;  4 2       alpha 2.67 
#&gt;  5 3       alpha 0.802
#&gt;  6 7       alpha 2.67 
#&gt;  7 9       alpha 0.802
#&gt;  8 10      beta  0.802
#&gt;  9 5       beta  0.802
#&gt; 10 6       beta  0.802
#&gt; 11 8       beta  0.802
#&gt; 12 4       delta 2.67

namedres &lt;- structure(res2$type, names = res2$cluster) # make named vector
namedres
#&gt;       0       1      11       2       3       7       9      10 
#&gt; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot;  &quot;beta&quot; 
#&gt;       5       6       8       4 
#&gt;  &quot;beta&quot;  &quot;beta&quot;  &quot;beta&quot; &quot;delta&quot;

# seurat idents are factors, which may have strange behavior if coerced to numbers/index
namedres[as.character(Idents(pan_smartseq2))][1:10] 
#&gt;      11       7      11      11      11      11      11      11 
#&gt; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot; &quot;alpha&quot; 
#&gt;      11      11 
#&gt; &quot;alpha&quot; &quot;alpha&quot;
pan_smartseq2@meta.data$celltype &lt;- namedres[as.character(Idents(pan_smartseq2))] 
DimPlot(pan_smartseq2, label = T, group.by = &quot;celltype&quot;) + NoLegend()</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-14-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<ol start="3" type="1">
<li>using <code>clustifyr</code>, and transcriptome profiles (from other scRNAseq, bulk RNAseq, or microarray), using ranked correlation of highly variable genes (<a href="https://github.com/hemberg-lab/scmap">scmap</a> is another similar package)</li>
</ol>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
# build a reference from expression matrix data
ref &lt;- average_clusters(mat = pan_celseq2@assays$RNA@data,
                        cluster_info = pan_celseq2@meta.data,
                        cluster_col = &quot;celltype&quot;)

res &lt;- clustify(input = pan_smartseq2@assays$RNA@data,
                ref_mat = ref,
                metadata = pan_smartseq2@meta.data,
                query_genes = pan_smartseq2@assays$RNA@var.features, # using Seurat computed variable genes
                cluster_col = &quot;seurat_clusters&quot;)
res2 &lt;- cor_to_call(res, threshold = 0.5) # anything below 0.5 correlation are labeled as &quot;unknown&quot;
res2
#&gt; # A tibble: 12 x 3
#&gt; # Groups:   cluster [12]
#&gt;    cluster type      r
#&gt;    &lt;chr&gt;   &lt;chr&gt; &lt;dbl&gt;
#&gt;  1 0       alpha 0.740
#&gt;  2 1       alpha 0.716
#&gt;  3 11      alpha 0.673
#&gt;  4 2       alpha 0.755
#&gt;  5 3       alpha 0.739
#&gt;  6 7       alpha 0.731
#&gt;  7 9       alpha 0.741
#&gt;  8 10      beta  0.726
#&gt;  9 5       beta  0.787
#&gt; 10 6       beta  0.805
#&gt; 11 8       beta  0.787
#&gt; 12 4       delta 0.801

# same as last section
namedres &lt;- structure(res2$type, names = res2$cluster) 
pan_smartseq2@meta.data$celltype &lt;- namedres[as.character(Idents(pan_smartseq2))] 
DimPlot(pan_smartseq2, label = T, group.by = &quot;celltype&quot;) + NoLegend()</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-15-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<p><code>clustifyr</code> also takes seurat objects as input directly, finds various needed data, and output another object with identities assigned.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
ref &lt;- seurat_ref(pan_celseq2,
                  cluster_col = &quot;celltype&quot;)

res &lt;- clustify(input = pan_smartseq2,
                ref_mat = ref,
                cluster_col = &quot;seurat_clusters&quot;,
                seurat_out = TRUE)

res
#&gt; An object of class Seurat 
#&gt; 34363 features across 1443 samples within 1 assay 
#&gt; Active assay: RNA (34363 features)
#&gt;  2 dimensional reductions calculated: pca, umap

# look at UMAP
DimPlot(res, label = TRUE, group.by = &quot;type&quot;) + NoLegend() # saved in &quot;type&quot; metadata column</code></pre>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-16-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<p>Also see more tutorials at <a href="https://rnabioco.github.io/clustifyr/"><code>clustifyr</code></a>, and prebuilt references from large scRNAseq/bulk RNAseq/microarray datasets at package <a href="https://github.com/rnabioco/clustifyrdata"><code>clustifyrdata</code></a>.</p>
<h3 id="cell-type-composition">cell type composition</h3>
<p>Insight into different samples can be gained from the proportion of cells that fall into each cell type. Unfortunately, no dedicated tools are available for statistical testing.</p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
tab1 &lt;- sobj_clusters@meta.data %&gt;% group_by(orig.ident, clusters) %&gt;% tally() # counting up all combinations
tab1
#&gt; # A tibble: 16 x 3
#&gt; # Groups:   orig.ident [2]
#&gt;    orig.ident clusters     n
#&gt;    &lt;fct&gt;      &lt;fct&gt;    &lt;int&gt;
#&gt;  1 control    0          553
#&gt;  2 control    1          243
#&gt;  3 control    2          176
#&gt;  4 control    3          190
#&gt;  5 control    4           74
#&gt;  6 control    5           80
#&gt;  7 control    6           14
#&gt;  8 control    7            6
#&gt;  9 treated    0          589
#&gt; 10 treated    1          250
#&gt; 11 treated    2          176
#&gt; 12 treated    3          150
#&gt; 13 treated    4           87
#&gt; 14 treated    5           78
#&gt; 15 treated    6           18
#&gt; 16 treated    7            6
tab2 &lt;- tab1 %&gt;% spread(key = clusters, value = n) # spread out into &quot;wide&quot; form
tab2
#&gt; # A tibble: 2 x 9
#&gt; # Groups:   orig.ident [2]
#&gt;   orig.ident   `0`   `1`   `2`   `3`   `4`   `5`   `6`   `7`
#&gt;   &lt;fct&gt;      &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
#&gt; 1 control      553   243   176   190    74    80    14     6
#&gt; 2 treated      589   250   176   150    87    78    18     6
tab3 &lt;- tab1 %&gt;% group_by(orig.ident) %&gt;%
  mutate(n = n/sum(n)) %&gt;% # convert counts to proportions first
  spread(key = clusters, value = n) 
tab3
#&gt; # A tibble: 2 x 9
#&gt; # Groups:   orig.ident [2]
#&gt;   orig.ident   `0`   `1`   `2`   `3`    `4`    `5`    `6`     `7`
#&gt;   &lt;fct&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
#&gt; 1 control    0.414 0.182 0.132 0.142 0.0554 0.0599 0.0105 0.00449
#&gt; 2 treated    0.435 0.185 0.130 0.111 0.0643 0.0576 0.0133 0.00443</code></pre>
</div>
<p>and save tables to disk with <code>write_csv</code></p>
<div class="layout-chunk" data-layout="l-body">
<pre class="r"><code>
write_csv(tab3, &quot;data/perc_clusters.csv&quot;)
tab4 &lt;- read_csv(&quot;data/perc_clusters.csv&quot;)</code></pre>
</div>
<h2 id="other-things-to-do-with-marker-genes">other things to do with marker genes</h2>
<ol type="1">
<li><p>gene list to pathway activity score, via <a href="https://github.com/aertslab/AUCell"><code>AUCell</code></a></p></li>
<li><p>if TF expression is too low, consider <a href="https://github.com/aertslab/SCENIC"><code>SCENIC</code></a> for TF activity inference</p></li>
<li><p>standard GO term enrichment tools <a href="https://biit.cs.ut.ee/gprofiler/page/r"><code>gProfiler2</code></a>, <a href="https://github.com/wjawaid/enrichR"><code>enrichR</code></a>, <a href="https://github.com/ctlab/fgsea%5D"><code>fgsea</code></a>, etc etc</p></li>
</ol>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="updates-and-corrections">Corrections</h3>
<p>If you see mistakes or want to suggest changes, please <a href="https://github.com/rnabioco/cellar/issues/new">create an issue</a> on the source repository.</p>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
