<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #545454; font-weight: bold; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #a1024a; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007faa; font-weight: bold; } /* ControlFlow */
code span.ch { color: #008000; } /* Char */
code span.cn { color: #d91e18; } /* Constant */
code span.co { color: #545454; } /* Comment */
code span.cv { color: #545454; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #aa5d00; } /* DataType */
code span.dv { color: #a1024a; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #a1024a; } /* Float */
code span.fu { color: #4254a7; } /* Function */
code span.im { } /* Import */
code span.in { color: #545454; font-weight: bold; } /* Information */
code span.kw { color: #007faa; font-weight: bold; } /* Keyword */
code span.op { color: #696969; } /* Operator */
code span.ot { color: #007faa; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #008000; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #008000; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #008000; } /* VerbatimString */
code span.wa { color: #545454; font-weight: bold; font-style: italic; } /* Warning */
</style>

  <!--radix_placeholder_meta_tags-->
  <title>Single-cell RNA-seq Workshop: &lt;b style="font-size:45px;"&gt;Cluster Markers and Cell Type Assignment&lt;/b&gt;&lt;/b&gt;</title>


  <link rel="icon" type="image/png" href="img/logo.png"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2020-11-11"/>
  <meta property="article:created" itemprop="dateCreated" content="2020-11-11"/>
  <meta name="article:author" content="Rui Fu, Kent Riemondy"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="Single-cell RNA-seq Workshop: &lt;b style=&quot;font-size:45px;&quot;&gt;Cluster Markers and Cell Type Assignment&lt;/b&gt;&lt;/b&gt;"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="Single-cell RNA-seq Workshop"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="Single-cell RNA-seq Workshop: &lt;b style=&quot;font-size:45px;&quot;&gt;Cluster Markers and Cell Type Assignment&lt;/b&gt;&lt;/b&gt;"/>
  <meta property="twitter:site" content="@rnabioco"/>

  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","author","date","output"]}},"value":[{"type":"character","attributes":{},"value":["<b style=\"font-size:45px;\">Cluster Markers and Cell Type Assignment<\/b><\/b>"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name"]}},"value":[{"type":"character","attributes":{},"value":["Rui Fu, Kent Riemondy"]}]}]},{"type":"character","attributes":{},"value":["November 11<sup>th<\/sup>, 2020"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["html_document"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["toc","toc_depth","self_contained"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"integer","attributes":{},"value":[1]},{"type":"logical","attributes":{},"value":[true]}]}]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  <meta name="distill:offset" content=""/>

  <script type="application/javascript">

    window.headroom_prevent_pin = false;

    window.document.addEventListener("DOMContentLoaded", function (event) {

      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');

      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });

      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });

      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>

  <style type="text/css">

  /* Theme (user-documented overrideables for nav appearance) */

  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #0F2E3D;
    font-size: 15px;
    font-weight: 300;
  }

  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }

  .distill-site-nav a:hover {
    color: white;
  }

  @media print {
    .distill-site-nav {
      display: none;
    }
  }

  .distill-site-header {

  }

  .distill-site-footer {

  }


  /* Site Header */

  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }

  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }


  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }

  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }

  .distill-site-header .title {
    font-size: 18px;
    min-width: 150px;
  }

  .distill-site-header .logo {
    padding: 0;
  }

  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }

  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }



  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }


  .distill-site-header .nav-toggle {
    display: none;
  }

  .nav-dropdown {
    display: inline-block;
    position: relative;
  }

  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }

  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }

  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .nav-dropdown-active {
    display: block;
  }

  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }

  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }

  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }

  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }

  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }


  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative; min-height: 500px; }
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }

  /* Site Footer */

  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }

  /* Headroom */

  d-title {
    padding-top: 6rem;
  }

  @media print {
    d-title {
      padding-top: 4rem;
    }
  }

  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }

  .headroom--transition {
    transition: all .4s ease-in-out;
  }

  .headroom--unpinned {
    top: -100px;
  }

  .headroom--pinned {
    top: 0;
  }

  </style>

  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <script src="site_libs/autocomplete-0.37.1/autocomplete.min.js"></script>
  <script src="site_libs/fuse-6.4.1/fuse.min.js"></script>

  <script type="application/javascript">

  function getMeta(metaName) {
    var metas = document.getElementsByTagName('meta');
    for (let i = 0; i < metas.length; i++) {
      if (metas[i].getAttribute('name') === metaName) {
        return metas[i].getAttribute('content');
      }
    }
    return '';
  }

  function offsetURL(url) {
    var offset = getMeta('distill:offset');
    return offset ? offset + '/' + url : url;
  }

  function createIndex() {
    var options = {
      keys: [
        "title",
        "categories",
        "description",
        "contents"
      ]
    };
    return new window.Fuse([],options);
  }

  function createFuseIndex() {

    // create fuse index
    var options = { keys: ["title", "description", "contents"] };
    var fuse = new window.Fuse([], options);

    // fetch the main search.json
    return fetch(offsetURL('search.json'))
      .then(function(response) {
        if (response.status == 200) {
          return response.json().then(function(json) {
            // index main articles
            json.articles.forEach(function(article) {
              fuse.add(article);
            });
            // download collections and index their articles
            return Promise.all(json.collections.map(function(collection) {
              return fetch(offsetURL(collection)).then(function(response) {
                if (response.status === 200) {
                  return response.json().then(function(articles) {
                    articles.forEach(function(article) {
                      fuse.add(article);
                    });
                  })
                } else {
                  return Promise.reject(
                    new Error('Unexpected status from search index request: ' +
                              response.status)
                  );
                }
              });
            })).then(function() {
              return fuse;
            });
          });

        } else {
          return Promise.reject(
            new Error('Unexpected status from search index request: ' +
                        response.status)
          );
        }
      });
  }

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // get search element (bail if we don't have one)
    var searchEl = window.document.getElementById('distill-search');
    if (!searchEl)
      return;

    createFuseIndex()
      .then(function(fuse) {

        // make search box visible
        searchEl.classList.remove('hidden');

        // initialize autocomplete
        var options = {
          autoselect: true,
          hint: false,
          minLength: 2,
        };
        window.autocomplete(searchEl, options, [{
          source: function(query, callback) {
            const searchOptions = {
              isCaseSensitive: false,
              shouldSort: true,
              minMatchCharLength: 2,
              limit: 10,
              keys: [
                { name: 'title', weight: 20 },
                { name: 'categories', weight: 15 },
                { name: 'description', weight: 10 },
                { name: 'contents', weight: 5 },
              ],
            };
            var results = fuse.search(query, searchOptions);
            callback(results
              .map(function(result) { return result.item; })
              .filter(function(item) { return !!item.description; })
            );
          },
          templates: {
            suggestion: function(suggestion) {
              var html = `
                <div class="search-item">
                  <h3>${suggestion.title}</h3>
                  <div class="search-item-description">
                    ${suggestion.description}
                  </div>
                  <div class="search-item-preview">
                    <img src="${suggestion.preview ? offsetURL(suggestion.preview) : ''}"</img>
                  </div>
                </div>
              `;
              return html;
            }
          }
        }]).on('autocomplete:selected', function(event, suggestion) {
          window.location.href = offsetURL(suggestion.path);
        });
        // remove inline display style on autocompleter (we want to
        // manage responsive display via css)
        $('.algolia-autocomplete').css("display", "");
      })
      .catch(function(error) {
        console.log(error);
      });

  });

  </script>

  <style type="text/css">

  /* Algolioa Autocomplete */

  .algolia-autocomplete {
    display: inline-block;
    margin-left: 10px;
    vertical-align: sub;
    background-color: white;
    color: black;
    padding: 6px;
    padding-top: 8px;
    padding-bottom: 0;
    border-radius: 6px;
    border: 1px #0F2E3D solid;
    width: 180px;
  }


  @media screen and (max-width: 768px) {
    .distill-site-nav .algolia-autocomplete {
      display: none;
      visibility: hidden;
    }
    .distill-site-nav.responsive .algolia-autocomplete {
      display: inline-block;
      visibility: visible;
    }
    .distill-site-nav.responsive .algolia-autocomplete .aa-dropdown-menu {
      margin-left: 0;
      width: 400px;
      max-height: 400px;
    }
  }

  .algolia-autocomplete .aa-input, .algolia-autocomplete .aa-hint {
    width: 90%;
    outline: none;
    border: none;
  }

  .algolia-autocomplete .aa-hint {
    color: #999;
  }
  .algolia-autocomplete .aa-dropdown-menu {
    width: 550px;
    max-height: 70vh;
    overflow-x: visible;
    overflow-y: scroll;
    padding: 5px;
    margin-top: 3px;
    margin-left: -150px;
    background-color: #fff;
    border-radius: 5px;
    border: 1px solid #999;
    border-top: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion {
    cursor: pointer;
    padding: 5px 4px;
    border-bottom: 1px solid #eee;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion:last-of-type {
    border-bottom: none;
    margin-bottom: 2px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item {
    overflow: hidden;
    font-size: 0.8em;
    line-height: 1.4em;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item h3 {
    margin-block-start: 0;
    margin-block-end: 5px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-description {
    display: inline-block;
    overflow: hidden;
    height: 2.8em;
    width: 80%;
    margin-right: 4%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview {
    display: inline-block;
    width: 15%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img {
    height: 3em;
    width: auto;
    display: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img[src] {
    display: initial;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion.aa-cursor {
    background-color: #eee;
  }
  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion em {
    font-weight: bold;
    font-style: normal;
  }

  </style>


  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: hidden;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="site_libs/header-attrs-2.5/header-attrs.js"></script>
  <script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
  <script src="site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <style type="text/css">
  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #000000;
      font-size: 15px;
    font-weight: 300;
  }

  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }

  .distill-site-nav a:hover {
    color: white;
    text-decoration:underline;
  }

  .distill-site-header {
  }

  .distill-site-footer {
  }

  @media print {
    .distill-site-nav {
      display: none;
    }
  }

  .l-body {
    display: none;
  }

  @media screen and (min-width: 768px) {
    .l-body {display: inline;}
  }

  </style>
  <!--/radix_placeholder_site_in_header-->

  <link rel="stylesheet" href="styles.css" type="text/css"/>

</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"<b style=\"font-size:45px;\">Cluster Markers and Cell Type Assignment<\/b><\/b>","authors":[{"author":"Rui Fu, Kent Riemondy","authorURL":"#","affiliation":"&nbsp;","affiliationURL":"#","orcidID":""}],"publishedDate":"2020-11-11T00:00:00.000-07:00","citationText":"Riemondy, 2020"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a href="index.html" class="title">Single-cell RNA-seq Workshop</a>
<a href="1_intro.html">Introduction</a>
<a href="previous.html">Previous-workshops</a>
</div>
<div class="nav-right">
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Cluster Markers and Cell Type Assignment</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->

</div>

<div class="d-byline">
  
  Rui Fu, Kent Riemondy
  
<br/>November 11<sup>th</sup>, 2020
</div>

<div class="d-article">
<div class="layout-chunk" data-layout="l-body">

</div>
<h2 id="finding-markers-and-differential-expression-analysis">finding markers and differential expression analysis</h2>
<p>After clustering, differential expression testing (DE analysis, similar to bulk RNA seq) finds gene signatures of each cluster/cell population, to give more insights into biology.</p>
<h3 id="many-proposed-methods">many proposed methods</h3>
<p>(In general, we recommend default “wilcox” or “t”, good balances between speed and accuracy)</p>
<figure>
<img src="docs/previous/2019/_vignettes/4_markers_files2/detests.jpg" width="600" alt="" /><figcaption>Soneson and Robinson, 2018</figcaption>
</figure>
<p>In Seurat, we can use the <code>FindAllMarkers()</code> function, which will perform the <code>wilcox.test()</code> by default . Specifically, the function will iterate through each cluster, comparing the cells in one cluster to the cells in all of the other clusters. The test is run for every gene that is detectable above certain thresholds. The output is then filtered to identify significant genes with a positive fold-change (i.e. higher in abundance the cluster).</p>
<p>We recommend using the <code>wilcoxauc()</code> function from the <code>presto</code> package (installed from github), which will do the same thing but in a (tiny) fraction of the time, due to implementation with C++ backend.</p>
<h3 id="important-considerations">important considerations</h3>
<ol type="1">
<li><p>Keep things simple first and look at the results (try not integrating, not regressing), look at the output, and determine what variables to regress out (batch, nCount_RNA, percent_mito, cell cycle, if needed) during scaling (note this only affects dimension reduction and clustering).</p></li>
<li><p>Note that the p-values from these tests will be absurdly low (near or at 0). This is in part due to the large number of samples tested (e.g. each cell is considered a sample) but also due to the data being clustered based on the gene expression. Testing for differential expression between clusters will always result in some differntially expressed genes… because that’s what makes them different clusters in the first place, hence the logical is somewhat circular and inflate p-values. Nevertheless we can use the p-values to rank the genes, but you shouldn’t just blindly trust that a p-value &lt; 0.05 is something meaningful in single cell data.</p></li>
</ol>
<p>A couple more words on preprocessing:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://satijalab.org/seurat'>Seurat</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://tidyverse.tidyverse.org'>tidyverse</a></span><span class='op'>)</span>
<span class='co'># load data from saved RDS</span>
<span class='va'>sobj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='st'>"/Users/rf/class/scRNAseq/cellar_v2/cellar/_vignettes/data/filtered_sobj.rds"</span><span class='op'>)</span>

<span class='co'># log normalize</span>
<span class='va'>sobj_l</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/NormalizeData.html'>NormalizeData</a></span><span class='op'>(</span><span class='va'>sobj</span><span class='op'>)</span> <span class='co'># &lt;- skip this step if using scran normalization</span>
<span class='va'>sobj_l</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/ScaleData.html'>ScaleData</a></span><span class='op'>(</span><span class='va'>sobj_l</span>,
                    vars.to.regress <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"nCount_RNA"</span>,
                                        <span class='st'>"percent_mito"</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># alternatively, sctransform is a one step normalization and scaling process</span>
<span class='va'>sobj_sc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/warning.html'>suppressWarnings</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/SCTransform.html'>SCTransform</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>sobj</span>, <span class='co'># sctransform has some unhelpful warnings </span>
                                        vars.to.regress <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"percent_mito"</span><span class='op'>)</span>, <span class='co'># already corrects for transcript number</span>
                                        verbose <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<ol start="3" type="1">
<li>Use normalized data for DE (slot is dependent on normalization method, also don’t use integrated assay)</li>
</ol>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># before normalization, "data" slot is the same as counts</span>
<span class='va'>sobj</span><span class='op'>@</span><span class='va'>assays</span><span class='op'>$</span><span class='va'>RNA</span><span class='op'>@</span><span class='va'>data</span><span class='op'>[</span><span class='fl'>101</span><span class='op'>:</span><span class='fl'>105</span>,<span class='fl'>101</span><span class='op'>:</span><span class='fl'>105</span><span class='op'>]</span> <span class='co'># raw counts</span>
</code></pre>
</div>
<pre><code>#&gt; 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;        ATCTTGACCTCCCA-1 ATCTTTCTTGTCCC-1 ATGAAGGACCTGTC-1
#&gt; PGD                   2                .                .
#&gt; APITD1                .                .                .
#&gt; DFFA                  .                .                .
#&gt; PEX14                 .                .                .
#&gt; CASZ1                 .                .                .
#&gt;        ATGACGTGATCGGT-1 ATGAGAGAAGTAGA-1
#&gt; PGD                   .                .
#&gt; APITD1                .                .
#&gt; DFFA                  .                .
#&gt; PEX14                 .                .
#&gt; CASZ1                 .                .</code></pre>
<div class="sourceCode">
<pre><code>
<span class='co'># after log normalization, results are stored in "data" slot</span>
<span class='va'>sobj_l</span><span class='op'>@</span><span class='va'>assays</span><span class='op'>$</span><span class='va'>RNA</span><span class='op'>@</span><span class='va'>data</span><span class='op'>[</span><span class='fl'>101</span><span class='op'>:</span><span class='fl'>105</span>,<span class='fl'>101</span><span class='op'>:</span><span class='fl'>105</span><span class='op'>]</span> <span class='co'># normalized</span>
</code></pre>
</div>
<pre><code>#&gt; 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;        ATCTTGACCTCCCA-1 ATCTTTCTTGTCCC-1 ATGAAGGACCTGTC-1
#&gt; PGD            2.727586                .                .
#&gt; APITD1         .                       .                .
#&gt; DFFA           .                       .                .
#&gt; PEX14          .                       .                .
#&gt; CASZ1          .                       .                .
#&gt;        ATGACGTGATCGGT-1 ATGAGAGAAGTAGA-1
#&gt; PGD                   .                .
#&gt; APITD1                .                .
#&gt; DFFA                  .                .
#&gt; PEX14                 .                .
#&gt; CASZ1                 .                .</code></pre>
<div class="sourceCode">
<pre><code>
<span class='co'># after sctransform, results are stored in new assay "SCT"</span>
<span class='va'>sobj_sc</span><span class='op'>@</span><span class='va'>assays</span><span class='op'>$</span><span class='va'>SCT</span><span class='op'>@</span><span class='va'>data</span><span class='op'>[</span><span class='fl'>101</span><span class='op'>:</span><span class='fl'>105</span>,<span class='fl'>101</span><span class='op'>:</span><span class='fl'>105</span><span class='op'>]</span> <span class='co'># normalized</span>
</code></pre>
</div>
<pre><code>#&gt; 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;        ATCTTGACCTCCCA-1 ATCTTTCTTGTCCC-1 ATGAAGGACCTGTC-1
#&gt; PGD            1.386294                .                .
#&gt; APITD1         .                       .                .
#&gt; DFFA           .                       .                .
#&gt; PEX14          .                       .                .
#&gt; CASZ1          .                       .                .
#&gt;        ATGACGTGATCGGT-1 ATGAGAGAAGTAGA-1
#&gt; PGD                   .                .
#&gt; APITD1                .                .
#&gt; DFFA                  .                .
#&gt; PEX14                 .                .
#&gt; CASZ1                 .                .</code></pre>
<div class="sourceCode">
<pre><code><span class='va'>sobj_sc</span><span class='op'>@</span><span class='va'>assays</span><span class='op'>$</span><span class='va'>RNA</span><span class='op'>@</span><span class='va'>data</span><span class='op'>[</span><span class='fl'>101</span><span class='op'>:</span><span class='fl'>105</span>,<span class='fl'>101</span><span class='op'>:</span><span class='fl'>105</span><span class='op'>]</span> <span class='co'># still same as counts, make sure not to use this!</span>
</code></pre>
</div>
<pre><code>#&gt; 5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;        ATCTTGACCTCCCA-1 ATCTTTCTTGTCCC-1 ATGAAGGACCTGTC-1
#&gt; PGD                   2                .                .
#&gt; APITD1                .                .                .
#&gt; DFFA                  .                .                .
#&gt; PEX14                 .                .                .
#&gt; CASZ1                 .                .                .
#&gt;        ATGACGTGATCGGT-1 ATGAGAGAAGTAGA-1
#&gt; PGD                   .                .
#&gt; APITD1                .                .
#&gt; DFFA                  .                .
#&gt; PEX14                 .                .
#&gt; CASZ1                 .                .</code></pre>
</div>
<ol start="4" type="1">
<li>Note that marker genes found is very dependent on clustering and the compared populations.</li>
</ol>
<h3 id="find-all-markers-for-each-cluster">find all markers for each cluster</h3>
<p><code>FindAllMarkers</code> compares cells in each cluster to all other cells in the dataset. Typically, focus is given to genes upregulated in each cluster, ie markers.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># load clustered object</span>
<span class='va'>sobj_clusters</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='st'>"/Users/rf/class/scRNAseq/cellar_v2/cellar/_vignettes/data/clustered_sobj.rds"</span><span class='op'>)</span>

<span class='co'># also check/set ident to the desired comparison</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/Idents.html'>Idents</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span><span class='op'>)</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
#&gt;                0                2                3                3 
#&gt; AAAGTTTGATCACG-1 AAATCATGACCACA-1 
#&gt;                2                5 
#&gt; Levels: 0 1 2 3 4 5 6 7</code></pre>
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/Idents.html'>Idents</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>sobj_clusters</span><span class='op'>@</span><span class='va'>meta.data</span><span class='op'>$</span><span class='va'>clusters</span>

<span class='va'>markers_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FindAllMarkers.html'>FindAllMarkers</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span>,
                             assay <span class='op'>=</span> <span class='st'>"RNA"</span>, <span class='co'># be careful with sobj_sc and intergrated objects</span>
                             slot <span class='op'>=</span> <span class='st'>"data"</span>,
                             only.pos <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>markers_df</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt;               p_val avg_logFC pct.1 pct.2     p_val_adj cluster  gene
#&gt; LDHB  7.282957e-247 1.1811710 0.926 0.483 9.156134e-243       0  LDHB
#&gt; RPS12 3.142126e-222 0.8524704 1.000 0.988 3.950281e-218       0 RPS12
#&gt; RPS25 1.258715e-211 0.8660816 1.000 0.966 1.582456e-207       0 RPS25
#&gt; CD3D  6.379927e-201 1.0471003 0.873 0.261 8.020845e-197       0  CD3D
#&gt; RPS27 1.100287e-196 0.7788876 0.999 0.990 1.383281e-192       0 RPS27
#&gt; RPS3  2.473657e-192 0.7175880 1.000 0.991 3.109882e-188       0  RPS3</code></pre>
</div>
<p>pct.1: fraction detected in the cluster in question, pct.2: fraction detected in the rest of the cells avg_logFC: expression comparison of the cluster in question (&gt; 0 means higher) vs rest of the cells</p>
<p>And the equivalent but faster way using <code>presto::wilcoxauc</code></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>presto</span><span class='op'>)</span>
<span class='va'>markers</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/presto/man/wilcoxauc.html'>wilcoxauc</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span>, <span class='st'>"clusters"</span><span class='op'>)</span> <span class='co'># compute markers</span>
<span class='va'>markers</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt;         feature group     avgExpr         logFC statistic       auc
#&gt; 1    AL627309.1     0 0.001629726 -0.0026757987  881460.0 0.4986152
#&gt; 2 RP11-206L10.2     0 0.001184927 -0.0005575694  882400.0 0.4991470
#&gt; 3     LINC00115     0 0.004489924 -0.0033339184  880359.5 0.4979927
#&gt; 4         NOC2L     0 0.098153914 -0.0002869219  878388.0 0.4968775
#&gt; 5        KLHL17     0 0.003237845 -0.0001995984  882809.0 0.4993783
#&gt; 6       PLEKHN1     0 0.002453311 -0.0006959034  883943.0 0.5000198
#&gt;        pval      padj     pct_in   pct_out
#&gt; 1 0.2190641 0.4027602 0.17513135 0.4521964
#&gt; 2 0.3101718 0.5006018 0.08756567 0.2583979
#&gt; 3 0.2069766 0.3875648 0.43782837 0.8397933
#&gt; 4 0.5860930 0.7458610 9.10683012 9.8191214
#&gt; 5 0.5812118 0.7417390 0.26269702 0.3875969
#&gt; 6 0.9843333 0.9886585 0.26269702 0.2583979</code></pre>
</div>
<h3 id="find-de-genes-for-specific-cell-groups">find DE genes for specific cell groups</h3>
<p>For more control in the comparisons, use <code>FindMarkers</code>. Positive average log (natural) fold change represents higher expression of the gene in cells of <code>ident.1</code>. pct1 and 2 are percent detected in each ident/population.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># DE analysis for 2 clusters</span>
<span class='va'>markers_df2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FindMarkers.html'>FindMarkers</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span>,
                           assay <span class='op'>=</span> <span class='st'>"RNA"</span>,
                           slot <span class='op'>=</span> <span class='st'>"data"</span>,
                           ident.1 <span class='op'>=</span> <span class='st'>"1"</span>,
                           ident.2 <span class='op'>=</span> <span class='st'>"2"</span>,
                           test.use <span class='op'>=</span> <span class='st'>"t"</span><span class='op'>)</span>
<span class='va'>markers_df2</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt;                p_val avg_logFC pct.1 pct.2     p_val_adj
#&gt; LYZ     0.000000e+00  4.586764 1.000 0.426  0.000000e+00
#&gt; TYROBP  0.000000e+00  3.544515 0.994 0.111  0.000000e+00
#&gt; CST3    0.000000e+00  3.459219 0.992 0.185  0.000000e+00
#&gt; LGALS1 7.140047e-315  3.066783 0.980 0.131 8.976467e-311
#&gt; S100A4 1.217470e-300  3.278204 1.000 0.358 1.530603e-296
#&gt; S100A9 7.302898e-282  4.927262 0.996 0.136 9.181203e-278</code></pre>
</div>
<p>pct.1: fraction detected in the cells of ident.1, pct.2: fraction detected in the cells of ident.2 avg_logFC: expression comparison of ident.1 vs ident.2 (&gt; 0 means higher in ident.1)</p>
<p><code>FindMarkers</code> defaults to the current active ident. To use other value groups, set idents to the intended column, or use the <code>group.by</code> argument.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># compare control vs treated in only cluster 1</span>
<span class='va'>markers_df3</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FindMarkers.html'>FindMarkers</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span>,
                           assay <span class='op'>=</span> <span class='st'>"RNA"</span>,
                           slot <span class='op'>=</span> <span class='st'>"data"</span>,
                           subset.ident <span class='op'>=</span> <span class='st'>"1"</span>, <span class='co'># &lt;- if needed, subset on current ident first, then switch idents</span>
                           group.by <span class='op'>=</span> <span class='st'>"orig.ident"</span>, <span class='co'># &lt;- grouping cells by this metadata column</span>
                           ident.1 <span class='op'>=</span> <span class='st'>"control"</span>,
                           ident.2 <span class='op'>=</span> <span class='st'>"treated"</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Genes of interest can then be visualized as violin plots or feature plots.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># violin plots</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/Idents.html'>Idents</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>sobj_clusters</span><span class='op'>@</span><span class='va'>meta.data</span><span class='op'>$</span><span class='va'>clusters</span> <span class='co'># plots are grouped by active ident</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/VlnPlot.html'>VlnPlot</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span>, <span class='st'>"LYZ"</span><span class='op'>)</span>
</code></pre>
</div>
<img src="4_markers_files/figure-html5/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/VlnPlot.html'>VlnPlot</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"LYZ"</span>, <span class='st'>"S100A4"</span><span class='op'>)</span><span class='op'>)</span> <span class='co'># can be a vector of gene names</span>
</code></pre>
</div>
<img src="4_markers_files/figure-html5/unnamed-chunk-7-2.png" width="576" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code>
<span class='co'># project on UMAP</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span>, <span class='st'>"LYZ"</span><span class='op'>)</span>
</code></pre>
</div>
<img src="4_markers_files/figure-html5/unnamed-chunk-7-3.png" width="576" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"LYZ"</span>, <span class='st'>"S100A4"</span><span class='op'>)</span><span class='op'>)</span> <span class='co'># can be a vector of gene names</span>
</code></pre>
</div>
<img src="4_markers_files/figure-html5/unnamed-chunk-7-4.png" width="576" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>sobj_clusters</span>, <span class='st'>"LYZ"</span>, split.by <span class='op'>=</span> <span class='st'>"orig.ident"</span><span class='op'>)</span> <span class='co'># &lt;- split into panels based on metadata column</span>
</code></pre>
</div>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-7-5.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<h3 id="sidebar-how-to-regress-out-cell-cycle-ribosomal-gene-heterogeneity-and-do-you-need-to">Sidebar: How to regress out cell cycle / ribosomal gene heterogeneity (and do you need to?)</h3>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># 1. assess phase and PCA</span>
<span class='co'># Seurat stores a list of cell cycle specific genes for humans</span>
<span class='va'>s.genes</span> <span class='op'>&lt;-</span> <span class='fu'>Seurat</span><span class='fu'>::</span><span class='va'><a href='https://rdrr.io/pkg/Seurat/man/cc.genes.html'>cc.genes</a></span><span class='op'>$</span><span class='va'>s.genes</span> <span class='co'># obviously unsuitable for nonhuman</span>
<span class='va'>g2m.genes</span> <span class='op'>&lt;-</span> <span class='fu'>Seurat</span><span class='fu'>::</span><span class='va'><a href='https://rdrr.io/pkg/Seurat/man/cc.genes.html'>cc.genes</a></span><span class='op'>$</span><span class='va'>g2m.genes</span>

<span class='va'>s.genes</span>
</code></pre>
</div>
<pre><code>#&gt;  [1] &quot;MCM5&quot;     &quot;PCNA&quot;     &quot;TYMS&quot;     &quot;FEN1&quot;     &quot;MCM2&quot;     &quot;MCM4&quot;    
#&gt;  [7] &quot;RRM1&quot;     &quot;UNG&quot;      &quot;GINS2&quot;    &quot;MCM6&quot;     &quot;CDCA7&quot;    &quot;DTL&quot;     
#&gt; [13] &quot;PRIM1&quot;    &quot;UHRF1&quot;    &quot;MLF1IP&quot;   &quot;HELLS&quot;    &quot;RFC2&quot;     &quot;RPA2&quot;    
#&gt; [19] &quot;NASP&quot;     &quot;RAD51AP1&quot; &quot;GMNN&quot;     &quot;WDR76&quot;    &quot;SLBP&quot;     &quot;CCNE2&quot;   
#&gt; [25] &quot;UBR7&quot;     &quot;POLD3&quot;    &quot;MSH2&quot;     &quot;ATAD2&quot;    &quot;RAD51&quot;    &quot;RRM2&quot;    
#&gt; [31] &quot;CDC45&quot;    &quot;CDC6&quot;     &quot;EXO1&quot;     &quot;TIPIN&quot;    &quot;DSCC1&quot;    &quot;BLM&quot;     
#&gt; [37] &quot;CASP8AP2&quot; &quot;USP1&quot;     &quot;CLSPN&quot;    &quot;POLA1&quot;    &quot;CHAF1B&quot;   &quot;BRIP1&quot;   
#&gt; [43] &quot;E2F8&quot;</code></pre>
<div class="sourceCode">
<pre><code><span class='va'>g2m.genes</span>
</code></pre>
</div>
<pre><code>#&gt;  [1] &quot;HMGB2&quot;   &quot;CDK1&quot;    &quot;NUSAP1&quot;  &quot;UBE2C&quot;   &quot;BIRC5&quot;   &quot;TPX2&quot;   
#&gt;  [7] &quot;TOP2A&quot;   &quot;NDC80&quot;   &quot;CKS2&quot;    &quot;NUF2&quot;    &quot;CKS1B&quot;   &quot;MKI67&quot;  
#&gt; [13] &quot;TMPO&quot;    &quot;CENPF&quot;   &quot;TACC3&quot;   &quot;FAM64A&quot;  &quot;SMC4&quot;    &quot;CCNB2&quot;  
#&gt; [19] &quot;CKAP2L&quot;  &quot;CKAP2&quot;   &quot;AURKB&quot;   &quot;BUB1&quot;    &quot;KIF11&quot;   &quot;ANP32E&quot; 
#&gt; [25] &quot;TUBB4B&quot;  &quot;GTSE1&quot;   &quot;KIF20B&quot;  &quot;HJURP&quot;   &quot;CDCA3&quot;   &quot;HN1&quot;    
#&gt; [31] &quot;CDC20&quot;   &quot;TTK&quot;     &quot;CDC25C&quot;  &quot;KIF2C&quot;   &quot;RANGAP1&quot; &quot;NCAPD2&quot; 
#&gt; [37] &quot;DLGAP5&quot;  &quot;CDCA2&quot;   &quot;CDCA8&quot;   &quot;ECT2&quot;    &quot;KIF23&quot;   &quot;HMMR&quot;   
#&gt; [43] &quot;AURKA&quot;   &quot;PSRC1&quot;   &quot;ANLN&quot;    &quot;LBR&quot;     &quot;CKAP5&quot;   &quot;CENPE&quot;  
#&gt; [49] &quot;CTCF&quot;    &quot;NEK2&quot;    &quot;G2E3&quot;    &quot;GAS2L3&quot;  &quot;CBX5&quot;    &quot;CENPA&quot;</code></pre>
<div class="sourceCode">
<pre><code>
<span class='co'># score and phase call is added to metadata</span>
<span class='va'>sobj_l</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/CellCycleScoring.html'>CellCycleScoring</a></span><span class='op'>(</span><span class='va'>sobj_l</span>, 
                           s.features <span class='op'>=</span> <span class='va'>s.genes</span>,
                           g2m.features <span class='op'>=</span> <span class='va'>g2m.genes</span>,
                           set.ident <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>sobj_l</span><span class='op'>@</span><span class='va'>meta.data</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt;                  orig.ident nCount_RNA nFeature_RNA percent_mito
#&gt; AAACATACAACCAC-1    control       2419          779     3.017776
#&gt; AAACATTGAGCTAC-1    control       4901         1350     3.795144
#&gt; AAACGCTGACCAGT-1    control       2174          781     3.817847
#&gt; AAACGCTGGTTCTT-1    control       2259          789     3.098716
#&gt; AAAGTTTGATCACG-1    control       1265          441     3.478261
#&gt; AAATCATGACCACA-1    control       4125         1365     4.581818
#&gt;                      S.Score   G2M.Score Phase
#&gt; AAACATACAACCAC-1  0.10140023 -0.05909799     S
#&gt; AAACATTGAGCTAC-1 -0.03669937 -0.05460303    G1
#&gt; AAACGCTGACCAGT-1 -0.04334459 -0.04401501    G1
#&gt; AAACGCTGGTTCTT-1 -0.02891103  0.01944167   G2M
#&gt; AAAGTTTGATCACG-1  0.09548679 -0.07351012     S
#&gt; AAATCATGACCACA-1 -0.06976057 -0.01901402    G1</code></pre>
<div class="sourceCode">
<pre><code>
<span class='co'># check PCA to see if cell cycle has strong effects</span>
<span class='va'>sobj_l</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html'>FindVariableFeatures</a></span><span class='op'>(</span><span class='va'>sobj_l</span><span class='op'>)</span>
<span class='va'>sobj_l</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/RunPCA.html'>RunPCA</a></span><span class='op'>(</span><span class='va'>sobj_l</span>, verbose <span class='op'>=</span> <span class='cn'>FALSE</span>, npcs <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>sobj_l</span>, group.by <span class='op'>=</span> <span class='st'>"Phase"</span><span class='op'>)</span> <span class='co'># again, in this case, cell cycle does not have strong effects</span>
</code></pre>
</div>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># 2. look for cell cycle-specific genes as main drivers of PCA</span>
<span class='va'>topPCAgenes</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span>
  <span class='va'>sobj_l</span><span class='op'>@</span><span class='va'>reductions</span><span class='op'>$</span><span class='va'>pca</span><span class='op'>@</span><span class='va'>feature.loadings</span>, <span class='co'># contribution of each gene to each PC is stored here</span>
  MARGIN <span class='op'>=</span> <span class='fl'>2</span>, 
  FUN <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/sort.html'>sort</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>abs</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>, decreasing <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span><span class='op'>]</span><span class='op'>)</span> <span class='co'># finds the most important genes in each PC</span>
  <span class='op'>)</span>

<span class='va'>topPCAgenes</span>
</code></pre>
</div>
<pre><code>#&gt;       PC_1     PC_2     PC_3       PC_4        PC_5           
#&gt;  [1,] &quot;CST3&quot;   &quot;NKG7&quot;   &quot;HLA-DQA1&quot; &quot;PF4&quot;       &quot;CKB&quot;          
#&gt;  [2,] &quot;S100A9&quot; &quot;CST7&quot;   &quot;CD79A&quot;    &quot;SDPR&quot;      &quot;FCGR3A&quot;       
#&gt;  [3,] &quot;TYROBP&quot; &quot;GZMA&quot;   &quot;CD79B&quot;    &quot;PPBP&quot;      &quot;MS4A7&quot;        
#&gt;  [4,] &quot;FTL&quot;    &quot;PRF1&quot;   &quot;HLA-DPB1&quot; &quot;HIST1H2AC&quot; &quot;RP11-290F20.3&quot;
#&gt;  [5,] &quot;LST1&quot;   &quot;GZMB&quot;   &quot;CD74&quot;     &quot;GNG11&quot;     &quot;RHOC&quot;         
#&gt;  [6,] &quot;FCN1&quot;   &quot;FGFBP2&quot; &quot;HLA-DPA1&quot; &quot;SPARC&quot;     &quot;SIGLEC10&quot;     
#&gt;  [7,] &quot;AIF1&quot;   &quot;GNLY&quot;   &quot;MS4A1&quot;    &quot;GP9&quot;       &quot;LILRA3&quot;       
#&gt;  [8,] &quot;LYZ&quot;    &quot;CTSW&quot;   &quot;HLA-DQB1&quot; &quot;NRGN&quot;      &quot;LGALS2&quot;       
#&gt;  [9,] &quot;FTH1&quot;   &quot;CD79A&quot;  &quot;HLA-DRB1&quot; &quot;RGS18&quot;     &quot;HMOX1&quot;        
#&gt; [10,] &quot;S100A8&quot; &quot;CCL4&quot;   &quot;PPBP&quot;     &quot;PTCRA&quot;     &quot;MS4A6A&quot;       
#&gt;       PC_6      PC_7      PC_8         PC_9       PC_10  
#&gt;  [1,] &quot;GZMK&quot;    &quot;CCL5&quot;    &quot;FCER1A&quot;     &quot;IFIT1&quot;    &quot;IFIT1&quot;
#&gt;  [2,] &quot;IL32&quot;    &quot;GZMK&quot;    &quot;CLEC10A&quot;    &quot;ISG15&quot;    &quot;ISG15&quot;
#&gt;  [3,] &quot;GZMB&quot;    &quot;FCER1A&quot;  &quot;ENHO&quot;       &quot;MX1&quot;      &quot;ANXA1&quot;
#&gt;  [4,] &quot;SPON2&quot;   &quot;KLRG1&quot;   &quot;SERPINF1&quot;   &quot;GZMK&quot;     &quot;PTTG1&quot;
#&gt;  [5,] &quot;VIM&quot;     &quot;CLEC10A&quot; &quot;CD1C&quot;       &quot;IFI6&quot;     &quot;GBP1&quot; 
#&gt;  [6,] &quot;S100A4&quot;  &quot;SELL&quot;    &quot;GZMK&quot;       &quot;OASL&quot;     &quot;LTB&quot;  
#&gt;  [7,] &quot;AKR1C3&quot;  &quot;LYAR&quot;    &quot;RBP7&quot;       &quot;TNFSF10&quot;  &quot;MX1&quot;  
#&gt;  [8,] &quot;B2M&quot;     &quot;NKG7&quot;    &quot;CLIC2&quot;      &quot;ACTG1&quot;    &quot;PCNA&quot; 
#&gt;  [9,] &quot;S100A10&quot; &quot;AKR1C3&quot;  &quot;RP6-91H8.3&quot; &quot;APOBEC3B&quot; &quot;HOPX&quot; 
#&gt; [10,] &quot;S100B&quot;   &quot;IGFBP7&quot;  &quot;S100A12&quot;    &quot;LYAR&quot;     &quot;CD27&quot;</code></pre>
<div class="sourceCode">
<pre><code>
<span class='va'>topPCAgenes</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/base/vector.html'>as.vector</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>intersect</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>s.genes</span>, <span class='va'>g2m.genes</span><span class='op'>)</span><span class='op'>)</span> <span class='co'># see how many S and G2M genes intersect with that list</span>
</code></pre>
</div>
<pre><code>#&gt; [1] &quot;PCNA&quot;</code></pre>
<div class="sourceCode">
<pre><code>
<span class='co'># or just check top variable genes</span>
</code></pre>
</div>
</div>
<h2 id="cluster-identities">cluster identities</h2>
<p>Without venturing into the realm of philosphical debates on what a “cell type” constitutes, standard pratice is to use certain gene expression features to classify cells. This is often done manually, by visual inspection of key genes. Automated approaches that utilize a broader range of features are currently being developed.</p>
<h3 id="manual-inspection-of-key-genes-using-expertise-in-the-lab">manual inspection of key genes using expertise in the lab</h3>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># from the Seurat pancreas example</span>
<span class='co'># we now have pan_celseq2, fully annotated in metadata column "celltype", and pan_smartseq2, ready to be annotated</span>
<span class='va'>data_url</span> <span class='op'>=</span> <span class='st'>"https://scrnaseq-workshop.s3-us-west-2.amazonaws.com"</span>
<span class='va'>pan_celseq2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/connections.html'>url</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>data_url</span>, <span class='st'>"pan_celseq2.rds"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>pan_celseq2</span><span class='op'>@</span><span class='va'>meta.data</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt;             orig.ident nCount_RNA nFeature_RNA    tech celltype
#&gt; D28-1_1  SeuratProject  23438.618         5448 celseq2    alpha
#&gt; D28-1_15 SeuratProject  27252.978         5918 celseq2    alpha
#&gt; D28-1_17 SeuratProject  16177.715         4522 celseq2    alpha
#&gt; D28-1_29 SeuratProject  39356.525         7416 celseq2    alpha
#&gt; D28-1_30 SeuratProject  28958.986         6121 celseq2    alpha
#&gt; D28-1_39 SeuratProject   7389.849         2746 celseq2    alpha
#&gt;          RNA_snn_res.0.8 seurat_clusters
#&gt; D28-1_1                8               8
#&gt; D28-1_15               8               8
#&gt; D28-1_17               0               0
#&gt; D28-1_29               8               8
#&gt; D28-1_30               8               8
#&gt; D28-1_39               0               0</code></pre>
<div class="sourceCode">
<pre><code><span class='va'>pan_smartseq2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/connections.html'>url</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='va'>data_url</span>, <span class='st'>"pan_smartseq2.rds"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>pan_smartseq2</span><span class='op'>@</span><span class='va'>meta.data</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt;           orig.ident nCount_RNA nFeature_RNA      tech celltype
#&gt; AZ_B9  SeuratProject     654549         4433 smartseq2    alpha
#&gt; AZ_A6  SeuratProject     753413         4414 smartseq2    alpha
#&gt; AZ_C1  SeuratProject    2044839         5069 smartseq2    alpha
#&gt; AZ_A11 SeuratProject     705927         3900 smartseq2    alpha
#&gt; AZ_C2  SeuratProject    1338503         5367 smartseq2    alpha
#&gt; AZ_B11 SeuratProject    3378011         7137 smartseq2    alpha
#&gt;        RNA_snn_res.0.8 seurat_clusters
#&gt; AZ_B9               11              11
#&gt; AZ_A6                7               7
#&gt; AZ_C1               11              11
#&gt; AZ_A11              11              11
#&gt; AZ_C2               11              11
#&gt; AZ_B11              11              11</code></pre>
<div class="sourceCode">
<pre><code><span class='va'>a</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>pan_smartseq2</span>, label <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/SeuratTheme.html'>NoLegend</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='va'>a</span>
</code></pre>
</div>
<img src="4_markers_files/figure-html5/unnamed-chunk-10-1.png" width="576" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>pan_smartseq2</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"IRX2"</span>,<span class='st'>"GC"</span><span class='op'>)</span><span class='op'>)</span> <span class='co'># marker genes for alpha</span>
</code></pre>
</div>
<img src="4_markers_files/figure-html5/unnamed-chunk-10-2.png" width="576" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>pan_smartseq2</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"IAPP"</span>,<span class='st'>"MAFA"</span><span class='op'>)</span><span class='op'>)</span> <span class='co'># marker genes for beta</span>
</code></pre>
</div>
<img src="4_markers_files/figure-html5/unnamed-chunk-10-3.png" width="576" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>pan_smartseq2</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"HHEX"</span>,<span class='st'>"LEPR"</span><span class='op'>)</span><span class='op'>)</span> <span class='co'># marker genes for delta</span>
</code></pre>
</div>
<img src="4_markers_files/figure-html5/unnamed-chunk-10-4.png" width="576" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code>
<span class='co'># optional, save the cluster numbers as "cluster.id"</span>
<span class='va'>pan_smartseq2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/Idents.html'>StashIdent</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>pan_smartseq2</span>, save.name <span class='op'>=</span> <span class='st'>"cluster.id"</span><span class='op'>)</span> <span class='co'># or use AddMetaData</span>

<span class='co'># Use RenameIdents to remap the idents from the current IDs to the new IDs </span>
<span class='va'>pan_smartseq2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/Idents.html'>RenameIdents</a></span><span class='op'>(</span><span class='va'>pan_smartseq2</span>,
                              <span class='st'>"0"</span> <span class='op'>=</span> <span class='st'>"alpha"</span>,
                              <span class='st'>"1"</span> <span class='op'>=</span> <span class='st'>"alpha"</span>,
                              <span class='st'>"2"</span> <span class='op'>=</span> <span class='st'>"alpha"</span>,
                              <span class='st'>"3"</span> <span class='op'>=</span> <span class='st'>"alpha"</span>,
                              <span class='st'>"7"</span> <span class='op'>=</span> <span class='st'>"alpha"</span>,
                              <span class='st'>"9"</span> <span class='op'>=</span> <span class='st'>"alpha"</span>,
                              <span class='st'>"11"</span> <span class='op'>=</span> <span class='st'>"alpha"</span>, 
                              <span class='st'>"5"</span> <span class='op'>=</span> <span class='st'>"beta"</span>,
                              <span class='st'>"6"</span> <span class='op'>=</span> <span class='st'>"beta"</span>,
                              <span class='st'>"8"</span> <span class='op'>=</span> <span class='st'>"beta"</span>,
                              <span class='st'>"10"</span> <span class='op'>=</span> <span class='st'>"beta"</span>,
                              <span class='st'>"4"</span> <span class='op'>=</span> <span class='st'>"delta"</span><span class='op'>)</span>

<span class='co'># Plot UMAP with new cluster IDs</span>
<span class='va'>b</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>pan_smartseq2</span>,
        label <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/SeuratTheme.html'>NoLegend</a></span><span class='op'>(</span><span class='op'>)</span>

<span class='fu'>cowplot</span><span class='fu'>::</span><span class='fu'><a href='https://wilkelab.org/cowplot/reference/plot_grid.html'>plot_grid</a></span><span class='op'>(</span><span class='va'>a</span>,<span class='va'>b</span><span class='op'>)</span>
</code></pre>
</div>
<img src="4_markers_files/figure-html5/unnamed-chunk-10-5.png" width="576" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code>
<span class='co'># remember to save the new idents into meta.data</span>
<span class='va'>pan_smartseq2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/Idents.html'>StashIdent</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>pan_smartseq2</span>, save.name <span class='op'>=</span> <span class='st'>"cluster_name"</span><span class='op'>)</span>
<span class='va'>pan_smartseq2</span><span class='op'>@</span><span class='va'>meta.data</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt;           orig.ident nCount_RNA nFeature_RNA      tech celltype
#&gt; AZ_B9  SeuratProject     654549         4433 smartseq2    alpha
#&gt; AZ_A6  SeuratProject     753413         4414 smartseq2    alpha
#&gt; AZ_C1  SeuratProject    2044839         5069 smartseq2    alpha
#&gt; AZ_A11 SeuratProject     705927         3900 smartseq2    alpha
#&gt; AZ_C2  SeuratProject    1338503         5367 smartseq2    alpha
#&gt; AZ_B11 SeuratProject    3378011         7137 smartseq2    alpha
#&gt;        RNA_snn_res.0.8 seurat_clusters cluster.id cluster_name
#&gt; AZ_B9               11              11         11        alpha
#&gt; AZ_A6                7               7          7        alpha
#&gt; AZ_C1               11              11         11        alpha
#&gt; AZ_A11              11              11         11        alpha
#&gt; AZ_C2               11              11         11        alpha
#&gt; AZ_B11              11              11         11        alpha</code></pre>
</div>
<h3 id="towards-a-more-automated-approach-of-identity-assignment">towards a more automated approach of identity assignment</h3>
<ol type="1">
<li>using <code>Seurat</code>, inference from previous seurat object (requires very similar seurat object), see <a href="https://satijalab.org/seurat/v3.2/integration.html">Seurat Integration Vignette</a> for more details.</li>
</ol>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># reset idents of pan_smartseq2</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/Idents.html'>Idents</a></span><span class='op'>(</span><span class='va'>pan_smartseq2</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='st'>"cluster.id"</span>

<span class='va'>pancreas</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FindTransferAnchors.html'>FindTransferAnchors</a></span><span class='op'>(</span>reference <span class='op'>=</span> <span class='va'>pan_celseq2</span>, query <span class='op'>=</span> <span class='va'>pan_smartseq2</span>, 
    dims <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>30</span><span class='op'>)</span>
<span class='va'>predictions</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/TransferData.html'>TransferData</a></span><span class='op'>(</span>anchorset <span class='op'>=</span> <span class='va'>pancreas</span>, refdata <span class='op'>=</span> <span class='va'>pan_celseq2</span><span class='op'>$</span><span class='va'>celltype</span>, 
    dims <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>30</span><span class='op'>)</span>
<span class='va'>pan_smartseq2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/AddMetaData.html'>AddMetaData</a></span><span class='op'>(</span><span class='va'>pan_smartseq2</span>, metadata <span class='op'>=</span> <span class='va'>predictions</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/Idents.html'>Idents</a></span><span class='op'>(</span><span class='va'>pan_smartseq2</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='st'>"predicted.id"</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>pan_smartseq2</span>, label <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/SeuratTheme.html'>NoLegend</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<ol start="2" type="1">
<li>using other classification packages such as <code>clustifyr</code>.</li>
</ol>
<p><code>clustifyr</code> works by comparing the average gene expression in each cluster to a reference matrix that contains average gene signatures of reference cell types. The reference can be built from other single cell data, bulk-rna-seq, or other sources. Ranked Spearman correlation is used to compare the reference to the clusters. Only the variable genes are used for the correlation.</p>
<p>In order to compare our dataset we need to use a publically available reference dataset. In this example we use the other annotated experiment.Please see <a href="https://github.com/rnabioco/clustifyrdatahub">clustifyrdatahub</a> for prebuilt general references available.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://github.com/rnabioco/clustifyr#readme'>clustifyr</a></span><span class='op'>)</span>
<span class='co'># build a reference from expression matrix data</span>
<span class='va'>ref</span> <span class='op'>&lt;-</span> <span class='fu'><a href='http://rnabioco.github.io/clustifyr/reference/average_clusters.html'>average_clusters</a></span><span class='op'>(</span><span class='va'>pan_celseq2</span><span class='op'>@</span><span class='va'>assays</span><span class='op'>$</span><span class='va'>RNA</span><span class='op'>@</span><span class='va'>data</span>,
                        metadata <span class='op'>=</span> <span class='va'>pan_celseq2</span><span class='op'>@</span><span class='va'>meta.data</span>,
                        cluster_col <span class='op'>=</span> <span class='st'>"celltype"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span><span class='op'>(</span><span class='va'>ref</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1] 34363     3</code></pre>
<div class="sourceCode">
<pre><code><span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'><a href='http://rnabioco.github.io/clustifyr/reference/clustify.html'>clustify</a></span><span class='op'>(</span>input <span class='op'>=</span> <span class='va'>pan_smartseq2</span><span class='op'>@</span><span class='va'>assays</span><span class='op'>$</span><span class='va'>RNA</span><span class='op'>@</span><span class='va'>data</span>,
                ref_mat <span class='op'>=</span> <span class='va'>ref</span>,
                metadata <span class='op'>=</span> <span class='va'>pan_smartseq2</span><span class='op'>@</span><span class='va'>meta.data</span>,
                cluster_col <span class='op'>=</span> <span class='st'>"cluster.id"</span>,
                query_genes <span class='op'>=</span> <span class='va'>pan_smartseq2</span><span class='op'>@</span><span class='va'>assays</span><span class='op'>$</span><span class='va'>RNA</span><span class='op'>@</span><span class='va'>var.features</span>, <span class='co'># using Seurat computed variable genes</span>
                <span class='op'>)</span>
<span class='va'>res</span> <span class='co'># a matrix of correlation values between each cluster and reference cell types</span>
</code></pre>
</div>
<pre><code>#&gt;        alpha      beta     delta
#&gt; 0  0.7382342 0.6399227 0.6088692
#&gt; 1  0.7106232 0.5793125 0.5715972
#&gt; 10 0.5854603 0.7216101 0.6418888
#&gt; 11 0.6708439 0.6172400 0.5833315
#&gt; 2  0.7496097 0.6165381 0.5963039
#&gt; 3  0.7320384 0.6300580 0.6072802
#&gt; 4  0.6102327 0.7002068 0.7932875
#&gt; 5  0.6098376 0.7814286 0.6615390
#&gt; 6  0.6251296 0.8029008 0.7063750
#&gt; 7  0.7232752 0.6290922 0.6093593
#&gt; 8  0.6437217 0.7805895 0.6762429
#&gt; 9  0.7269138 0.5570430 0.5184847</code></pre>
<div class="sourceCode">
<pre><code><span class='va'>res2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='http://rnabioco.github.io/clustifyr/reference/cor_to_call.html'>cor_to_call</a></span><span class='op'>(</span><span class='va'>res</span>, 
                    metadata <span class='op'>=</span> <span class='va'>pan_smartseq2</span><span class='op'>@</span><span class='va'>meta.data</span>,
                    cluster_col <span class='op'>=</span> <span class='st'>"cluster.id"</span>,
                    threshold <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='co'># anything below 0.5 correlation are labeled as "unknown"</span>
<span class='va'>res2</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 12 x 3
#&gt; # Groups:   cluster.id [12]
#&gt;    cluster.id type      r
#&gt;    &lt;chr&gt;      &lt;chr&gt; &lt;dbl&gt;
#&gt;  1 0          alpha 0.738
#&gt;  2 1          alpha 0.711
#&gt;  3 11         alpha 0.671
#&gt;  4 2          alpha 0.750
#&gt;  5 3          alpha 0.732
#&gt;  6 7          alpha 0.723
#&gt;  7 9          alpha 0.727
#&gt;  8 10         beta  0.722
#&gt;  9 5          beta  0.781
#&gt; 10 6          beta  0.803
#&gt; 11 8          beta  0.781
#&gt; 12 4          delta 0.793</code></pre>
<div class="sourceCode">
<pre><code>
<span class='co'># same as last section</span>
<span class='va'>new_meta</span> <span class='op'>&lt;-</span> <span class='fu'><a href='http://rnabioco.github.io/clustifyr/reference/call_to_metadata.html'>call_to_metadata</a></span><span class='op'>(</span><span class='va'>res2</span>, 
                             metadata <span class='op'>=</span> <span class='va'>pan_smartseq2</span><span class='op'>@</span><span class='va'>meta.data</span>,
                             cluster_col <span class='op'>=</span> <span class='st'>"cluster.id"</span><span class='op'>)</span>
<span class='va'>pan_smartseq2</span><span class='op'>@</span><span class='va'>meta.data</span> <span class='op'>&lt;-</span> <span class='va'>new_meta</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>pan_smartseq2</span>, label <span class='op'>=</span> <span class='cn'>T</span>, group.by <span class='op'>=</span> <span class='st'>"type"</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/SeuratTheme.html'>NoLegend</a></span><span class='op'>(</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-12-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<p><code>clustifyr</code> also takes seurat objects as input directly, finds various needed data, and output another object with identities assigned.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>ref2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='http://rnabioco.github.io/clustifyr/reference/seurat_ref.html'>seurat_ref</a></span><span class='op'>(</span><span class='va'>pan_celseq2</span>,
                  cluster_col <span class='op'>=</span> <span class='st'>"celltype"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/identical.html'>identical</a></span><span class='op'>(</span><span class='va'>ref</span>,<span class='va'>ref2</span><span class='op'>)</span> <span class='co'># exactly the same as chunk above</span>
</code></pre>
</div>
<pre><code>#&gt; [1] TRUE</code></pre>
<div class="sourceCode">
<pre><code>
<span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'><a href='http://rnabioco.github.io/clustifyr/reference/clustify.html'>clustify</a></span><span class='op'>(</span>input <span class='op'>=</span> <span class='va'>pan_smartseq2</span>,
                ref_mat <span class='op'>=</span> <span class='va'>ref2</span>,
                cluster_col <span class='op'>=</span> <span class='st'>"cluster.id"</span>,
                seurat_out <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>

<span class='va'>res</span> <span class='co'># returns a new seurat object</span>
</code></pre>
</div>
<pre><code>#&gt; An object of class Seurat 
#&gt; 34363 features across 1443 samples within 1 assay 
#&gt; Active assay: RNA (34363 features, 2000 variable features)
#&gt;  2 dimensional reductions calculated: pca, umap</code></pre>
<div class="sourceCode">
<pre><code>
<span class='co'># look at UMAP</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>res</span>, label <span class='op'>=</span> <span class='cn'>TRUE</span>, group.by <span class='op'>=</span> <span class='st'>"type"</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/SeuratTheme.html'>NoLegend</a></span><span class='op'>(</span><span class='op'>)</span> <span class='co'># saved in "type" metadata column</span>
</code></pre>
</div>
<p><img src="4_markers_files/figure-html5/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<p>Also see more tutorials at <a href="https://rnabioco.github.io/clustifyr/"><code>clustifyr</code></a>.</p>
<h3 id="cell-type-composition">cell type composition</h3>
<p>Insight into different samples can be gained from the proportion of cells that fall into each cell type. Unfortunately, no dedicated tools are available for statistical testing.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>tab1</span> <span class='op'>&lt;-</span> <span class='va'>sobj_clusters</span><span class='op'>@</span><span class='va'>meta.data</span> <span class='op'>%&gt;%</span> <span class='fu'>group_by</span><span class='op'>(</span><span class='va'>orig.ident</span>, <span class='va'>clusters</span><span class='op'>)</span> <span class='op'>%&gt;%</span> <span class='fu'>tally</span><span class='op'>(</span><span class='op'>)</span> <span class='co'># counting up all combinations</span>
<span class='va'>tab1</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 16 x 3
#&gt; # Groups:   orig.ident [2]
#&gt;    orig.ident clusters     n
#&gt;    &lt;fct&gt;      &lt;fct&gt;    &lt;int&gt;
#&gt;  1 control    0          553
#&gt;  2 control    1          243
#&gt;  3 control    2          176
#&gt;  4 control    3          190
#&gt;  5 control    4           74
#&gt;  6 control    5           80
#&gt;  7 control    6           14
#&gt;  8 control    7            6
#&gt;  9 treated    0          589
#&gt; 10 treated    1          250
#&gt; 11 treated    2          176
#&gt; 12 treated    3          150
#&gt; 13 treated    4           87
#&gt; 14 treated    5           78
#&gt; 15 treated    6           18
#&gt; 16 treated    7            6</code></pre>
<div class="sourceCode">
<pre><code><span class='va'>tab2</span> <span class='op'>&lt;-</span> <span class='va'>tab1</span> <span class='op'>%&gt;%</span> <span class='fu'>spread</span><span class='op'>(</span>key <span class='op'>=</span> <span class='va'>clusters</span>, value <span class='op'>=</span> <span class='va'>n</span><span class='op'>)</span> <span class='co'># spread out into "wide" form</span>
<span class='va'>tab2</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 2 x 9
#&gt; # Groups:   orig.ident [2]
#&gt;   orig.ident   `0`   `1`   `2`   `3`   `4`   `5`   `6`   `7`
#&gt;   &lt;fct&gt;      &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
#&gt; 1 control      553   243   176   190    74    80    14     6
#&gt; 2 treated      589   250   176   150    87    78    18     6</code></pre>
<div class="sourceCode">
<pre><code><span class='va'>tab3</span> <span class='op'>&lt;-</span> <span class='va'>tab1</span> <span class='op'>%&gt;%</span> <span class='fu'>group_by</span><span class='op'>(</span><span class='va'>orig.ident</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>mutate</span><span class='op'>(</span>n <span class='op'>=</span> <span class='va'>n</span><span class='op'>/</span><span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> <span class='co'># convert counts to proportions first</span>
  <span class='fu'>spread</span><span class='op'>(</span>key <span class='op'>=</span> <span class='va'>clusters</span>, value <span class='op'>=</span> <span class='va'>n</span><span class='op'>)</span> 
<span class='va'>tab3</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 2 x 9
#&gt; # Groups:   orig.ident [2]
#&gt;   orig.ident   `0`   `1`   `2`   `3`    `4`    `5`    `6`     `7`
#&gt;   &lt;fct&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
#&gt; 1 control    0.414 0.182 0.132 0.142 0.0554 0.0599 0.0105 0.00449
#&gt; 2 treated    0.435 0.185 0.130 0.111 0.0643 0.0576 0.0133 0.00443</code></pre>
</div>
<p>and save tables to disk with <code>write_csv</code></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'>write_csv</span><span class='op'>(</span><span class='va'>tab3</span>, <span class='st'>"data/perc_clusters.csv"</span><span class='op'>)</span>
<span class='va'>tab4</span> <span class='op'>&lt;-</span> <span class='fu'>read_csv</span><span class='op'>(</span><span class='st'>"data/perc_clusters.csv"</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h2 id="other-things-to-do-with-marker-genes">other things to do with marker genes</h2>
<ol type="1">
<li><p>gene list to pathway activity score, via <a href="https://github.com/aertslab/AUCell"><code>AUCell</code></a></p></li>
<li><p>if TF expression is too low, consider <a href="https://github.com/aertslab/SCENIC"><code>SCENIC</code></a> for TF activity inference</p></li>
<li><p>standard GO term enrichment tools <a href="https://biit.cs.ut.ee/gprofiler/page/r"><code>gProfiler2</code></a>, <a href="https://github.com/wjawaid/enrichR"><code>enrichR</code></a>, <a href="https://github.com/ctlab/fgsea%5D"><code>fgsea</code></a>, etc etc</p></li>
</ol>
<div class="sourceCode" id="cb25"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="updates-and-corrections">Corrections</h3>
<p>If you see mistakes or want to suggest changes, please <a href="https://github.com/rnabioco/cellar/issues/new">create an issue</a> on the source repository.</p>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
