<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>

  <!--radix_placeholder_meta_tags-->
  <title>Single-cell RNA-seq Workshop: Trajectory inference</title>


  <link rel="icon" type="image/png" href="img/logo.png"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2020-11-17"/>
  <meta property="article:created" itemprop="dateCreated" content="2020-11-17"/>
  <meta name="article:author" content="Kent Riemondy, modified by Caitlin Winkler"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="Single-cell RNA-seq Workshop: Trajectory inference"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="Single-cell RNA-seq Workshop"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="Single-cell RNA-seq Workshop: Trajectory inference"/>
  <meta property="twitter:site" content="@rnabioco"/>

  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","author","date","output"]}},"value":[{"type":"character","attributes":{},"value":["Trajectory inference"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name"]}},"value":[{"type":"character","attributes":{},"value":["Kent Riemondy, modified by Caitlin Winkler"]}]}]},{"type":"character","attributes":{},"value":["2020-11-17"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["toc","toc_depth","self_contained","highlight"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"integer","attributes":{},"value":[1]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["tango"]}]}]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  <meta name="distill:offset" content=""/>

  <script type="application/javascript">

    window.headroom_prevent_pin = false;

    window.document.addEventListener("DOMContentLoaded", function (event) {

      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');

      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });

      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });

      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>

  <style type="text/css">

  /* Theme (user-documented overrideables for nav appearance) */

  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #0F2E3D;
    font-size: 15px;
    font-weight: 300;
  }

  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }

  .distill-site-nav a:hover {
    color: white;
  }

  @media print {
    .distill-site-nav {
      display: none;
    }
  }

  .distill-site-header {

  }

  .distill-site-footer {

  }


  /* Site Header */

  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }

  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }


  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }

  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }

  .distill-site-header .title {
    font-size: 18px;
    min-width: 150px;
  }

  .distill-site-header .logo {
    padding: 0;
  }

  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }

  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }



  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }


  .distill-site-header .nav-toggle {
    display: none;
  }

  .nav-dropdown {
    display: inline-block;
    position: relative;
  }

  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }

  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }

  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .nav-dropdown-active {
    display: block;
  }

  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }

  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }

  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }

  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }

  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }


  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative; min-height: 500px; }
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }

  /* Site Footer */

  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }

  /* Headroom */

  d-title {
    padding-top: 6rem;
  }

  @media print {
    d-title {
      padding-top: 4rem;
    }
  }

  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }

  .headroom--transition {
    transition: all .4s ease-in-out;
  }

  .headroom--unpinned {
    top: -100px;
  }

  .headroom--pinned {
    top: 0;
  }

  </style>

  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <script src="site_libs/autocomplete-0.37.1/autocomplete.min.js"></script>
  <script src="site_libs/fuse-6.4.1/fuse.min.js"></script>

  <script type="application/javascript">

  function getMeta(metaName) {
    var metas = document.getElementsByTagName('meta');
    for (let i = 0; i < metas.length; i++) {
      if (metas[i].getAttribute('name') === metaName) {
        return metas[i].getAttribute('content');
      }
    }
    return '';
  }

  function offsetURL(url) {
    var offset = getMeta('distill:offset');
    return offset ? offset + '/' + url : url;
  }

  function createIndex() {
    var options = {
      keys: [
        "title",
        "categories",
        "description",
        "contents"
      ]
    };
    return new window.Fuse([],options);
  }

  function createFuseIndex() {

    // create fuse index
    var options = { keys: ["title", "description", "contents"] };
    var fuse = new window.Fuse([], options);

    // fetch the main search.json
    return fetch(offsetURL('search.json'))
      .then(function(response) {
        if (response.status == 200) {
          return response.json().then(function(json) {
            // index main articles
            json.articles.forEach(function(article) {
              fuse.add(article);
            });
            // download collections and index their articles
            return Promise.all(json.collections.map(function(collection) {
              return fetch(offsetURL(collection)).then(function(response) {
                if (response.status === 200) {
                  return response.json().then(function(articles) {
                    articles.forEach(function(article) {
                      fuse.add(article);
                    });
                  })
                } else {
                  return Promise.reject(
                    new Error('Unexpected status from search index request: ' +
                              response.status)
                  );
                }
              });
            })).then(function() {
              return fuse;
            });
          });

        } else {
          return Promise.reject(
            new Error('Unexpected status from search index request: ' +
                        response.status)
          );
        }
      });
  }

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // get search element (bail if we don't have one)
    var searchEl = window.document.getElementById('distill-search');
    if (!searchEl)
      return;

    createFuseIndex()
      .then(function(fuse) {

        // make search box visible
        searchEl.classList.remove('hidden');

        // initialize autocomplete
        var options = {
          autoselect: true,
          hint: false,
          minLength: 2,
        };
        window.autocomplete(searchEl, options, [{
          source: function(query, callback) {
            const searchOptions = {
              isCaseSensitive: false,
              shouldSort: true,
              minMatchCharLength: 2,
              limit: 10,
              keys: [
                { name: 'title', weight: 20 },
                { name: 'categories', weight: 15 },
                { name: 'description', weight: 10 },
                { name: 'contents', weight: 5 },
              ],
            };
            var results = fuse.search(query, searchOptions);
            callback(results
              .map(function(result) { return result.item; })
              .filter(function(item) { return !!item.description; })
            );
          },
          templates: {
            suggestion: function(suggestion) {
              var html = `
                <div class="search-item">
                  <h3>${suggestion.title}</h3>
                  <div class="search-item-description">
                    ${suggestion.description}
                  </div>
                  <div class="search-item-preview">
                    <img src="${suggestion.preview ? offsetURL(suggestion.preview) : ''}"</img>
                  </div>
                </div>
              `;
              return html;
            }
          }
        }]).on('autocomplete:selected', function(event, suggestion) {
          window.location.href = offsetURL(suggestion.path);
        });
        // remove inline display style on autocompleter (we want to
        // manage responsive display via css)
        $('.algolia-autocomplete').css("display", "");
      })
      .catch(function(error) {
        console.log(error);
      });

  });

  </script>

  <style type="text/css">

  /* Algolioa Autocomplete */

  .algolia-autocomplete {
    display: inline-block;
    margin-left: 10px;
    vertical-align: sub;
    background-color: white;
    color: black;
    padding: 6px;
    padding-top: 8px;
    padding-bottom: 0;
    border-radius: 6px;
    border: 1px #0F2E3D solid;
    width: 180px;
  }


  @media screen and (max-width: 768px) {
    .distill-site-nav .algolia-autocomplete {
      display: none;
      visibility: hidden;
    }
    .distill-site-nav.responsive .algolia-autocomplete {
      display: inline-block;
      visibility: visible;
    }
    .distill-site-nav.responsive .algolia-autocomplete .aa-dropdown-menu {
      margin-left: 0;
      width: 400px;
      max-height: 400px;
    }
  }

  .algolia-autocomplete .aa-input, .algolia-autocomplete .aa-hint {
    width: 90%;
    outline: none;
    border: none;
  }

  .algolia-autocomplete .aa-hint {
    color: #999;
  }
  .algolia-autocomplete .aa-dropdown-menu {
    width: 550px;
    max-height: 70vh;
    overflow-x: visible;
    overflow-y: scroll;
    padding: 5px;
    margin-top: 3px;
    margin-left: -150px;
    background-color: #fff;
    border-radius: 5px;
    border: 1px solid #999;
    border-top: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion {
    cursor: pointer;
    padding: 5px 4px;
    border-bottom: 1px solid #eee;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion:last-of-type {
    border-bottom: none;
    margin-bottom: 2px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item {
    overflow: hidden;
    font-size: 0.8em;
    line-height: 1.4em;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item h3 {
    margin-block-start: 0;
    margin-block-end: 5px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-description {
    display: inline-block;
    overflow: hidden;
    height: 2.8em;
    width: 80%;
    margin-right: 4%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview {
    display: inline-block;
    width: 15%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img {
    height: 3em;
    width: auto;
    display: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img[src] {
    display: initial;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion.aa-cursor {
    background-color: #eee;
  }
  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion em {
    font-weight: bold;
    font-style: normal;
  }

  </style>


  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: hidden;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="site_libs/header-attrs-2.5/header-attrs.js"></script>
  <script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
  <script src="site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <style type="text/css">
  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #000000;
      font-size: 15px;
    font-weight: 300;
  }

  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }

  .distill-site-nav a:hover {
    color: white;
    text-decoration:underline;
  }

  .distill-site-header {
  }

  .distill-site-footer {
  }

  @media print {
    .distill-site-nav {
      display: none;
    }
  }

  .l-body {
    display: none;
  }

  @media screen and (min-width: 768px) {
    .l-body {display: inline;}
  }

  </style>
  <!--/radix_placeholder_site_in_header-->

  <link rel="stylesheet" href="styles.css" type="text/css"/>

</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Trajectory inference","authors":[{"author":"Kent Riemondy, modified by Caitlin Winkler","authorURL":"#","affiliation":"&nbsp;","affiliationURL":"#","orcidID":""}],"publishedDate":"2020-11-17T00:00:00.000-07:00","citationText":"Winkler, 2020"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a href="index.html" class="title">Single-cell RNA-seq Workshop</a>
<a href="prerequisite.html">Prerequisite-Material</a>
<a href="tba.html">Introduction</a>
<a href="2_filtering_QC.html">Data Filtering and Quality Control</a>
<a href="tba.html">Clustering and Visualization</a>
<a href="4_markers.html">Cluster Markers and Cell Type Assignment</a>
<a href="5_trajectories.html">Trajectories</a>
<a href="6_alignment.html">Alignment</a>
<a href="7_multimodal.html">Multi-modal</a>
<a href="8_vdj.html">VDJ</a>
<a href="previous.html">Previous-workshops</a>
<a href="r_install.html">Installing R</a>
</div>
<div class="nav-right">
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Trajectory inference</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->

</div>

<div class="d-byline">
  
  Kent Riemondy, modified by Caitlin Winkler
  
<br/>2020-11-17
</div>

<div class="d-article">
<p>Rmarkdown: <a href="https://raw.githubusercontent.com/rnabioco/cellar/master/5_trajectories.Rmd" style="text-decoration: none;"> <i class="fa fa-file-code-o"></i> </a></p>
<h2 id="trajectory-inference-aka-pseudotime">Trajectory inference, aka pseudotime</h2>
<p>Trajectory inference assumes that cells are sampled during various stages of a transition from a cell type or state to another type or state. By identifying trajectories that connect cells based on similarilty in gene expression, one can gain insights into lineage relationships and developmental trajectories. To quantitate the measure of biological progress outside of defined time-points we use a metric called pseudotime, which is defined as a distance metric between the starting cell and ending cell along the trajectory.</p>
<p>Analysis of discrete clusters can hide interesting continous behaviors in cell populations. By ordering cells based on an expression trajectory one can uncover novel patterns of gene expression.</p>
<h2 id="there-are-many-methods">There are many methods</h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'>knitr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/knitr/man/include_graphics.html'>include_graphics</a></span><span class='op'>(</span><span class='st'>"img/dynverse.webp"</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="img/dynverse.webp" style="display: block; margin: auto;" /></p>
</div>
<p>Check out the <a href="https://dynverse.org/"><code>dynverse</code></a> for help with algorithm selection. They benchmarked &gt; 60 methods, and offer some tools to run multiple algorithms on the same data.</p>
<h2 id="simplest-method-pca">Simplest method (PCA)</h2>
<p>In some datasets, particularly developmental datasets it is often the case that a principal component may separate out cells based on known developmental time. This can be used a pseudotime metric, which is simply the PC1 scores for each cell. As an example, scRNA-seq analyses of <a href="http://dx.doi.org/10.1038/nature22796">liver development</a> and <a href="http://dx.doi.org/10.1016/j.cmet.2017.04.014">Beta-cell maturation</a> used PC1 to examine pseudotime.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://satijalab.org/seurat'>Seurat</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://tidyverse.tidyverse.org'>tidyverse</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://Matrix.R-forge.R-project.org/'>Matrix</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>slingshot</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://statomics.github.io/tradeSeq/index.html'>tradeSeq</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>SingleCellExperiment</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>RColorBrewer</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>gam</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://wilkelab.org/cowplot/'>cowplot</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://scales.r-lib.org'>scales</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>gridExtra</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://theislab.github.io/destiny/'>destiny</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/jokergoo/ComplexHeatmap'>ComplexHeatmap</a></span><span class='op'>)</span>
<span class='fu'>theme_set</span><span class='op'>(</span><span class='fu'><a href='https://wilkelab.org/cowplot/reference/theme_cowplot.html'>theme_cowplot</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>For this first analysis we will use a dataset of mouse early embryogenesis that was curated into a SingleCellExperiment object from here <a href="https://hemberg-lab.github.io/scRNA.seq.datasets/" class="uri">https://hemberg-lab.github.io/scRNA.seq.datasets/</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># read in data</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/connections.html'>url</a></span><span class='op'>(</span><span class='st'>"https://scrnaseq-public-datasets.s3.amazonaws.com/scater-objects/deng-reads.rds"</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># convert SingleCellExperiment to Seurat</span>
<span class='va'>sobj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html'>CreateSeuratObject</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/SingleCellExperiment/man/assays.html'>counts</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, meta.data <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/as.data.frame.html'>as.data.frame</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>sobj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/NormalizeData.html'>NormalizeData</a></span><span class='op'>(</span><span class='va'>sobj</span><span class='op'>)</span>
<span class='va'>sobj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html'>FindVariableFeatures</a></span><span class='op'>(</span><span class='va'>sobj</span><span class='op'>)</span>
<span class='va'>sobj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/ScaleData.html'>ScaleData</a></span><span class='op'>(</span><span class='va'>sobj</span><span class='op'>)</span>
<span class='va'>sobj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/RunPCA.html'>RunPCA</a></span><span class='op'>(</span><span class='va'>sobj</span>, verbose <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/Idents.html'>Idents</a></span><span class='op'>(</span><span class='va'>sobj</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='st'>"cell_type1"</span>

<span class='co'># plot the PCA reduction</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>sobj</span>, reduction <span class='op'>=</span> <span class='st'>"pca"</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-2-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>Next we will plot out the PC1 scores for each cell type (developmental stage), using the meta data stored in the <code>cell_type</code> column of <code>meta.data</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># get the PC1 values for each cell, and store in a data frame</span>
<span class='va'>plt_dat</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FetchData.html'>FetchData</a></span><span class='op'>(</span><span class='va'>sobj</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"PC_1"</span>, <span class='st'>"cell_type1"</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># reorder cell_type based on known developmental time</span>
<span class='va'>cell_type</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>plt_dat</span><span class='op'>$</span><span class='va'>cell_type1</span>,
                    levels <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"zygote"</span>,
                               <span class='st'>"2cell"</span>,
                               <span class='st'>"4cell"</span>,
                               <span class='st'>"8cell"</span>,
                               <span class='st'>"16cell"</span>,
                               <span class='st'>"blast"</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># add the reordered cell_type info as a new column to plt_data</span>
<span class='va'>plt_dat</span><span class='op'>$</span><span class='va'>cell_type</span> <span class='op'>&lt;-</span> <span class='va'>cell_type</span>

<span class='co'># plot the cells ordered by pseudotime (PC1)</span>
<span class='fu'>ggplot</span><span class='op'>(</span><span class='va'>plt_dat</span>, <span class='fu'>aes</span><span class='op'>(</span><span class='va'>cell_type</span>, <span class='va'>PC_1</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_jitter</span><span class='op'>(</span><span class='fu'>aes</span><span class='op'>(</span>color <span class='op'>=</span> <span class='va'>cell_type</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>labs</span><span class='op'>(</span>y <span class='op'>=</span> <span class='st'>"PC1 (aka pseudotime)"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"Cells ordered by PC1"</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-3-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># Visualize the top genes associated with PC1 (pseudotime)</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/VizDimLoadings.html'>VizDimLoadings</a></span><span class='op'>(</span><span class='va'>sobj</span>, dims <span class='op'>=</span> <span class='fl'>1</span>, balanced <span class='op'>=</span> <span class='cn'>T</span><span class='op'>)</span>
</code></pre>
</div>
<img src="5_trajectories_files/figure-html5/unnamed-chunk-4-1.png" width="624" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='co'># Plot example genes</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FeatureScatter.html'>FeatureScatter</a></span><span class='op'>(</span><span class='va'>sobj</span>, <span class='st'>"PC_1"</span>, <span class='st'>"Actb"</span><span class='op'>)</span>
</code></pre>
</div>
<img src="5_trajectories_files/figure-html5/unnamed-chunk-4-2.png" width="624" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FeatureScatter.html'>FeatureScatter</a></span><span class='op'>(</span><span class='va'>sobj</span>, <span class='st'>"PC_1"</span>, <span class='st'>"Parp12"</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-4-3.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>We can overlay the pseudotime scores onto a projection by storing them as a <code>meta.data</code> column.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># Add UMAP dimensional reduction</span>
<span class='va'>sobj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/RunUMAP.html'>RunUMAP</a></span><span class='op'>(</span><span class='va'>sobj</span>, dims <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>15</span><span class='op'>)</span>

<span class='co'># get PC1 values and rank to generate a 'pseudotime'</span>
<span class='va'>ptime</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FetchData.html'>FetchData</a></span><span class='op'>(</span><span class='va'>sobj</span>, <span class='st'>"PC_1"</span><span class='op'>)</span>
<span class='va'>ptime</span><span class='op'>$</span><span class='va'>ptime</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/rank.html'>rank</a></span><span class='op'>(</span><span class='va'>ptime</span><span class='op'>$</span><span class='va'>PC_1</span><span class='op'>)</span>

<span class='co'># add pseudotime to the metadata</span>
<span class='va'>sobj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/AddMetaData.html'>AddMetaData</a></span><span class='op'>(</span><span class='va'>sobj</span>, 
                    <span class='va'>ptime</span><span class='op'>$</span><span class='va'>ptime</span>, 
                    col.name <span class='op'>=</span> <span class='st'>"Pseudotime"</span><span class='op'>)</span>

<span class='co'># color single cells on a UMAP plot according to pseudotime</span>
<span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/FeaturePlot.html'>FeaturePlot</a></span><span class='op'>(</span><span class='va'>sobj</span>, <span class='st'>"Pseudotime"</span><span class='op'>)</span>

<span class='co'># color single cells on a UMAP plot according to cell_type (i.e., timepoint)</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>sobj</span>, 
             group.by  <span class='op'>=</span> <span class='st'>"cell_type1"</span>, 
             reduction <span class='op'>=</span> <span class='st'>"umap"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"Cell type"</span><span class='op'>)</span>

<span class='fu'><a href='https://wilkelab.org/cowplot/reference/plot_grid.html'>plot_grid</a></span><span class='op'>(</span><span class='va'>p1</span>, <span class='va'>p2</span>, align <span class='op'>=</span> <span class='st'>"hv"</span>, ncol <span class='op'>=</span> <span class='fl'>2</span>, rel_widths <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-5-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>sobj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/RunTSNE.html'>RunTSNE</a></span><span class='op'>(</span><span class='va'>sobj</span>, dims <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>15</span><span class='op'>)</span>

<span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>sobj</span>, 
             group.by  <span class='op'>=</span> <span class='st'>"cell_type1"</span>, 
             reduction <span class='op'>=</span> <span class='st'>"tsne"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"tSNE"</span><span class='op'>)</span>

<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>sobj</span>, 
             group.by  <span class='op'>=</span> <span class='st'>"cell_type1"</span>, 
             reduction <span class='op'>=</span> <span class='st'>"umap"</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"UMAP"</span><span class='op'>)</span>

<span class='fu'><a href='https://wilkelab.org/cowplot/reference/plot_grid.html'>plot_grid</a></span><span class='op'>(</span><span class='va'>p1</span>, <span class='va'>p2</span>, align <span class='op'>=</span> <span class='st'>"hv"</span>, ncol <span class='op'>=</span> <span class='fl'>2</span>, rel_widths <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-6-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<h2 id="destiny">Destiny</h2>
<p>Non-linear dynamics are not always captured by PCA. Diffusion maps offer an alternative, non-linear transformation approach that infers the temporal order of differentiating cells by modeling it as a diffusion process. This can better resolve complex trajectories and branching than linear methods like PCA. Well use the R package <code>destiny</code> to look at diffusion maps and diffusion pseudotime.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># Generating a diffusion map requires a matrix of expression data:</span>
<span class='co'># prepare a counts matrix</span>
<span class='va'>logcounts</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/GetAssayData.html'>GetAssayData</a></span><span class='op'>(</span><span class='va'>sobj</span>, <span class='st'>"data"</span><span class='op'>)</span>

<span class='co'># transpose matrix (genes as columns, cells as rows)</span>
<span class='va'>input_matrix</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>logcounts</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/VariableFeatures.html'>VariableFeatures</a></span><span class='op'>(</span><span class='va'>sobj</span><span class='op'>)</span>, <span class='op'>]</span><span class='op'>)</span>

<span class='co'># generate a diffusion map</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>42</span><span class='op'>)</span>
<span class='va'>dm</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://theislab.github.io/destiny/reference/DiffusionMap-class.html'>DiffusionMap</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>input_matrix</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># store the diffusion map as a custom dimensional reduction in the Seurat object</span>
<span class='va'>sobj</span><span class='op'>[[</span><span class='st'>"DM"</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/CreateDimReducObject.html'>CreateDimReducObject</a></span><span class='op'>(</span>embeddings <span class='op'>=</span> <span class='va'>dm</span><span class='op'>@</span><span class='va'>eigenvectors</span>, key <span class='op'>=</span> <span class='st'>"DM_"</span>, assay <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DefaultAssay.html'>DefaultAssay</a></span><span class='op'>(</span><span class='va'>sobj</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># plot the diffusion map</span>
<span class='fu'><a href='https://rdrr.io/pkg/Seurat/man/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>sobj</span>, reduction <span class='op'>=</span> <span class='st'>"DM"</span>, group.by <span class='op'>=</span> <span class='st'>"cell_type1"</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"Diffusion Map"</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-7-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>Next we calculate diffusion pseudotime using <code>DPT</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># calculate the diffusion pseudotime (DPT)</span>
<span class='va'>dpt</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://theislab.github.io/destiny/reference/DPT.html'>DPT</a></span><span class='op'>(</span><span class='va'>dm</span><span class='op'>)</span>

<span class='co'># color single cells on diffusion map plot according to DPT</span>
<span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/graphics/plot.html'>plot</a></span><span class='op'>(</span><span class='va'>dpt</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"Diffusion Pseudotime (DPT)"</span><span class='op'>)</span>

<span class='co'># create data.frame for easy plotting of DPT</span>
<span class='va'>tmp</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>DC1 <span class='op'>=</span> <span class='va'>dm</span><span class='op'>$</span><span class='va'>DC1</span>,
                  DC2 <span class='op'>=</span> <span class='va'>dm</span><span class='op'>$</span><span class='va'>DC2</span>,
                  timepoint <span class='op'>=</span> <span class='va'>cell_type</span>,
                  dpt <span class='op'>=</span> <span class='va'>dpt</span><span class='op'>$</span><span class='va'>DPT1</span><span class='op'>)</span>

<span class='co'># color single cells on diffusion map plot according to timepoint</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>ggplot</span><span class='op'>(</span><span class='va'>tmp</span>, <span class='fu'>aes</span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>DC1</span>, y <span class='op'>=</span> <span class='va'>DC2</span>, colour <span class='op'>=</span> <span class='va'>timepoint</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'>geom_point</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'>theme_classic</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
    <span class='fu'>ggtitle</span><span class='op'>(</span><span class='st'>"Timepoint"</span><span class='op'>)</span>

<span class='fu'><a href='https://wilkelab.org/cowplot/reference/plot_grid.html'>plot_grid</a></span><span class='op'>(</span><span class='va'>p1</span>, <span class='va'>p2</span>, align <span class='op'>=</span> <span class='st'>"hv"</span>, ncol <span class='op'>=</span> <span class='fl'>2</span>, rel_widths <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-8-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># plot timepoint vs. DPT</span>
<span class='fu'>ggplot</span><span class='op'>(</span><span class='va'>tmp</span>, <span class='fu'>aes</span><span class='op'>(</span><span class='va'>timepoint</span>, <span class='va'>dpt</span>, colour <span class='op'>=</span> <span class='va'>timepoint</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>geom_point</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span> 
  <span class='fu'>geom_jitter</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'>ggtitle</span> <span class='op'>(</span><span class='st'>"Timepoint vs. DPT"</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-9-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>PCA and diffusion maps can be good for simple trajectories or pathway activation systems e.g.T-cell activation, drug pertubation, single cell type responses.</p>
<h2 id="slingshot">Slingshot</h2>
<p>For more complex trajectories, simply using a single PC or a diffusion componenet may not be sufficient. Systems with bi or tri-furcating trajectories wont be well fit within a single dimension.</p>
<p>For this next analysis we will use a dataset taken from a single cell RNA-seq study of <a href="https://doi.org/10.1002/hep.29353">hepatocyte development</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>sobj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='st'>"obj/slingshot.rds"</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-11-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<p><a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-4772-0">Slingshot</a> is a <a href="https://bioconductor.org/packages/release/bioc/html/slingshot.html">Bioconductor package</a> that draws curved trajectories through a low dimensional embedding to infer developmental dynamics. It provides functionality for computing pseudotimes through multiple trajectories.</p>
<div class="layout-chunk" data-layout="l-body">
<p><img src="img/slingshot.png" width="952" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'>#convert to SingleCellExperiment</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'>as.SingleCellExperiment</span><span class='op'>(</span><span class='va'>sobj</span><span class='op'>)</span>

<span class='co'>#subset to only a few PCA dimensions</span>
<span class='fu'>reducedDim</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>reducedDim</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span><span class='op'>]</span>

<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/warning.html'>suppressWarnings</a></span><span class='op'>(</span><span class='fu'>slingshot</span><span class='op'>(</span>
  <span class='va'>sce</span>,
  reducedDim <span class='op'>=</span> <span class='st'>'PCA'</span>,
  clusterLabels <span class='op'>=</span> <span class='st'>'seurat_clusters'</span>,
  start.clus <span class='op'>=</span> <span class='st'>"4"</span>
<span class='op'>)</span><span class='op'>)</span>

<span class='co'># extract info about pseudotimes from sce</span>
<span class='va'>sds</span> <span class='op'>&lt;-</span> <span class='fu'>SlingshotDataSet</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>
<span class='va'>sds</span>
</code></pre>
</div>
<pre><code>#&gt; class: SlingshotDataSet 
#&gt; 
#&gt;  Samples Dimensions
#&gt;      447         10
#&gt; 
#&gt; lineages: 3 
#&gt; Lineage1: 4  2  6  1  3  
#&gt; Lineage2: 4  2  6  1  0  
#&gt; Lineage3: 4  2  6  1  5  
#&gt; 
#&gt; curves: 3 
#&gt; Curve1: Length: 82.362   Samples: 305.31
#&gt; Curve2: Length: 122.06   Samples: 310.18
#&gt; Curve3: Length: 84.424   Samples: 267.11</code></pre>
</div>
<p><code>slingshot</code> provides some minimal plotting utilities in base R plotting:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># Plot clusters with identified lineages overlayed</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>xpd<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mar<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>4.5</span>,<span class='fl'>5.5</span>,<span class='fl'>2</span>,<span class='fl'>7</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.html'>plot</a></span><span class='op'>(</span><span class='fu'>reducedDims</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>PCA</span>, col <span class='op'>=</span> <span class='fu'>brewer.pal</span><span class='op'>(</span><span class='fl'>9</span>,<span class='st'>'Set1'</span><span class='op'>)</span><span class='op'>[</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>seurat_clusters</span><span class='op'>]</span>, pch<span class='op'>=</span><span class='fl'>16</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/lines.html'>lines</a></span><span class='op'>(</span><span class='fu'>SlingshotDataSet</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, lwd<span class='op'>=</span><span class='fl'>2</span>, col<span class='op'>=</span><span class='st'>'black'</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span>x<span class='op'>=</span><span class='fl'>25</span>, y<span class='op'>=</span><span class='fl'>20</span>, legend<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>seurat_clusters</span><span class='op'>)</span>, fill<span class='op'>=</span><span class='fu'>brewer.pal</span><span class='op'>(</span><span class='fl'>11</span>,<span class='st'>'Set1'</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/integer.html'>as.integer</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>seurat_clusters</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>
</code></pre>
</div>
<img src="5_trajectories_files/figure-html5/unnamed-chunk-14-1.png" width="624" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre><code><span class='co'># Plot pseudotime for a lineage (we'll plot lineage #3 in this example)</span>
<span class='va'>colors</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/colorRamp.html'>colorRampPalette</a></span><span class='op'>(</span><span class='fu'>brewer.pal</span><span class='op'>(</span><span class='fl'>11</span>,<span class='st'>'Spectral'</span><span class='op'>)</span><span class='op'>[</span><span class='op'>-</span><span class='fl'>6</span><span class='op'>]</span><span class='op'>)</span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span>
<span class='va'>plotcol</span> <span class='op'>&lt;-</span> <span class='va'>colors</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/cut.html'>cut</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>$</span><span class='va'>slingPseudotime_3</span>, breaks<span class='op'>=</span><span class='fl'>100</span><span class='op'>)</span><span class='op'>]</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.html'>plot</a></span><span class='op'>(</span><span class='fu'>reducedDims</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>PCA</span>, col <span class='op'>=</span> <span class='va'>plotcol</span>, pch<span class='op'>=</span><span class='fl'>16</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/lines.html'>lines</a></span><span class='op'>(</span><span class='fu'>SlingshotDataSet</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span>, lwd<span class='op'>=</span><span class='fl'>2</span>, col<span class='op'>=</span><span class='st'>'black'</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-14-2.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<h3 id="plotting-slingshot-pseudotime-onto-umap">Plotting <code>slingshot</code> pseudotime onto UMAP</h3>
<p>The lineage trees are generated in the same dimensionality as the input data. To visualize the pseudotime estimates in UMAP space we can add these to the meta.data slot of the seurat object.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># add slingshot pseudotimes to seurat object `meta.data`</span>
<span class='va'>sobj</span><span class='op'>$</span><span class='va'>pseudotime1</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>slingPseudotime_1</span>
<span class='va'>sobj</span><span class='op'>$</span><span class='va'>pseudotime2</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>slingPseudotime_2</span>
<span class='va'>sobj</span><span class='op'>$</span><span class='va'>pseudotime3</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>slingPseudotime_3</span>

<span class='fu'>FeaturePlot</span><span class='op'>(</span><span class='va'>sobj</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"pseudotime1"</span>, <span class='st'>"pseudotime2"</span>, <span class='st'>"pseudotime3"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-15-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<h2 id="identifying-differentially-expressed-genes-along-a-trajectory">Identifying differentially expressed genes along a trajectory</h2>
<p>We can use a <code>GAM</code> (or other models) to compare pseudotimes to gene expression values to identify genes that vary significantly with pseudotime. Using a GAM can uncover non-linear expression patterns (i.e.not just up or down). See this <a href="https://multithreaded.stitchfix.com/blog/2015/07/30/gam/">blog</a> for a nice description.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># select the ptime values </span>
<span class='va'>ptime</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>slingPseudotime_2</span>

<span class='co'># get cells in that lineage</span>
<span class='va'>lineage_cells</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/NA.html'>is.na</a></span><span class='op'>(</span><span class='va'>ptime</span><span class='op'>)</span><span class='op'>]</span>

<span class='co'># remove values for cells not in the lineage</span>
<span class='va'>ptime</span> <span class='op'>&lt;-</span> <span class='va'>ptime</span><span class='op'>[</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/NA.html'>is.na</a></span><span class='op'>(</span><span class='va'>ptime</span><span class='op'>)</span><span class='op'>]</span>

<span class='co'># just test variable genes to save some time</span>
<span class='va'>genes_to_test</span> <span class='op'>&lt;-</span> <span class='fu'>VariableFeatures</span><span class='op'>(</span><span class='va'>sobj</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>1000</span><span class='op'>]</span>

<span class='co'># get log normalized data to test</span>
<span class='va'>cnts</span> <span class='op'>&lt;-</span> <span class='fu'>logcounts</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='va'>genes_to_test</span>, <span class='va'>lineage_cells</span><span class='op'>]</span>

<span class='co'># fit a GAM with a loess term for pseudotime</span>
<span class='va'>gam.pval</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>cnts</span>, <span class='fl'>1</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>z</span><span class='op'>)</span><span class='op'>{</span>
    <span class='va'>d</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>z <span class='op'>=</span> <span class='va'>z</span>, ptime <span class='op'>=</span> <span class='va'>ptime</span><span class='op'>)</span>
    <span class='va'>tmp</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/warning.html'>suppressWarnings</a></span><span class='op'>(</span><span class='fu'>gam</span><span class='op'>(</span><span class='va'>z</span> <span class='op'>~</span> <span class='fu'>lo</span><span class='op'>(</span><span class='va'>ptime</span><span class='op'>)</span>, data<span class='op'>=</span><span class='va'>d</span><span class='op'>)</span><span class='op'>)</span>
    <span class='va'>p</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>tmp</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>4</span><span class='op'>]</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span><span class='op'>[</span><span class='fl'>1</span>, <span class='fl'>5</span><span class='op'>]</span>
    <span class='va'>p</span>
<span class='op'>}</span><span class='op'>)</span>

<span class='co'># adjust pvalues </span>
<span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'>tibble</span><span class='op'>(</span>
  id <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>gam.pval</span><span class='op'>)</span>,
  pvals <span class='op'>=</span> <span class='va'>gam.pval</span>,
  qval <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/p.adjust.html'>p.adjust</a></span><span class='op'>(</span><span class='va'>gam.pval</span>, method <span class='op'>=</span> <span class='st'>"fdr"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'>arrange</span><span class='op'>(</span><span class='va'>qval</span><span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>res</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; # A tibble: 6 x 3
#&gt;   id                pvals      qval
#&gt;   &lt;chr&gt;             &lt;dbl&gt;     &lt;dbl&gt;
#&gt; 1 Tinagl1       8.73e-151 8.72e-148
#&gt; 2 Spon1         2.61e-150 1.30e-147
#&gt; 3 Pdzk1ip1      8.85e-145 2.95e-142
#&gt; 4 Gc            2.58e-139 6.44e-137
#&gt; 5 1700011H14Rik 1.67e-127 3.35e-125
#&gt; 6 Adamts1       6.47e-127 1.08e-124</code></pre>
</div>
<p>Finally make a heatmap to plot out these genes that vary over pseudotime.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># get log normalized counts </span>
<span class='va'>to_plot</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='fu'>logcounts</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='va'>res</span><span class='op'>$</span><span class='va'>id</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>100</span><span class='op'>]</span>, <span class='va'>lineage_cells</span><span class='op'>]</span><span class='op'>)</span>

<span class='co'># arrange cells by pseudotime</span>
<span class='va'>ptime_order</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>to_plot</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='va'>ptime</span><span class='op'>)</span><span class='op'>]</span>

<span class='co'># add useful annotations</span>
<span class='va'>annotations</span> <span class='op'>&lt;-</span> <span class='fu'>colData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='va'>lineage_cells</span>, 
                            <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"slingPseudotime_2"</span>, 
                            <span class='st'>"seurat_clusters"</span>, 
                            <span class='st'>"timepoint"</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>%&gt;%</span> 
  <span class='fu'><a href='https://rdrr.io/r/base/as.data.frame.html'>as.data.frame</a></span><span class='op'>(</span><span class='op'>)</span>

<span class='va'>ha</span> <span class='op'>&lt;-</span> <span class='fu'>HeatmapAnnotation</span><span class='op'>(</span>df <span class='op'>=</span> <span class='va'>annotations</span><span class='op'>)</span>

<span class='fu'>Heatmap</span><span class='op'>(</span><span class='va'>to_plot</span>,
        column_order <span class='op'>=</span> <span class='va'>ptime_order</span>,
        show_column_names <span class='op'>=</span> <span class='cn'>FALSE</span>,
        show_row_names <span class='op'>=</span> <span class='cn'>FALSE</span>,
        top_annotation <span class='op'>=</span> <span class='va'>ha</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-17-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<h2 id="tradeseq">tradeSeq</h2>
<p><code>tradeSeq</code> is a package that allows analysis of gene expression along trajectories (such as those found with <code>Slingshot</code>). For every gene in the dataset, <code>tradeSeq</code> fits a GAM (as explained in the previous section). It then allows statistical inference on the GAM by assessing contrasts of the parameters of the fitted GAM, aiding in interpreting complex datasets. For more details, see the <code>tradeSeq</code> <a href="https://www.nature.com/articles/s41467-020-14766-3">paper</a>. The following is from the <code>tradeSeq</code> <a href="https://bioconductor.org/packages/release/bioc/vignettes/tradeSeq/inst/doc/tradeSeq.html">tutorial</a>.</p>
<p>Lets use <code>tradeSeq</code> to look at our <code>Slingshot</code> data set from the previous section:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># tradeSeq needs a counts matrix and lineage curves as minimal input</span>
<span class='va'>crv</span> <span class='op'>&lt;-</span> <span class='va'>sds</span>
<span class='va'>counts</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='fu'>assays</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>counts</span><span class='op'>)</span>
<span class='va'>clusters</span> <span class='op'>&lt;-</span> <span class='va'>sce</span><span class='op'>$</span><span class='va'>seurat_clusters</span>
</code></pre>
</div>
</div>
<p>The following two code blocks we will <strong>NOT</strong> run, as they both take several minutes. These two code blocks are fitting the <code>tradeSeq</code> negative binomial generalized additive model (NB-GAM). For an in-depth description of whats going on, see <a href="https://statomics.github.io/tradeSeq/articles/fitGAM.html">this vignette</a>. Basically, we are using the NB-GAM framework in <code>tradeSeq</code> to smooth each genes expression in each lineage. Smoothers can be decomposed into a set of basis functions, which are joined together at knot points (simply called knots).</p>
<p>We run the <code>evaluateK</code> function in order to determine the correct number of knots. This returns the AIC (Akaike Informaction Criterion) for each gene fitted with each number of knots defined in the function (typically the interesting range is 3 to 10 knots) in a series of diagnostic plots.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>5</span><span class='op'>)</span>
<span class='va'>icMat</span> <span class='op'>&lt;-</span> <span class='fu'>evaluateK</span><span class='op'>(</span>counts <span class='op'>=</span> <span class='va'>counts</span>, sds <span class='op'>=</span> <span class='va'>crv</span>, k <span class='op'>=</span> <span class='fl'>3</span><span class='op'>:</span><span class='fl'>10</span>, nGenes <span class='op'>=</span> <span class='fl'>200</span>, verbose <span class='op'>=</span> <span class='cn'>T</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<p><img src="img/evaluateK.png" width="690" style="display: block; margin: auto;" /></p>
</div>
<p>The middles panels show that the drop in AIC levels off if the number of knots is increased beyond 6, so well choose that number of knots to fit the <code>tradeSeq</code> models using the <code>fitGAM</code> function. <code>nknots = 6</code> is also the default for the <code>fitGAM</code> function:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>7</span><span class='op'>)</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'>fitGAM</span><span class='op'>(</span>counts <span class='op'>=</span> <span class='va'>counts</span>, sds <span class='op'>=</span> <span class='va'>sds</span>, nknots <span class='op'>=</span> <span class='fl'>6</span>, genes <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>1000</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span>, sce<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># Load in the SingleCellExperiment created by running `fitGAM` above</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='st'>"obj/tradeSeq.rds"</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># plot our Slingshot lineage trajectories, this time illustrating the new tradeSeq knots</span>
<span class='fu'>plotGeneCount</span><span class='op'>(</span>curve <span class='op'>=</span> <span class='va'>sds</span>, counts <span class='op'>=</span> <span class='va'>counts</span>,
              clusters <span class='op'>=</span> <span class='va'>clusters</span>,
              models <span class='op'>=</span> <span class='va'>sce</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-23-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>With <code>tradeSeq</code>, we can examine differential gene expression in a variety of ways. We can check if a gene is associated with a particular lineage using <code>associationTest</code>, discover progenitor marker genes using <code>startVsEndTest</code>, compare specific pseudotime values with a lineage, discover differentiated cell type markers using <code>diffEndTest</code>, discover genes with different expression patterns using <code>patternTest</code>, and identify early drivers of differentiation using <code>earlyDETest</code>.</p>
<h3 id="discovering-differentiated-cell-type-markers">Discovering differentiated cell type markers</h3>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='co'># discover marker genes for the differentiated cell types</span>
<span class='va'>endRes</span> <span class='op'>&lt;-</span> <span class='fu'>diffEndTest</span><span class='op'>(</span><span class='va'>sce</span>, pairwise <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>endRes</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt;         waldStat df       pvalue waldStat_1vs2 df_1vs2  pvalue_1vs2
#&gt; Gnai3   1.262924  2 0.5318136472     0.3567923       1 5.502931e-01
#&gt; Cdc45  16.362748  2 0.0002798172     2.0114514       1 1.561159e-01
#&gt; H19   120.357216  2 0.0000000000     0.1184053       1 7.307705e-01
#&gt; Scml2   7.014859  2 0.0299738613     1.1724870       1 2.788908e-01
#&gt; Apoh  347.651164  2 0.0000000000    39.2279070       1 3.771115e-10
#&gt; Narf    6.657228  2 0.0358427555     6.6347775       1 1.000067e-02
#&gt;       waldStat_1vs3 df_1vs3  pvalue_1vs3 waldStat_2vs3 df_2vs3
#&gt; Gnai3      1.252185       1 2.631355e-01      0.249168       1
#&gt; Cdc45     16.165901       1 5.802937e-05      4.148820       1
#&gt; H19       98.180587       1 0.000000e+00     88.028317       1
#&gt; Scml2      6.891550       1 8.660416e-03      3.467197       1
#&gt; Apoh     331.912714       1 0.000000e+00    162.163896       1
#&gt; Narf       1.486989       1 2.226842e-01      2.483978       1
#&gt;       pvalue_2vs3    logFC1_2   logFC1_3    logFC2_3
#&gt; Gnai3  0.61766151  0.05497623  0.1031472  0.04817095
#&gt; Cdc45  0.04166355  1.71145519  4.1113875  2.39993230
#&gt; H19    0.00000000  0.06696979 -1.7552750 -1.82224479
#&gt; Scml2  0.06259731  1.09854160  4.0596169  2.96107529
#&gt; Apoh   0.00000000 -3.24297255 -9.7584555 -6.51548295
#&gt; Narf   0.11501103 -0.91400830 -0.3579889  0.55601937</code></pre>
</div>
<p>We can plot the most significant gene using the <code>plotSmoothers</code> function.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>o</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='va'>endRes</span><span class='op'>$</span><span class='va'>waldStat</span>, decreasing <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>sigGene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='va'>o</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>
<span class='fu'>plotSmoothers</span><span class='op'>(</span><span class='va'>sce</span>, <span class='va'>counts</span>, <span class='va'>sigGene</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-25-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<p>Alternatively, we can color the cells in PC space with that genes expression.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'>plotGeneCount</span><span class='op'>(</span><span class='va'>crv</span>, <span class='va'>counts</span>, gene <span class='op'>=</span> <span class='va'>sigGene</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-26-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<h3 id="discovering-early-drivers-of-differentiation">Discovering early drivers of differentiation</h3>
<p>We might want to find a list of genes that are differentially expressed between lineages at a particular region, e.g.around the separation of two or multiple lineages.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='va'>earlyDERes</span> <span class='op'>&lt;-</span> <span class='fu'>earlyDETest</span><span class='op'>(</span><span class='va'>sce</span>, pairwise <span class='op'>=</span> <span class='cn'>TRUE</span>, knots <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>4</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>oEarly</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='va'>earlyDERes</span><span class='op'>$</span><span class='va'>waldStat</span>, decreasing <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>earlyDERes</span><span class='op'>)</span><span class='op'>[</span><span class='va'>oEarly</span><span class='op'>]</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>#&gt; [1] &quot;Rps11&quot; &quot;Cp&quot;    &quot;Apoe&quot;  &quot;Apoh&quot;  &quot;Rpl13&quot; &quot;Rpl8&quot;</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'>plotSmoothers</span><span class='op'>(</span><span class='va'>sce</span>, <span class='va'>counts</span>, gene <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>earlyDERes</span><span class='op'>)</span><span class='op'>[</span><span class='va'>oEarly</span><span class='op'>]</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-28-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre><code><span class='fu'>plotGeneCount</span><span class='op'>(</span><span class='va'>crv</span>, <span class='va'>counts</span>, gene <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>earlyDERes</span><span class='op'>)</span><span class='op'>[</span><span class='va'>oEarly</span><span class='op'>]</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>
</code></pre>
</div>
<p><img src="5_trajectories_files/figure-html5/unnamed-chunk-29-1.png" width="624" style="display: block; margin: auto;" /></p>
</div>
<h2 id="paga">PAGA</h2>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x">PAGA</a> (partition-based graph abstraction) is a python module available in the python counterpart to seurat known as <code>scanpy</code>. <code>PAGA</code> provides both a very nice visualization technique, and a method for trajectory analysis. Whereas PC and DPT analyses are good for simple trajectories, and <code>Slingshot</code> is good for bi- and multifuracting trajectories, PAGA is good for more complex trajectories.</p>
<div class="layout-chunk" data-layout="l-body">
<p><img src="img/paga.webp" style="display: block; margin: auto;" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<p><img src="img/paga_graph.png" width="2073" style="display: block; margin: auto;" /></p>
</div>
<h2 id="how-to-deal-with-multiple-timepoints">How to deal with multiple timepoints</h2>
<p><a href="https://github.com/farrellja/URD">URD</a> <a href="https://github.com/wagnerde/STITCH">STITCH</a> <a href="https://broadinstitute.github.io/wot/">Waddington-OT</a></p>
<div class="layout-chunk" data-layout="l-body">
<p><img src="img/urd.jpg" width="570" style="display: block; margin: auto;" /></p>
</div>
<h2 id="rna-velocity">RNA velocity</h2>
<p>Based on a cool concept that while the exonic transcriptome represents the current state of a cell, the intronic transcriptome represents what the cell is going to be in the near future. By introducing a transcriptional dynamic model, RNA velocity analysis predicts the directional flow of cell state transition, which greatly expands the application of scRNA-seq to capture dynamics of molecular changes. There is a python and R package implementation for performing analysis, both of which requires processing a bam file to quantitate unspliced and spliced mRNAs per cell (which we cant do within the constraints of the workshop). Recently a scalable (python) toolkit for RNA velocity analysis in single cells called <code>scVelo</code> was published, which can do additional analysis such as estimating the directional velocity pseudotime.</p>
<p><a href="http://velocyto.org/" class="uri">http://velocyto.org/</a> <a href="https://github.com/velocyto-team/velocyto.R" class="uri">https://github.com/velocyto-team/velocyto.R</a> <a href="https://scvelo.readthedocs.io/" class="uri">https://scvelo.readthedocs.io/</a></p>
<p><img src="https://user-images.githubusercontent.com/31883718/80227452-eb822480-864d-11ea-9399-56886c5e2785.gif" /></p>
<div class="layout-chunk" data-layout="l-body">
<p><img src="img/rna_velo.jpg" width="3963" style="display: block; margin: auto;" /></p>
</div>
<h2 id="recommendations">Recommendations:</h2>
<ol type="1">
<li><p>Limit analysis to relevant cell populations (i.e.dont try to infer trajectories where you know they dont exist)</p></li>
<li><p>Start with simple methods and trajectories.</p></li>
<li><p>Complement pseudotime trajectories with RNA velocity analyses to provide directionality and orthogonal validation.</p></li>
</ol>
<h2 id="credits">Credits:</h2>
<p>Much of this material was either copied or motivated by the <code>slingshot</code>, <code>destiny</code>, and <code>tradeSeq</code> vignettes. Some wording was also borrowed/copied from <a href="https://github.com/quadbiolab/scRNAseq_analysis_vignette/blob/master/Tutorial.pdf">this</a> scRNA-seq analyis vignette (which I recommend).</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="updates-and-corrections">Corrections</h3>
<p>If you see mistakes or want to suggest changes, please <a href="https://github.com/rnabioco/cellar/issues/new">create an issue</a> on the source repository.</p>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
