<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Ryan Sheridan" />


<title>Working with Multi-modal Data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="1_intro.html">Introduction</a>
</li>
<li>
  <a href="2_filtering_QC.html">Quality Control</a>
</li>
<li>
  <a href="3_norm_viz_clustering.html">Clustering</a>
</li>
<li>
  <a href="4_markers.html">Cell Type Annotation</a>
</li>
<li>
  <a href="5_trajectories.html">Pseudotime</a>
</li>
<li>
  <a href="6_alignment.html">Dataset Alignment</a>
</li>
<li>
  <a href="7_multimodal.html">Multi-modal Data</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore"><b style="font-size:45px;">Working with Multi-modal Data</b></h1>
<h4 class="author">Ryan Sheridan</h4>
<h4 class="date">August 15<sup>th</sup>, 2019</h4>

</div>


<p><br></p>
<p><br></p>
<div id="experimental-design" class="section level2">
<h2>Experimental design</h2>
<div id="cite-seq-reagents" class="section level3">
<h3>CITE-seq reagents</h3>
<p>Biolegend is the main company selling CITE-seq and cell hashing antibodies (<a href="https://www.biolegend.com/en-us/totalseq">TotalSeq</a>). Biolegend reagents are divided into three product lines:</p>
<ul>
<li>TotalSeq-A: 3’ gene expression, v2 and v3 chemistry</li>
<li>TotalSeq-B: 3’ gene expression, v3 chemistry</li>
<li>TotalSeq-C: 5’ gene expression and V(D)J</li>
</ul>
<p><img src="7_multimodal_files2/Totalseq_compatibility_figure_V1.png" /></p>
</div>
<div id="cell-hashing-reagents" class="section level3">
<h3>Cell hashing reagents</h3>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1603-1">Cell hashing</a> allows for sample multiplexing and “super-loaded” runs with &gt;10,000 captured cells. Super-loading results in higher doublet rates (~10% for 10,000 captured cells), but these doublets can be removed by identifying cell barcodes that are associated with multiple hashtag oligos.</p>
<p><img src="7_multimodal_files2/cell_hashing_diagram.png" /></p>
<p>Biolegend cell hashing reagents for human cells include a mix of two antibodies that recognize CD298 and β2 microglobulin. Mouse hashing antibodies recognize CD45 and H-2 MHC class I.</p>
<p>TotalSeq-A reagents use a different PCR handle for CITE-seq and cell hashing antibodies. This means two separate libraries have to be prepared. To ensure that the correct libraries are created, it is import to tell the sequencing core which types of antibodies were included in the experiment.</p>
<p>TotalSeq-C reagents use the same PCR handle for CITE-seq and cell hashing antibodies, which means that only a single library will be prepared. However, to ensure that the correct libraries are created the core should be notified of all reagents used for the experiment.</p>
<p><a href="https://www.nature.com/articles/s41592-019-0433-8">MULTI-seq</a> uses lipid- and cholesterol-modified oligonucleotides.</p>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="creating-a-seurat-object-with-multiple-assays" class="section level2">
<h2>Creating a Seurat object with multiple assays</h2>
<div id="loading-counts-matrices" class="section level3">
<h3>Loading counts matrices</h3>
<p>The <code>Read10X</code> function can be used with the output directory generated by Cell Ranger. However, our count data is stored as comma-separated files, which we can load as data.frames and then convert to sparse matrices.</p>
<pre class="r"><code># Data URL
data_url &lt;- &quot;https://scrnaseq-workshop.s3-us-west-2.amazonaws.com&quot;

# Function to import counts
import_counts &lt;- function(file_name, file_url = data_url) {
  mtx &lt;- file.path(file_url, file_name) %&gt;%
    read_csv() %&gt;%
    column_to_rownames(&quot;X1&quot;) %&gt;%
    as.sparse()

  mtx
}

# Import gene expression matrix
rna_mtx &lt;- import_counts(&quot;CITEseq_cDNA.csv.gz&quot;)

# Import CITE-seq matrix
adt_mtx &lt;- import_counts(&quot;CITEseq_ADT.csv.gz&quot;)

rownames(adt_mtx) &lt;- str_c(&quot;adt-&quot;, rownames(adt_mtx))
adt_mtx[, 1:10]</code></pre>
<pre><code>#&gt; 8 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                               
#&gt; adt-CD14   11   2  14   1   .   1   1  . .   1
#&gt; adt-CD19    9   . 165 295   .   3   6  4 2   1
#&gt; adt-CD3     .  19  79   2   3   6   9  3 3   4
#&gt; adt-CD4    69 360 293   7 159 321   1  7 7   4
#&gt; adt-CD45    .   .   3   4   .   1   .  . 1   .
#&gt; adt-CD45RA 25   3 246  16   8   5  18  6 4   9
#&gt; adt-CD45RO  6   4 108   6   .   2   1  2 1   3
#&gt; adt-CD8     7   .  36   6   .   . 473 21 7 197</code></pre>
<pre class="r"><code># Import HTO matrix
hto_mtx &lt;- import_counts(&quot;CITEseq_HTO.csv.gz&quot;)

hto_mtx[, 1:10]</code></pre>
<pre><code>#&gt; 4 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#&gt;                                             
#&gt; HTO28 351   .   .  .   2 161   .   .   .   .
#&gt; HTO29   6   2   3  1   1   2   2 107   2   .
#&gt; HTO30   . 131 177 60   .   .   .   . 239 155
#&gt; HTO44   1   .   .  . 122   . 172   .   1   .</code></pre>
<p><br></p>
</div>
<div id="creating-a-seurat-object" class="section level3">
<h3>Creating a Seurat object</h3>
<p>When adding multiple assays to a Seurat object, we first must identify cell barcodes that are present in all of the datasets. If one of the assays has a different number of cell barcodes Seurat will throw an error.</p>
<pre class="r"><code># Get list of common cell barcodes
rna_bcs &lt;- colnames(rna_mtx)
adt_bcs &lt;- colnames(adt_mtx)
hto_bcs &lt;- colnames(hto_mtx)

merged_bcs &lt;- rna_bcs %&gt;%
  intersect(adt_bcs) %&gt;%
  intersect(hto_bcs)

# Create Seurat object
sobj &lt;- rna_mtx[, merged_bcs] %&gt;%
  CreateSeuratObject(min.cells = 5)

# Add CITE-seq and cell hashing data to Seurat object
sobj[[&quot;ADT&quot;]] &lt;- adt_mtx[, merged_bcs] %&gt;%
  CreateAssayObject()

sobj[[&quot;HTO&quot;]] &lt;- hto_mtx[, merged_bcs] %&gt;%
  CreateAssayObject()

sobj</code></pre>
<pre><code>#&gt; An object of class Seurat 
#&gt; 15032 features across 4292 samples within 3 assays 
#&gt; Active assay: RNA (15020 features)
#&gt;  2 other assays present: ADT, HTO</code></pre>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="demultiplexing-hashed-samples" class="section level2">
<h2>Demultiplexing hashed samples</h2>
<div id="normalizing-hto-counts" class="section level3">
<h3>Normalizing HTO counts</h3>
<p>To account for differences in antibody binding efficiencies, CITE-seq and cell hashing data can be normalized by performing a centered log-ratio transformation for each individual antibody.</p>
<pre class="r"><code># Normalize HTO counts
sobj &lt;- sobj %&gt;%
  NormalizeData(
    assay = &quot;HTO&quot;,
    normalization.method = &quot;CLR&quot;
  )</code></pre>
<p><br></p>
</div>
<div id="sample-demultiplexing-and-identification-of-doublets" class="section level3">
<h3>Sample demultiplexing and identification of doublets</h3>
<p>To demultiplex hashed samples, the <code>HTODemux</code> function uses the normalized HTO counts for k-medoids clustering. This results in a cluster for each HTO A background signal is then calculated for each HTO using cells that are not present in the HTO-specific cluster. Outlier cells from this background signal are then classified as being “positive” for the HTO.</p>
<p>Cells that are positive for multiple HTOs are classified as doublets and cells that are not positive for any HTO are classified as “negative” cells.</p>
<p>The <code>HTODemux</code> function automatically adds several columns to the meta.data table.</p>
<ul>
<li>HTO_classification: shows positive HTOs that were identified for the cell</li>
<li>HTO_classification.global: singlet classification (singlet, doublet, negative)</li>
<li>hash.ID: final HTO assignment including doublet and negative classifications</li>
</ul>
<pre class="r"><code># Demultiplex samples
# By default HTODemux will look for the &quot;HTO&quot; assay
sobj &lt;- sobj %&gt;%
  HTODemux(positive.quantile = 0.97)

sobj@meta.data %&gt;%
  head()</code></pre>
<pre><code>#&gt;                     orig.ident nCount_RNA nFeature_RNA nCount_ADT
#&gt; AAACCTGAGACAAAGG SeuratProject       3054         1217        388
#&gt; AAACCTGAGCTACCGC SeuratProject       2411          989        337
#&gt; AAACCTGCACCAGGTC SeuratProject       4182         1394        509
#&gt; AAACCTGCACTCTGTC SeuratProject       3681         1456         43
#&gt; AAACCTGGTATGAAAC SeuratProject       2503         1255        530
#&gt; AAACCTGGTGAGTATA SeuratProject       1241          692        352
#&gt;                  nFeature_ADT nCount_HTO nFeature_HTO HTO_maxID
#&gt; AAACCTGAGACAAAGG            5        358            3     HTO28
#&gt; AAACCTGAGCTACCGC            8        180            2     HTO30
#&gt; AAACCTGCACCAGGTC            7        163            2     HTO28
#&gt; AAACCTGCACTCTGTC            6        174            2     HTO44
#&gt; AAACCTGGTATGAAAC            6        638            2     HTO28
#&gt; AAACCTGGTGAGTATA            6         71            2     HTO44
#&gt;                  HTO_secondID HTO_margin HTO_classification
#&gt; AAACCTGAGACAAAGG        HTO29   3.211027              HTO28
#&gt; AAACCTGAGCTACCGC        HTO29   3.448779              HTO30
#&gt; AAACCTGCACCAGGTC        HTO29   2.802755              HTO28
#&gt; AAACCTGCACTCTGTC        HTO29   3.148890              HTO44
#&gt; AAACCTGGTATGAAAC        HTO29   3.649778              HTO28
#&gt; AAACCTGGTGAGTATA        HTO29   2.285112              HTO44
#&gt;                  HTO_classification.global hash.ID
#&gt; AAACCTGAGACAAAGG                   Singlet   HTO28
#&gt; AAACCTGAGCTACCGC                   Singlet   HTO30
#&gt; AAACCTGCACCAGGTC                   Singlet   HTO28
#&gt; AAACCTGCACTCTGTC                   Singlet   HTO44
#&gt; AAACCTGGTATGAAAC                   Singlet   HTO28
#&gt; AAACCTGGTGAGTATA                   Singlet   HTO44</code></pre>
<pre class="r"><code># Summarize cell classifications
table(sobj$HTO_classification.global)</code></pre>
<pre><code>#&gt; 
#&gt;  Doublet Negative  Singlet 
#&gt;      386        1     3905</code></pre>
<pre class="r"><code># Create ridge plots showing HTO signal
sobj %&gt;%
  RidgePlot(
    assay    = &quot;HTO&quot;,
    features = rownames(hto_mtx),
    ncol     = 2
  )</code></pre>
<p><img src="7_multimodal_files/figure-html/Demultiplex%20samples-1.png" width="768" style="display: block; margin: auto;" /></p>
<p><br></p>
</div>
<div id="exercise-compare-the-number-of-cells-with-each-hash.id-and-calculate-the-doublet-rate" class="section level3">
<h3>EXERCISE: Compare the number of cells with each hash.ID and calculate the doublet rate</h3>
<pre class="r"><code># Use the meta.data table to compare the number of cells with each hash.ID

# sobj@meta.data %&gt;%</code></pre>
<p><br></p>
</div>
<div id="answer" class="section level3">
<h3>ANSWER</h3>
<pre><code>#&gt; # A tibble: 6 x 2
#&gt;   hash.ID     fract
#&gt;   &lt;fct&gt;       &lt;dbl&gt;
#&gt; 1 Doublet  0.0899  
#&gt; 2 HTO28    0.251   
#&gt; 3 HTO29    0.229   
#&gt; 4 HTO30    0.199   
#&gt; 5 HTO44    0.230   
#&gt; 6 Negative 0.000233</code></pre>
<p><img src="7_multimodal_files/figure-html/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="filtering-data-and-assessing-quality" class="section level2">
<h2>Filtering data and assessing quality</h2>
<div id="assessing-data-quality" class="section level3">
<h3>Assessing data quality</h3>
<pre class="r"><code># Add mitochondrial percentage to meta.data table
sobj &lt;- sobj %&gt;%
  PercentageFeatureSet(
    assay    = &quot;RNA&quot;,
    pattern  = &quot;^MT-&quot;, 
    col.name = &quot;percent_mito&quot;
  )

# Create violin plots for gene expression data
sobj %&gt;%
  VlnPlot(
    features = c(&quot;nCount_RNA&quot;, &quot;nFeature_RNA&quot;, &quot;percent_mito&quot;), 
    ncol     = 3,
    pt.size  = 0.25
  )</code></pre>
<p><img src="7_multimodal_files/figure-html/Create%20RNA%20violin%20plots-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Aim to sequence CITE-seq libraries at 2k-5k reads/cell, cell hashing 1k-2k reads/cell

sobj %&gt;%
  VlnPlot(
    features = c(&quot;nCount_ADT&quot;, &quot;nCount_HTO&quot;),
    ncol    = 2,
    pt.size = 0.25,
    log     = T
  )</code></pre>
<p><img src="7_multimodal_files/figure-html/Create%20ADT%20and%20HTO%20violin%20plots-1.png" width="384" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Filter cells based on HTO class, number of genes, and percent mito UMIs
filt_so &lt;- sobj %&gt;%
  subset(
    nFeature_RNA &gt; 250 &amp;   # Remove cells with &lt; 250 detected genes
    nFeature_RNA &lt; 2500 &amp;  # Remove cells with &gt; 2500 detected genes (could be doublets)
    percent_mito &lt; 15 &amp;    # Remove cells with &gt; 0.15 mito/total reads
    HTO_classification.global == &quot;Singlet&quot;
  )

filt_so</code></pre>
<pre><code>#&gt; An object of class Seurat 
#&gt; 15032 features across 3686 samples within 3 assays 
#&gt; Active assay: RNA (15020 features)
#&gt;  2 other assays present: ADT, HTO</code></pre>
<pre class="r"><code># Rename cell identities with sample names
filt_so &lt;- filt_so %&gt;%
  RenameIdents(
    &quot;HTO28&quot; = &quot;PBMC-1&quot;,
    &quot;HTO29&quot; = &quot;PBMC-2&quot;,
    &quot;HTO30&quot; = &quot;PBMC-3&quot;,
    &quot;HTO44&quot; = &quot;PBMC-4&quot;
  )

# Add sample names to meta.data table
filt_so &lt;- filt_so %&gt;%
  AddMetaData(
    metadata = Idents(filt_so),
    col.name = &quot;sample&quot;
  )

filt_so@meta.data %&gt;%
  head()</code></pre>
<pre><code>#&gt;                     orig.ident nCount_RNA nFeature_RNA nCount_ADT
#&gt; AAACCTGAGACAAAGG SeuratProject       3054         1217        388
#&gt; AAACCTGCACCAGGTC SeuratProject       4182         1394        509
#&gt; AAACCTGGTATGAAAC SeuratProject       2503         1255        530
#&gt; AAACGGGTCACCTTAT SeuratProject       2434         1165        360
#&gt; AAAGATGAGTGTACCT SeuratProject       3851         1183        273
#&gt; AAAGATGCACAGCCCA SeuratProject       2657         1381        111
#&gt;                  nFeature_ADT nCount_HTO nFeature_HTO HTO_maxID
#&gt; AAACCTGAGACAAAGG            5        358            3     HTO28
#&gt; AAACCTGCACCAGGTC            7        163            2     HTO28
#&gt; AAACCTGGTATGAAAC            6        638            2     HTO28
#&gt; AAACGGGTCACCTTAT            4        172            1     HTO28
#&gt; AAAGATGAGTGTACCT            5         72            3     HTO28
#&gt; AAAGATGCACAGCCCA            7        508            1     HTO28
#&gt;                  HTO_secondID HTO_margin HTO_classification
#&gt; AAACCTGAGACAAAGG        HTO29   3.211027              HTO28
#&gt; AAACCTGCACCAGGTC        HTO29   2.802755              HTO28
#&gt; AAACCTGGTATGAAAC        HTO29   3.649778              HTO28
#&gt; AAACGGGTCACCTTAT        HTO29   3.096152              HTO28
#&gt; AAAGATGAGTGTACCT        HTO30   2.044904              HTO28
#&gt; AAAGATGCACAGCCCA        HTO29   4.148771              HTO28
#&gt;                  HTO_classification.global hash.ID percent_mito sample
#&gt; AAACCTGAGACAAAGG                   Singlet   HTO28     7.105435 PBMC-1
#&gt; AAACCTGCACCAGGTC                   Singlet   HTO28     7.747489 PBMC-1
#&gt; AAACCTGGTATGAAAC                   Singlet   HTO28     7.311227 PBMC-1
#&gt; AAACGGGTCACCTTAT                   Singlet   HTO28     5.258833 PBMC-1
#&gt; AAAGATGAGTGTACCT                   Singlet   HTO28     5.556998 PBMC-1
#&gt; AAAGATGCACAGCCCA                   Singlet   HTO28     4.516372 PBMC-1</code></pre>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="normalizing-gene-expression-and-antibody-data" class="section level2">
<h2>Normalizing gene expression and antibody data</h2>
<p>Like the cell hashing data, CITE-seq counts can be normalized by performing a centered log-ratio transformation.</p>
<pre class="r"><code># Normalize expression data
filt_so &lt;- filt_so %&gt;%
  NormalizeData(
    normalization.method = &quot;LogNormalize&quot;,
    verbose = FALSE
  ) %&gt;%
  FindVariableFeatures(verbose = FALSE) %&gt;%
  ScaleData(verbose = FALSE)

# Normalize CITE-seq data
filt_so &lt;- filt_so %&gt;%
  NormalizeData(
    assay = &quot;ADT&quot;,
    normalization.method = &quot;CLR&quot;   # Centered log-ratio normalization
  ) %&gt;%
  ScaleData(
    assay   = &quot;ADT&quot;,
    verbose = FALSE
  )</code></pre>
<p><br></p>
<p><br></p>
</div>
<div id="clustering-cells-based-on-gene-expression" class="section level2">
<h2>Clustering cells based on gene expression</h2>
<div id="perform-pricipal-component-analysis" class="section level3">
<h3>Perform pricipal component analysis</h3>
<pre class="r"><code># Perform PCA
filt_so &lt;- filt_so %&gt;%
  RunPCA(verbose = FALSE)       # By default only variable features are used for PCA

# Plot standard deviations of principal components
elbow_plot &lt;- filt_so %&gt;%
  ElbowPlot(ndims = 50)

# Create scatter plot comparing PC-1 and PC-2
PCA_plot &lt;- filt_so %&gt;%
  DimPlot(reduction = &quot;pca&quot;)

plot_grid(
  elbow_plot, PCA_plot,
  nrow       = 1,
  rel_widths = c(0.45, 0.55)
)</code></pre>
<p><img src="7_multimodal_files/figure-html/Perform%20PCA-1.png" width="960" style="display: block; margin: auto;" /></p>
<p><br></p>
</div>
<div id="cluster-cells-using-gene-expression-data" class="section level3">
<h3>Cluster cells using gene expression data</h3>
<pre class="r"><code># Cluster cells
filt_so &lt;- filt_so %&gt;%
  FindNeighbors(dims = 1:20) %&gt;%
  FindClusters(
    resolution = 0.6,
    verbose    = FALSE
  )

# Run UMAP
# Use same PCs as clustering, by default reduction = &quot;pca&quot;
filt_so &lt;- filt_so %&gt;%
  RunUMAP(dims = 1:20)

# Add RNA clusters to meta.data
filt_so &lt;- filt_so %&gt;%
  AddMetaData(
    metadata = Idents(filt_so),
    col.name = &quot;RNA_clusters&quot;
  )

# Create UMAP
filt_so %&gt;%
  DimPlot(reduction = &quot;umap&quot;)</code></pre>
<p><img src="7_multimodal_files/figure-html/Cluster%20cells%20and%20run%20UMAP-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="clustering-cells-based-on-antibody-signal" class="section level2">
<h2>Clustering cells based on antibody signal</h2>
<div id="cluster-cells-using-antibody-signal" class="section level3">
<h3>Cluster cells using antibody signal</h3>
<p>Since there are only a few antibodies and not many dimensions, instead of performing PCA we can just create a distance matrix to use for clustering. To do this we can use the <code>dist</code> function which computes distances between the rows. To calculate distances between cells, we need to transpose the matrix.</p>
<p>We can then use the resulting distance matrix as input to the <code>FindNeighbors</code> function, which creates a Shared Nearest Neighbor (SNN) graph that is used for clustering.</p>
<pre class="r"><code># Create a standard euclidean distance matrix for clustering
adt_data &lt;- filt_so %&gt;%
  GetAssayData(
    assay = &quot;ADT&quot;, 
    slot = &quot;data&quot;
  )

adt_data[, 1:5]

adt_dist &lt;- adt_data %&gt;%
  t() %&gt;%
  dist()
  
# Find clusters using distance matrix
adt_graphs &lt;- FindNeighbors(adt_dist)

filt_so[[&quot;ADT_snn&quot;]] &lt;- adt_graphs$snn

filt_so &lt;- filt_so %&gt;%
  FindClusters(
    resolution = 0.2,
    graph.name = &quot;ADT_snn&quot;
  )

# Add ADT clusters to meta.data
filt_so &lt;- filt_so %&gt;%
  AddMetaData(
    metadata = Idents(filt_so),
    col.name = &quot;ADT_clusters&quot;
  )</code></pre>
<p><br></p>
</div>
<div id="run-umap-using-the-antibody-signal" class="section level3">
<h3>Run UMAP using the antibody signal</h3>
<pre class="r"><code># Run UMAP
filt_so &lt;- filt_so %&gt;%
  RunUMAP(
    reduction.name = &quot;adt_umap&quot;,
    reduction.key  = &quot;ADTUMAP_&quot;,
    graph          = &quot;ADT_snn&quot;
  )

# Create UMAP
filt_so %&gt;%
  DimPlot(
    reduction = &quot;adt_umap&quot;, 
    group.by = &quot;sample&quot;, 
    split.by = &quot;sample&quot;, 
    ncol = 2
  )</code></pre>
<p><img src="7_multimodal_files/figure-html/Run%20UMAP%20for%20antibody%20data-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><br></p>
</div>
<div id="identify-marker-proteins" class="section level3">
<h3>Identify marker proteins</h3>
<pre><code>#&gt;                     p_val avg_logFC pct.1 pct.2     p_val_adj cluster
#&gt; adt-CD8      0.000000e+00 1.9569546 1.000 0.936  0.000000e+00       0
#&gt; adt-CD4      0.000000e+00 2.5652784 1.000 0.903  0.000000e+00       1
#&gt; adt-CD14    3.266554e-144 1.6929125 0.885 0.346 2.613243e-143       3
#&gt; adt-CD45    3.678499e-102 0.8132439 0.934 0.612 2.942799e-101       3
#&gt; adt-CD45RO  2.822220e-101 1.3871472 0.987 0.865 2.257776e-100       3
#&gt; adt-CD19     2.468598e-78 0.3683039 0.993 0.912  1.974879e-77       3
#&gt; adt-CD45RA   5.105446e-09 0.4612004 0.987 0.982  4.084357e-08       3
#&gt; adt-CD191    1.720165e-81 3.4396729 1.000 0.916  1.376132e-80       4
#&gt; adt-CD45RA1  8.828050e-20 0.4139237 0.992 0.982  7.062440e-19       4
#&gt; adt-CD45RO1  1.411546e-17 0.5103029 0.984 0.871  1.129237e-16       4
#&gt;                   gene
#&gt; adt-CD8        adt-CD8
#&gt; adt-CD4        adt-CD4
#&gt; adt-CD14      adt-CD14
#&gt; adt-CD45      adt-CD45
#&gt; adt-CD45RO  adt-CD45RO
#&gt; adt-CD19      adt-CD19
#&gt; adt-CD45RA  adt-CD45RA
#&gt; adt-CD191     adt-CD19
#&gt; adt-CD45RA1 adt-CD45RA
#&gt; adt-CD45RO1 adt-CD45RO</code></pre>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="visualizing-antibody-signal" class="section level2">
<h2>Visualizing antibody signal</h2>
<div id="overlay-antibody-signal-on-umaps" class="section level3">
<h3>Overlay antibody signal on UMAPs</h3>
<pre class="r"><code># Set active.assay to ADT
filt_so@active.assay &lt;- &quot;ADT&quot;

# Overlay antibody signal on gene expression UMAP
filt_so %&gt;%
  FeaturePlot(
    reduction = &quot;umap&quot;,
    features  = c(&quot;adt-CD4&quot;, &quot;CD4&quot;, &quot;adt-CD8&quot;, &quot;CD8A&quot;)
  )</code></pre>
<p><img src="7_multimodal_files/figure-html/Overlay%20antibody%20signal%20on%20UMAPs-1.png" width="960" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Overlay antibody signal on antibody UMAP
filt_so %&gt;%
  FeaturePlot(
    reduction = &quot;adt_umap&quot;,
    features  = c(&quot;adt-CD4&quot;, &quot;CD4&quot;, &quot;adt-CD8&quot;, &quot;CD8A&quot;)
  )</code></pre>
<p><img src="7_multimodal_files/figure-html/Overlay%20antibody%20signal%20on%20UMAPs-2.png" width="960" style="display: block; margin: auto;" /></p>
<p><br></p>
</div>
<div id="ridge-plots" class="section level3">
<h3>Ridge plots</h3>
<pre class="r"><code># Create ridge plot
filt_so %&gt;%
  RidgePlot(features = c(
    &quot;adt-CD14&quot;, &quot;adt-CD45&quot;,
    &quot;adt-CD19&quot;, &quot;adt-CD3&quot;,  
    &quot;adt-CD4&quot;,  &quot;adt-CD8&quot;
  ))</code></pre>
<p><img src="7_multimodal_files/figure-html/Create%20ridge%20plots-1.png" width="960" style="display: block; margin: auto;" /></p>
<p><br></p>
</div>
<div id="heatmaps" class="section level3">
<h3>Heatmaps</h3>
<pre class="r"><code>filt_so %&gt;%
  DoHeatmap(
    features = rownames(filt_so), 
    angle    = 0
  ) + 
  NoLegend()</code></pre>
<p><img src="7_multimodal_files/figure-html/Create%20heatmaps-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><br></p>
</div>
<div id="violin-plots" class="section level3">
<h3>Violin plots</h3>
<pre class="r"><code>filt_so %&gt;%
  VlnPlot(
    features = c(&quot;adt-CD4&quot;, &quot;CD4&quot;, &quot;adt-CD19&quot;, &quot;CD19&quot;),
    ncol     = 2
  )</code></pre>
<p><img src="7_multimodal_files/figure-html/Create%20violinplots-1.png" width="768" style="display: block; margin: auto;" /></p>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="classifying-cells-based-on-antibody-signal" class="section level2">
<h2>Classifying cells based on antibody signal</h2>
<div id="identify-cd19-cells" class="section level3">
<h3>Identify CD19+ cells</h3>
<pre class="r"><code># Plot CD3 and CD19 signal
filt_so %&gt;%
  FeatureScatter(&quot;adt-CD3&quot;, &quot;adt-CD19&quot;)</code></pre>
<p><img src="7_multimodal_files/figure-html/Filter%20using%20antibody%20signal-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Identify CD19+ cells using antibody signal
CD19_cells &lt;- filt_so %&gt;%
  subset(`adt-CD19` &gt; 2.5 &amp; `adt-CD3` &lt; 1) %&gt;%
  Cells()

CD19_cells %&gt;%
  head()</code></pre>
<pre><code>#&gt; [1] &quot;CCGTGGAAGCTCCCAG&quot; &quot;AACTTTCCAGCCTTGG&quot; &quot;AAGGCAGTCGCCTGAG&quot;
#&gt; [4] &quot;AGATCTGCAGGACCCT&quot; &quot;AGATCTGGTCATCGGC&quot; &quot;AGGTCATTCTTGTACT&quot;</code></pre>
<pre class="r"><code># Set cell identities
labeled_so &lt;- filt_so %&gt;%
  SetIdent(value = &quot;Other&quot;) %&gt;%
  SetIdent(
    value = &quot;CD19+&quot;,
    cells = CD19_cells
  )

labeled_so@active.ident %&gt;%
  head()</code></pre>
<pre><code>#&gt; AAACCTGAGACAAAGG AAACCTGCACCAGGTC AAACCTGGTATGAAAC AAACGGGTCACCTTAT 
#&gt;            Other            Other            Other            Other 
#&gt; AAAGATGAGTGTACCT AAAGATGCACAGCCCA 
#&gt;            Other            Other 
#&gt; Levels: CD19+ Other</code></pre>
<pre class="r"><code># Label UMAP with new cell identities
labeled_so %&gt;%
  DimPlot(reduction = &quot;adt_umap&quot;)</code></pre>
<p><img src="7_multimodal_files/figure-html/Filter%20using%20antibody%20signal-2.png" width="672" style="display: block; margin: auto;" /></p>
<p><br></p>
</div>
<div id="filter-cells-using-cellselector" class="section level3">
<h3>Filter cells using CellSelector()</h3>
<pre class="r"><code># Plot CD3 and CD19 signal
CD19_plot &lt;- filt_so %&gt;%
  FeatureScatter(&quot;adt-CD3&quot;, &quot;adt-CD19&quot;)

CD19_plot

# Identify CD19+ cells using antibody signal
labeled_so &lt;- filt_so %&gt;%
  SetIdent(value = &quot;Other&quot;)

labeled_so &lt;- CellSelector(
  plot   = CD19_plot,
  object = labeled_so,
  ident  = &quot;B cells&quot;
)

labeled_so@active.ident %&gt;%
  head()

# Label UMAP with new cell identities
labeled_so %&gt;%
  DimPlot(reduction = &quot;adt_umap&quot;)</code></pre>
<p><br></p>
</div>
<div id="exercise-identify-cd4-and-cd8-cells" class="section level3">
<h3>EXERCISE: Identify CD4+ and CD8+ cells</h3>
<pre class="r"><code># Plot CD4 and CD8 signal

# filt_so %&gt;%
#   FeatureScatter(&quot;adt_CD4&quot;, &quot;adt_CD8&quot;)

# Identify CD4+ and CD8+ cells using antibody signal

# Set cell identities

# Label UMAP with new cell identities</code></pre>
<p><br></p>
</div>
<div id="answer-1" class="section level3">
<h3>ANSWER</h3>
<p><img src="7_multimodal_files/figure-html/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /><img src="7_multimodal_files/figure-html/unnamed-chunk-5-2.png" width="672" style="display: block; margin: auto;" /></p>
<p><br></p>
<p><br></p>
</div>
</div>
<div id="viewing-results-with-the-ucsc-cell-browser" class="section level2">
<h2>Viewing results with the UCSC Cell Browser</h2>
<div id="merge-gene-expression-and-antibody-matrices" class="section level3">
<h3>Merge gene expression and antibody matrices</h3>
<pre class="r"><code># Combine RNA and ADT matrices
merged_so   &lt;- labeled_so
RNA_data    &lt;- merged_so@assays$RNA@data
ADT_data    &lt;- merged_so@assays$ADT@data
merged_data &lt;- rbind(RNA_data, ADT_data)

# Add merged matrix to Seurat object
merged_so@assays$RNA@data &lt;- merged_data

# Set active assay
merged_so@active.assay &lt;- &quot;RNA&quot;</code></pre>
<p><br></p>
</div>
<div id="create-cell-browser-files" class="section level3">
<h3>Create Cell Browser files</h3>
<pre class="r"><code># Create Cell Browser directories
dir.create(
  path = &quot;cellbrowser/RNA&quot;,
  recursive = T
)

# Create Cell Browser files for gene expression data
merged_so %&gt;%
  ExportToCellbrowser(
    dir          = &quot;cellbrowser/RNA&quot;,
    reductions   = &quot;umap&quot;,
    dataset.name = &quot;RNA&quot;,

    sample       = &quot;Sample&quot;,
    RNA_clusters = &quot;Cluster&quot;,
    ADT_clusters = &quot;ADT cluster&quot;,
    cell_label   = &quot;Cell label&quot;,
    
    nCount_RNA   = &quot;RNA UMI count&quot;,
    nFeature_RNA = &quot;Gene count&quot;,
    percent_mito = &quot;Percent mito UMIs&quot;,
    nCount_ADT   = &quot;ADT UMI count&quot;,
    nFeature_ADT = &quot;Antibody count&quot;,
    nCount_HTO   = &quot;HTO UMI count&quot;,
    nFeature_HTO = &quot;HTO count&quot;
  )</code></pre>
<pre class="r"><code># Create Cell Browser directories
dir.create(
  path = &quot;cellbrowser/ADT&quot;,
  recursive = T
)

# Export marker genes
ADT_markers %&gt;%
  rename(score = p_val_adj) %&gt;%
  select(cluster, gene, score) %&gt;%
  write_tsv(&quot;cellbrowser/ADT/ADT_markers.tsv&quot;)

# Create Cell Browser files for antibody data
merged_so %&gt;%
  ExportToCellbrowser(
    dir          = &quot;cellbrowser/ADT&quot;,
    reductions   = &quot;adt_umap&quot;,
    dataset.name = &quot;ADT&quot;,
    markers.file = &quot;cellbrowser/ADT/ADT_markers.tsv&quot;,
    
    sample       = &quot;Sample&quot;,
    ADT_clusters = &quot;Cluster&quot;,
    RNA_clusters = &quot;RNA cluster&quot;,
    cell_label   = &quot;Cell label&quot;,
    
    nCount_RNA   = &quot;RNA UMI count&quot;,
    nFeature_RNA = &quot;Gene count&quot;,
    percent_mito = &quot;Percent mito UMIs&quot;,
    nCount_ADT   = &quot;ADT UMI count&quot;,
    nFeature_ADT = &quot;Antibody count&quot;,
    nCount_HTO   = &quot;HTO UMI count&quot;,
    nFeature_HTO = &quot;HTO count&quot;
  )</code></pre>
<p><br></p>
</div>
<div id="build-cell-browser-session" class="section level3">
<h3>Build Cell Browser session</h3>
<pre class="bash"><code>
mkdir -p cellbrowser/html

cbBuild \
  -i cellbrowser/RNA/cellbrowser.conf \
  -i cellbrowser/ADT/cellbrowser.conf \
  -o cellbrowser/html \
  -p 8888
</code></pre>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
